// @stdlib/utils/parse-json/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var noop = require( '@stdlib/utils/noop' );
var parseJSON = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof parseJSON, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		3.14,
		NaN,
		true,
		null,
		undefined,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			parseJSON( value );
		};
	}
});

tape( 'the function throws if not provided a string (reviver)', function test( t ) {
	var values;
	var i;

	values = [
		3.14,
		NaN,
		true,
		null,
		undefined,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			parseJSON( value, noop );
		};
	}
});

tape( 'the function throws if provided a reviver argument which is not a function', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		3.14,
		NaN,
		true,
		null,
		undefined,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			parseJSON( '{"a":"b"}', value );
		};
	}
});

tape( 'the function returns a JSON value if provided valid JSON', function test( t ) {
	var expected;
	var actual;

	expected = {
		'beep': 'boop'
	};
	actual = parseJSON( '{"beep":"boop"}' );
	t.deepEqual( actual, expected, 'deep equal' );

	expected = null;
	actual = parseJSON( 'null' );
	t.equal( actual, expected, 'returns null' );

	expected = true;
	actual = parseJSON( 'true' );
	t.equal( actual, expected, 'returns true' );

	expected = 3.14;
	actual = parseJSON( '3.14' );
	t.equal( actual, expected, 'returns 3.14' );

	t.end();
});

tape( 'the function returns an error if provided invalid JSON', function test( t ) {
	var out = parseJSON( '{"beep":"boop}' );
	t.equal( out instanceof Error, true, 'returns an error' );
	t.end();
});

tape( 'the function supports providing a custom reviver function', function test( t ) {
	var expected;
	var actual;
	var str;

	str = '{"beep":"boop","a":"b"}';
	expected = {
		'beep': 'boop'
	};
	actual = parseJSON( str, reviver );

	t.deepEqual( actual, expected, 'supports custom reviver' );

	t.end();

	function reviver( key, value ) {
		if ( key === '' ) {
			return value;
		}
		if ( key === 'beep' ) {
			return value;
		}
	}
});
