/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import parseComplex128 = require( './index' );


// TESTS //

// The function parses a string representation of a complex number...
{
	const str = '5 + 3j';
	parseComplex128( str ); // $ExpectType Complex128
}

// The function parses a string representation of a complex number (only real part)...
{
    const str = '3';
    parseComplex128( str ); // $ExpectType Complex128
}

// The function parses a string representation of a complex number (only imaginary part)...
{
    const str = '8j';
    parseComplex128( str ); // $ExpectType Complex128
}

// The function throws an error if the input is not recognized as a complex number...
{
	const str = 'beep boop';
	parseComplex128( str ); // $ExpectError
}

// The compiler throws an error if the function is provided a first argument that is not a string...
{
	parseComplex128( true ); // $ExpectError
	parseComplex128( false ); // $ExpectError
	parseComplex128( null ); // $ExpectError
	parseComplex128( undefined ); // $ExpectError
	parseComplex128( 5 ); // $ExpectError
	parseComplex128( [] ); // $ExpectError
	parseComplex128( {} ); // $ExpectError
	parseComplex128( ( x: number ): number => x ); // $ExpectError
}
