<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-multiple-empty-lines

> [ESLint rule][eslint-rules] to enforce that that code does not contain multiple blank lines.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-multiple-empty-lines' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that code does not contain multiple blank lines. Up to two blank lines are allowed before section header comments such as `// MAIN //`.

**Bad**:

<!-- eslint-disable stdlib/no-multiple-empty-lines -->

```javascript
/**
* Returns the maximum absolute value.
*
* @param {number} [x] - first number
* @param {number} [y] - second number
* @param {...number} [args] - numbers
* @returns {number} maximum absolute value
*/
function maxabs( x, y ) {
    var nargs;
    var args;
    var i;


    nargs = arguments.length;
    if ( nargs === 0 ) {
        return PINF;
    }
    if ( nargs === 2 ) {
        return max( abs( x ), abs( y ) );
    }
    args = new Array( nargs );
    for ( i = 0; i < nargs; i++ ) {
        args[ i ] = abs( arguments[ i ] );
    }


    return max.apply( null, args );
}
```

**Good**:

```javascript
/**
* Returns the maximum absolute value.
*
* @param {number} [x] - first number
* @param {number} [y] - second number
* @param {...number} [args] - numbers
* @returns {number} maximum absolute value
*/
function maxabs( x, y ) {
    var nargs;
    var args;
    var i;

    nargs = arguments.length;
    if ( nargs === 0 ) {
        return PINF;
    }
    if ( nargs === 2 ) {
        return max( abs( x ), abs( y ) );
    }
    args = new Array( nargs );
    for ( i = 0; i < nargs; i++ ) {
        args[ i ] = abs( arguments[ i ] );
    }
    return max.apply( null, args );
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-multiple-empty-lines' );

var linter = new Linter();
var result;
var code;

var opts = {
    'rules': {
        'no-multiple-empty-lines': 'error'
    }
};
linter.defineRule( 'no-multiple-empty-lines', rule );

code = [
    '\'use strict\';',
    '',
    '// MODULES //',
    '',
    'var os = require( \'os\' );',
    '',
    '',
    '// MAIN //',
    '',
    '/**',
    '* Returns the number of CPUs.',
    '*',
    '* @private',
    '* @returns {PositiveInteger} number of CPUs',
    '*',
    '* @example',
    '* var n = numCPUs();',
    '* // returns <number>',
    '*/',
    'function numCPUs() {',
    '  ',
    '  ',
    '  return os.cpus().length;',
    '}',
    '',
    '',
    '',
    '// EXPORTS //',
    '',
    'module.exports = numCPUs();'
].join( '\n' );

result = linter.verify( code, opts );
console.log( result );
/* =>
    [
        {
            'ruleId': 'no-multiple-empty-lines',
            'severity': 2,
            'message': 'Only one blank line is allowed',
            'line': 20,
            'column': 1,
            'nodeType': null,
            'source': 'function numCPUs() {',
            'endLine': 22,
            'endColumn': 1
        },
        {
            'ruleId': 'no-multiple-empty-lines',
            'severity': 2,
            'message': 'Not more than two blank lines are allowed before a section header',
            'line': 24,
            'column': 1,
            'nodeType': null,
            'source': '}',
            'endLine': 27,
            'endColumn': 1
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
