// @stdlib/math/stats/ztest/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var normal = require( '@stdlib/math/base/random/normal' ).factory;
var isArray = require( '@stdlib/assert/is-array' );
var NINF = require( '@stdlib/math/constants/float64-ninf' );
var PINF = require( '@stdlib/math/constants/float64-pinf' );
var ztest = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof ztest, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the `x` argument is not a numeric array', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		[ 1, 2, '5' ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ztest( value, 1.0 );
		};
	}
});

tape( 'the function throws an error if the `x` array has less than two elements', function test( t ) {
	var values;
	var i;

	values = [
		[],
		[ 1 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), Error, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ztest( value, 1.0 );
		};
	}
});

tape( 'the function throws an error when the standard deviation `sigma` is not a positive number', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		-1.0,
		0.0,
		[ 1, 2, '5' ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ztest( [ 2, 3, 4, 3, 4, 3 ], value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		5,
		true,
		undefined,
		null,
		NaN,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}

	t.throws(
		badValue( 'not one of less, greater or two-sided' ),
		Error,
		'throws an error when provided not less, greater or two-sided'
	);
	t.end();

	function badValue( value ) {
		return function badValue() {
			ztest( [ 0.2, 0.5, 0.75 ], 0.5, {
				'alternative': value
			});
		};
	}
});

tape( 'the function throws an error if the `alpha` option is a numeric value outside `[0,1]`', function test( t ) {
	var values;
	var i;

	values = [
		1.1,
		-0.1
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ztest( [ 0.2, 0.5, 0.75 ], 0.5, {
				'alpha': value
			});
		};
	}
});

tape( 'the function computes a one-sample two-sided z-test', function test( t ) {
	var rnorm;
	var out;
	var i;
	var x;

	rnorm = normal({
		'seed': 338
	});

	x = new Array( 100 );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = rnorm( 0.0, 1.0 );
	}
	out = ztest( x, 1.0 );

	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );

	// TODO: Add test fixtures

	t.end();
});

tape( 'the function computes a one-sample one-sided z-test', function test( t ) {
	var rnorm;
	var out;
	var i;
	var x;

	rnorm = normal({
		'seed': 338
	});

	x = new Array( 100 );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = rnorm( 0.0, 1.0 );
	}

	// Alternative: mu > 0.0
	out = ztest( x, 1.0, {
		'alternative': 'greater'
	});
	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );
	t.strictEqual( out.ci[1], PINF, 'CI upper bound is +infinity' );

	// Alternative: mu < 0.0
	out = ztest( x, 1.0, {
		'alternative': 'less'
	});
	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );
	t.strictEqual( out.ci[0], NINF, 'CI lower bound is -infinity' );

	// TODO: Add test fixtures

	t.end();
});

tape( 'the function computes a one-sample two-sided z-test with a custom significance level', function test( t ) {
	var rnorm;
	var out;
	var i;
	var x;

	rnorm = normal({
		'seed': 383
	});

	x = new Array( 100 );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = rnorm( 0.0, 1.0 );
	}

	out = ztest( x, 1.0, {
		'alpha': 0.1
	});

	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );

	// TODO: Add test fixtures

	t.end();
});

tape( 'the function correctly computes a one-sample one-sided z-test with a custom significance level', function test( t ) {
	var rnorm;
	var out;
	var i;
	var x;

	rnorm = normal({
		'seed': 338
	});

	x = new Array( 100 );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = rnorm( 0.0, 1.0 );
	}

	// Alternative: mu > 0.0
	out = ztest( x, 1.0, {
		'alternative': 'greater',
		'alpha': 0.1
	});

	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );
	t.strictEqual( out.ci[1], PINF, 'CI upper bound is +infinity' );

	// Alternative: mu < 0.0
	out = ztest( x, 1.0, {
		'alternative': 'less',
		'alpha': 0.01
	});

	t.strictEqual( typeof out.rejected, 'boolean', '`rejected` property is a boolean' );
	t.strictEqual( typeof out.pValue, 'number', '`pValue` property is a boolean' );
	t.strictEqual( typeof out.statistic, 'number', '`statistic` property is a boolean' );
	t.strictEqual( isArray( out.ci ), true, '`ci` property is an array' );
	t.strictEqual( out.ci[0], NINF, 'CI lower bound is -infinity' );

	// TODO: Add test fixtures

	t.end();
});

tape( 'the function returns an object with a .print() method for generating a formatted output of results of the one-sample z-test', function test( t ) {
	var table;
	var out;
	var x;

	x = [ 0, 1, 2, 3, 4 ];
	out = ztest( x, 2.0 );
	table = out.print();
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );

	out = ztest( x, 2.0, {
		'alternative': 'less'
	});
	table = out.print();
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );

	out = ztest( x, 2.0, {
		'alternative': 'greater',
		'mu': 2
	});
	table = out.print();
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that accepts a `numdigits` option to control the number of decimal digits displayed', function test( t ) {
	var table;
	var out;
	var x;

	x = [ 0, 1, 2, 3, 4 ];
	out = ztest( x, 2.0 );
	table = out.print({
		'numdigits': 6
	});
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that accepts an `options` object', function test( t ) {
	var table;
	var out;
	var x;

	x = [ 0, 1, 2, 3, 4 ];
	out = ztest( x, 2.0 );
	table = out.print( {} );

	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that throws an error if `options` is not a simple object', function test( t ) {
	var values;
	var out;
	var i;
	var x;

	x = [ 0, 1, 2, 3, 4 ];
	out = ztest( x, 2.0 );

	values = [
		'abc',
		4,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print( value );
		};
	}
});

tape( 'the function returns an object with a .print() method that throws an error if the `numdigits` option is not a positive integer', function test( t ) {
	var values;
	var out;
	var i;
	var x;

	x = [ 0, 1, 2, 3, 4 ];
	out = ztest( x, 2.0 );

	values = [
		'abc',
		2.4,
		0.0,
		-1.0,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print({
				'numdigits': value
			});
		};
	}
});
