// @stdlib/utils/curry-right/lib/curry_right.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );
var isPositiveInteger = require( '@stdlib/assert/is-positive-integer' ).isPrimitive;


// MAIN //

/**
* Transforms a function into a sequence of functions each accepting a single argument.
*
* @param {Function} fcn - function to curry
* @param {PositiveInteger} [arity=fcn.length] - number of parameters
* @param {*} [thisArg] - evaluation context
* @throws {TypeError} first argument must be a function
* @throws {TypeError} `arity` argument must be a positive integer
* @returns {Function} curry function
*
* @example
* function add( x, y ) {
*     return x + y;
* }
*
* var f = curryRight( add );
*
* var sum = f( 2 )( 3 );
* // returns 5
*/
function curryRight( fcn, arity, thisArg ) {
	var context;
	var len;
	if ( !isFunction( fcn ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `' + fcn + '`.' );
	}
	if ( arguments.length < 2 ) {
		len = fcn.length;
	}
	else if ( arguments.length > 2 ) {
		len = arity;
		context = thisArg;
		if ( !isPositiveInteger( len ) ) {
			throw new TypeError( 'invalid input argument. Arity argument must be positive integer. Value: `' + len + '`.' );
		}
	}
	else if ( isPositiveInteger( arity ) ) {
		len = arity;
	}
	else {
		len = fcn.length;
		context = arity;
	}
	return createCurried( [] );

	/**
	* Returns a curry function.
	*
	* @private
	* @param {Array} args - arguments
	* @returns {Function} curry function
	*/
	function createCurried( args ) {
		return curried;

		/**
		* Curry function.
		*
		* @private
		* @param {*} v - curried function parameter
		* @returns {(Function|*)} partially applied curry function or curried function result
		*/
		function curried( v ) {
			var cargs = args.slice();
			cargs.unshift( v );
			if ( cargs.length < len ) {
				return createCurried( cargs );
			}
			return fcn.apply( context, cargs );
		} // end FUNCTION curried()
	} // end FUNCTION createCurried()
} // end FUNCTION curryRight()


// EXPORTS //

module.exports = curryRight;
