// @stdlib/assert/is-node/test/test.is_node.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var isNode = require( './../lib/is_node.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof isNode, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if runtime is Node.js (Node.js < v7)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});

	t.equal( isNode(), true, 'returns true' );
	t.end();
});

tape( 'the function returns `true` if runtime is Node.js (Node.js >= v7)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version >=7 (https://github.com/nodejs/node/issues/9274):
		return '[object Object]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});

	t.equal( isNode(), true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`global` variable does not equal the detected global variable)', function test( t ) {
	var isNode;

	function getGlobal() {
		return {};
	}

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'system.global': getGlobal,
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`global` variable has wrong class)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		return '[object beeeeeeep]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`global` variable is not the global scope)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': false
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`process` variable has wrong class)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object beeeeeeep]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`process.versions` is not an object)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function isObject( val ) {
		if ( val === process.versions ) {
			return false;
		}
		return ( typeof val === 'object' );
	}

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': isObject,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`process.versions.node` is not a string primitive)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = isPrimitive;

	function isPrimitive( val ) {
		if ( val === process.versions.node ) {
			return false;
		}
		return ( typeof val === 'string' );
	}

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	t.equal( isNode(), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`process.release` is not an object; Node >=v3.0.0)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = alwaysTrue;

	function isObject( val ) {
		if ( process.release && val === process.release ) {
			return false;
		}
		return ( typeof val === 'object' );
	}

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': isObject,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	// `process.release` was added in Node v3.0.0
	if ( process.release ) {
		t.equal( isNode(), false, 'returns false' );
	} else {
		t.equal( isNode(), true, 'returns true' );
	}
	t.end();
});

tape( 'the function returns `false` if runtime is not Node.js (`process.release.name` is not a string primitive; Node >=v3.0.0)', function test( t ) {
	var isNode;

	function nativeClass( val ) {
		if ( val === process ) {
			return '[object process]';
		}
		// Node.js version <7 (https://github.com/nodejs/node/issues/9274):
		return '[object global]';
	}

	function isString(){}
	isString.isPrimitive = isPrimitive;

	function isPrimitive( val ) {
		if ( process.release && val === process.release.name ) {
			return false;
		}
		return ( typeof val === 'string' );
	}

	function alwaysTrue() {
		return true;
	}

	isNode = proxyquire( './../lib/is_node.js', {
		'@stdlib/utils/native-class': nativeClass,
		'@stdlib/assert/is-plain-object': alwaysTrue,
		'@stdlib/assert/is-string': isString,
		'./global_scope.js': true
	});
	// `process.release` was added in Node v3.0.0
	if ( process.release ) {
		t.equal( isNode(), false, 'returns false' );
	} else {
		t.equal( isNode(), true, 'returns true' );
	}
	t.end();
});
