// @stdlib/utils/try-then/lib/try_then.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );


// MAIN //

/**
* If a function does not throw, returns the function return value; otherwise, returns the return value of a second function `y`.
*
* @param {Function} x - function to try invoking
* @param {Function} y - function to invoke if a function throws
* @throws {TypeError} first argument must be a function
* @returns {*} the return value of either `x` or `y`
*
* @example
* var randu = require( '@stdlib/math/base/random/randu' );
*
* function x() {
*     if ( randu() < 0.5 ) {
*         throw new Error( 'beep' );
*     }
*     return 1.0;
* }
*
* function y() {
*     return randu();
* }
*
* var z = trythen( x, -1.0 );
* // returns <number>
*/
function trythen( x, y ) {
	if ( !isFunction( x ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `'+x+'`.' );
	}
	if ( !isFunction( y ) ) {
		throw new TypeError( 'invalid input argument. Second argument must be a function. Value: `'+y+'`.' );
	}
	try {
		return x();
	} catch ( error ) { // eslint-disable-line no-unused-vars
		return y();
	}
} // end FUNCTION trythen()


// EXPORTS //

module.exports = trythen;
