<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-immediate-require

> [ESLint rule][eslint-rules] enforcing that `require()` expressions are not immediately invoked.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-immediate-require' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that `require()` expressions are not immediately invoked.

**Bad**:

<!-- eslint-disable stdlib/no-immediate-require -->

```javascript
var debug = require( 'debug' )( 'stdlib' );
```

**Good**:

```javascript
var logger = require( 'debug' );

var debug = logger( 'stdlib' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-immediate-require' );

var linter = new Linter();
var result;
var code;

code = 'var pdf = require( \'@stdlib/stats/base/dists/normal/pdf\' ).factory( 0.0, 1.0 );';

linter.defineRule( 'no-immediate-require', rule );

result = linter.verify( code, {
    'rules': {
        'no-immediate-require': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-immediate-require',
            'severity': 2,
            'message': 'require() calls should only use string literals',
            line': 2,
            'column': 15,
            'nodeType': 'CallExpression',
            'source': 'var pdf = require( \'@stdlib/stats/base/dists/normal/pdf\' ).factory( 0.0, 1.0 );',
            'endLine': 2,
            'endColumn': 29
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
