// @stdlib/iter/cartesian-square/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var iterCartesianSquare = require( './../lib/main.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof iterCartesianSquare, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the first argument is not an array-like object', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		true,
		false,
		null,
		undefined,
		{},
		function namedFunction() {},
		NaN,
		Infinity,
		-Infinity
	];

	function badValue(value) {
		return function thrower() {
			iterCartesianSquare(value);
		};
	}

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue(values[i]), TypeError, 'throws an error when provided ' + values[i] );
	}
	t.end();
});

tape( 'the function returns an iterator which generates the Cartesian square of an input array', function test( t ) {
	var expected;
	var actual;
	var iter;

	iter = iterCartesianSquare( [ 'a', 'b' ] );

	expected = [ 'a', 'a' ];
	actual = iter.next().value;
	t.deepEqual( actual, expected, 'returns expected value' );

	expected = [ 'a', 'b' ];
	actual = iter.next().value;
	t.deepEqual( actual, expected, 'returns expected value' );

	expected = [ 'b', 'a' ];
	actual = iter.next().value;
	t.deepEqual( actual, expected, 'returns expected value' );

	expected = [ 'b', 'b' ];
	actual = iter.next().value;
	t.deepEqual( actual, expected, 'returns expected value' );

	t.strictEqual( iter.next().done, true, 'returns expected value' );

	t.end();
});

tape('the finish function returns an object with done property set to true', function test(t) {
	var iterator = iterCartesianSquare([1, 2]);
	var result;
	iterator.next(); // consume a value to ensure the iterator is not at the start
	result = iterator.return(); // call the finish function
	t.deepEqual(result, {
		'done': true
	}, 'returns expected value');
	t.end();
});

tape('the factory function returns a new iterator', function test(t) {
	var iterator1 = iterCartesianSquare([1, 2]);
	var iterator2 = iterator1[Symbol.iterator](); // call the factory function
	t.notEqual(iterator1, iterator2, 'returns new iterator');
	t.end();
});

tape('the factory function returns an iterator that generates the same sequence', function test(t) {
	var iterator1 = iterCartesianSquare([1, 2]);
	var iterator2 = iterator1[Symbol.iterator](); // call the factory function
	var result1 = [];
	var result2 = [];
	var next1;
	var next2;

	next1 = iterator1.next();
	while (!next1.done) {
		result1.push(next1.value);
		next1 = iterator1.next();
	}

	next2 = iterator2.next();
	while (!next2.done) {
		result2.push(next2.value);
		next2 = iterator2.next();
	}

	t.deepEqual(result1, result2, 'returns iterator that generates the same sequence');
	t.end();
});
