// @stdlib/assert/has-prototype/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var inherit = require( '@stdlib/utils/inherit' );
var hasPrototype = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof hasPrototype, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if a prototype argument which is neither an object (except null) or a function', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		undefined
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			hasPrototype( {}, value );
		};
	}
});

tape( 'the function returns `false` if provided a test value which is neither an object (except null) or a function', function test( t ) {
	var values;
	var bool;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		undefined
	];
	for ( i = 0; i < values.length; i++ ) {
		bool = hasPrototype( values[i], {} );
		t.strictEqual( bool, false, 'returns false when provided '+values[i] );
	}
	t.end();
});

tape( 'the function returns `true` if a provided prototype is part of a value\'s prototype chain', function test( t ) {
	var bool;
	var bar;

	function Foo() {
		return this;
	}

	function Bar() {
		return this;
	}
	inherit( Bar, Foo );

	bar = new Bar();
	bool = hasPrototype( bar, Foo.prototype );

	t.strictEqual( bool, true, 'returns true' );
	t.strictEqual( bar instanceof Bar, true, 'is instance of Bar' );
	t.strictEqual( bar instanceof Foo, true, 'is instance of Foo' );
	t.end();
});

tape( 'the function returns `false` if a provided prototype is not part of a value\'s prototype chain', function test( t ) {
	var bool;
	var bar;

	function Foo() {
		return this;
	}

	function Bar() {
		return this;
	}

	bar = new Bar();
	bool = hasPrototype( bar, Foo.prototype );

	t.strictEqual( bool, false, 'returns false' );
	t.strictEqual( bar instanceof Bar, true, 'is instance of Bar' );
	t.strictEqual( bar instanceof Foo, false, 'is not instance of Foo' );
	t.end();
});

tape( 'the function supports situations where constructors are never used', function test( t ) {
	var superProto;
	var subProto;
	var bool;
	var v;

	function createObject( proto ) {
		function Ctor() {}
		Ctor.prototype = proto;
		return new Ctor();
	}
	superProto = {
		'beep': 'beep'
	};

	subProto = createObject( superProto );
	subProto.boop = 'boop';

	v = createObject( subProto );

	t.throws( isInstance, TypeError, 'throws type error if using `instanceof`' );

	bool = hasPrototype( v, superProto );
	t.strictEqual( bool, true, 'returns true' );

	t.end();

	function isInstance() {
		var bool = ( v instanceof superProto );
		return bool;
	}
});

tape( 'the function returns `false` if provided primitives and their corresponding object constructors', function test( t ) {
	var bool;

	bool = hasPrototype( true, Boolean.prototype );
	t.strictEqual( bool, false, 'returns false' );

	bool = hasPrototype( 'beep', String.prototype );
	t.strictEqual( bool, false, 'returns false' );

	bool = hasPrototype( 5, Number.prototype );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});
