/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import { typedndarray } from '@stdlib/types/ndarray';
import fill = require( './index' );

/**
* Mock function to create an ndarray-like object.
*
* @return ndarray-like object
*/
function array(): typedndarray<number> {
	const obj: typedndarray<number> = {
		'byteLength': 80,
		'BYTES_PER_ELEMENT': 8,
		'data': new Float64Array( 10 ),
		'dtype': 'float64',
		'flags': {
			'ROW_MAJOR_CONTIGUOUS': true,
			'COLUMN_MAJOR_CONTIGUOUS': false
		},
		'length': 10,
		'ndims': 1,
		'offset': 0,
		'order': 'row-major',
		'shape': [ 10 ],
		'strides': [ 1 ],
		'get': (): number => 0,
		'set': (): typedndarray<number> => obj
	};
	return obj;
}

// TESTS //

// The function returns `undefined`...
{
	const x = array();
	const value = 10.0;

	fill( x, value ); // $ExpectType void
}

// The compiler throws an error if the function is provided a first argument which is not an array-like object containing ndarray-like objects...
{
	fill( '5', 10.0 ); // $ExpectError
	fill( 5, 10.0 ); // $ExpectError
	fill( true, 10.0 ); // $ExpectError
	fill( false, 10.0 ); // $ExpectError
	fill( null, 10.0 ); // $ExpectError
	fill( undefined, 10.0 ); // $ExpectError
	fill( {}, 10.0 ); // $ExpectError
	fill( [ 1 ], 10.0 ); // $ExpectError
	fill( ( x: number ): number => x, 10.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = array();
	const value = 10.0;

	fill(); // $ExpectError
	fill( x, value, {} ); // $ExpectError
}
