<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Definition Spacing

> [ESLint rule][eslint-rules] to prevent consecutive whitespace in Markdown definition labels in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-definition-spacing' );
```

#### rule

[ESLint rule][eslint-rules] to prevent consecutive whitespace in Markdown definition labels in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-definition-spacing, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep [boop][beep  boop].
*
* [beep  boop]: http://example.com
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Beep [boop][beep boop].
*
* [beep boop]: https://example.com
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-definition-spacing' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep [boop][beep    boop].',
    '*',
    '* [beep    boop]: https://example.com',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-definition-spacing', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-definition-spacing': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-definition-spacing',
            'severity': 2,
            'message': 'Do not use consecutive white-space in definition labels',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* [beep    boop]: http://example.com',
            'endLine': 13,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
