/*
* @license Apache-2.0
*
* Copyright (c) 2023 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import truncate = require( './index' );


// TESTS //

// The function returns a string...
{
	truncate( 'abcdefghi', 10, '|' ); // $ExpectType string
}

// The compiler throws an error if the function is provided a value other than a string as its first argument...
{
	truncate( true, 1, '...' ); // $ExpectError
	truncate( false, 1, '...' ); // $ExpectError
	truncate( null, 1, '...' ); // $ExpectError
	truncate( undefined, 1, '...' ); // $ExpectError
	truncate( 5, 1, '...' ); // $ExpectError
	truncate( [], 1, '...' ); // $ExpectError
	truncate( {}, 1, '...' ); // $ExpectError
	truncate( ( x: number ): number => x, 1, '...' ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument that is not a number...
{
	truncate( 'abc', true, '...' ); // $ExpectError
	truncate( 'abc', false, '...' ); // $ExpectError
	truncate( 'abc', null, '...' ); // $ExpectError
	truncate( 'abc', 'abc', '...' ); // $ExpectError
	truncate( 'abc', [], '...' ); // $ExpectError
	truncate( 'abc', {}, '...' ); // $ExpectError
	truncate( 'abc', ( x: number ): number => x, '...' ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument that is not a string...
{
	truncate( 'beep boop', 4, true ); // $ExpectError
	truncate( 'beep boop', 4, false ); // $ExpectError
	truncate( 'beep boop', 4, null ); // $ExpectError
	truncate( 'beep boop', 4, 123 ); // $ExpectError
	truncate( 'beep boop', 4, [], 0 ); // $ExpectError
	truncate( 'beep boop', 4, {}, 0 ); // $ExpectError
	truncate( 'beep boop', 4, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	truncate(); // $ExpectError
	truncate( 'abc' ); // $ExpectError
	truncate( 'abc', 1, '.', true ); // $ExpectError
}
