<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-private-annotation

> [ESLint rule][eslint-rules] enforcing that JSDoc comments of unassigned function declarations have a `@private` tag.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-private-annotation' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that JSDoc comments of unassigned function declarations have a `@private` tag.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-private-annotation -->

```javascript
/**
* Creates a function which always returns the same value.
*
* @param {*} [value] - value to always return
* @returns {Function} constant function
*
* @example
* var fcn = wrap( 3.14 );
*
* var v = fcn();
* // returns 3.14
*/
function wrap( value ) {
    return constantFunction;

    /**
    * Constant function.
    *
    * @returns {*} constant value
    */
    function constantFunction() {
        return value;
    }
}


// EXPORTS //

module.exports = wrap;
```

**Good**:

```javascript
/**
* Creates a function which always returns the same value.
*
* @param {*} [value] - value to always return
* @returns {Function} constant function
*
* @example
* var fcn = wrap( 3.14 );
*
* var v = fcn();
* // returns 3.14
*/
function wrap( value ) {
    return constantFunction;

    /**
    * Constant function.
    *
    * @private
    * @returns {*} constant value
    */
    function constantFunction() {
        return value;
    }
}


// EXPORTS //

module.exports = wrap;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-private-annotation' );

var linter = new Linter();
var result;

var code = [
    '/**',
    '* Comparator function for sorting characters in ascending order.',
    '*',
    '* @param {string} a - character',
    '* @param {string} b - character',
    '* @returns {number} comparison value',
    '*/',
    'function ascending( a, b ) {',
    '  if ( a < b ) {',
    '    return -1;',
    '  }',
    '  if ( a === b ) {',
    '    return 0;',
    '  }',
    '  return 1;',
    '}',
    '',
    'var arr = [ 3, 1, 6, 12, 9, 2 ];',
    '',
    'module.exports = arr.sort( ascending );'
].join( '\n' );

linter.defineRule( 'jsdoc-private-annotation', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-private-annotation': [ 'error' ]
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-private-annotation',
            'severity': 2,
            'message': '`ascending` is missing a @private annotation',
            'line': 8,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'source': 'function ascending( a, b ) {',
            'endLine': 16,
            'endColumn': 2
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
