// @stdlib/_tools/scaffold/validate-js/test/test.js
'use strict';

// MODULES //

var resolve = require( 'path' ).resolve;
var tape = require( 'tape' );
var readFileSync = require( '@stdlib/fs/read-file' ).sync;
var create = require( './../lib' );


// FIXTURES //

var MAIN1 = readFileSync( resolve( __dirname, 'fixtures', 'main.js_1.txt' ) );
var VALIDATE1 = readFileSync( resolve( __dirname, 'fixtures', 'validate.js_1.txt' ) );
var MAIN2 = readFileSync( resolve( __dirname, 'fixtures', 'main.js_2.txt' ) );
var VALIDATE2 = readFileSync( resolve( __dirname, 'fixtures', 'validate.js_2.txt' ) );
var MAIN3 = readFileSync( resolve( __dirname, 'fixtures', 'main.js_3.txt' ) );
var VALIDATE3 = readFileSync( resolve( __dirname, 'fixtures', 'validate.js_3.txt' ) );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof create, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a primitive string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		undefined,
		true,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			create( value );
		};
	}
});

tape( 'the function throws an error if the source code does not contain a JSDoc comment for a validate function', function test( t ) {
	var values;
	var i;

	values = [
		'// Only a single-line comment',
		[
			'/**',
			'* Not a JSDoc comment for a validate function.',
			'*/'
		].join( '\n' ),
		'function validate() {}'
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), Error, 'throws a error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			create( value );
		};
	}
});

tape( 'the function returns `validate.js` contents', function test( t ) {
	var expected;
	var actual;

	actual = create( MAIN1 );
	expected = VALIDATE1.toString();
	t.strictEqual( actual, expected, 'returns expected contents' );
	t.end();
});

tape( 'the function returns `validate.js` contents (union type)', function test( t ) {
	var expected;
	var actual;

	actual = create( MAIN2 );
	expected = VALIDATE2.toString();
	t.strictEqual( actual, expected, 'returns expected contents' );
	t.end();
});

tape( 'the function returns `validate.js` contents (wildcard)', function test( t ) {
	var expected;
	var actual;

	actual = create( MAIN3 );
	expected = VALIDATE3.toString();
	t.strictEqual( actual, expected, 'returns expected contents' );
	t.end();
});
