// @stdlib/_tools/github/following/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var get = require( './../lib/main.js' );


// FIXTURES //

var getOpts = require( './fixtures/opts.js' );
var data = require( './fixtures/results.json' );
var info = require( './fixtures/info.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof get, 'function', 'export is a function' );
	t.end();
});

tape( 'function returns an error to a provided callback if an error is encountered when fetching who a user is following', function test( t ) {
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	opts = getOpts();
	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk({
					'status': 404,
					'message': 'beep'
				});
			}
		};
	}

	function done( error ) {
		t.equal( error.status, 404, 'equal status' );
		t.equal( error.message, 'beep', 'equal message' );
		t.end();
	}
});

tape( 'functions returns response data to a provided callback', function test( t ) {
	var expected;
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = data;

	opts = getOpts();
	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, data, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function returns rate limit info to a provided callback', function test( t ) {
	var expected;
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = info;

	opts = getOpts();
	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, data, info );
			}
		};
	}

	function done( error, data, info ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( info, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function supports providing only a `username` option (no authentication)', function test( t ) {
	var expected;
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = data;

	opts = getOpts();
	delete opts.token;

	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, data, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function supports providing only a `token` option (get who the authenticated user is following)', function test( t ) {
	var expected;
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = data;

	opts = getOpts();
	delete opts.username;

	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, data, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function supports providing both `token` and `username` options to get who another user is following (increased rate limits)', function test( t ) {
	var expected;
	var opts;
	var get;

	get = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = data;

	opts = getOpts();
	opts.token = 'boopbeep';
	opts.username = 'kgryte';

	get( opts, done );

	function factory( opts, clbk ) {
		return function following() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, data, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});
