<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# require-order

> [ESLint rule][eslint-rules] enforcing a specified order of `require()` calls.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-order' );
```

#### rule

[ESLint rule][eslint-rules] enforcing a specified order of `require()` calls. By default, require statements should appear in the following order:

-   **builtin**: built-in Node.js modules.
-   **external**: external packages.
-   **path**: relative or absolute paths (strings beginning with `./`, `/`, or `../`).

**Bad**:

<!-- run-disable -->

<!-- eslint-disable stdlib/require-order -->

```javascript
// 'order': [ 'builtin', 'external', '/^@stdlib/', 'path' ]
var validate = require( './validate.js' );
var math = require( '@stdlib/math' );
var debug = require( 'debug' );
var fs = require( 'fs' );
```

**Good**:

<!-- run-disable -->

```javascript
// 'order': [ 'builtin', 'external', '/^@stdlib/', 'path' ]
var fs = require( 'fs' );
var debug = require( 'debug' );
var math = require( '@stdlib/math' );
var validate = require( './validate.js' );
```

To enforce a custom ordering, set the `order` option. In addition to `builtin`, `external`, and `path`, the `order` option supports regular expression strings such as `'/^@stdlib/'`.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-order' );

var linter = new Linter();
var result;

var code = [
    'var validate = require( \'./validate.js\' )',
    'var fs = require( \'fs\' )'
].join( '\n' );

linter.defineRule( 'require-order', rule );

result = linter.verify( code, {
    'rules': {
        'require-order': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'require-order',
            'severity': 2,
            'message': '`require( \'fs\' )` should come before `require( \'./validate.js\' )`',
            'line': 2,
            'column': 10,
            'nodeType': null,
            'source': 'var fs = require( \'fs\' )',
            'endLine': 2,
            'endColumn': 25
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
