// @stdlib/utils/function-name/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var functionName = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof functionName, 'function', 'export is a function' );
	t.end();
});

tape( 'if not provided a function, the function will throw a type error', function test( t ) {
	var values;
	var i;

	values = [
		'beep',
		5,
		NaN,
		null,
		true,
		undefined,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a TypeError when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			functionName( value );
		};
	}
});

tape( 'the function returns a function\'s name', function test( t ) {
	function beep() {
		return 'boop';
	}
	t.equal( functionName( beep ), 'beep', 'returns beep' );
	t.equal( functionName( Date ), 'Date', 'returns Date' );
	t.equal( functionName( Buffer ), 'Buffer', 'returns Buffer' );
	t.equal( functionName( Number ), 'Number', 'returns Number' );
	t.equal( functionName( Math.sqrt ), 'sqrt', 'returns sqrt' );
	t.equal( functionName( Int8Array ), 'Int8Array', 'returns Int8Array' );
	t.equal( functionName( Boolean ), 'Boolean', 'returns Boolean' );
	t.equal( functionName( String ), 'String', 'returns String' );
	t.equal( functionName( Function ), 'Function', 'returns Function' );
	t.end();
});

tape( 'if provided an anonymous function, the function returns an empty string or "anonymous" (ES2015 and some browsers)', function test( t ) {
	/* jshint evil:true */
	var name;

	/*
	* See
	*   - http://www.2ality.com/2015/09/function-names-es6.html
	*   - https://bugs.webkit.org/show_bug.cgi?id=7726
	*   - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name (which claims that functions created using the `Function` constructor should have an empty string as a name)
	*     - "Functions created with the syntax new Function(...) or just Function(...) have their name property set to an empty string."
	*
	* In Node.js <v6.0, the returned name is an empty string. For Node.js v6.0, the returned name when using the `Function` constructor is `"anonymous"`.
	*/

	name = functionName( function(){} );
	t.equal( check( name ), true, 'returns an empty string (or "anonymous")' );

	name = functionName( new Function('a', 'b', 'return') );
	t.equal( check( name ), true, 'returns an empty string (or "anonymous")' );

	t.end();

	function check() {
		if ( name === '' || name === 'anonymous' ) {
			return true;
		}
		return false;
	}
});

tape( 'the function supports returning a function name in ES5 and earlier environments', function test( t ) {
	var functionName = proxyquire( './../lib', {
		'@stdlib/utils/detect-function-name-support': stub
	});
	t.equal( functionName( beep ), 'beep', 'returns beep' );
	t.end();

	function beep() {
		return 'boop';
	}

	// Earlier environments did not have a `name` property:
	function stub() {
		return false;
	}
});
