// @stdlib/math/generics/statistics/chi2gof/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var poisson = require( '@stdlib/math/base/random/poisson' ).factory;
var abs = require( '@stdlib/math/base/special/abs' );
var EPS = require( '@stdlib/math/constants/float64-eps' );
var chi2gof = require( './../lib' );


// FUNCTIONS //

function almostEqual( a, b, tol ) {
	var delta = abs( a - b );
	tol *= abs( b );
	return ( delta <= tol );
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof chi2gof, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the `x` argument is not an array of nonnegative integers', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		[ 1, 2, 3, 'not a numeric array, hehe' ],
		[ 1.2, 0.8, 3, 4 ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( value, [ 0.25, 0.25, 0.25, 0.25 ] );
		};
	}
});

tape( 'the function throws an error if the `p` argument is not an array of nonnegative numbers or array of probabilities summing to one', function test( t ) {
	var values;
	var i;

	values = [
		5,
		[ 0.25, 0.25, 0.25, 'not a numeric array, hehe' ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( [ 20, 48, 10, 20 ], value, {
				'rescale': true
			});
		};
	}
});

tape( 'the function throws an error if `x` and `y` are arrays of different lengths', function test( t ) {
	var values;
	var i;

	values = [
		[ [ 20, 10 ], [ 7.5, 7.5, 7.5, 7.5 ] ],
		[ [ 20, 10, 30, 90, 80 ], [ 0.25, 0.25, 0.25, 0.25 ] ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), Error, 'throws an error when provided ['+values[i][0]+'] and ['+values[i][1]+']' );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( value[ 0 ], value[ 1 ] );
		};
	}
});

tape( 'the function throws an error if `x` does not contain at least one positive element', function test( t ) {
	var value = [ 0, 0, 0, 0 ];

	t.throws( badValue(), Error, 'throws an error when provided ['+value+']' );
	t.end();

	function badValue() {
		return function badValue() {
			chi2gof( value, [ 0.25, 0.25, 0.25, 0.25 ] );
		};
	}
});

tape( 'the function throws an error if the supplied distribution parameter(s) are not numeric values', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( [ 100, 60, 40, 30, 20, 10, 5, 2 ], 'poisson', value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		5,
		'abc',
		undefined,
		null,
		NaN,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( [ 30, 40, 50, 80 ], [ 0.2, 0.2, 0.3, 0.3 ], {
				'simulate': value
			});
		};
	}
});

tape( 'the function throws an error if provided a significance level `alpha` outside `[0,1]`', function test( t ) {
	var values;
	var i;

	values = [
		1.2,
		-0.2
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			chi2gof( [ 30, 40, 50, 80 ], [ 0.2, 0.2, 0.3, 0.3 ], {
				'alpha': value
			});
		};
	}
});

tape( 'the function correctly computes the chi-square goodness of fit test (when `y` is array of probabilities)', function test( t ) {
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.04059404334478121, 3.0*EPS ), 'returns correct value' );
	t.ok( almostEqual( out.statistic, 9.990143369175627, EPS ), 'returns correct value' );

	t.end();
});

tape( 'the function correctly computes the chi-square goodness of fit test (when `y` is array of expected counts)', function test( t ) {
	var out;
	var x;
	var y;

	x = [ 89, 37, 30, 28, 2 ];
	y = [ 74.4, 37.2, 37.2, 27.9, 9.3 ];

	out = chi2gof( x, y );

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.04059404334478121, 3.0*EPS ), 'returns correct value' );
	t.ok( almostEqual( out.statistic, 9.990143369175627, EPS ), 'returns correct value' );

	t.end();
});

tape( 'the function computes the chi-square goodness-of-fit test (when `y` is a distribution name)', function test( t ) {
	var lambda;
	var pValue;
	var freq;
	var pmf;
	var j;
	var x;

	lambda = 30;
	pmf = poisson( lambda, {
		'seed': 23
	});

	freq = new Array( 300 );
	for ( j = 0; j < freq.length; j++ ) {
		freq[ j ] = 0;
	}
	for ( j = 0; j < 1000; j++ ) {
		x = pmf();
		freq[ x ] += 1;
	}
	// Test using chi-square goodness-of-fit test:
	pValue = chi2gof( freq, 'poisson', lambda, {
		'simulate': true,
		'iterations': 500
	}).pValue;
	t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
	t.end();
});

tape( 'the function allows to correct the used degrees of freedom via the `ddof` option', function test( t ) {
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p, {
		'ddof': 2
	});

	t.equal( out.df, 2, 'df is equal to two' );

	t.end();
});

tape( 'the function computes the p-value via Monte Carlo simulation when `simulate` option is set to `true`', function test( t ) {
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p, {
		'simulate': true,
		'iterations': 500
	});

	t.equal( typeof out.pValue, 'number', 'returns a number' );
	t.end();
});

tape( 'the function returns the correct decision giving different values of `alpha`', function test( t ) {
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p, {
		'alpha': 0.05
	});

	t.equal( out.rejected, true, 'rejects null hypothesis at alpha = 0.05' );

	out = chi2gof( x, p, {
		'alpha': 0.01
	});

	t.equal( out.rejected, false, 'does not reject null hypothesis at alpha = 0.01' );

	t.end();
});

tape( 'the function returns an object with a .print() method for printing a formatted output table', function test( t ) {
	var table;
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );
	table = out.print();
	/*
		Chi-square goodness-of-fit test

		Null hypothesis: population probabilities are equal to those in p

			pValue: 0.0406
			statistic: 9.9901
			degrees of freedom: 4

		Test Decision: Reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a string' );

	out = chi2gof( x, p, {
		'alpha': 0.01
	});
	table = out.print();
	/*
		Chi-square goodness-of-fit test

		Null hypothesis: population probabilities are equal to those in p

			pValue: 0.0406
			statistic: 9.9901
			degrees of freedom: 4

		Test Decision: Fail to reject null in favor of alternative at 1% significance level
	*/
	t.equal( typeof table, 'string', 'returns a string' );

	t.end();
});

tape( 'the function returns an object with a .print() method that accepts a `numdigits` option to control the number of decimal degits displayed', function test( t ) {
	var table;
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );
	table = out.print({
		'numdigits': 6
	});
	/*
		Chi-square goodness-of-fit test

		Null hypothesis: population probabilities are equal to those in p

			pValue: 0.040594
			statistic: 9.990143
			degrees of freedom: 4

		Test Decision: Reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that accepts an `options` object', function test( t ) {
	var table;
	var out;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );
	table = out.print( {} );

	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});


tape( 'the function returns an object with a .print() method that throws an error if `options` is not a simple object', function test( t ) {
	var values;
	var out;
	var x;
	var i;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );

	values = [
		'abc',
		4,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print( value );
		};
	}
});

tape( 'the function returns an object with a .print() method that throws an error if the `numdigits` option is not a positive integer', function test( t ) {
	var values;
	var out;
	var i;
	var x;
	var p;

	x = [ 89, 37, 30, 28, 2 ];
	p = [ 0.40, 0.20, 0.20, 0.15, 0.05 ];

	out = chi2gof( x, p );

	values = [
		'abc',
		2.4,
		0.0,
		-1.0,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print({
				'numdigits': value
			});
		};
	}
});
