// @stdlib/_tools/pkgs/includes/test/test.sync.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var resolve = require( 'path' ).resolve;
var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
var isArray = require( '@stdlib/assert/is-array' );
var findIncludes = require( './../lib/sync.js' );


// VARIABLES //

var dir = resolve( __dirname, '..', '..' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof findIncludes, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	t.throws( foo, TypeError, 'throws error' );
	t.end();
	function foo() {
		var opts = {
			'dir': null
		};
		findIncludes( opts );
	}
});

tape( 'the function returns a string array (if at least one package containing an `include` directory is detected)', function test( t ) {
	var findIncludes;
	var count;
	var dirs;
	var opts;

	findIncludes = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		},
		'@stdlib/fs/read-json': {
			'sync': readJSON
		}
	});

	count = 0;
	opts = {
		'dir': dir
	};
	dirs = findIncludes( opts );

	t.strictEqual( isStringArray( dirs ), true, 'returns a string array' );
	t.strictEqual( dirs.length, 1, 'expected length' );

	t.end();

	function glob() {
		return [
			'/beep/boop/package.json',
			'/a/b/c/d/e/f/g/h/beep/boop/bop/package.json'
		];
	}

	function readJSON() {
		count += 1;
		if ( count < 2 ) {
			return {
				'directories': {
					'include': './include'
				}
			};
		}
		return {};
	}
});

tape( 'the function returns an empty array if unable to resolve any packages containing an `include` directory', function test( t ) {
	var findIncludes;
	var dirs;

	findIncludes = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		},
		'@stdlib/fs/read-json': {
			'sync': readJSON
		}
	});

	dirs = findIncludes();

	t.strictEqual( isArray( dirs ), true, 'returns an array' );
	t.strictEqual( dirs.length, 0, 'returns an empty array' );

	t.end();

	function glob() {
		return [ '/beep/boop/package.json' ];
	}

	function readJSON() {
		return {};
	}
});

tape( 'the function returns an empty array if unable to resolve any `include` directories', function test( t ) {
	var findIncludes;
	var dirs;

	findIncludes = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		}
	});

	dirs = findIncludes();

	t.strictEqual( isArray( dirs ), true, 'returns an array' );
	t.strictEqual( dirs.length, 0, 'returns an empty array' );

	t.end();

	function glob() {
		return [];
	}
});
