// @stdlib/tools/disposable-http-server/lib/server.js
'use strict';

// MODULES //

var debug = require( 'debug' )( 'disposable-http-server' );
var path = require( 'path' );
var createServer = require( '@stdlib/tools/http-server' );
var readFileSync = require( '@stdlib/utils/fs/read-file' ).sync;
var isString = require( '@stdlib/utils/is-string' ).isPrimitive;
var isFunction = require( '@stdlib/utils/is-function' );
var openURL = require( '@stdlib/utils/open-url' );
var noop = require( '@stdlib/utils/noop' );
var validate = require( './validate.js' );
var serverOpts = require( './opts.js' );


// DISPOSABLE HTTP SERVER //

/**
* Creates a disposable HTTP server.
*
* @param {Options} options - server options
* @param {(Buffer|string)} [options.html] - HTML content to serve
* @param {(Buffer|string)} [options.javascript] - JavaScript to serve
* @param {NonNegativeInteger} [options.port=0] - server port
* @param {NonNegativeInteger} [options.maxport] - max server port
* @param {string} [options.hostname] - server hostname
* @param {string} [options.address="0.0.0.0"] - server address
* @param {boolean} [options.open=false] - boolean indicating whether to launch a web browser
* @param {Callback} [clbk] - callback to invoke upon creating a server
* @throws {TypeError} must provide valid options
* @throws {Error} error encountered when starting server
*
* @example
* var opts = {
*     'html': '<h1>Beep</h1>'
* };
*
* httpServer( opts, onReady );
*
* function onReady( error, server ) {
*     if ( error ) {
*         throw error;
*     }
*     server.close();
* }
*/
function httpServer( options ) {
	var server;
	var sopts;
	var fpath;
	var clbk;
	var boot;
	var opts;
	var err;

	opts = {};
	if ( arguments.length > 1 ) {
		clbk = arguments[ 1 ];
		if ( !isFunction( clbk ) ) {
			throw new TypeError( 'invalid input argument. Callback argument must be a function. Value: `' + clbk + '`.' );
		}
	} else {
		clbk = noop;
	}
	err = validate( opts, options );
	if ( err ) {
		throw err;
	}
	// If provided HTML and/or JavaScript as `strings`, convert to `buffers`...
	if (
		opts.html &&
		isString( opts.html )
	) {
		opts.html = new Buffer( opts.html );
	}
	if (
		opts.javascript &&
		isString( opts.javascript )
	) {
		opts.javascript = new Buffer( opts.javascript );
	}
	// Extract server options:
	sopts = serverOpts( options );

	debug( 'Serving provided content.' );
	if ( !opts.html ) {
		debug( 'No HTML content provided.' );
		debug( 'Loading a boilerplate HTML page...' );
		fpath = path.resolve( __dirname, '../static/index.html' );
		opts.html = readFileSync( fpath );
	}
	// Create a function to boot a server...
	boot = createServer( sopts, requestListener );

	debug( 'Starting server...' );
	boot( onServer );

	/**
	* Callback invoked upon creating a server.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {Server} server - server instance
	* @throws {Error} error encountered when starting server
	*/
	function onServer( error, _server ) {
		var addr;
		if ( error ) {
			throw error;
		}
		debug( 'Server started.' );
		server = _server;
		server.once( 'close', onClose );
		if ( opts.open ) {
			addr = server.address();
			openURL( 'http://'+addr.address+':'+addr.port );
		}
		clbk( null, server );
	} // end FUNCTION onServer()

	/**
	* Callback invoked upon receiving an HTTP request for provided content.
	*
	* @private
	* @param {IncomingMessage} request - HTTP request object
	* @param {ServerResponse} response - HTTP response object
	*/
	function requestListener( request, response ) {
		debug( 'Received a request for %s', request.url );

		if ( !opts.javascript ) {
			response.once( 'finish', onFinish );
		} else if ( request.url === '/bundle.js' ) {
			response.once( 'finish', onFinish );
			return sendJavaScript( request, response );
		}
		sendHTML( request, response );
	} // end FUNCTION requestListener()

	/**
	* Sends HTML content in response to a client request.
	*
	* @private
	* @param {IncomingMessage} request - HTTP request object
	* @param {ServerResponse} response - HTTP response object
	*/
	function sendHTML( request, response ) {
		debug( 'Sending HTML...' );
		response.statusCode = 200;
		response.setHeader( 'Content-Type', 'text/html' );
		response.setHeader( 'Content-Length', opts.html.length );
		response.end( opts.html );
	} // end FUNCTION sendHTML()

	/**
	* Sends JavaScript content in response to a client request.
	*
	* @private
	* @param {IncomingMessage} request - HTTP request object
	* @param {ServerResponse} response - HTTP response object
	*/
	function sendJavaScript( request, response ) {
		debug( 'Sending JavaScript...' );
		response.statusCode = 200;
		response.setHeader( 'Content-Type', 'text/javascript' );
		response.setHeader( 'Content-Length', opts.javascript.length );
		response.end( opts.javascript );
	} // end FUNCTION sendJavaScript()

	/**
	* Callback invoked once the server should close and the process exit.
	*
	* @private
	*/
	function onFinish() {
		debug( 'Finishing serving content.' );
		debug( 'Closing the server...' );
		server.close();
	} // end FUNCTION onFinish()

	/**
	* Callback invoked once a server closes.
	*
	* @private
	*/
	function onClose() {
		debug( 'Server closed.' );
	} // end FUNCTION onClose()
} // end FUNCTION httpServer()


// EXPORTS //

module.exports = httpServer;
