<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# new-cap-regexp

> [ESLint rule][eslint-rules] enforcing that the `RegExp` constructor is invoked with the `new` keyword.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/new-cap-regexp' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that the `RegExp` constructor is invoked with the `new` keyword.

**Bad**:

<!-- eslint-disable stdlib/new-cap-regexp -->

```javascript
var re = RegExp( '[0-9]' );
```

**Good**:

```javascript
var re = new RegExp( '[0-9]' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/new-cap-regexp' );

var linter = new Linter();
var result;

var code = 'var re = RegExp( \'ab+c\', \'i\' );';

linter.defineRule( 'new-cap-regexp', rule );

result = linter.verify( code, {
    'rules': {
        'new-cap-regexp': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'new-cap-regexp',
            'severity': 2,
            'message': '`RegExp` constructor is invoked without the `new` keyword',
            'line': 1,
            'column': 10,
            'nodeType': 'CallExpression',
            'source': 'var re = RegExp( \'ab+c\', \'i\' );',
            'endLine': 1,
            'endColumn': 31
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
