// @stdlib/string/base/replace-after/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2023 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var replaceAfter = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof replaceAfter, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a string as its first argument', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			replaceAfter( value, 'foo', 'e' );
		};
	}
});

tape( 'the function throws an error if not provided a string as its second argument', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			replaceAfter( 'beep', value, 'e' );
		};
	}
});

tape( 'the function throws an error if not provided a string as its third argument', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			replaceAfter( 'beep', 'foo', value );
		};
	}
});

tape( 'the function throws an error if provided a non-integer value as its fourth argument', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			replaceAfter( 'beep', 'foo', 'e', value );
		};
	}
});

tape( 'the function replaces the substring after the first occurrence of a specified search string', function test( t ) {
	var replacement;
	var expected;
	var actual;
	var str;

	str = 'beep boop';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, ' ' );
	expected = 'beep foo';
	t.strictEqual( actual, expected, 'returns expected value' );

	str = 'beep boop';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'p' );
	expected = 'beepfoo';
	t.strictEqual( actual, expected, 'returns expected value' );

	str = 'Hello, World!';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'o' );
	expected = 'Hellofoo';
	t.strictEqual( actual, expected, 'returns expected value' );

	t.end();
});

tape( 'the function replaces the substring after the first occurrence of a specified search string (Unicode characters)', function test( t ) {
	var replacement;
	var expected;
	var actual;
	var str;

	str = 'beep ðŸ˜€ boop ðŸ˜€ baz';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'ðŸ˜€' );
	expected = 'beep ðŸ˜€foo';
	t.strictEqual( actual, expected, 'returns expected value' );

	str = 'ðŸ¤– Robot army ðŸ¤–!';
	actual = replaceAfter( str, replacement, 'ðŸ¤–' );
	expected = 'ðŸ¤–foo';

	str = 'ðŸº Wolf brothers ðŸº';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'o' );
	expected = 'ðŸº Wofoo';

	t.end();
});

tape( 'the function replaces the substring after the first occurrence of a specified search string (custom start index)', function test( t ) {
	var replacement;
	var expected;
	var actual;
	var str;

	str = 'beep boop baz';
	replacement = 'foo';
	actual = replaceAfter( str, ' ', 6 );
	expected = 'beep boop foo';
	t.strictEqual( actual, expected, 'returns expected value' );

	str = 'beep boop baz';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'p', 6 );
	expected = 'beep boopfoo';

	str = 'beep boop baz';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'beep', -2 );
	expected = 'beepfoo';

	str = 'beep boop baz';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'beep', 20 );
	expected = '';

	t.end();
});

tape( 'the function returns the empty string if the search string is not found', function test( t ) {
	var replacement;
	var expected;
	var actual;
	var str;

	str = 'beep boop';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'z' );
	expected = '';
	t.strictEqual( actual, expected, 'returns expected value' );

	str = 'beep boop';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, 'baz' );
	expected = '';
	t.strictEqual( actual, expected, 'returns expected value' );

	t.end();
});

tape( 'the function returns the replacement string if the search string is the empty string', function test( t ) {
	var replacement;
	var expected;
	var actual;
	var str;

	str = 'beep boop';
	replacement = 'foo';
	actual = replaceAfter( str, replacement, '' );
	expected = 'foo';
	t.strictEqual( actual, expected, 'returns expected value' );

	t.end();
});
