// @stdlib/_tools/esm/create-dist/lib/copy_sources.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var fs = require( 'fs' );
var path = require( 'path' );
var mkdirp = require( 'mkdirp' ).sync;
var logger = require( 'debug' );
var forEach = require( '@stdlib/utils/async/for-each' );
var objectInverse = require( '@stdlib/utils/object-inverse' );
var mapKeys = require( '@stdlib/utils/map-keys' );


// VARIABLES //

var debug = logger( 'esm-create-dist:copy-sources' );


// FUNCTIONS //

/**
 * Expand implicit index paths in object keys.
 *
 * @private
 * @param {String} key - alias map key
 * @returns {String} mapped key
 */
function expandAlias( key ) {
	if ( /lib$/.test( key ) ) {
		return path.resolve( key, 'index.js' );
	}

	return key;
}

/**
 * Construct a function for iterating over a package file.
 *
 * @private
 * @param {Object} deps - package deps, returned from @stdlib/_tools/pkgs/browser-deps
 * @param {String} outputDir - base output directory
 * @param {Object} aliases - expanded browser file aliases
 * @param {Array} files - mutable running list of outputted files
 * @returns {Function} file iteration function
 */
function onFileFactory( deps, outputDir, aliases, files ) {
	/**
	 * Iterate over a file to be copied.
	 *
	 * @private
	 * @param {String} from - path to input file
	 * @param {Number} i - iteration index
	 * @param {Callback} clbk - callback executed on completion
	 * @returns {void}
	 */
	function onFile( from, i, clbk ) {
		var relFrom;
		var relDst;
		var alias;
		var dst;
		var to;

		relFrom = path.relative( deps.dir, from );
		alias = aliases[ from ];

		if ( typeof alias === 'undefined' ) {
			dst = from;
		} else {
			debug( 'Found alias while copying file %s', relFrom );
			dst = alias;
		}

		relDst = path.relative( deps.dir, dst );

		to = path.resolve( outputDir, deps.pkg, relDst );

		if ( /lib$/.test( to ) ) {
			to = path.resolve( to, 'index.js' );
		}

		mkdirp( path.dirname( to ) );

		files.push( to );

		debug( 'Copy from <src>/%s to <buildtmp>/%s', relFrom, relDst );

		fs.copyFile( from, to, clbk );
	}

	return onFile;
}


// MAIN //

/**
 * Copy browser package sources, applying aliases.
 *
 * @param {Array} pkgsDeps - package deps, returned from @stdlib/_tools/pkgs/browser-deps
 * @param {String} outputDir - Base output directory name
 * @param {Callback} clbk - callback called upon completion
 * @returns {void}
 */
function copySources( pkgsDeps, outputDir, clbk ) {
	var files;
	var opts;

	files = [];

	opts = {
		'limit': 5
	};

	forEach( pkgsDeps, opts, onPkgDeps, onComplete );

	/**
	 * Function for iterating over packages.
	 *
	 * @private
	 * @param {Object} deps - package deps, from @stdlib/_tools/pkgs/browser-deps
	 * @param {Callback} next - function called on completion
	 * @returns {void}
	 */
	function onPkgDeps( deps, next ) {
		var invertedAliases;
		var aliases;
		var onFile;
		var opts;

		aliases = mapKeys( deps.aliases, expandAlias );
		invertedAliases = objectInverse( aliases );

		onFile = onFileFactory( deps, outputDir, invertedAliases, files );

		opts = {
			'limit': 5
		};

		forEach( deps.files, opts, onFile, next );
	}

	/**
	 * Function executed on completion of copying.
	 *
	 * @private
	 * @param {Error|null} err - error, if encountered during copying
	 * @returns {void}
	 */
	function onComplete( err ) {
		if ( err instanceof Error ) {
			clbk( err );
			return;
		}

		clbk( null, files );
	}
}


// EXPORTS //

module.exports = copySources;
