// @stdlib/blas/base/dtrsm/test/test.dtrsm.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable max-len */

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var EPS = require( '@stdlib/constants/float64/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var dtrsm = require( './../lib/dtrsm.js' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof dtrsm, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 12', function test( t ) {
	t.strictEqual( dtrsm.length, 12, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided an invalid first argument', function test( t ) {
	var values;
	var A;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dtrsm( value, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
		};
	}
});

tape( 'the function throws an error if provided an invalid second argument', function test( t ) {
	var values;
	var A;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dtrsm( 'column-major', value, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
		};
	}
});

tape( 'the function throws an error if provided an invalid third argument', function test( t ) {
	var values;
	var A;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dtrsm( 'column-major', 'left', value, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
		};
	}
});

tape( 'the function throws an error if provided an invalid fourth argument', function test( t ) {
	var values;
	var A;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dtrsm( 'column-major', 'left', 'upper', value, 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
		};
	}
});

tape( 'the function throws an error if provided an invalid fifth argument', function test( t ) {
	var values;
	var A;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dtrsm( 'column-major', 'left', 'upper', 'no-transpose', value, 2, 2, 6.0, A, 2, B, 2 );
		};
	}
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, -12.0, 0.0, 12.0 ] );

	out = dtrsm( 'column-major', 'left', 'lower', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ -1.5, 10.5, -36, 12 ] );

	out = dtrsm( 'column-major', 'left', 'lower', 'transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, 0.0, 6.0, 12.0 ] );

	out = dtrsm( 'column-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, -22.5, 42, -19.5 ] );

	out = dtrsm( 'column-major', 'left', 'upper', 'transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ 30, 6, 0, 12 ] );

	out = dtrsm( 'column-major', 'right', 'lower', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ 30, 0, -12, 12 ] );

	out = dtrsm( 'column-major', 'right', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, 0.0, -12.0, 12.0 ] );

	out = dtrsm( 'row-major', 'left', 'lower', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ -1.5, -36, 10.5, 12 ] );

	out = dtrsm( 'row-major', 'left', 'lower', 'transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, 6.0, 0.0, 12.0 ] );

	out = dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ 30.0, 42.0, -22.5, -19.5 ] );

	out = dtrsm( 'row-major', 'left', 'upper', 'transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 3.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 0.0, 7.0, 8.0 ] );

	expected = new Float64Array( [ 30, 0, 6, 12 ] );

	out = dtrsm( 'row-major', 'right', 'lower', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	expected = new Float64Array( [ 30, -12, 0, 12 ] );

	out = dtrsm( 'row-major', 'right', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 2.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 0.0, 5.0, 6.0, 0.0, 0.0, 9.0 ] );
	B = new Float64Array( [ 10.0, 11.0, 12.0, 0.0, 14.0, 15.0, 0.0, 0.0, 18.0 ] );

	expected = new Float64Array( [ 60, -10.8, -4.8, 0, 16.8, -1.20, 0, 0, 12 ] );

	out = dtrsm( 'column-major', 'left', 'lower', 'no-transpose', 'non-unit', 3, 3, 6.0, A, 3, B, 3 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 10.0 );

	t.end();
});

tape( 'the function solves matrix equation `op(A) * X = alpha * B` or `X * op(A) = alpha * B` where `alpha` is a scalar, `X` and `B` are m by n matrices, `A` is a unit, or non-unit,  upper or lower triangular matrix  and  `op( A )`  is one  of `op( A) = A` or `op( A ) = A**T`. The matrix `X` is overwritten on `B` (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var B;

	A = new Float64Array( [ 1.0, 0.0, 0.0, 2.0, 5.0, 0.0, 3.0, 6.0, 9.0 ] );
	B = new Float64Array( [ 10.0, 0.0, 0.0, 11.0, 14.0, 0.0, 12.0, 15.0, 18.0 ] );

	expected = new Float64Array( [ 60, 0.0, 0.0, -10.8, 16.8, 0.0, -4.8, -1.20, 12 ] );

	out = dtrsm( 'row-major', 'left', 'lower', 'no-transpose', 'non-unit', 3, 3, 6.0, A, 3, B, 3 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 10.0 );

	t.end();
});
