/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import copysignf = require( './index' );


// TESTS //

// The function returns a number...
{
	copysignf( -3.14, 10 ); // $ExpectType number
	copysignf( 3.14, -10 ); // $ExpectType number
}

// The compiler throws an error if the function is provided values other than two numbers...
{
	copysignf( true, 3 ); // $ExpectError
	copysignf( false, 2 ); // $ExpectError
	copysignf( '5', 1 ); // $ExpectError
	copysignf( [], 1 ); // $ExpectError
	copysignf( {}, 2 ); // $ExpectError
	copysignf( ( x: number ): number => x, 2 ); // $ExpectError

	copysignf( -9, true ); // $ExpectError
	copysignf( -9, false ); // $ExpectError
	copysignf( -5, '5' ); // $ExpectError
	copysignf( -8, [] ); // $ExpectError
	copysignf( -9, {} ); // $ExpectError
	copysignf( -8, ( x: number ): number => x ); // $ExpectError

	copysignf( [], true ); // $ExpectError
	copysignf( {}, false ); // $ExpectError
	copysignf( false, '5' ); // $ExpectError
	copysignf( {}, [] ); // $ExpectError
	copysignf( '5', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	copysignf(); // $ExpectError
	copysignf( -3 ); // $ExpectError
}
