# List Item Content Indent

> [ESLint rule][eslint-rules] to prevent mixed indentation in Markdown list items in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-list-item-content-indent' );
```

#### rule

[ESLint rule][eslint-rules] to prevent mixed indentation in Markdown list items in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-list-item-content-indent, stdlib/jsdoc-list-item-indent, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep.
*
* 1. Item
*
*     1. Item
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

<!-- eslint-disable stdlib/jsdoc-list-item-indent, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep.
*
* 1. Item
*
*    1. Item
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-list-item-content-indent' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* 1. Item',
    '*',
    '*     1. Item',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-list-item-content-indent', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-list-item-content-indent': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-list-item-content-indent',
            'severity': 2,
            'message': 'Donâ€™t use mixed indentation for children, remove 1 space',
            'line': 6,
            'column': 7,
            'nodeType': null,
            'source': '*     1. Item',
            'endLine': 15,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
