<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# section-header-empty-lines

> [ESLint rule][eslint-rules] to enforce that section header comments are padded by empty lines.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/section-header-empty-lines' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that section header comments are padded by empty lines. Specifically, the rule requires that section headers are preceded by two empty lines and have one successive empty line. Exceptions are the beginning of a file, where no empty lines are required, and section headers following comments or the `'use strict'` directive, in which case only one empty line is needed.

**Bad**:

<!-- eslint-disable stdlib/section-header-empty-lines -->

```javascript
// FUNCTIONS //

function add( x, y ) {
    return x + y;
}

// VARIABLES //
var TWO_PI = 2.0 * 3.141592653589793;


// EXPORTS //

module.exports = TWO_PI;
```

**Good**:

```javascript
// FUNCTIONS //

function add( x, y ) {
    return x + y;
}


// VARIABLES //

var TWO_PI = 2.0 * 3.141592653589793;


// EXPORTS //

module.exports = TWO_PI;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/section-header-empty-lines' );

var linter = new Linter();
var result;
var code;

var opts = {
    'rules': {
        'section-header-empty-lines': 'error'
    }
};
linter.defineRule( 'section-header-empty-lines', rule );

code = [
    '\'use strict\';',
    '',
    '// MODULES //',
    'var os = require( \'os\' );',
    '',
    '',
    '// MAIN //',
    '',
    '/**',
    '* Returns the number of CPUs.',
    '*',
    '* @private',
    '* @returns {PositiveInteger} number of CPUs',
    '*',
    '* @example',
    '* var n = numCPUs();',
    '* // returns <number>',
    '*/',
    'function numCPUs() {',
    '  return os.cpus().length;',
    '}',
    '',
    '// EXPORTS //',
    '',
    'module.exports = numCPUs();'
].join( '\n' );

result = linter.verify( code, opts );
console.log( result );
/* =>
    [
        {
            'ruleId': 'section-header-empty-lines',
            'severity': 2,
            'message': 'Header comment must be followed by an empty line',
            'line': 3,
            'column': 1,
            'nodeType': null,
            'source': '// MODULES //',
            'endLine': 3,
            'endColumn': 14
        },
        {
            'ruleId': 'section-header-empty-lines',
            'severity': 2,
            'message': 'Header comment must be preceded by one additional empty line',
            'line': 23,
            'column': 1,
            'nodeType': null,
            'source': '// EXPORTS //',
            'endLine': 23,
            'endColumn': 14
        },
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
