// @stdlib/blas/base/ssymv/test/test.ssymv.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float32Array = require( '@stdlib/array/float32' );
var EPS = require( '@stdlib/constants/float32/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var ssymv = require( './../lib/ssymv.js' );


// FIXTURES //

var stridexoyt = require( './fixtures/strides_xoyt.json' );
var stridexpyp = require( './fixtures/strides_xpyp.json' );
var stridexnyp = require( './fixtures/strides_xnyp.json' );
var stridexpyn = require( './fixtures/strides_xpyn.json' );
var stridexnyn = require( './fixtures/strides_xnyn.json' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof ssymv, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 11', function test( t ) {
	t.strictEqual( ssymv.length, 11, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations matrix-vector operations `y = alpha*A*x + beta*y` where alpha and beta are scalars, x and y are n element vectors and A is an n by n symmetric matrix (sx=1, sy=1)', function test( t ) {
	var expected;
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexpyp.order;
	uplo = stridexpyp.uplo;

	n = stridexpyp.N;
	lda = stridexpyp.lda;

	strideX = stridexpyp.strideX;
	strideY = stridexpyp.strideY;

	alpha = stridexpyp.alpha;
	beta = stridexpyp.beta;

	a = new Float32Array( stridexpyp.A );
	x = new Float32Array( stridexpyp.x );
	y = new Float32Array( stridexpyp.y );

	expected = new Float32Array( stridexpyp.y_out );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations matrix-vector operations `y = alpha*A*x + beta*y` where alpha and beta are scalars, x and y are n element vectors and A is an n by n symmetric matrix (sx=1, sy=2)', function test( t ) {
	var expected;
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexoyt.order;
	uplo = stridexoyt.uplo;

	n = stridexoyt.N;
	lda = stridexoyt.lda;

	strideX = stridexoyt.strideX;
	strideY = stridexoyt.strideY;

	alpha = stridexoyt.alpha;
	beta = stridexoyt.beta;

	a = new Float32Array( stridexoyt.A );
	x = new Float32Array( stridexoyt.x );
	y = new Float32Array( stridexoyt.y );

	expected = new Float32Array( stridexoyt.y_out );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations matrix-vector operations `y = alpha*A*x + beta*y` where alpha and beta are scalars, x and y are n element vectors and A is an n by n symmetric matrix (sx=1, sy=-1)', function test( t ) {
	var expected;
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexpyn.order;
	uplo = stridexpyn.uplo;

	n = stridexpyn.N;
	lda = stridexpyn.lda;

	strideX = stridexpyn.strideX;
	strideY = stridexpyn.strideY;

	alpha = stridexpyn.alpha;
	beta = stridexpyn.beta;

	a = new Float32Array( stridexpyn.A );
	x = new Float32Array( stridexpyn.x );
	y = new Float32Array( stridexpyn.y );

	expected = new Float32Array( stridexpyn.y_out );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations matrix-vector operations `y = alpha*A*x + beta*y` where alpha and beta are scalars, x and y are n element vectors and A is an n by n symmetric matrix (sx=-1, sy=1)', function test( t ) {
	var expected;
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexnyp.order;
	uplo = stridexnyp.uplo;

	n = stridexnyp.N;
	lda = stridexnyp.lda;

	strideX = stridexnyp.strideX;
	strideY = stridexnyp.strideY;

	alpha = stridexnyp.alpha;
	beta = stridexnyp.beta;

	a = new Float32Array( stridexnyp.A );
	x = new Float32Array( stridexnyp.x );
	y = new Float32Array( stridexnyp.y );

	expected = new Float32Array( stridexnyp.y_out );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function returns a reference to the second input array', function test( t ) {
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexpyp.order;
	uplo = stridexpyp.uplo;

	n = stridexpyp.N;
	lda = stridexpyp.lda;

	strideX = stridexpyp.strideX;
	strideY = stridexpyp.strideY;

	alpha = stridexpyp.alpha;
	beta = stridexpyp.beta;

	a = new Float32Array( stridexpyp.A );
	x = new Float32Array( stridexpyp.x );
	y = new Float32Array( stridexpyp.y );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function supports complex access patterns', function test( t ) {
	var expected;
	var strideX;
	var strideY;
	var alpha;
	var order;
	var beta;
	var uplo;
	var lda;
	var out;
	var a;
	var n;
	var x;
	var y;

	order = stridexnyn.order;
	uplo = stridexnyn.uplo;

	n = stridexnyn.N;
	lda = stridexnyn.lda;

	strideX = stridexnyn.strideX;
	strideY = stridexnyn.strideY;

	alpha = stridexnyn.alpha;
	beta = stridexnyn.beta;

	a = new Float32Array( stridexnyn.A );
	x = new Float32Array( stridexnyn.x );
	y = new Float32Array( stridexnyn.y );

	expected = new Float32Array( stridexnyn.y_out );

	out = ssymv( order, uplo, n, alpha, a, lda, x, strideX, beta, y, strideY );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});
