<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Checkbox Character Style

> [ESLint rule][eslint-rules] to enforce that Markdown checkboxes follow a specified style in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-checkbox-character-style' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that Markdown checkboxes follow a specified style in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-checkbox-character-style, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep.
*
* -   [x] Item 1
* -   [X] Item 2
* -   [ ] Item 3
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Beep.
*
* -   [x] Item 1
* -   [x] Item 2
* -   [ ] Item 3
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-checkbox-character-style].

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- lint disable no-tabs -->

<!-- eslint no-undef: "error" -->

<!-- eslint-disable no-tabs -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-checkbox-character-style' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* - [X] First Item',
    '* - [ ] Second Item',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-checkbox-character-style', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-checkbox-character-style': [
            'error',
            {
                'checked': 'x',
                'unchecked': '	' // <- TAB
            }
        ]
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-checkbox-character-style',
            'severity': 2,
            'message': 'Checked checkboxes should use `x` as a marker',
            'line': 4,
            'column': 6,
            'nodeType': null,
            'source': '* - [X] Second Item',
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-checkbox-character-style',
            'severity': 2,
            'message': 'Unchecked checkboxes should use `\t` as a marker',
            'line': 5,
            'column': 6,
            'nodeType': null,
            'source': '* - [ ] Third Item',
            'endLine': 14,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-checkbox-character-style]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-checkbox-character-style

</section>

<!-- /.links -->
