<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# require-file-extensions

> [ESLint rule][eslint-rules] enforcing that `require()` calls of files end with a whitelisted file extension.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-file-extensions' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that `require()` calls of files end with a whitelisted file extension.

**Bad**:

<!-- run-disable -->

<!-- eslint-disable stdlib/require-file-extensions -->

```javascript
// Missing file extension:
var debug = require( 'debug/src/browser' );

// Invalid file extension:
var readme = require( '@stdlib/array/int32/README.md' );
```

**Good**:

```javascript
var debug = require( 'debug/src/browser.js' );

var Int32Array = require( '@stdlib/array/int32' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-file-extensions' );

var linter = new Linter();
var result;
var code;

var opts = {
    'rules': {
        'require-file-extensions': 'error'
    }
};

linter.defineRule( 'require-file-extensions', rule );

code = 'var readme = require( \'debug/README.md\' )';
result = linter.verify( code, opts );
/* returns
    [
        {
            'ruleId': 'require-file-extensions',
            'severity': 2,
            'message': 'require statement of file does not end with one of the whitelisted file extensions (.js,.json,.node). Value: debug/README.md',
            'line': 1,
            'column': 14,
            'nodeType': 'CallExpression',
            'source': 'var readme = require( \'debug/README.md\' )',
            'endLine': 1,
            'endColumn': 42
        }
    ]
*/

code = 'var beep = require( \'@stdlib/beep\' )';
result = linter.verify( code, opts );
/* returns
    [
        {
            'ruleId': 'require-file-extensions',
            'severity': 2,
            'message': 'cannot resolve module: "@stdlib/beep"',
            'line': 1,
            'column': 12,
            'nodeType': 'CallExpression',
            'source': 'var beep = require( \'@stdlib/beep\' )',
            'endLine': 1,
            'endColumn': 37
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
