# require-file-extensions

> [ESLint rule][eslint-rules] enforcing that `require()` calls of files end with a whitelisted file extension.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-file-extensions' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that `require()` calls of files end with a whitelisted file extension.

**Bad**:

<!-- run-disable -->

<!-- eslint-disable stdlib/require-file-extensions -->

```javascript
// Missing file extension:
var debug = require( 'debug/src/browser' );

// Invalid file extension:
var readme = require( '@stdlib/array/int32/README.md' );
```

**Good**:

``` javascript 
var debug = require( 'debug/src/browser.js' );

var Int32Array = require( '@stdlib/array/int32' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-file-extensions' );

var linter = new Linter();
var result;
var code;

var opts = {
    'rules': {
        'require-file-extensions': 'error'
    }
};

linter.defineRule( 'require-file-extensions', rule );

code = 'var readme = require( \'debug/README.md\' )';
result = linter.verify( code, opts );
/* returns
    [
        {
            'ruleId': 'require-file-extensions',
            'severity': 2,
            'message': 'require statement of file does not end with one of the whitelisted file extensions (.js,.json,.node). Value: debug/README.md',
            'line': 1,
            'column': 14,
            'nodeType': 'CallExpression',
            'source': 'var readme = require( \'debug/README.md\' )',
            'endLine': 1,
            'endColumn': 42
        }
    ]
*/

code = 'var beep = require( \'@stdlib/beep\' )';
result = linter.verify( code, opts );
/* returns
    [
        {
            'ruleId': 'require-file-extensions',
            'severity': 2,
            'message': 'cannot resolve module: "@stdlib/beep"',
            'line': 1,
            'column': 12,
            'nodeType': 'CallExpression',
            'source': 'var beep = require( \'@stdlib/beep\' )',
            'endLine': 1,
            'endColumn': 37
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
