/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dlaqge = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectType string
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 5, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( true, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( false, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( null, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( void 0, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( [], 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( {}, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( ( x: number ): number => x, 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', '5', 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', true, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', false, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', null, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', void 0, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', [], 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', {}, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', ( x: number ): number => x, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, '5', A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, true, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, false, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, null, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, void 0, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, [], A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, {}, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, ( x: number ): number => x, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a Float64Array...
{
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, '5', 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, 5, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, true, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, false, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, null, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, void 0, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, [], 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, {}, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, ( x: number ): number => x, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, '5', R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, true, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, false, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, null, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, void 0, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, [], R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, {}, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, ( x: number ): number => x, R, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, '5', C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, 5, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, true, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, false, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, null, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, void 0, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, [], C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, {}, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, ( x: number ): number => x, C, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, R, '5', 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, 5, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, true, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, false, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, null, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, void 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, [], 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, {}, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, ( x: number ): number => x, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, R, C, '5', 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, true, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, false, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, null, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, void 0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, [], 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, {}, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, ( x: number ): number => x, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, '5', 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, true, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, false, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, null, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, void 0, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, [], 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, {}, 4.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, ( x: number ): number => x, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, '5' ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, true ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, false ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, null ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, void 0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, [] ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, {} ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge(); // $ExpectError
	dlaqge( 'row-major' ); // $ExpectError
	dlaqge( 'row-major', 2 ); // $ExpectError
	dlaqge( 'row-major', 2, 2 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0 ); // $ExpectError
	dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectType string
}

// The compiler throws an error if the function is provided a first argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( '5', 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( true, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( false, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( null, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( void 0, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( [], 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( {}, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( ( x: number ): number => x, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, '5', A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, true, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, false, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, null, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, void 0, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, [], A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, {}, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, ( x: number ): number => x, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a Float64Array...
{
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, '5', 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, 5, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, true, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, false, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, null, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, void 0, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, [], 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, {}, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, ( x: number ): number => x, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, '5', 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, true, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, false, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, null, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, void 0, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, [], 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, {}, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, ( x: number ): number => x, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, '5', 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, true, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, false, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, null, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, void 0, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, [], 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, {}, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, ( x: number ): number => x, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, '5', R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, true, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, false, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, null, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, void 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, [], R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, {}, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, ( x: number ): number => x, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, '5', 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, 5, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, true, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, false, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, null, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, void 0, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, [], 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, {}, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, ( x: number ): number => x, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, '5', 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, true, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, false, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, null, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, void 0, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, [], 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, {}, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, ( x: number ): number => x, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, '5', C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, true, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, false, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, null, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, void 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, [], C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, {}, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, ( x: number ): number => x, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, '5', 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, 5, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, true, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, false, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, null, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, void 0, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, [], 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, {}, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, ( x: number ): number => x, 1, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, '5', 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, true, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, false, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, null, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, void 0, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, [], 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, {}, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, ( x: number ): number => x, 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, '5', 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, true, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, false, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, null, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, void 0, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, [], 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, {}, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, ( x: number ): number => x, 1.0/11.0, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a thirteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, '5', 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, true, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, false, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, null, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, void 0, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, [], 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, {}, 1.0/11.0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, ( x: number ): number => x, 1.0/11.0, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, '5', 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, true, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, false, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, null, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, void 0, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, [], 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, {}, 4.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, ( x: number ): number => x, 4.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, '5' ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, true ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, false ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, null ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, void 0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, [] ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, {} ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );
	const R = new Float64Array( [ 1.0, 11.0 ] );
	const C = new Float64Array( [ 1.0, 11.0 ] );

	dlaqge.ndarray(); // $ExpectError
	dlaqge.ndarray( 2 ); // $ExpectError
	dlaqge.ndarray( 2, 2 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0 ); // $ExpectError
	dlaqge.ndarray( 2, 2, A, 2, 1, 0, R, 1, 0, C, 1, 0, 1.0/11.0, 1.0/11.0, 4.0, 10 ); // $ExpectError
}
