# no-internal-require

> [ESLint rule][eslint-rules] disallowing `require()` calls into internals of another stdlib package.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-internal-require' );
```

#### rule

[ESLint rule][eslint-rules] disallowing `require()` calls into internals of another stdlib package.

**Bad**:

<!-- eslint-disable stdlib/no-internal-require -->

```javascript
var betainc = require( '@stdlib/math/base/special/betainc/lib/betainc.js' );
```

**Good**:

``` javascript 
var betainc = require( '@stdlib/math/base/special/betainc' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-internal-require' );

var linter = new Linter();
var result;

var code = 'var betainc = require( \'@stdlib/math/base/special/betainc/lib/betainc.js\' )';

linter.defineRule( 'no-internal-require', rule );

result = linter.verify( code, {
    'rules': {
        'no-internal-require': 'error'
    }
});
/* returns
    [
        {
            ruleId: 'no-internal-require',
            severity: 2,
            message: 'require() call contains a path which loads a module internal to a stdlib package, and, thus, should be considered private',
            line: 1,
            column: 15,
            nodeType: 'CallExpression',
            source: 'var betainc = require( \'@stdlib/math/base/special/betainc/lib/betainc.js\' )',
            endLine: 1,
            endColumn: 76
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
