<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-internal-require

> [ESLint rule][eslint-rules] disallowing `require()` calls into internals of another stdlib package.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-internal-require' );
```

#### rule

[ESLint rule][eslint-rules] disallowing `require()` calls into internals of another stdlib package.

**Bad**:

<!-- eslint-disable stdlib/no-internal-require -->

```javascript
var betainc = require( '@stdlib/math/base/special/betainc/lib/betainc.js' );
```

**Good**:

```javascript
var betainc = require( '@stdlib/math/base/special/betainc' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-internal-require' );

var linter = new Linter();
var result;

var code = 'var betainc = require( \'@stdlib/math/base/special/betainc/lib/betainc.js\' )';

linter.defineRule( 'no-internal-require', rule );

result = linter.verify( code, {
    'rules': {
        'no-internal-require': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-internal-require',
            'severity': 2,
            'message': 'require() call contains a path which loads a module internal to a stdlib package, and, thus, should be considered private',
            'line': 1,
            'column': 15,
            'nodeType': 'CallExpression',
            'source': 'var betainc = require( \'@stdlib/math/base/special/betainc/lib/betainc.js\' )',
            'endLine': 1,
            'endColumn': 76
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
