<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# First Heading Level

> [ESLint rule][eslint-rules] to require the first Markdown heading level be a specified level in JSDoc comments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-first-heading-level' );
```

#### rule

[ESLint rule][eslint-rules] to require the first Markdown heading level be a specified level in JSDoc comments.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-first-heading-level, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* # Beep
*
* Boop.
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

<!-- eslint-disable stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* ## Beep
*
* Boop.
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-first-heading-level]. By default, the rule requires that the first heading level is `1`.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-first-heading-level' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* # Beep',
    '*',
    '* Boop.',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-first-heading-level', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-first-heading-level': [ 'error', 2 ]
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-first-heading-level',
            'severity': 2,
            'message': 'First heading level should be `2`',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* # Beep',
            'endLine': 14,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-first-heading-level]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-first-heading-level

</section>

<!-- /.links -->
