<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# capitalized-comments

> [ESLint rule][eslint-rules] to enforce that comments start with an uppercase letter.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/capitalized-comments' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that comments start with an uppercase letter. The rule does not apply to JSDoc comments, trailing line comments, and block comments with preceding and subsequent tokens on the same line. The rule also allows words at the beginning of comments to be uncapitalized if they are identifiers for variables in the current scope.

**Bad**:

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
function square( x ) {
    var out;

    // square the number:
    out = x*x;
    return out;
}
```

**Good**:

```javascript
function square( x ) {
    var out;

    // Square the number:
    out = x*x; // x^2
    return out;
}
```

To allow uncapitalized words at the beginning of a comment, set the `whitelist` option to a list of uncapitalized exceptions.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/capitalized-comments' );

var linter = new Linter();
var result;
var code;

code = [
    'function pow2( x ) {',
    '  var out;',
    '',
    '  // square the number:',
    '  out = x*x; // x^2',
    '  return out;',
    '}'
].join( '\n' );

linter.defineRule( 'capitalized-comments', rule );

result = linter.verify( code, {
    'rules': {
        'capitalized-comments': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'capitalized-comments',
            'severity': 2,
            'message': 'Comments should begin with an uppercase character',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '  // square the number:',
            'endLine': 4,
            'endColumn': 24
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
