<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# module-exports-last

> [ESLint rule][eslint-rules] enforcing that export statements appear at the end of a file.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/module-exports-last' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that export statements appear at the end of a file.

**Bad**:

<!-- eslint-disable stdlib/module-exports-last -->

```javascript
var gammaln = require( '@stdlib/math/base/special/gammaln' );

module.exports = exports = {
    'gammaln': gammaln
};

var beta = require( '@stdlib/math/base/special/beta' );

exports.beta = beta;
```

**Good**:

```javascript
var gammaln = require( '@stdlib/math/base/special/gammaln' );
var betainc = require( '@stdlib/math/base/special/betainc' );

exports.gammaln = gammaln;
exports.betainc = betainc;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/module-exports-last' );

var linter = new Linter();
var result;
var code;

code = [
    'var betainc = require( \'./betainc.js\' );',
    '',
    'module.exports = betainc;',
    '',
    'var gamma = require( \'./gamma.js\' );',
    '',
    'module.exports = gamma;'
].join( '\n' );

linter.defineRule( 'module-exports-last', rule );

result = linter.verify( code, {
    'rules': {
        'module-exports-last': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'module-exports-last',
            'severity': 2,
            'message': 'Export statements should appear at the end of the file',
            'line': 3,
            'column': 15,
            'nodeType': 'ExpressionStatement',
            'source': 'module.exports = betainc;',
            'endLine': 3,
            'endColumn': 26
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
