// @stdlib/_tools/esm/transform-vendor-requires-file-list/test/test.sync.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var IS_BROWSER = require( '@stdlib/assert/is-browser' );
var readFileSync = require( '@stdlib/fs/read-file' ).sync;
var writeFileSync = require( '@stdlib/fs/write-file' ).sync;
var unlinkSync = require( '@stdlib/fs/unlink' ).sync;
var exists = require( '@stdlib/fs/exists' ).sync;
var licenseHeader = require( '@stdlib/_tools/licenses/header' );
var removeHeader = require( './../lib/sync.js' );


// VARIABLES //

var opts = {
	'skip': IS_BROWSER
};
var TMP = join( __dirname, 'fixtures', 'tmp.js.txt' );
var EXPECTED = join( __dirname, 'fixtures', 'expected.js.txt' );
var EXPECTED_SHEBANG = join( __dirname, 'fixtures', 'expected_shebang.js.txt' );
var SINGLE_LINE_HEADER = join( __dirname, 'fixtures', 'single_line_header.js.txt' );
var MULTI_LINE_HEADER = join( __dirname, 'fixtures', 'multiline_header.js.txt' );
var SHEBANG_SINGLE_LINE_HEADER = join( __dirname, 'fixtures', 'shebang_single_line_header.js.txt' ); // eslint-disable-line id-length
var SHEBANG_MULTI_LINE_HEADER = join( __dirname, 'fixtures', 'shebang_multiline_header.js.txt' ); // eslint-disable-line id-length


// FUNCTIONS //

/**
* Creates a temporary file.
*
* @private
* @param {boolean} bool - boolean indicating whether a temporary file should include a "shebang"
* @param {string} type - header type ('single-line' or 'multi-line')
* @param {string} [f] - filename
* @throws {Error} unexpected error
* @returns {string} temporary file filename
*/
function tmp( bool, type ) {
	var err;
	var out;
	var f;
	if ( arguments.length > 2 ) {
		out = arguments[ 2 ];
	} else {
		out = TMP;
	}
	if ( bool ) {
		if ( type === 'single-line' ) {
			f = SHEBANG_SINGLE_LINE_HEADER;
		} else {
			f = SHEBANG_MULTI_LINE_HEADER;
		}
		err = writeFileSync( out, readFileSync( f ) );
	} else {
		if ( type === 'single-line' ) {
			f = SINGLE_LINE_HEADER;
		} else {
			f = MULTI_LINE_HEADER;
		}
		err = writeFileSync( out, readFileSync( f ) );
	}
	if ( err ) {
		throw err;
	}
	return out;
}

/**
* Removes a temporary file.
*
* @private
* @param {string} [f] - filename
* @throws {Error} unexpected error
*/
function cleanup() {
	var err;
	var f;
	if ( arguments.length ) {
		f = arguments[ 0 ];
	} else {
		f = TMP;
	}
	err = unlinkSync( f );
	if ( err ) {
		throw err;
	}
}


// TESTS //

tape( 'main export is a function', opts, function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof removeHeader, 'function', 'main export is a function' );
	t.end();
});

tape( 'if provided a first argument which is not a list of files, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			removeHeader( value, {} );
		};
	}
});

tape( 'if provided a header argument which is not either a string, regular expression, or an object mapping file types to license header strings or regular expressions, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{
			'foo': null
		},
		{
			'js': []
		},
		{
			'js': {}
		},
		{
			'awk': '# @license Apache-2.0',
			'js': '// @license Apache-2.0',
			'jl': 3.14
		},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			removeHeader( [ __filename ], value );
		};
	}
});

tape( 'the function returns an error if an error is encountered when reading a file', opts, function test( t ) {
	var header;
	var files;
	var err;

	files = [ 'dakfjdlfjsdfjasdjf/aldfkajdlfjsadl.jsdfad' ]; // File should not exist!
	header = {
		'jsdfad': 'beep boop'
	};
	err = removeHeader( files, header );
	t.strictEqual( err instanceof Error, true, 'returns an error' );

	t.end();
});

tape( 'the function returns an error if an error is encountered when writing a file', opts, function test( t ) {
	var removeHeader;
	var header;
	var files;
	var err;

	removeHeader = proxyquire( './../lib/sync.js', {
		'@stdlib/fs/write-file': {
			'sync': writeFile
		}
	});

	files = [ __filename ];
	header = {
		'js': '// @license Apache-2.0\n'
	};
	err = removeHeader( files, header );
	t.strictEqual( err instanceof Error, true, 'returns an error' );

	t.end();

	function writeFile() {
		return new Error( 'beep' );
	}
});

tape( 'the function returns `null` if provided an empty file array', opts, function test( t ) {
	var err = removeHeader( [], {} );
	t.strictEqual( err, null, 'returns expected value' );
	t.end();
});

tape( 'the function supports providing an empty header string', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create an empty header string:
	header = '';

	// Attempt to remove headers:
	err = removeHeader( [ f, f, f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has not changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( SINGLE_LINE_HEADER, 'utf8' );
	t.strictEqual( actual, expected, 'file is unchanged' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing an empty header object', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create an empty header object:
	header = {};

	// Attempt to remove headers:
	err = removeHeader( [ f, f, f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has not changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( SINGLE_LINE_HEADER, 'utf8' );
	t.strictEqual( actual, expected, 'file is unchanged' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header string', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header regular expression', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header regular expression:
	header = /\/\/ @license Apache-2\.0\n\n/;

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header string (shebang)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Create a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header regular expression (shebang)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Create a header regular expression:
	header = /\n\/\/ @license Apache-2\.0\n\n/;

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header string (file lacking filename extension)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file lacking a filename extension:
	f = tmp( false, 'single-line', join( __dirname, 'fixtures', 'tmp' ) );

	// Create a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup( f );
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'multi-line' );

	// Create a header object:
	header = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object (regular expression)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header object:
	header = {
		'txt': /\/\/ @license Apache-2\.0\n\n/
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object (shebang)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'multi-line' );

	// Create a header object:
	header = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object (shebang; regular expression)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Create a header object:
	header = {
		'txt': /\n\/\/ @license Apache-2\.0\n\n/
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object (file lacking filename extension)', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file lacking a filename extension:
	f = tmp( false, 'multi-line', join( __dirname, 'fixtures', 'tmp' ) );

	// Create a header object:
	header = {
		'tmp': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup( f );
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object containing a `default` header', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp( false, 'multi-line' );

	// Create a header object:
	header = {
		'f': licenseHeader( 'Apache-2.0', 'f', {
			'year': 2050,
			'copyright': 'The Stdlib Authors and Contributors'
		}),
		'default': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to remove headers:
	err = removeHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFileSync( f, 'utf8' );
	expected = readFileSync( EXPECTED, 'utf8' );
	t.strictEqual( actual, expected, 'file does not contain header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});
