// @stdlib/utils/delay/lib/main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2022 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

// TODO: Load required modules...


// MAIN //

/**
* Invokes a function after a specified duration.
*
* @param {Function} fcn - function to invoke
* @param {(NonNegativeInteger|string)} duration - duration
* @param {...*} [args] - arguments to invoke function with
* @returns {Object} timeout object
*
* @example
* function beep() {
*     console.log( 'beep' );
* }
*
* var t = delay( beep, 3000 );
* // returns <Object>
*
* @example
* function beep() {
*     console.log( 'beep' );
* }
*
* var t = delay( beep, '3s' );
* // returns <Object>
*
* @example
* function add( x, y ) {
*     var sum = x + y;
*     console.log( 'sum: %d', sum );
* }
*
* var t = delay( add, 3000, 2, 3 );
* // returns <Object>
*/
function delay( fcn, duration ) {
	// TODO: Add implementation...
}


// EXPORTS //

module.exports = delay;
