# require-typed-arrays

> [ESLint rule][eslint-rules] enforcing that typed array constructors are explicitly required.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-typed-arrays' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that typed array constructors are explicitly required.

**Bad**:

<!-- eslint-disable stdlib/require-typed-arrays -->

```javascript
var arr = new Float32Array();
```

**Good**:

``` javascript 
var Float32Array = require( '@stdlib/array/float32' );

var arr = new Float32Array();
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-typed-arrays' );

var linter = new Linter();
var result;

var code = 'var arr = new Int32Array( [1,2,3] );';

linter.defineRule( 'require-typed-arrays', rule );

result = linter.verify( code, {
    'rules': {
        'require-typed-arrays': 'error'
    }
});
console.log( result );
/* returns
    [
        {
            ruleId: 'require-typed-arrays',
            severity: 2,
            message: 'Typed array constructor is invoked without loading it beforehand via require()',
            line: 1,
            column: 11,
            nodeType: 'NewExpression',
            source: 'var arr = new Int32Array( [1,2,3] );',
            endLine: 1,
            endColumn: 36
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
