// @stdlib/blas/base/sspmv/test/test.ndarray.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable max-len */

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float32Array = require( '@stdlib/array/float32' );
var ones = require( '@stdlib/array/ones' );
var sspmv = require( './../lib/ndarray.js' );


// FIXTURES //

var rl = require( './fixtures/row_major_lower.json' );
var ru = require( './fixtures/row_major_upper.json' );
var rxpyp = require( './fixtures/row_major_xpyp.json' );
var rxnyp = require( './fixtures/row_major_xnyp.json' );
var rxpyn = require( './fixtures/row_major_xpyn.json' );
var rxnyn = require( './fixtures/row_major_xnyn.json' );
var rsap = require( './fixtures/row_major_sap.json' );
var rsapn = require( './fixtures/row_major_sapn.json' );
var roap = require( './fixtures/row_major_oap.json' );
var rox = require( './fixtures/row_major_ox.json' );
var roy = require( './fixtures/row_major_oy.json' );
var rcap = require( './fixtures/row_major_complex_access_pattern.json' );

var cl = require( './fixtures/column_major_lower.json' );
var cu = require( './fixtures/column_major_upper.json' );
var cxpyp = require( './fixtures/column_major_xpyp.json' );
var cxnyp = require( './fixtures/column_major_xnyp.json' );
var cxpyn = require( './fixtures/column_major_xpyn.json' );
var cxnyn = require( './fixtures/column_major_xnyn.json' );
var csap = require( './fixtures/column_major_sap.json' );
var csapn = require( './fixtures/column_major_sapn.json' );
var coap = require( './fixtures/column_major_oap.json' );
var cox = require( './fixtures/column_major_ox.json' );
var coy = require( './fixtures/column_major_oy.json' );
var ccap = require( './fixtures/column_major_complex_access_pattern.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof sspmv, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 14', function test( t ) {
	t.strictEqual( sspmv.length, 14, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided an invalid first argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( value, data.uplo, data.N, data.alpha, new Float32Array( data.AP ), data.strideAP, data.offsetAP, new Float32Array( data.x ), data.strideX, data.offsetX, data.beta, new Float32Array( data.y ), data.strideY, data.offsetY );
		};
	}
});

tape( 'the function throws an error if provided an invalid second argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( data.order, value, data.N, data.alpha, new Float32Array( data.AP ), data.strideAP, data.offsetAP, new Float32Array( data.x ), data.strideX, data.offsetX, data.beta, new Float32Array( data.y ), data.strideY, data.offsetY );
		};
	}
});

tape( 'the function throws an error if provided an invalid third argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		-1,
		-2,
		-3
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( data.order, data.uplo, value, data.alpha, new Float32Array( data.AP ), data.strideAP, data.offsetAP, new Float32Array( data.x ), data.strideX, data.offsetX, data.beta, new Float32Array( data.y ), data.strideY, data.offsetY );
		};
	}
});

tape( 'the function throws an error if provided an invalid sixth argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		0
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( data.order, data.uplo, data.N, data.alpha, new Float32Array( data.AP ), value, data.offsetAP, new Float32Array( data.x ), data.strideX, data.offsetX, data.beta, new Float32Array( data.y ), data.strideY, data.offsetY );
		};
	}
});

tape( 'the function throws an error if provided an invalid ninth argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		0
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( data.order, data.uplo, data.N, data.alpha, new Float32Array( data.AP ), data.strideAP, data.offsetAP, new Float32Array( data.x ), value, data.offsetX, data.beta, new Float32Array( data.y ), data.strideY, data.offsetY );
		};
	}
});

tape( 'the function throws an error if provided an invalid thirteenth argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		0
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspmv( data.order, data.uplo, data.N, data.alpha, new Float32Array( data.AP ), data.strideAP, data.offsetAP, new Float32Array( data.x ), data.strideX, data.offsetX, data.beta, new Float32Array( data.y ), value, data.offsetY );
		};
	}
});

tape( 'the function performs the matrix-vector operation `y = α*A*x + β*y (row-major, lower)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the matrix-vector operation `y = α*A*x + β*y (column-major, lower)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the matrix-vector operation `y = α*A*x + β*y (row-major, upper)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = ru;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the matrix-vector operation `y = α*A*x + β*y (column-major, upper)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cu;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function returns a reference to the second input vector', function test( t ) {
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );

	t.end();
});

tape( 'if `N` is zero or the scalar constants are zero and unity, respectively, the function returns the second input vector unchanged (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y );

	out = sspmv( data.order, data.uplo, 0, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	out = sspmv( data.order, data.uplo, data.N, 0.0, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, 1.0, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'if `N` is zero or the scalar constants are zero and unity, respectively, the function returns the second input vector unchanged (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y );

	out = sspmv( data.order, data.uplo, 0, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	out = sspmv( data.order, data.uplo, data.N, 0.0, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, 1.0, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'when `α` is zero, the function scales the second input vector (row-major, upper)', function test( t ) {
	var expected;
	var out;
	var ap;
	var x;
	var y;

	ap = ones( 6, 'float32' );
	x = ones( 3, 'float32' );
	y = ones( 3, 'float32' );

	expected = new Float32Array( [ 5.0, 5.0, 5.0 ] );

	out = sspmv( 'row-major', 'upper', 3, 0.0, ap, 1, 0, x, 1, 0, 5.0, y, 1, 0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	expected = new Float32Array( [ 0.0, 0.0, 0.0 ] );

	out = sspmv( 'row-major', 'upper', 3, 0.0, ap, 1, 0, x, 1, 0, 0.0, y, -1, 2 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'when `α` is zero, the function scales the second input vector (row-major, lower)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = ones( 6, 'float32' );
	x = ones( 3, 'float32' );
	y = ones( 3, 'float32' );

	expected = new Float32Array( [ 5.0, 5.0, 5.0 ] );

	out = sspmv( 'row-major', 'lower', 3, 0.0, ap, 1, 0, x, 1, 0, 5.0, y, -1, 2 );
	t.strictEqual( out, y, 'returns expected value', data.offsetY );
	t.deepEqual( y, expected, 'returns expected value' );

	expected = new Float32Array( [ 0.0, 0.0, 0.0 ] );

	out = sspmv( 'row-major', 'lower', 3, 0.0, ap, 1, 0, x, 1, 0, 0.0, y, 1, 0 );
	t.strictEqual( out, y, 'returns expected value', data.offsetY );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'when `α` is zero, the function scales the second input vector (column-major, upper)', function test( t ) {
	var expected;
	var out;
	var ap;
	var x;
	var y;

	ap = ones( 6, 'float32' );
	x = ones( 3, 'float32' );
	y = ones( 3, 'float32' );

	expected = new Float32Array( [ 5.0, 5.0, 5.0 ] );

	out = sspmv( 'column-major', 'upper', 3, 0.0, ap, 1, 0, x, 1, 0, 5.0, y, 1, 0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	expected = new Float32Array( [ 0.0, 0.0, 0.0 ] );

	out = sspmv( 'column-major', 'upper', 3, 0.0, ap, 1, 0, x, 1, 0, 0.0, y, -1, 2 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'when `α` is zero, the function scales the second input vector (column-major, lower)', function test( t ) {
	var expected;
	var out;
	var ap;
	var x;
	var y;

	ap = ones( 6, 'float32' );
	x = ones( 3, 'float32' );
	y = ones( 3, 'float32' );

	expected = new Float32Array( [ 5.0, 5.0, 5.0 ] );

	out = sspmv( 'column-major', 'lower', 3, 0.0, ap, 1, 0, x, 1, 0, 5.0, y, -1, 2 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	expected = new Float32Array( [ 0.0, 0.0, 0.0 ] );

	out = sspmv( 'column-major', 'lower', 3, 0.0, ap, 1, 0, x, 1, 0, 0.0, y, 1, 0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( y, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying `x` and `y` strides (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxpyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying `x` and `y` strides (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxpyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `y` stride (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxpyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `y` stride (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxpyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `x` stride (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxnyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `x` stride (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxnyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying negative strides for `x` and `y` (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxnyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying negative strides for `x` and `y` (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxnyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying an `x` offset (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rox;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying an `x` offset (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cox;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a `y` offset (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = roy;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a `y` offset (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = coy;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a stride for `AP` (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rsap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a stride for `AP` (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = csap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative stride for `AP` (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rsapn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative stride for `AP` (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = csapn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying an offset for `AP` (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = roap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying an offset for `AP` (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = coap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying complex access patterns (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rcap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying complex access patterns (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = ccap;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.y_out );

	out = sspmv( data.order, data.uplo, data.N, data.alpha, ap, data.strideAP, data.offsetAP, x, data.strideX, data.offsetX, data.beta, y, data.strideY, data.offsetY );
	t.strictEqual( out, y, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});
