/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dspr2 = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 10, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( true, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( false, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( null, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( void 0, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( [], 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( {}, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( ( x: number ): number => x, 'upper', 10, 1.0, x, 1, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 10, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', true, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', false, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', null, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', void 0, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', [ '1' ], 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', {}, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', ( x: number ): number => x, 10, 1.0, x, 1, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', '10', 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', true, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', false, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', null, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', void 0, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', [], 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', {}, 1.0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', ( x: number ): number => x, 1.0, x, 1, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, '10', x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, true, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, false, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, null, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, void 0, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, [], x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, {}, x, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, ( x: number ): number => x, x, 1, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a Float64Array...
{
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, 1.0, 10, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, '10', 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, true, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, false, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, null, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, void 0, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, [ '1' ], 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, {}, 1, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, ( x: number ): number => x, 1, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, 1.0, x, '10', y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, true, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, false, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, null, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, void 0, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, [], y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, {}, y, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, ( x: number ): number => x, y, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float64Array...
{
	const x = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, 10, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, '10', 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, true, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, false, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, null, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, void 0, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, [ '1' ], 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, {}, 1, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, ( x: number ): number => x, 1, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, '10', A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, true, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, false, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, null, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, void 0, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, [], A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, {}, A ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, ( x: number ): number => x, A ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a Float64Array...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );

	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, 10 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, '10' ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, true ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, false ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, null ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, void 0 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, [ '1' ] ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, {} ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2(); // $ExpectError
	dspr2( 'row-major' ); // $ExpectError
	dspr2( 'row-major', 'upper' ); // $ExpectError
	dspr2( 'row-major', 'upper', 10 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1 ); // $ExpectError
	dspr2( 'row-major', 'upper', 10, 1.0, x, 1, y, 1, A, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 10, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( true, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( false, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( null, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( void 0, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( [], 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( {}, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( ( x: number ): number => x, 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 10, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', true, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', false, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', null, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', void 0, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', [ '1' ], 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', {}, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', ( x: number ): number => x, 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', '10', 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', true, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', false, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', null, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', void 0, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', [], 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', {}, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', ( x: number ): number => x, 1.0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, '10', x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, true, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, false, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, null, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, void 0, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, [], x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, {}, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, ( x: number ): number => x, x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a Float64Array...
{
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, 10, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, '10', 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, true, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, false, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, null, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, void 0, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, [ '1' ], 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, {}, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, ( x: number ): number => x, 1, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, '10', 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, true, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, false, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, null, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, void 0, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, [], 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, {}, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, ( x: number ): number => x, 0, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, '10', y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, true, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, false, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, null, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, void 0, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, [], y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, {}, y, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, ( x: number ): number => x, y, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a Float64Array...
{
	const x = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, 10, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, '10', 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, true, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, false, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, null, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, void 0, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, [ '1' ], 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, {}, 1, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, ( x: number ): number => x, 1, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, '10', 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, true, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, false, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, null, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, void 0, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, [], 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, {}, 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, ( x: number ): number => x, 0, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, '10', A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, true, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, false, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, null, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, void 0, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, [], A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, {}, A, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, ( x: number ): number => x, A, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a Float64Array...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, 10, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, '10', 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, true, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, false, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, null, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, void 0, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, [ '1' ], 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, {}, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, '10', 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, true, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, false, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, null, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, void 0, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, [], 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, {}, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a thirteenth argument which is not a number...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, '10' ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, true ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, false ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, null ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, void 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, [] ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, {} ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );
	const A = new Float64Array( 55 );

	dspr2.ndarray(); // $ExpectError
	dspr2.ndarray( 'row-major' ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper' ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1 ); // $ExpectError
	dspr2.ndarray( 'row-major', 'upper', 10, 1.0, x, 1, 0, y, 1, 0, A, 1, 0, 10 ); // $ExpectError
}
