// @stdlib/_tools/static-analysis/js/summarize-file-list/test/test.async.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var getKeys = require( 'object-keys' ).shim();
var instanceOf = require( '@stdlib/assert/instance-of' );
var isPlainObject = require( '@stdlib/assert/is-plain-object' );
var isPlainObjectArray = require( '@stdlib/assert/is-plain-object-array' );
var noop = require( '@stdlib/utils/noop' );
var analyze = require( './../lib/main.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof analyze, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a list of files', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		[ 5 ],
		[ '5', null ],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			analyze( value, noop );
		};
	}
});

tape( 'the function throws an error if not provided a list of files (options)', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		[ 5 ],
		[ '5', null ],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			analyze( value, {}, noop );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			analyze( [ 'a', 'b' ], value, noop );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var opts = {
				'cumulative': value
			};
			analyze( [ 'a', 'b' ], opts, noop );
		};
	}
});

tape( 'the function throws an error if not provided a callback argument', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			analyze( [ 'a', 'b' ], value );
		};
	}
});

tape( 'the function throws an error if not provided a callback argument (options)', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			analyze( [ 'a', 'b' ], {}, value );
		};
	}
});

tape( 'the function returns an error to a provided callback if an error is encountered when reading files', function test( t ) {
	var doesNotExist = 'akfjdalfkjsdlfjdasfjadsfjdslfjldajfasdjfasjfjasdlfjasdlfj.js';
	analyze( [ doesNotExist ], clbk );

	function clbk( error ) {
		t.strictEqual( instanceOf( error, Error ), true, 'returns an error' );
		t.end();
	}
});

tape( 'the function returns an error to a provided callback if an error is encountered when reading files (options)', function test( t ) {
	var doesNotExist = 'akfjdalfkjsdlfjdasfjadsfjdslfjldajfasdjfasjfjasdlfjasdlfj.js';
	analyze( [ doesNotExist ], {}, clbk );

	function clbk( error ) {
		t.strictEqual( instanceOf( error, Error ), true, 'returns an error' );
		t.end();
	}
});

tape( 'by default, the function analyzes input files and returns a cumulative static summary to a provided callback', function test( t ) {
	var files = [ __filename, __filename ];
	analyze( files, clbk );

	function clbk( error, o ) {
		if ( error ) {
			t.fail( error.message );
		} else {
			t.strictEqual( isPlainObject( o ), true, 'returns a plain object' );
			t.strictEqual( getKeys( o ).length > 0, true, 'has key-value pairs' );
		}
		t.end();
	}
});

tape( 'if the `cumulative` option is `true`, the function analyzes input files and returns a cumulative static summary to a provided callback', function test( t ) {
	var files;
	var opts;

	files = [ __filename, __filename ];
	opts = {
		'cumulative': true
	};
	analyze( files, opts, clbk );

	function clbk( error, o ) {
		if ( error ) {
			t.fail( error.message );
		} else {
			t.strictEqual( isPlainObject( o ), true, 'returns a plain object' );
			t.strictEqual( getKeys( o ).length > 0, true, 'has key-value pairs' );
		}
		t.end();
	}
});

tape( 'if the `cumulative` option is `false`, the function analyzes input files and returns an array of program summaries to a provided callback', function test( t ) {
	var files;
	var opts;

	files = [ __filename, __filename ];
	opts = {
		'cumulative': false
	};
	analyze( files, opts, clbk );

	function clbk( error, o ) {
		var i;
		if ( error ) {
			t.fail( error.message );
		} else {
			t.strictEqual( isPlainObjectArray( o ), true, 'returns an array of plain objects' );
			for ( i = 0; i < o.length; i++ ) {
				t.strictEqual( getKeys( o[ i ] ).length > 0, true, 'has key-value pairs' );
			}
		}
	}
	t.end();
});
