// @stdlib/utils/try-catch-async/lib/try_catch_async.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );


// MAIN //

/**
* If a function does not return an error, invokes a callback with the function result; otherwise, invokes a callback with a value `y`.
*
* @param {Function} x - function to invoke
* @param {*} y - value to return if `x` returns an error
* @param {Function} done - callback to invoke upon completion
* @throws {TypeError} first argument must be a function
* @throws {TypeError} last argument must be a function
*
* @example
* var randu = require( '@stdlib/math/base/random/randu' );
*
* function x( clbk ) {
*     setTimeout( onTimeout, 0 );
*     function onTimeout() {
*         if ( randu() > 0.5 ) {
*             return clbk( null, 1.0 );
*         }
*         clbk( new Error( 'beep' ) );
*     }
* }
*
* function done( error, result ) {
*     if ( error ) {
*         console.log( error.message );
*     }
*     console.log( result );
* }
*
* trycatchAsync( x, -1.0, done );
*/
function trycatchAsync( x, y, done ) {
	if ( !isFunction( x ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `'+x+'`.' );
	}
	if ( !isFunction( done ) ) {
		throw new TypeError( 'invalid input argument. Last argument must be a function. Value: `'+done+'`.' );
	}
	x( clbk );

	/**
	* Callback invoked by `x`.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {*} result - result
	* @returns {void}
	*/
	function clbk( error, result ) {
		if ( error ) {
			return done( error, y );
		}
		done( null, result );
	} // end FUNCTION clbk()
} // end FUNCTION trycatchAsync()


// EXPORTS //

module.exports = trycatchAsync;
