<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# ternary-condition-parentheses

> [ESLint rule][eslint-rules] requiring parentheses around ternary conditions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/ternary-condition-parentheses' );
```

#### rule

[ESLint rule][eslint-rules] requiring parentheses around ternary conditions.

**Bad**:

<!-- eslint-disable stdlib/ternary-condition-parentheses -->

```javascript
var randu = require( '@stdlib/random/base/randu' );

var bool = randu() > 0.5 ? 1 : 0;
```

**Good**:

```javascript
var randu = require( '@stdlib/random/base/randu' );

var bool = ( randu() > 0.5 ) ? 1 : 0;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/ternary-condition-parentheses' );

var linter = new Linter();
var result;

var code = 'var bool = randu() > 0.5 ? 1 : 0;';

linter.defineRule( 'ternary-condition-parentheses', rule );

result = linter.verify( code, {
    'rules': {
        'ternary-condition-parentheses': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'ternary-condition-parentheses',
            'severity': 2,
            'message': 'parentheses required around ternary conditions',
            'line': 1,
            'column': 12,
            'nodeType': 'BinaryExpression',
            'source': 'var bool = randu() > 0.5 ? 1 : 0;',
            'endLine': 1,
            'endColumn': 25
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
