/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import stpsv = require( './index' );


// TESTS //

// The function returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 10, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( true, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( false, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( null, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( undefined, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( [], 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( {}, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( ( x: number ): number => x, 'upper', 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 10, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', true, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', false, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', null, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', undefined, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', [], 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', {}, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', ( x: number ): number => x, 'no-transpose', 'unit', 10, AP, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 10, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', true, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', false, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', null, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', undefined, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', [], 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', {}, 'unit', 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', ( x: number ): number => x, 'unit', 10, AP, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 'no-transpose', 10, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', true, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', false, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', null, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', undefined, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', [], 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', {}, 10, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', ( x: number ): number => x, 10, AP, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', '10', AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', true, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', false, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', null, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', undefined, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', [], AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', {}, AP, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', ( x: number ): number => x, AP, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );

	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, 10, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, '10', x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, true, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, false, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, null, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, undefined, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, [ '1' ], x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, {}, x, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, ( x: number ): number => x, x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float32Array...
{
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 10, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, '10', 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, true, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, false, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, null, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, undefined, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, [ '1' ], 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, {}, 1 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, ( x: number ): number => x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, '10' ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, true ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, false ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, null ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, undefined ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, [] ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, {} ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv(); // $ExpectError
	stpsv( 'row-major' ); // $ExpectError
	stpsv( 'row-major', 'upper' ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose' ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit' ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10 ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x ); // $ExpectError
	stpsv( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, x, 1, 1 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 10, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( true, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( false, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( null, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( undefined, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( [], 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( {}, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( ( x: number ): number => x, 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 10, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', true, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', false, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', null, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', undefined, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', [], 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', {}, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', ( x: number ): number => x, 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 10, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', true, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', false, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', null, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', undefined, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', [], 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', {}, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', ( x: number ): number => x, 'unit', 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 10, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', true, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', false, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', null, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', undefined, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', [], 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', {}, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', ( x: number ): number => x, 10, AP, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', '10', AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', true, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', false, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', null, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', undefined, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', [], AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', {}, AP, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', ( x: number ): number => x, AP, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, 10, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, '10', 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, true, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, false, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, null, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, undefined, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, [ '1' ], 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, {}, 1, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, ( x: number ): number => x, 1, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, '10', 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, true, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, false, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, null, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, undefined, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, [], 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, {}, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, ( x: number ): number => x, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, '10', 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, true, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, false, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, null, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, undefined, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, [], 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, {}, 0, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, ( x: number ): number => x, 0, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, '10', x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, true, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, false, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, null, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, undefined, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, [], x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, {}, x, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, ( x: number ): number => x, x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a Float32Array...
{
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, 10, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, '10', 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, true, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, false, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, null, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, undefined, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, [ '1' ], 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, {}, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, '10', 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, true, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, false, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, null, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, undefined, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, [], 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, {}, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, '10' ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, true ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, false ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, null ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, undefined ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, [] ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, {} ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const AP = new Float32Array( 55 );

	stpsv.ndarray(); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper' ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose' ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit' ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1 ); // $ExpectError
	stpsv.ndarray( 'row-major', 'upper', 'no-transpose', 'unit', 10, AP, 1, 0, x, 1, 0, 10 ); // $ExpectError
}
