// @stdlib/math/base/random/beta/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/utils/env' );
var kstest = require( '@stdlib/math/generics/statistics/kstest' );
var round = require( '@stdlib/math/base/special/round' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var beta;
	var r;
	var i;

	// When binding distribution parameters...
	beta = factory( 2.0, 2.0 );
	for ( i = 0; i < 400; i++ ) {
		r = beta();
		t.equal( typeof r, 'number', 'returns a number' );
	}
	beta = factory( 30.0, 30.0 );
	for ( i = 0; i < 400; i++ ) {
		r = beta();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameters...
	beta = factory();
	for ( i = 0; i < 100; i++ ) {
		r = beta( 2.5, 2.0 );
		t.equal( typeof r, 'number', 'returns a number' );
	}

	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator (alpha,beta <= 1.0)', function test( t ) {
	var beta1;
	var beta2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	beta1 = factory( 0.5, 0.5, {
		'seed': seed
	});
	beta2 = factory( 0.5, 0.5, {
		'seed': seed
	});

	t.notEqual( beta1, beta2, 'separate generators' );

	for ( i = 0; i < 1e3; i++ ) {
		r1 = beta1();
		r2 = beta2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator (alpha = beta > 1.5)', function test( t ) {
	var beta1;
	var beta2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	beta1 = factory( 3.14, 3.14, {
		'seed': seed
	});
	beta2 = factory( 3.14, 3.14, {
		'seed': seed
	});

	t.notEqual( beta1, beta2, 'separate generators' );

	for ( i = 0; i < 1e3; i++ ) {
		r1 = beta1();
		r2 = beta2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator (alpha != beta > 1.0)', function test( t ) {
	var beta1;
	var beta2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	beta1 = factory( 3.14, 1.25, {
		'seed': seed
	});
	beta2 = factory( 3.14, 1.25, {
		'seed': seed
	});

	t.notEqual( beta1, beta2, 'separate generators' );

	for ( i = 0; i < 1e3; i++ ) {
		r1 = beta1();
		r2 = beta2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var beta = factory();
	t.equal( beta.NAME, 'beta', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var beta = factory();
	t.equal( typeof beta.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var beta = factory({
		'seed': 12345
	});
	t.equal( typeof beta.SEED, 'number', 'has `SEED` property' );
	t.equal( beta.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'the function throws an error if `alpha` is not a positive number', function test( t ) {
	var values;
	var i;

	values = [
		-2.0,
		0.0,
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, 2.0 );
		};
	}
});

tape( 'the function throws an error if `beta` is not a positive number', function test( t ) {
	var values;
	var i;

	values = [
		-2.0,
		0.0,
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 2.0, value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 1.0, 1.0, value );
		};
	}
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a first argument which is `NaN`', function test( t ) {
	var beta;
	var r;

	beta = factory();
	r = beta( NaN, 1.0 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a second argument which is `NaN`', function test( t ) {
	var beta;
	var r;

	beta = factory();
	r = beta( 1.0, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided `NaNs`', function test( t ) {
	var beta;
	var r;

	beta = factory();
	r = beta( NaN, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a first argument which is nonpositive', function test( t ) {
	var beta;
	var r;

	beta = factory();

	r = beta( 0.0, 1.0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = beta( -1.0, 1.0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a second argument which is nonpositive', function test( t ) {
	var beta;
	var r;

	beta = factory();

	r = beta( 1.0, 0.0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = beta( 1.0, -1.0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from a beta distribution (when `alpha > 1` and `beta > 1`)', opts, function test( t ) {
	var threshold;
	var count;
	var npass;
	var alpha;
	var beta;
	var N;
	var x;

	threshold = 0.10;

	alpha = 2.0;
	beta = 4.0;

	x = new Array( 1e3 );
	N = 300;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var pValue;
		var rbeta;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			rbeta = factory( alpha, beta );
			t.ok( true, 'seed: '+rbeta.SEED );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = rbeta();
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'beta', alpha, beta ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Beta('+alpha+','+beta+')) is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});

tape( 'the function returns a PRNG for generating random numbers from a beta distribution (when `alpha <= 1` and `beta <= 1`)', opts, function test( t ) {
	var threshold;
	var count;
	var npass;
	var alpha;
	var beta;
	var N;
	var x;

	threshold = 0.10;

	alpha = 0.1;
	beta = 0.5;

	x = new Array( 1e3 );
	N = 300;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var pValue;
		var rbeta;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			rbeta = factory( alpha, beta );
			t.ok( true, 'seed: '+rbeta.SEED );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = rbeta();
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'beta', alpha, beta ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Beta('+alpha+','+beta+')) is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});
