// @stdlib/utils/if-else-async/lib/if_else_async.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );


// MAIN //

/**
* If a predicate function returns a truthy value, returns `x`; otherwise, returns `y`.
*
* @param {Function} predicate - predicate function
* @param {*} x - value to return if a condition is truthy
* @param {*} y - value to return if a condition is falsy
* @param {Function} done - callback to invoke upon completion
* @throws {TypeError} first argument must be a function
* @throws {TypeError} last argument must be a function
*
* @example
* var randu = require( '@stdlib/math/base/random/randu' );
*
* function predicate( clbk ) {
*     setTimeout( onTimeout, 0 );
*     function onTimeout() {
*         clbk( null, randu() > 0.5 );
*     }
* }
*
* function done( error, result ) {
*     if ( error ) {
*         throw error;
*     }
*     console.log( result );
* }
* ifelseAsync( predicate, 1.0, -1.0, done );
*/
function ifelseAsync( predicate, x, y, done ) {
	if ( !isFunction( predicate ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `'+predicate+'`.' );
	}
	if ( !isFunction( done ) ) {
		throw new TypeError( 'invalid input argument. Last argument must be a function. Value: `'+done+'`.' );
	}
	predicate( clbk );

	/**
	* Callback invoked by a predicate function.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {boolean} bool - condition
	* @returns {void}
	*/
	function clbk( error, bool ) {
		if ( error ) {
			return done( error );
		}
		if ( bool ) {
			return done( null, x );
		}
		done( null, y );
	} // end FUNCTION clbk()
} // end FUNCTION ifelseAsync()


// EXPORTS //

module.exports = ifelseAsync;
