// @stdlib/_tools/search/create/lib/index.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var path = require( 'path' );
var readFileList = require( '@stdlib/fs/read-file-list' );
var isFunction = require( '@stdlib/assert/is-function' );
var copy = require( '@stdlib/utils/copy' );
var cwd = require( '@stdlib/process/cwd' );
var findPkgs = require( '@stdlib/_tools/pkgs/find' );
var createIndex = require( './create_index.js' );
var getExisting = require( './get_existing.js' );
var defaults = require( './defaults.json' );
var validate = require( './validate.js' );


// MAIN //

/**
* Creates a serialized search index for all packages.
*
* @param {Options} [options] - function options
* @param {string} [options.dir] - root directory at which to start indexing
* @param {string} [options.pattern='**\/package.json'] - glob pattern
* @param {StringArray} [options.ignore] - glob pattern(s) to exclude matches
* @param {Callback} clbk - callback function
*/
function createSearchIndex() {
	var options;
	var opts;
	var clbk;
	var err;

	opts = copy( defaults );
	if ( arguments.length < 2 ) {
		clbk = arguments[ 0 ];
	} else {
		options = arguments[ 0 ];
		clbk = arguments[ 1 ];
		err = validate( opts, options );
		if ( err ) {
			throw err;
		}
	}
	if ( !isFunction( clbk ) ) {
		throw new TypeError( 'invalid input argument. Callback argument must be a function. Value: `' + clbk + '`.' );
	}
	if ( opts.dir ) {
		opts.dir = path.resolve( cwd(), opts.dir );
	} else {
		opts.dir = cwd();
	}
	findPkgs( opts, onPkgs );

	/**
	* Callback invoked after retrieving packages.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {StringArray} pkgs - list of package paths
	* @returns {void}
	*/
	function onPkgs( error, pkgs ) {
		var i;
		if ( error ) {
			return done( error );
		}
		for ( i = 0; i < pkgs.length; i++ ) {
			pkgs[ i ] = path.join( pkgs[ i ], 'README.md' );
		}
		getExisting( pkgs, onReadmes );
	}

	/**
	* Callback invoked after retrieving existing README.md paths.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {StringArray} readmes - list of existing README.md paths
	* @returns {void}
	*/
	function onReadmes( error, readmes ) {
		if ( error ) {
			return done( error );
		}
		readFileList( readmes, {
			'encoding': 'utf-8'
		}, onFiles );
	}

	/**
	* Callback invoked after retrieving file contents.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {ObjectArray} files - list of files
	* @returns {void}
	*/
	function onFiles( error, files ) {
		var idx;
		if ( error ) {
			return done( error );
		}
		idx = createIndex( files );
		done( null, idx );
	}

	/**
	* Callback invoked upon completion.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {Object} idx - serialized search index
	* @returns {void}
	*/
	function done( error, idx ) {
		if ( error ) {
			return clbk( error );
		}
		clbk( null, idx );
	}
}


// EXPORTS //

module.exports = createSearchIndex;
