/*
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import omitBy = require( './index' );


// TESTS //

// The function returns an object...
{
	omitBy( { 'a': 1, 'b': 2 }, ( _: string, value: number ): boolean => value > 1 ); // $ExpectType Partial<{ a: number; b: number; }>
	omitBy( { 'a': true, 'b': 3, 'c': 'foo' }, ( _: string, value: string | number | boolean ): boolean => !!value ); // $ExpectType Partial<{ a: boolean; b: number; c: string; }>
}

// The compiler throws an error if the function is provided a second argument which is not a predicate function...
{
	omitBy( {}, 3.12 ); // $ExpectError
	omitBy( {}, 'abc' ); // $ExpectError
	omitBy( {}, true ); // $ExpectError
	omitBy( {}, false ); // $ExpectError
	omitBy( {}, {} ); // $ExpectError
	omitBy( {}, [] ); // $ExpectError
}

// The compiler throws an error if the function is provided an incorrect number of arguments...
{
	omitBy(); // $ExpectError
	omitBy( {} ); // $ExpectError
	omitBy( {}, 'foo', 2 ); // $ExpectError
}
