// @stdlib/_tools/utils/tex-equation-to-svg/test/test.factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var noop = require( '@stdlib/utils/noop' );
var factory = require( './../lib/factory.js' );


// FIXTURES //

var TEST1 = require( './fixtures/1.js' );
var TEST3 = require( './fixtures/3.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an options argument which is not an object', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		null,
		NaN,
		true,
		false,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	t.throws( foo, TypeError, 'throws type error' );
	t.end();

	function foo() {
		factory({
			'ex': null
		});
	}
});

tape( 'the function returns a function which throws an error if not provided an input string', function test( t ) {
	var tex2svg;
	var values;
	var i;

	values = [
		5,
		null,
		NaN,
		true,
		false,
		void 0,
		[],
		{},
		function noop() {}
	];

	tex2svg = factory( {} );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			tex2svg( value, noop );
		};
	}
});

tape( 'the function returns a function which throws an error if provided a callback argument which is not a function', function test( t ) {
	var tex2svg;
	var values;
	var i;

	values = [
		'5',
		5,
		null,
		NaN,
		true,
		void 0,
		[],
		{}
	];

	tex2svg = factory( {} );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			tex2svg( 'y = mx + b', value );
		};
	}
});

tape( 'the function returns a function which returns any errors to a provided callback', function test( t ) {
	var tex2svg = factory( {} );

	// Non-existent LaTeX command...
	tex2svg( '\\infinity', clbk );

	function clbk( error ) {
		t.ok( error, 'returns an error' );
		t.end();
	}
});

tape( 'the function returns a function which converts an input string to an SVG', function test( t ) {
	var tex2svg = factory( {} );
	tex2svg( '\\operatorname{erf}(x) = \\frac{2}{\\sqrt\\pi}\\int_0^x e^{-t^2}\\,\\mathrm dt.', clbk );

	function clbk( error, svg ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.strictEqual( typeof svg, 'string', 'returns a string' );
		t.strictEqual( svg.substring( 0, 4 ), '<svg', 'svg tag' );
		t.strictEqual( svg.substring( svg.length-6 ), '</svg>', 'closing svg tag' );
		t.strictEqual( svg, TEST1, 'expected value' );
		t.end();
	}
});

tape( 'the function returns a function which converts an input string to an SVG (options)', function test( t ) {
	var tex2svg;
	var opts;

	opts = {
		'width': 200,
		'ex': 2,
		'inline': true,
		'linebreaks': false
	};
	tex2svg = factory( opts );
	tex2svg( '\\operatorname{erf}(x) = \\frac{2}{\\sqrt\\pi}\\int_0^x e^{-t^2}\\,\\mathrm dt.', clbk );

	function clbk( error, svg ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.strictEqual( typeof svg, 'string', 'returns a string' );
		t.strictEqual( svg.substring( 0, 4 ), '<svg', 'svg tag' );
		t.strictEqual( svg.substring( svg.length-6 ), '</svg>', 'closing svg tag' );
		t.strictEqual( svg, TEST3, 'expected value' );
		t.end();
	}
});
