// @stdlib/repl/lib/completer_preview.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable no-restricted-syntax,  no-underscore-dangle, no-invalid-this */

'use strict';

// MODULES //

var readline = require( 'readline' );
var logger = require( 'debug' );
var setNonEnumerableReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
var repeat = require( '@stdlib/string/repeat' );
var longestCommonPrefix = require( './longest_common_prefix.js' );


// VARIABLES //

var debug = logger( 'repl:preview_completer' );


// MAIN //

/**
* Constructor for creating a preview completer.
*
* @private
* @constructor
* @param {Object} rli - readline instance
* @param {Function} completer - function for generating possible completions
* @param {WritableStream} ostream - writable stream
* @returns {PreviewCompleter} completer instance
*/
function PreviewCompleter( rli, completer, ostream ) {
	if ( !(this instanceof PreviewCompleter) ) {
		return new PreviewCompleter( rli, completer, ostream );
	}
	debug( 'Creating a preview completer...' );

	// Cache a reference to the provided readline interface:
	this._rli = rli;

	// Cache a reference to the output writable stream:
	this._ostream = ostream;

	// Cache a reference to the provided completer:
	this._completer = completer;

	// Create a callback for processing potential preview completions:
	this._onCompletions = this._completionCallback();

	// Initialize a buffer containing the currently displayed completion preview:
	this._preview = '';

	return this;
}

/**
* Returns a callback for processing potential preview completions.
*
* @private
* @name _completionCallback
* @memberof PreviewCompleter.prototype
* @returns {Function} completion callback
*/
setNonEnumerableReadOnly( PreviewCompleter.prototype, '_completionCallback', function completionCallback() {
	var self = this;
	return clbk;

	/**
	* Callback invoked upon resolving potential preview completions.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {Array} completions - completion results
	* @returns {void}
	*/
	function clbk( error, completions ) {
		var prefix;
		var list;
		var line;
		var N;

		// Check whether we encountered an error when generating completions...
		if ( error ) {
			debug( 'Encountered an error when generating completions. Unable to display a preview completion...' );
			return;
		}
		list = completions[ 0 ];
		if ( list.length === 0 ) {
			debug( 'No potential preview completion. Unable to display a preview completion...' );
			self.clear();
			return;
		}
		// If multiple completions are possible, do not display a completion preview...
		if ( list.length > 1 ) {
			debug( 'Multiple potential preview completions. Unable to display a preview completion...' );
			self.clear();
			return;
		}
		// Extract the preview completion substring:
		line = completions[ 1 ];
		prefix = longestCommonPrefix( list[ 0 ], line );
		self._preview = list[ 0 ].substring( prefix.length );

		// If the substring is empty, nothing to display...
		if ( self._preview === '' ) {
			debug( 'Exact match. Unable to display a preview completion...' );
			return;
		}
		debug( 'Preview completion: %s', self._preview );

		// Compute the number of character until the end of the line from the current cursor position:
		N = self._rli.line.length - self._rli.cursor;

		// Move the cursor to the end of the line:
		readline.moveCursor( self._ostream, N );

		// Append the preview completion to the current line (using ASCII color escape codes for displaying grey text):
		self._ostream.write( '\x1b[90m' + self._preview + '\x1b[0m' );

		// Move the cursor back to previous position:
		readline.moveCursor( self._ostream, -self._preview.length-N );
	}
});

/**
* Clears a completion preview.
*
* @name clear
* @memberof PreviewCompleter.prototype
* @returns {void}
*/
setNonEnumerableReadOnly( PreviewCompleter.prototype, 'clear', function clear() {
	var preview;
	var N;

	preview = this._preview;

	// If no preview currently displayed, nothing to clear...
	if ( preview === '' ) {
		return;
	}
	debug( 'Clearing completion preview...' );

	// Compute the number of character until the end of the line from the current cursor position:
	N = this._rli.line.length - this._rli.cursor;

	// Move the cursor to the end of the line:
	readline.moveCursor( this._ostream, N );

	// Replace the current display text with whitespace:
	this._ostream.write( repeat( ' ', preview.length ) );

	// Reset the cursor:
	readline.moveCursor( this._ostream, -preview.length-N );

	// Reset the completion preview buffer:
	this._preview = '';
});

/**
* Callback for handling a "keypress" event.
*
* @name onKeypress
* @memberof PreviewCompleter.prototype
* @returns {void}
*/
setNonEnumerableReadOnly( PreviewCompleter.prototype, 'onKeypress', function onKeypress() {
	this._completer( this._rli.line, this._onCompletions );
});

/**
* Callback which should be invoked **before** a "keypress" event is processed by a readline interface.
*
* @name beforeKeypress
* @memberof PreviewCompleter.prototype
* @param {Object} key - key object
* @returns {void}
*/
setNonEnumerableReadOnly( PreviewCompleter.prototype, 'beforeKeypress', function beforeKeypress( key ) {
	if ( !key || this._preview === '' ) {
		return;
	}
	// Handle the case where the user is not at the end of the line...
	if ( this._rli.cursor !== this._rli.line.length ) {
		// If a user is in the middle of a line and presses ENTER, clear the preview string, as the preview was not accepted prior to executing the expression...
		if ( key.name === 'return' || key.name === 'enter' ) {
			return this.clear();
		}
		return;
	}
	// When the user is at the end of the line, auto-complete the line with the preview completion when a user presses RETURN or the RIGHT arrow key (note: pressing ENTER will result in both completion AND execution)...
	if ( key.name === 'return' || key.name === 'enter' || key.name === 'right' ) {
		this._rli.write( this._preview );
		this._preview = '';
	}
});


// EXPORTS //

module.exports = PreviewCompleter;
