// @stdlib/repl/lib/tokenizer.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var parse = require( 'acorn-loose' ).parse;
var walk = require( 'acorn-walk' );
var linkedList = require( '@stdlib/utils/linked-list' );
var contains = require('@stdlib/array/base/assert/contains');
var commands = require( './commands.js' );


// VARIABLES //

var COMMANDS = commands();
var RE_PUNCTUATION = /(\{|\}|\[|\]|\(|\)|,|;|:|\.|\?|\?\.|=>|\.\.\.|`|\${)/;


// MAIN //

/**
* Tokenizes the input line based on ECMAScript specification.
*
* @private
* @name tokenizer
* @type {Function}
* @param {string} line - input line
* @param {Object} context - REPL context
* @returns {Array} array of tokens
*/
function tokenizer( line, context ) {
	var VISITORS;
	var tokens = [];
	var ast;

	// Parse the given line into meaningful tokens...
	ast = parse( line, {
		'ecmaVersion': 'latest',
		'onToken': onToken,
		'onComment': onComment
	});

	/**
	* Callback for handling a `Token` event when parsing.
	*
	* @private
	* @param {Object} token - token object
	*/
	function onToken( token ) {
		// Ignore non-existent tokens:
		if ( token.start === token.end ) {
			return;
		}
		if ( token.type.isLoop || contains( [ 'if', 'else', 'switch', 'case', 'catch', 'finally', 'try' ], token.type.keyword ) ) {
			// Control flow keywords - `for`, `while`, `do`, `if`, `else` etc:
			token.type = 'control';
			tokens.push( token );
			return;
		}
		if ( contains( [ 'this', 'super' ], token.type.keyword ) ) {
			// Special identifiers - `this`, `super`:
			token.type = 'specialIdentifier';
			tokens.push( token );
			return;
		}
		if ( contains( [ 'null', 'true', 'false' ], token.type.keyword ) || ( token.type.label === 'name' && token.value === 'undefined' ) ) {
			// Built-in literals - `true`, `false`, `null`, `undefined`:
			token.type = 'literal';
			tokens.push( token );
			return;
		}
		if ( token.type.keyword || ( token.type.label === 'name' && contains( [ 'async', 'await', 'let' ], token.value ) ) ) {
			// Keywords - `function`, `import`, `var`, `const`, `let` etc.:
			token.type = 'keyword';
			tokens.push( token );
			return;
		}
		if ( token.type.label === 'string' || token.type.label === 'template' ) {
			// Strings and template string literals:
			token.type = 'string';
			tokens.push( token );
			return;
		}
		if ( token.type.label === 'regexp' ) {
			// Regex expressions:
			token.type = 'regexp';
			tokens.push( token );
			return;
		}
		if ( token.type.label === 'num' ) {
			// Numeric literals:
			token.type = 'number';
			tokens.push( token );
			return;
		}
		if ( token.type.binop || token.type.prefix || token.type.postfix || token.type.isAssign ) { // eslint-disable-line max-len
			// Operator symbols - `+`, `=`, `++` etc:
			token.type = 'operator';
			tokens.push( token );
			return;
		}
		if ( RE_PUNCTUATION.test( token.type.label ) ) {
			// Punctuation symbols - `,`, `(`, `;` etc:
			token.value = token.type.label;
			token.type = 'punctuation';
			tokens.push( token );
		}
	}

	/**
	* Callback for handling a `Comment` event when parsing.
	*
	* @private
	* @param {Object} block - true if the comment is a block comment, false if it is a line comment
	* @param {string} text - comment value
	* @param {number} start - start index
	* @param {number} end - end index
	*/
	function onComment( block, text, start, end ) {
		if ( block ) {
			// TODO: add support for highlighting multi-line comments.
			return;
		}
		tokens.push({
			'type': 'comment',
			'start': start,
			'end': end
		});
	}

	// Resolve global identifiers from given line...
	VISITORS = {
		'Identifier': Identifier,
		'MemberExpression': MemberExpression
	};
	walk.simple( ast, VISITORS );

	/**
	* Handles `Identifier` nodes.
	*
	* @private
	* @param {Object} node - `Identifier` node.
	*/
	function Identifier( node ) {
		var identifier;
		var command;
		var i;

		// If identifier is a REPL command, push it as a token...
		for ( i = 0; i < COMMANDS.length; i++ ) {
			command = COMMANDS[ i ];
			if ( node.name === command[ 0 ] ) {
				tokens.push( {
					'value': node.name,
					'type': 'command',
					'start': node.start,
					'end': node.end
				});
				return;
			}
		}
		// If identifier is found in global context, push it as a token...
		identifier = context[ node.name ];
		if ( identifier ) {
			tokens.push( {
				'value': node.name,
				'type': typeof identifier + 'Identifier',
				'start': node.start,
				'end': node.end
			});
		}
	}

	/**
	* Handles `MemberExpression` nodes.
	*
	* @private
	* @param {Object} node - `MemberExpression` node.
	*/
	function MemberExpression( node ) {
		var properties = linkedList();
		var property;
		var obj = context;

		/**
		* Recursively traverses the `MemberExpression` node and populates the `properties` array.
		*
		* @private
		* @param {Object} node - node to be traversed
		*/
		function traverse( node ) {
			if ( node.object.type === 'Identifier' ) {
				// Reached a resolvable MemberExpression (a.b in a.b.c.d), save property(b) & object(a) and exit:
				properties.unshift( node.property );
				properties.unshift( node.object );
			} else if ( node.object.type === 'MemberExpression' ) {
				// Found node with another MemberExpression as object, save property and traverse it:
				properties.unshift( node.property );
				traverse( node.object );
			}
		}
		traverse( node );

		if ( !properties.first() ) {
			// Not an object:
			return;
		}
		// Enter first property's namespace as the first property is already getting tokenized by the `Identifier` handler:
		properties = properties.iterator();
		obj = obj[ properties.next().value.name ];
		if ( !obj ) {
			// Object not defined in context:
			return;
		}
		property = properties.next();
		while ( !property.done ) {
			// Fetch identifier value from context:
			obj = obj[ property.value.name ];
			if ( !obj ) {
				break;
			}
			// Push token if identifier exists in context:
			tokens.push({
				'value': property.value.name,
				'type': typeof obj + 'Identifier',
				'start': property.value.start,
				'end': property.value.end
			});
			property = properties.next();
		}
	}

	return tokens;
}


// EXPORTS //

module.exports = tokenizer;
