<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-doctest

> [ESLint rule][eslint-rules] to ensure that return annotations in example code match the actual output.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest' );
```

#### rule

[ESLint rule][eslint-rules] to ensure that return annotations in example code match the actual output. Return annotations may start with `returns`, `throws`, or `=>`. `returns` follow variable declarations or assignment expressions, whereas `=>` only follow `console.log` calls.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-doctest -->

```javascript
/**
* Squares a number.
*
* @param {number} x - input value
* @returns {number} x*x
*
* @example
* var y = square( 3.0 );
* // return 12.0
*
* y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input value
* @returns {number} x*x
*
* @example
* var y = square( 3.0 );
* // return 9.0
*
* y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '*',
    '* @param {number} x - input value',
    '* @returns {number} x*x',
    '*',
    '* @example',
    '* var y = square( 3.0 );',
    '* // returns 3.0',
    '*',
    '* y = square( 1.0 );',
    '* // returns 2.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-doctest', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-doctest': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-doctest',
            'severity': 2,
            'message': 'Displayed return value is `3.0`, but function returns `9` instead',
            'line': 14,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'endLine': 16,
            'endColumn': 2 },
        {
            'ruleId': 'jsdoc-doctest',
            'severity': 2,
            'message': 'Displayed return value is `2.0`, but function 'returns `1` instead',
            'line': 14,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'endLine': 16,
            'endColumn': 2
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
