/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import arrayShape = require( './index' );


// TESTS //

// The function returns an array of numbers..
{
	arrayShape( [ [ 1, 2, 3 ], [ 4, 5, 6 ], [ 7, 8, 9 ] ] ); // $ExpectType number[]
}

// The compiler throws an error if not provided an array-like object...
{
	arrayShape( 5 ); // $ExpectError
	arrayShape( false ); // $ExpectError
	arrayShape( true ); // $ExpectError
	arrayShape( null ); // $ExpectError
	arrayShape( undefined ); // $ExpectError
	arrayShape( {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	arrayShape(); // $ExpectError
	arrayShape( [ 1, 2, 3 ], 3 ); // $ExpectError
}
