<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# List Item Spacing

> [ESLint rule][eslint-rules] to enforce consistent list item spacing in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-list-item-spacing' );
```

#### rule

[ESLint rule][eslint-rules] to enforce consistent list item spacing in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-list-item-spacing, stdlib/jsdoc-list-item-indent, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep boop.
*
* ## A tight list
*
* -   Wrapped
*     item
* -   item 2
* -   item 3
*
* ## A loose list
*
* -   item 1
*
* -   item 2
*
* -   item 3
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

<!-- eslint-disable stdlib/jsdoc-list-item-indent, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep boop.
*
* ## A tight list
*
* -   item 1
* -   item 2
* -   item 3
*
* ## A loose list
*
* -   Wrapped
*     item
*
* -   item 2
*
* -   item 3
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-list-item-spacing]. 

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-list-item-spacing' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* 1. Item',
    '*',
    '* 2. Item',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-list-item-spacing', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-list-item-spacing': 'error'
    }
});
console.log( result );
/* =>
    [
         {
            'ruleId': 'jsdoc-list-item-spacing',
            'severity': 2,
            'message': 'Extraneous new line after list item',
            'line': 5,
            'column': 3,
            'nodeType': null,
            'source': '*',
            'endLine': 15,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[remark-lint-list-item-spacing]: https://github.com/remarkjs/remark-lint/tree/master/packages/remark-lint-list-item-spacing

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
