/*
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import boxcox1pinv = require( './index' );


// TESTS //

// The function returns a number...
{
	boxcox1pinv( 8, 2 ); // $ExpectType number
}

// The compiler throws an error if the function is provided values other than two numbers...
{
	boxcox1pinv( true, 3 ); // $ExpectError
	boxcox1pinv( false, 2 ); // $ExpectError
	boxcox1pinv( '5', 1 ); // $ExpectError
	boxcox1pinv( [], 1 ); // $ExpectError
	boxcox1pinv( {}, 2 ); // $ExpectError
	boxcox1pinv( ( x: number ): number => x, 2 ); // $ExpectError

	boxcox1pinv( 9, true ); // $ExpectError
	boxcox1pinv( 9, false ); // $ExpectError
	boxcox1pinv( 5, '5' ); // $ExpectError
	boxcox1pinv( 8, [] ); // $ExpectError
	boxcox1pinv( 9, {} ); // $ExpectError
	boxcox1pinv( 8, ( x: number ): number => x ); // $ExpectError

	boxcox1pinv( [], true ); // $ExpectError
	boxcox1pinv( {}, false ); // $ExpectError
	boxcox1pinv( false, '5' ); // $ExpectError
	boxcox1pinv( {}, [] ); // $ExpectError
	boxcox1pinv( '5', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	boxcox1pinv(); // $ExpectError
	boxcox1pinv( 3 ); // $ExpectError
}
