<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-tag-order

> [ESLint rule][eslint-rules] to enforce that JSDoc tags follow a specified order.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-order' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that JSDoc tags follow a specified order. By default, tags must appear in the following order:

-   @private
-   @module
-   @namespace
-   @constructor
-   @constant
-   @function
-   @name
-   @memberof
-   @readonly
-   @type
-   @param
-   @throws
-   @returns
-   @default
-   @see
-   @example

To specify a tag order, set the `order` option.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-tag-order, valid-jsdoc -->

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @returns {number} `x` squared
* @throws {TypeError} must provide a number
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    if ( typeof x !== 'number' ) {
        throw new TypeError( 'invalid argument. Must provide a number. Value: '+x );
    }
    return x * x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @throws {TypeError} must provide number
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    if ( typeof x !== 'number' ) {
        throw new TypeError( 'invalid argument. Must provide a number. Value: '+x );
    }
    return x * x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-order' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '* ',
    '* @returns {number} x squared',
    '* @param {number} x - input number',
    '*',
    '* @example',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-tag-order', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-tag-order': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-tag-order',
            'severity': 2,
            'message': '"@param" tag should not follow "@returns',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'source': '/**',
            'endLine': 10,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
