// @stdlib/random/base/beta/lib/factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var defineProperty = require( '@stdlib/utils/define-property' );
var setReadOnly = require( '@stdlib/utils/define-read-only-property' );
var isObject = require( '@stdlib/assert/is-plain-object' );
var isUint32Array = require( '@stdlib/assert/is-uint32array' );
var minstd = require( '@stdlib/random/base/minstd-shuffle' ).factory;
var randn = require( '@stdlib/random/base/improved-ziggurat' ).factory;
var randu = require( '@stdlib/random/base/mt19937' ).factory;
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var gcopy = require( '@stdlib/blas/base/gcopy' ).ndarray;
var Uint32Array = require( '@stdlib/array/uint32' );
var validate = require( './validate.js' );
var beta0 = require( './beta.js' );


// MAIN //

/**
* Returns a pseudorandom number generator for generating beta distributed random numbers.
*
* @param {PositiveNumber} [alpha] - first shape parameter
* @param {PositiveNumber} [beta] - second shape parameter
* @param {Options} [options] - function options
* @param {*} [options.seed] - pseudorandom number generator seed
* @throws {TypeError} `alpha` must be a positive number
* @throws {TypeError} `beta` must be a positive number
* @throws {TypeError} `options` must be an object
* @returns {Function} pseudorandom number generator
*
* @example
* var beta = factory( 2.0, 1.0 );
* var v = beta();
* // returns <number>
*
* @example
* var beta = factory( 2.0, 2.0, {
*     'seed': 297
* });
* var v = beta();
* // returns <number>
*/
function factory() {
	var randint;
	var rnorm;
	var alpha;
	var beta;
	var opts;
	var rand;
	var prng;
	var err;
	var N;
	var M;

	if ( arguments.length === 0 ) {
		rand = randu();
	} else if ( arguments.length === 1 ) {
		opts = arguments[ 0 ];
		if ( !isObject( opts ) ) {
			throw new TypeError( 'invalid argument. Options argument must be an object. Value: `' + opts + '`.' );
		}
		rand = randu( opts );
	} else {
		alpha = arguments[ 0 ];
		beta = arguments[ 1 ];
		err = validate( alpha, beta );
		if ( err ) {
			throw err;
		}
		if ( arguments.length > 2 ) {
			opts = arguments[ 2 ];
			if ( !isObject( opts ) ) {
				throw new TypeError( 'invalid argument. Options argument must be an object. Value: `' + opts + '`.' );
			}
			rand = randu( opts );
		} else {
			rand = randu();
		}
	}
	rand = rand.normalized;

	randint = minstd({
		'seed': ( rand.SEED%(minstd.MAX+1) ) || 1
	});
	rnorm = randn({
		'seed': randint()
	});

	N = rand.STATE_LENGTH;
	M = rnorm.STATE_LENGTH;

	if ( alpha === void 0 ) {
		prng = beta2;
	} else {
		prng = beta1;
	}
	setReadOnly( prng, 'NAME', 'beta' );
	setReadOnly( prng, 'SEED', rand.SEED );
	defineProperty( prng, 'STATE', {
		'configurable': false,
		'enumerable': true,
		'get': getState,
		'set': setState
	});
	setReadOnly( prng, 'STATE_LENGTH', N+M );
	setReadOnly( prng, 'STATE_SIZE', prng.STATE_LENGTH*4 ); // 4 bytes per uint32
	setReadOnly( prng, 'PRNG', rand );

	return prng;

	/**
	* Returns the current pseudorandom number generator state.
	*
	* @private
	* @returns {Uint32Array} current state
	*/
	function getState() {
		var out;
		var s1;
		var s2;

		s1 = rand.STATE;
		s2 = rnorm.STATE;

		out = new Uint32Array( N+M );
		gcopy( N, s1, 1, 0, out, 1, 0 );
		gcopy( M, s2, 1, 0, out, 1, N );

		return out;
	}

	/**
	* Sets the pseudorandom number generator state.
	*
	* @private
	* @param {Uint32Array} s - generator state
	* @throws {TypeError} must provide a `Uint32Array`
	* @throws {RangeError} must provide a `Uint32Array` having an expected length
	*/
	function setState( s ) {
		var buf;
		var o;
		if ( !isUint32Array( s ) ) {
			throw new TypeError( 'invalid argument. Must provide a Uint32Array. Value: `' + s + '`.' );
		}
		if ( s.length !== N+M ) {
			throw new RangeError( 'invalid argument. The provided array has an unexpected length.' );
		}
		buf = s.buffer;
		o = s.byteOffset;
		rand.STATE = new Uint32Array( buf, o, N );
		rnorm.STATE = new Uint32Array( buf, o+(N*s.BYTES_PER_ELEMENT), M );
	}

	/**
	* Returns a random number drawn from a beta distribution with bound parameter values.
	*
	* @private
	* @returns {Probability} pseudorandom number
	*
	* @example
	* var v = beta1();
	* // returns <number>
	*/
	function beta1() {
		return beta0( rand, rnorm, alpha, beta );
	}

	/**
	* Returns a random number drawn from a beta distribution.
	*
	* @private
	* @param {PositiveNumber} alpha - first shape parameter
	* @param {PositiveNumber} beta - second shape parameter
	* @returns {Probability} pseudorandom number
	*
	* @example
	* var v = beta2( 2.0, 3.0 );
	* // returns <number>
	*/
	function beta2( alpha, beta ) {
		if (
			isnan( alpha ) ||
			isnan( beta ) ||
			alpha <= 0.0 ||
			beta <= 0.0
		) {
			return NaN;
		}
		return beta0( rand, rnorm, alpha, beta );
	}
}


// EXPORTS //

module.exports = factory;
