/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import promotionRules = require( './index' );


// TESTS //

// The function returns an object when not provided any arguments...
{
	promotionRules(); // $ExpectType Table
}

// The function returns a promoted data type promotion rule when provided recognized data types...
{
	promotionRules( 'float32', 'uint32' ); // $ExpectType PromotionRule
	promotionRules( 'int32', 'generic' ); // $ExpectType PromotionRule
}

// The function returns null when provided unrecognized data types...
{
	promotionRules( 'float32', 'foo' ); // $ExpectType null
	promotionRules( 'bar', 'foo' ); // $ExpectType null
	promotionRules( 'bar', 'float32' ); // $ExpectType null
}

// The compiler throws an error if the function is provided a first argument that is not a string...
{
	promotionRules( 123, 'float64' ); // $ExpectError
	promotionRules( true, 'float64' ); // $ExpectError
	promotionRules( false, 'float64' ); // $ExpectError
	promotionRules( null, 'float64' ); // $ExpectError
	promotionRules( {}, 'float64' ); // $ExpectError
	promotionRules( ( x: number ): number => x, 'float64' ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument that is not a string...
{
	promotionRules( 'int32', 123 ); // $ExpectError
	promotionRules( 'int32', true ); // $ExpectError
	promotionRules( 'int32', false ); // $ExpectError
	promotionRules( 'int32', null ); // $ExpectError
	promotionRules( 'int32', {} ); // $ExpectError
	promotionRules( 'int32', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided more than one argument...
{
	promotionRules( 'float32' ); // $ExpectError
	promotionRules( 'float32', 'int32', {} ); // $ExpectError
}
