<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-doctest-decimal-point

> [ESLint rule][eslint-rules] to ensure that only return annotations for real-valued return values always contain decimal points.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest-decimal-point' );
```

#### rule

[ESLint rule][eslint-rules] to ensure that only return annotations for real-valued return values always contain decimal points.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-doctest-decimal-point, stdlib/no-builtin-math -->

```javascript
/**
* Rounds a numeric value to the nearest integer.
*
* @param {number} x - input value
* @returns {integer} function value
*
* @example
* var v = round( -4.2 );
* // returns -4.0
*
* @example
* var v = round( 9.5 );
* // returns 10.0
*
* @example
* var v = round( NaN );
* // returns NaN
*/
var round = Math.round;
```

**Good**:

<!-- eslint-disable stdlib/no-builtin-math -->

```javascript
/**
* Rounds a numeric value to the nearest integer.
*
* @param {number} x - input value
* @returns {integer} function value
*
* @example
* var v = round( -4.2 );
* // returns -4
*
* @example
* var v = round( 9.5 );
* // returns 10
*
* @example
* var v = round( NaN );
* // returns NaN
*/
var round = Math.round;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest-decimal-point' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '*',
    '* @param {number} x - input value',
    '* @returns {number} x*x',
    '*',
    '* @example',
    '* var y = square( 3.0 );',
    '* // returns 9',
    '*',
    '* console.log( square( 2.0 ) );',
    '* // => 4',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-doctest-decimal-point', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-doctest-decimal-point': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-doctest-decimal-point',
            'severity': 2,
            'message: '`// returns 9` should be `// returns 9.0` (return annotations for values of type `number` must always include a decimal point)',
            'line': 14,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'endLine': 16,
            'endColumn': 2
        },
        {
            'ruleId': 'jsdoc-doctest-decimal-point',
            'severity': 2,
            'message': '`// => 4` should be `// => 4.0` (return annotations for values of type `number` must always include a decimal point)',
            'line': 14,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'endLine': 16,
            'endColumn': 2
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
