<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-empty-line-before-example

> [ESLint rule][eslint-rules] to enforce empty lines before JSDoc `@example` tags.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-empty-line-before-example' );
```

#### rule

[ESLint rule][eslint-rules] to enforce empty lines before JSDoc `@example` tags.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-empty-line-before-example -->

```javascript
/**
* Rounds a numeric value toward positive infinity.
*
* @param {number} x - input value
* @returns {number} rounded value
* @example
* var v = ceil( -4.2 );
* // returns -4.0
* @example
* var v = ceil( 9.99999 );
* // returns 10.0
*/
var ceil = require( '@stdlib/math/base/special/ceil' );
```

**Good**:

```javascript
/**
* Rounds a numeric value toward positive infinity.
*
* @param {number} x - input value
* @returns {number} rounded value
*
* @example
* var v = ceil( -4.2 );
* // returns -4.0
*
* @example
* var v = ceil( 9.99999 );
* // returns 10.0
*/
var ceil = require( '@stdlib/math/base/special/ceil' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-empty-line-before-example' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Rounds a numeric value toward negative infinity.',
    '*',
    '* @param {number} x - input value',
    '* @returns {number} rounded value',
    '* @example',
    '* var v = floor( -4.2 );',
    '* // returns -5.0',
    '* @example',
    '* var v = floor( 9.99999 );',
    '* // returns 9.0',
    '* @example',
    '* var v = floor( NaN );',
    '* // returns NaN',
    '*/',
    'var floor = Math.floor;'
].join( '\n' );

linter.defineRule( 'jsdoc-empty-line-before-example', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-empty-line-before-example': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-empty-line-before-example',
            'severity': 2,
            'message': 'Missing empty lines before JSDoc @example tags',
            'line': 16,
            'column': 1,
            'nodeType': 'VariableDeclaration',
            'source': 'var floor = Math.floor;',
            'endLine': 16,
            'endColumn': 24
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
