// @stdlib/math/base/random/binomial/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/utils/env' );
var chi2gof = require( '@stdlib/math/generics/statistics/chi2gof' );
var round = require( '@stdlib/math/base/special/round' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var binomial;
	var r;
	var i;

	// When binding distribution parameters...
	binomial = factory( 80, 0.05 );
	for ( i = 0; i < 100; i++ ) {
		r = binomial();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameters...
	binomial = factory();
	for ( i = 0; i < 100; i++ ) {
		r = binomial( 120, 0.75 );
		t.equal( typeof r, 'number', 'returns a number' );
	}

	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var binomial1;
	var binomial2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	binomial1 = factory( 20, 0.3, {
		'seed': seed
	});
	binomial2 = factory( 20, 0.3, {
		'seed': seed
	});

	t.notEqual( binomial1, binomial2, 'separate generators' );

	for ( i = 0; i < 200; i++ ) {
		r1 = binomial1();
		r2 = binomial2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	seed = round( Date.now() / 1000 );

	binomial1 = factory( 100, 0.9, {
		'seed': seed
	});
	binomial2 = factory( 100, 0.9, {
		'seed': seed
	});

	t.notEqual( binomial1, binomial2, 'separate generators' );

	for ( i = 0; i < 200; i++ ) {
		r1 = binomial1();
		r2 = binomial2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	seed = round( Date.now() / 1000 );

	binomial1 = factory( 900, 0.01, {
		'seed': seed
	});
	binomial2 = factory( 900, 0.01, {
		'seed': seed
	});

	t.notEqual( binomial1, binomial2, 'separate generators' );

	for ( i = 0; i < 200; i++ ) {
		r1 = binomial1();
		r2 = binomial2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}

	seed = round( Date.now() / 1000 );

	binomial1 = factory( 900, 0.5, {
		'seed': seed
	});
	binomial2 = factory( 900, 0.5, {
		'seed': seed
	});

	t.notEqual( binomial1, binomial2, 'separate generators' );

	for ( i = 0; i < 200; i++ ) {
		r1 = binomial1();
		r2 = binomial2();
		t.equal( r1, r2, 'both return same number: '+r1 );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var binomial = factory();
	t.equal( binomial.NAME, 'binomial', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var binomial = factory();
	t.equal( typeof binomial.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var binomial = factory({
		'seed': 12345
	});
	t.equal( typeof binomial.SEED, 'number', 'has `SEED` property' );
	t.equal( binomial.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'the function throws an error if the `n` value is not a positive integer', function test( t ) {
	var values;
	var i;

	values = [
		-1.0,
		-0.5,
		0.0,
		0.5,
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, 0.5 );
		};
	}
});

tape( 'the function throws an error if the `p` value is not a probability', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		-1.0,
		3.14,
		2,
		NaN,
		null,
		true,
		undefined,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 10, value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		NaN,
		null,
		true,
		undefined,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		NaN,
		null,
		true,
		undefined,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 10, 0.3, value );
		};
	}
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a first argument which is `NaN`', function test( t ) {
	var binomial;
	var r;

	binomial = factory();
	r = binomial( NaN, 0.5 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a second argument which is `NaN`', function test( t ) {
	var binomial;
	var r;

	binomial = factory();
	r = binomial( 2, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided `NaNs`', function test( t ) {
	var binomial;
	var r;

	binomial = factory();
	r = binomial( NaN, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a first argument which is not a positive integer', function test( t ) {
	var binomial;
	var r;

	binomial = factory();

	r = binomial( 3.14, 0.5 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = binomial( 0.0, 0.5 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = binomial( -1.0, 0.5 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` if provided a second argument which is not a probability', function test( t ) {
	var binomial;
	var r;

	binomial = factory();

	r = binomial( 2, 3.14 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = binomial( 2, -1.0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = binomial( 2, 1.5 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from a binomial distribution', opts, function test( t ) {
	var threshold;
	var repeats;
	var count;
	var npass;
	var freq;
	var N;
	var n;
	var p;
	var x;

	threshold = 0.15;

	n = 20;
	p = 0.67;

	freq = new Array( n+1 );
	repeats = 100;
	N = 1e3;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var binomial;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < repeats; i++ ) {
			binomial = factory( n, p );
			t.ok( true, 'seed: '+binomial.SEED );

			// Reset the `freq` array...
			for ( j = 0; j < freq.length; j++ ) {
				freq[ j ] = 0;
			}
			for ( j = 0; j < N; j++ ) {
				x = binomial();
				freq[ x ] += 1;
			}
			// Test using chi-square goodness-of-fit test:
			pValue = chi2gof( freq, 'binomial', n, p, {
				'simulate': true,
				'iterations': 500
			}).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample sizes and few repeats...
		bool = ( rejected / repeats < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}
	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Binomial('+n+','+p+') is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+repeats+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});

tape( 'the function returns a PRNG for generating random numbers from a binomial distribution', opts, function test( t ) {
	var threshold;
	var repeats;
	var count;
	var npass;
	var freq;
	var N;
	var n;
	var p;
	var x;

	threshold = 0.15;

	n = 8;
	p = 0.39;

	freq = new Array( n+1 );
	repeats = 100;
	N = 1e3;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var binomial;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < repeats; i++ ) {
			binomial = factory( n, p );
			t.ok( true, 'seed: '+binomial.SEED );

			// Reset the `freq` array...
			for ( j = 0; j < freq.length; j++ ) {
				freq[ j ] = 0;
			}
			for ( j = 0; j < N; j++ ) {
				x = binomial();
				freq[ x ] += 1;
			}
			// Test using chi-square goodness-of-fit test:
			pValue = chi2gof( freq, 'binomial', n, p, {
				'simulate': true,
				'iterations': 500
			}).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample sizes and few repeats...
		bool = ( rejected / repeats < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Binomial('+n+','+p+') is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+repeats+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});
