// @stdlib/random/base/mt19937/test/test.factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/process/env' );
var UINT32_MAX = require( '@stdlib/constants/math/uint32-max' );
var FLOAT64_MAX_SAFE_INTEGER = require('@stdlib/constants/math/float64-max-safe-integer');
var isPositiveInteger = require( '@stdlib/math/base/assert/is-positive-integer' );
var isUint32Array = require( '@stdlib/assert/is-uint32array' );
var kstest = require( '@stdlib/stats/kstest' );
var minstd = require( '@stdlib/random/base/minstd-shuffle' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// FIXTURES //

var UINT32_INTEGER_SEED = require( './fixtures/c/uint32_integer_seed.json' );
var UINT32_ARRAY_SEED = require( './fixtures/c/uint32_array_seed.json' );
var FLOAT64_INTEGER_SEED = require( './fixtures/c/float64_integer_seed.json' );
var FLOAT64_ARRAY_SEED = require( './fixtures/c/float64_array_seed.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'if provided an options argument which is not an object, the factory function throws an error', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		3,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'if provided a `seed` which is not a positive integer or a non-empty array-like object, the factory function throws an error', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		3.14,
		0.0,
		-5.0,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory({
				'seed': value
			});
		};
	}
});

tape( 'the function throws a range error if provided an integer greater than the maximum unsigned 32-bit integer', function test( t ) {
	var values;
	var i;

	values = [
		UINT32_MAX + 1,
		UINT32_MAX + 2,
		UINT32_MAX + 3
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws a range error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory({
				'seed': value
			});
		};
	}
});

tape( 'the function returns a pseudorandom number generator (no options)', function test( t ) {
	var mt19937;
	var v;
	var i;

	mt19937 = factory();
	for ( i = 0; i < 1e3; i++ ) {
		v = mt19937();
		t.equal( typeof v, 'number', 'returns a number' );
		t.equal( isPositiveInteger( v ), true, 'returns a positive integer' );
		t.equal( v >= 1 && v <= UINT32_MAX, true, 'returns an integer between 1 and 2^32-1 (inclusive)' );
	}
	t.end();
});

tape( 'the function returns a pseudorandom number generator (options; no seed)', function test( t ) {
	var mt19937;
	var v;
	var i;

	mt19937 = factory( {} );
	for ( i = 0; i < 1e3; i++ ) {
		v = mt19937();
		t.equal( typeof v, 'number', 'returns a number' );
		t.equal( isPositiveInteger( v ), true, 'returns a positive integer' );
		t.equal( v >= 1 && v <= UINT32_MAX, true, 'returns an integer between 1 and 2^32-1 (inclusive)' );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator (integer seed)', function test( t ) {
	var mt19937a;
	var mt19937b;
	var seed;
	var v1;
	var v2;
	var i;

	seed = minstd();

	mt19937a = factory({
		'seed': seed
	});
	mt19937b = factory({
		'seed': seed
	});

	t.notEqual( mt19937a, mt19937b, 'separate generators' );

	for ( i = 0; i < 1e3; i++ ) {
		v1 = mt19937a();
		v2 = mt19937b();
		t.equal( v1, v2, 'both return same number' );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator (array seed)', function test( t ) {
	var mt19937a;
	var mt19937b;
	var seed;
	var v1;
	var v2;
	var i;

	seed = [ minstd(), minstd(), minstd(), minstd() ];

	mt19937a = factory({
		'seed': seed
	});
	mt19937b = factory({
		'seed': seed
	});

	t.notEqual( mt19937a, mt19937b, 'separate generators' );

	for ( i = 0; i < 1e3; i++ ) {
		v1 = mt19937a();
		v2 = mt19937b();
		t.equal( v1, v2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.NAME, 'mt19937', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed (integer seed)', function test( t ) {
	var mt19937 = factory({
		'seed': 12345
	});
	t.equal( typeof mt19937.SEED, 'number', 'has `SEED` property' );
	t.equal( mt19937.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'attached to the returned function is the generator seed (array seed)', function test( t ) {
	var mt19937;
	var actual;
	var seed;
	var i;

	seed = [ minstd(), minstd(), minstd(), minstd(), minstd() ];
	mt19937 = factory({
		'seed': seed
	});

	actual = mt19937.SEED;
	t.equal( isUint32Array( actual ), true, 'has `SEED` property' );
	for ( i = 0; i < seed.length; i++ ) {
		t.equal( actual[ i ], seed[ i ], 'returns expected value for word '+i );
	}
	t.end();
});

tape( 'attached to the returned function is the minimum possible generated number', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.MIN, 1, 'has `MIN` property' );
	t.end();
});

tape( 'attached to the returned function is the maximum possible generated number', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.MAX, UINT32_MAX, 'has `MAX` property' );
	t.end();
});

tape( 'attached to the returned function is a `normalized` method for generating pseudorandom numbers strictly between 0 (inclusive) and 1 (exclusive)', function test( t ) {
	var mt19937;
	var v;
	var i;

	mt19937 = factory();
	for ( i = 0; i < 1e3; i++ ) {
		v = mt19937.normalized();
		t.equal( typeof v, 'number', 'returns a number' );
		t.equal( v >= 0.0 && v < 1.0, true, 'returns a number between 0 (inclusive) and 1 (exclusive)' );
	}
	t.end();
});

tape( 'attached to the `normalized` method is the generator name', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.normalized.NAME, 'mt19937', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the `normalized` method is the generator seed (integer seed)', function test( t ) {
	var mt19937 = factory({
		'seed': 12345
	});
	t.equal( typeof mt19937.normalized.SEED, 'number', 'has `SEED` property' );
	t.equal( mt19937.normalized.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'attached to the `normalized` method is the generator seed (array seed)', function test( t ) {
	var mt19937;
	var actual;
	var seed;
	var i;

	seed = [ minstd(), minstd(), minstd(), minstd(), minstd() ];
	mt19937 = factory({
		'seed': seed
	});

	actual = mt19937.normalized.SEED;
	t.equal( isUint32Array( actual ), true, 'has `SEED` property' );
	for ( i = 0; i < seed.length; i++ ) {
		t.equal( actual[ i ], seed[ i ], 'returns expected value for word '+i );
	}
	t.end();
});

tape( 'attached to the `normalized` method is the minimum possible generated number', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.normalized.MIN, 0.0, 'has `MIN` property' );
	t.end();
});

tape( 'attached to the `normalized` method is the maximum possible generated number', function test( t ) {
	var mt19937 = factory();
	t.equal( mt19937.normalized.MAX, FLOAT64_MAX_SAFE_INTEGER/(FLOAT64_MAX_SAFE_INTEGER+1), 'has `MAX` property' );
	t.end();
});

tape( 'the `normalized` method returns pseudorandom numbers drawn from a uniform distribution', opts, function test( t ) {
	var threshold;
	var count;
	var npass;
	var N;
	var x;

	threshold = 0.10;

	x = new Array( 1e3 );
	N = 500;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var mt19937;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			mt19937 = factory();
			t.ok( true, 'seed: '+mt19937.SEED );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = mt19937.normalized();
				if ( x[ j ] < 0.0 || x[ j ] > 1.0 ) {
					t.ok( false, 'returned a number outside support: '+x[ j ] );
				}
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'uniform', 0.0, 1.0 ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Uniform(0,1)) is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});

tape( 'the function returns pseudorandom numbers from a Mersenne Twister PRNG (integer seed)', function test( t ) {
	var expected;
	var rand;
	var i;

	expected = UINT32_INTEGER_SEED.expected;
	rand = factory({
		'seed': 1234
	});
	for ( i = 0; i < expected.length; i++ ) {
		t.equal( rand(), expected[ i ], 'returns expected value for iteration '+i );
	}
	t.end();
});

tape( 'the function returns pseudorandom numbers from a Mersenne Twister PRNG (array seed)', function test( t ) {
	var expected;
	var rand;
	var i;

	expected = UINT32_ARRAY_SEED.expected;
	rand = factory({
		'seed': [ 291, 564, 837, 1110 ]
	});
	for ( i = 0; i < expected.length; i++ ) {
		t.equal( rand(), expected[ i ], 'returns expected value for iteration '+i );
	}
	t.end();
});

tape( 'the `normalized` method returns pseudorandom double-precision floating-point numbers from a Mersenne Twister PRNG (integer seed)', function test( t ) {
	var expected;
	var opts;
	var rand;
	var i;

	expected = FLOAT64_INTEGER_SEED.expected;

	opts = {
		'seed': 1234
	};
	rand = factory( opts ).normalized;
	for ( i = 0; i < expected.length; i++ ) {
		t.equal( rand(), expected[ i ], 'returns expected value for iteration '+i );
	}
	t.end();
});

tape( 'the `normalized` method returns pseudorandom double-precision floating-point numbers from a Mersenne Twister PRNG (array seed)', function test( t ) {
	var expected;
	var opts;
	var rand;
	var i;

	expected = FLOAT64_ARRAY_SEED.expected;

	opts = {
		'seed': [ 291, 564, 837, 1110 ]
	};
	rand = factory( opts ).normalized;
	for ( i = 0; i < expected.length; i++ ) {
		t.equal( rand(), expected[ i ], 'returns expected value for iteration '+i );
	}
	t.end();
});
