// @stdlib/lapack/base/dlaqge/test/test.dlaqge.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var abs = require( '@stdlib/math/base/special/abs' );
var EPS = require( '@stdlib/constants/float64/eps' );
var Float64Array = require( '@stdlib/array/float64' );
var dlaqge = require( './../lib/dlaqge.js' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof dlaqge, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 10', function test( t ) {
	t.strictEqual( dlaqge.length, 10, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided an invalid first argument', function test( t ) {
	var values;
	var A;
	var R;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	R = new Float64Array( 4 );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dlaqge( value, 2, 2, A, 2, R, R, 1.0, 1.0, 4.0 );
		};
	}
});

tape( 'the function scales both row and column of a square matrix (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	R = new Float64Array( [ 1.0, 11.0 ] );
	C = new Float64Array( [ 1.0, 11.0 ] );

	out = dlaqge( 'row-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 );
	expected = new Float64Array( [ 1.0, 22.0, 33.0, 484.0 ] );
	t.strictEqual( out, 'B', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales both row and column of a square matrix (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 3.0, 2.0, 4.0 ] );
	R = new Float64Array( [ 1.0, 11.0 ] );
	C = new Float64Array( [ 1.0, 11.0 ] );

	out = dlaqge( 'column-major', 2, 2, A, 2, R, C, 1.0/11.0, 1.0/11.0, 4.0 );
	expected = new Float64Array( [ 1.0, 33.0, 22.0, 484.0 ] );
	t.strictEqual( out, 'B', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales both row and column of a rectangular matrix (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 11.0, 13.0 ] );
	C = new Float64Array( [ 1.0, 11.0, 13.0 ] );

	out = dlaqge( 'row-major', 2, 3, A, 3, R, C, 1.0/13.0, 1.0/13.0, 4.0 );
	expected = new Float64Array( [ 1.0, 22.0, 39.0, 44.0, 605.0, 858.0 ] );
	t.strictEqual( out, 'B', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales both row and column of a rectangular matrix (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 11.0, 13.0 ] );
	C = new Float64Array( [ 1.0, 11.0, 13.0 ] );

	out = dlaqge( 'column-major', 2, 3, A, 3, R, C, 1.0/13.0, 1.0/13.0, 4.0 );
	expected = new Float64Array( [ 1.0, 22.0, 39.0, 44.0, 605.0, 858.0 ] );
	t.strictEqual( out, 'B', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales column of a rectangular matrix (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	C = new Float64Array( [ 1.0, 11.0, 13.0 ] );

	out = dlaqge( 'column-major', 2, 3, A, 3, R, C, 1.0, 1.0/13.0, 4.0 );
	expected = new Float64Array( [ 1.0, 22.0, 39.0, 4.0, 55.0, 78.0 ] );
	t.strictEqual( out, 'C', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales column of a rectangular matrix (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	C = new Float64Array( [ 1.0, 11.0, 13.0 ] );

	out = dlaqge( 'row-major', 2, 3, A, 3, R, C, 1.0, 1.0/13.0, 4.0 );
	expected = new Float64Array( [ 1.0, 22.0, 39.0, 4.0, 55.0, 78.0 ] );
	t.strictEqual( out, 'C', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales row of a rectangular matrix (row-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 11.0, 13.0 ] );
	C = new Float64Array( [ 1.0, 1.0, 1.0 ] );

	out = dlaqge( 'row-major', 2, 3, A, 3, R, C, 1.0/13.0, 1.0, 4.0 );
	expected = new Float64Array( [ 1.0, 2.0, 3.0, 44.0, 55.0, 66.0 ] );
	t.strictEqual( out, 'R', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});

tape( 'the function scales column of a rectangular matrix (column-major)', function test( t ) {
	var expected;
	var out;
	var A;
	var R;
	var C;

	A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
	R = new Float64Array( [ 1.0, 11.0, 13.0 ] );
	C = new Float64Array( [ 1.0, 1.0, 1.0 ] );

	out = dlaqge( 'column-major', 2, 3, A, 3, R, C, 1.0/13.0, 1.0, 4.0 );
	expected = new Float64Array( [ 1.0, 2.0, 3.0, 44.0, 55.0, 66.0 ] );
	t.strictEqual( out, 'R', 'returns expected value' );
	isApprox( t, A, expected, 2.0 );
	t.end();
});
