// @stdlib/utils/map-keys-async/lib/factory.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );
var PINF = require( '@stdlib/math/constants/float64-pinf' );
var validate = require( './validate.js' );
var limit = require( './limit.js' );


// MAIN //

/**
* Returns a function to map keys from one object to a new object having the same values.
*
* ## Notes
*
* * If a provided function calls the provided callback with a truthy error argument, the function suspends execution and immediately calls the `done` callback for subsequent error handling.
* * This function does __not__ guarantee that execution is asynchronous. To do so, wrap the `done` callback in a function which either executes at the end of the current stack (e.g., `nextTick`) or during a subsequent turn of the event loop (e.g., `setIntermediate`, `setTimeout`).
* * Iteration and insertion order are __not__ guaranteed.
* * The function only operates on own properties, not inherited properties.
*
*
* @param {Options} [options] - function options
* @param {*} [options.thisArg] - execution context
* @param {PositiveInteger} [options.limit] - maximum number of pending invocations at any one time
* @param {boolean} [options.series=false] - boolean indicating whether to wait for a previous invocation to complete before invoking a provided function for the next own property
* @param {Function} transform - transform function
* @throws {TypeError} options argument must be an object
* @throws {TypeError} must provide valid options
* @throws {TypeError} last argument must be a function
* @returns {Function} function which maps keys from one object to a new object having the same values
*
* @example
* var readFile = require( '@stdlib/fs/read-file' );
*
* function read( key, value, next ) {
*     var opts = {
*         'encoding': 'utf8'
*     };
*     readFile( value, opts, onFile );
*
*     function onFile( error ) {
*         if ( error ) {
*             return next( null, key+':unreadable' );
*         }
*         next( null, key+':readable' );
*     }
* }
*
* var opts = {
*     'series': true
* };
*
* // Create a reusable function:
* var mapKeysAsync = factory( opts, read );
*
* // Create a dictionary of file names:
* var files = {
*     'file1': './beep.js',
*     'file2': './boop.js'
* };
*
* // Define a callback which handles errors:
* function done( error, out ) {
*     if ( error ) {
*         throw error;
*     }
*     console.log( out );
* }
*
* // Process each file in `files`:
* mapKeysAsync( files, done );
*/
function factory( options, transform ) {
	var opts;
	var err;
	var f;

	opts = {};
	if ( arguments.length > 1 ) {
		err = validate( opts, options );
		if ( err ) {
			throw err;
		}
		f = transform;
	} else {
		f = options;
	}
	if ( !isFunction( f ) ) {
		throw new TypeError( 'invalid input argument. Last argument must be a function. Value: `'+f+'`.' );
	}
	if ( opts.series ) {
		opts.limit = 1;
	} else if ( !opts.limit ) {
		opts.limit = PINF;
	}
	return mapKeysAsync;

	/**
	* Maps keys from one object to a new object having the same values.
	*
	* @private
	* @param {Object} obj - source object
	* @param {Callback} done - function to invoke upon completion
	* @throws {TypeError} first argument must be an object
	* @throws {TypeError} last argument must be a function
	* @returns {void}
	*/
	function mapKeysAsync( obj, done ) {
		if ( typeof obj !== 'object' || obj === null ) {
			throw new TypeError( 'invalid input argument. First argument must be an object. Value: `'+obj+'.`' );
		}
		if ( !isFunction( done ) ) {
			throw new TypeError( 'invalid input argument. Last argument must be a function. Value: `'+done+'`.' );
		}
		return limit( obj, opts, f, clbk );

		/**
		* Callback invoked upon completion.
		*
		* @private
		* @param {*} [error] - error
		* @param {Object} [out] - output object
		* @returns {void}
		*/
		function clbk( error, out ) {
			if ( error ) {
				return done( error );
			}
			done( null, out );
		} // end FUNCTION cblk()
	} // end FUNCTION mapKeysAsync()
} // end FUNCTION factory()


// EXPORTS //

module.exports = factory;
