<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-tag-names

> [ESLint rule][eslint-rules] to enforce that only allowed JSDoc tags are used.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-names' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that only allowed JSDoc tags are used. By default, the following tags are whitelisted:

-   @constant
-   @constructor
-   @customfunction
-   @default
-   @example
-   @function
-   @license
-   @memberof
-   @module
-   @name
-   @namespace
-   @param
-   @private
-   @readonly
-   @returns
-   @see
-   @throws
-   @type

To change the default list of allowed tags, set the `tags` option of the rule.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-tag-names, valid-jsdoc -->

```javascript
/**
* Squares a number.
*
* @arg {number} x - input number
* @return {number} x squared
*
* @examples
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-names' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '* ',
    '* @params {number} x - input number',
    '* @returns {number} x squared',
    '*',
    '* @examples',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-tag-names', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-tag-names': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-tag-names',
            'severity': 2,
            'message': '"params" is not one of the allowed tags (...)',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'source': '/**',
            'endLine': 10,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-tag-names',
            'severity': 2,
            'message': '"examples" is not one of the allowed tags (...)',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'source': '/**',
            'endLine': 10,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
