// @stdlib/utils/papply/lib/papply.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );


// MAIN //

/**
* Returns a function of smaller arity by partially applying arguments.
*
* @param {Function} fcn - function to partially apply
* @param {...*} [args] - arguments to partially apply
* @throws {TypeError} first argument must be a function
* @returns {Function} partially applied function
*
* @example
* function add( x, y ) {
*     return x + y;
* }
*
* var add2 = papply( add, 2 );
*
* var sum = f( 3 );
* // returns 5
*/
function papply( fcn ) {
	var pargs;
	var i;
	if ( !isFunction( fcn ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `' + fcn + '`.' );
	}
	pargs = new Array( arguments.length-1 );
	for ( i = 1; i < arguments.length; i++ ) {
		pargs[ i-1 ] = arguments[ i ];
	}
	return papplied;

	/**
	* Partially applied function.
	*
	* @private
	* @param {...*} [args] - function arguments
	* @returns {*} partially applied function result
	*/
	function papplied() {
		var args;
		var j;
		args = pargs.slice();
		for ( j = 0; j < arguments.length; j++ ) {
			args.push( arguments[ j ] );
		}
		return fcn.apply( null, args );
	} // end FUNCTION papplied()
} // end FUNCTION papply()


// EXPORTS //

module.exports = papply;
