<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-typedef-typos

> [ESLint rule][eslint-rules] to catch typos in type definitions of JSDoc tags.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-typedef-typos' );
```

#### rule

[ESLint rule][eslint-rules] to catch typos in type definitions of JSDoc tags. To change the default list of type definitions to test against, set the `types` option of the rule.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-typedef-typos, valid-jsdoc -->

```javascript
/**
* Squares a number.
*
* @param {numbers} x - input number
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x*x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-typedef-typos' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '* ',
    '* @param {number} x - input number',
    '* @returns {numbr} x squared',
    '*',
    '* @examples',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-typedef-typos', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-typedef-typos': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-typedef-typos',
            'severity': 2,
            'message': '"numbr" should be "number"',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'endLine': 10,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
