/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import iterCartesianSquare = require( './index' );

// TESTS //

// The function returns an iterator...
{
	iterCartesianSquare( [ 'a', 'b', 'c' ] ); // $ExpectType Iterator
	iterCartesianSquare( [ 1, 2, 3 ] ); // $ExpectType Iterator
}

// The compiler throws an error if the function is provided an argument which is not an array-like object...
{
	iterCartesianSquare( 5 ); // $ExpectError
	iterCartesianSquare( true ); // $ExpectError
	iterCartesianSquare( false ); // $ExpectError
	iterCartesianSquare( null ); // $ExpectError
	iterCartesianSquare( {} ); // $ExpectError
	iterCartesianSquare( ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	iterCartesianSquare(); // $ExpectError
}

