// @stdlib/_tools/utils/jsdelivr-url/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var jsdelivr = require( './../lib' );


// FUNCTIONS //

function setup() {
	return {
		'slug': 'stdlib-js/stdlib/develop',
		'file': 'README.md'
	};
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof jsdelivr, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an invalid `options` argument', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		null,
		void 0,
		true,
		false,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			jsdelivr( value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		false,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var opts = setup();
			opts.file = value;
			jsdelivr( opts );
		};
	}
});

tape( 'the function throws an error if not provided a repository slug', function test( t ) {
	t.throws( foo, TypeError, 'throws error' );
	t.end();
	function foo() {
		var opts = setup();
		delete opts.slug;
		jsdelivr( opts );
	}
});

tape( 'the function throws an error if not provided a file path', function test( t ) {
	t.throws( foo, TypeError, 'throws error' );
	t.end();
	function foo() {
		var opts = setup();
		delete opts.file;
		jsdelivr( opts );
	}
});

tape( 'the function returns a jsDelivr URL', function test( t ) {
	var expected;
	var actual;
	var opts;

	opts = setup();

	expected = 'https://cdn.jsdelivr.net/gh/stdlib-js/stdlib@develop/README.md';
	actual = jsdelivr( opts );

	t.strictEqual( actual, expected, 'returns expected value' );
	t.end();
});

tape( 'the function accepts slugs having a dangling `/` and file paths prefixed with `./`', function test( t ) {
	var expected;
	var actual;
	var opts;

	opts = setup();
	opts.slug += '/';
	opts.file = './' + opts.file;

	expected = 'https://cdn.jsdelivr.net/gh/stdlib-js/stdlib@develop/README.md';
	actual = jsdelivr( opts );

	t.strictEqual( actual, expected, 'returns expected value' );
	t.end();
});
