// @stdlib/fs/read-json/test/test.sync.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var join = require( 'path' ).join;
var isBrowser = require( '@stdlib/assert/is-browser' );
var readJSON = require( './../lib/sync.js' );


// FIXME //

// Don't run tests in the browser...for now...
var opts = {
	'skip': isBrowser
};


// FIXTURES //

var goodJSON = join( __dirname, 'fixtures', 'good.json' );
var badJSON = join( __dirname, 'fixtures', 'bad.json.txt' );
var bomJSON = join( __dirname, 'fixtures', 'bom.json.txt' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof readJSON, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an options argument which is neither a string or object', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		null,
		undefined,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			readJSON( goodJSON, value );
		};
	}
});

tape( 'if the function encounters an error when attempting to read a file, the function returns the error', opts, function test( t ) {
	var out = readJSON( 'beepboopbapbop' );

	t.equal( out instanceof Error, true, 'returns an error' );

	t.end();
});

tape( 'if the function encounters an error when attempting to read a file, the function returns the error (options)', opts, function test( t ) {
	var out;

	out = readJSON( 'beepboopbapbop', 'utf8' );
	t.equal( out instanceof Error, true, 'returns an error' );


	out = readJSON( 'beepboopbapbop', {
		'encoding': 'utf8'
	});
	t.equal( out instanceof Error, true, 'returns an error' );

	t.end();
});

tape( 'if the function encounters an error when attempting to parse file contents as JSON, the function returns the error', opts, function test( t ) {
	var out = readJSON( badJSON );
	t.equal( out instanceof Error, true, 'returns an error' );
	t.end();
});

tape( 'if the function encounters an error when attempting to parse file contents as JSON, the function returns the error (options)', opts, function test( t ) {
	var out;

	out = readJSON( badJSON, 'utf8' );
	t.equal( out instanceof Error, true, 'returns an error' );

	out = readJSON( badJSON, {
		'encoding': 'utf8'
	});
	t.equal( out instanceof Error, true, 'returns an error' );

	t.end();
});

tape( 'the function reads a file as JSON', opts, function test( t ) {
	var expected;
	var actual;

	expected = require( goodJSON );
	actual = readJSON( goodJSON );

	t.deepEqual( actual, expected, 'returns a file as JSON' );

	t.end();
});

tape( 'the function reads a file as JSON using provided options', opts, function test( t ) {
	var expected;
	var actual;

	// String options:
	expected = require( goodJSON );
	actual = readJSON( goodJSON, 'utf8' );

	t.deepEqual( expected, actual, 'returns a file as JSON' );

	// Object options:
	actual = readJSON( goodJSON, {
		'encoding': 'utf8'
	});

	t.deepEqual( actual, expected, 'returns a file as JSON' );

	t.end();
});

tape( 'the function supports reading a UTF-8 encoded file having a byte order mark (BOM) as JSON', opts, function test( t ) {
	var expected;
	var actual;

	expected = require( goodJSON );
	actual = readJSON( bomJSON, {
		'encoding': 'utf8'
	});

	t.deepEqual( actual, expected, 'returns a file as JSON' );

	t.end();
});

tape( 'the function supports providing a JSON reviver', opts, function test( t ) {
	var expected;
	var actual;

	expected = {
		'beep': 'boop',
		'bop': [ 1, 2, 3 ],
		'bool': true,
		'bap': 'woot'
	};
	actual = readJSON( goodJSON, {
		'reviver': reviver
	});

	t.deepEqual( actual, expected, 'deep equal' );
	t.end();

	function reviver( key, value ) {
		if ( key === 'bap' ) {
			return 'woot';
		}
		return value;
	}
});
