// @stdlib/utils/is-boolean/test/test.object.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var hasSymbols = require( '@stdlib/utils/detect-symbol-support' )();
var hasToStringTag = require( '@stdlib/utils/detect-tostringtag-support' )();
var isBoolean = require( './../lib/object.js' );


// VARIABLES //

var opts = {
	'skip': !hasSymbols
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof isBoolean, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if provided a boolean object', function test( t ) {
	t.equal( isBoolean( new Boolean( false ) ), true, 'returns true' );
	t.equal( isBoolean( new Boolean( true ) ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if provided a primitive boolean', function test( t ) {
	t.equal( isBoolean( true ), false, 'returns false' );
	t.equal( isBoolean( false ), false, 'returns false' );
	t.end();
});

tape( 'if `Symbol.toStringTag` is supported, the function guards against objects masquerading as `Boolean` objects', opts, function test( t ) {
	var isBoolean;
	var mock;

	isBoolean = proxyquire( './../lib/object.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	mock = {
		'toString': toString,
		'valueOf': valueOf
	};
	if ( hasToStringTag ) {
		mock[ Symbol.toStringTag ] = 'Boolean';
	}
	t.equal( isBoolean( mock ), false, 'returns false' );
	t.end();

	function detect() {
		return true;
	}
	function toString() {
		return 'true';
	}
	function valueOf() {
		return true;
	}
});

tape( 'if `Symbol.toStringTag` is not supported, the function attempts to determine the native class', function test( t ) {
	var isBoolean = proxyquire( './../lib/object.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	t.equal( isBoolean( new Boolean( true ) ), true, 'returns true' );
	t.equal( isBoolean( {} ), false, 'returns false' );

	t.end();

	function detect() {
		return false;
	}
});

tape( 'function returns `false` if not provided a boolean', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		5,
		NaN,
		undefined,
		[],
		{},
		new Date(),
		/./,
		new RegExp( '.' ),
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.equal( isBoolean( values[i] ), false, 'returns false when provided '+values[i] );
	}
	t.end();
});
