// @stdlib/utils/detect-uint32array-support/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var detect = require( './../lib' );


// VARIABLES //

var hasUint32Array = ( typeof Uint32Array === 'function' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof detect, 'function', 'main export is a function' );
	t.end();
});

tape( 'feature detection result is a boolean', function test( t ) {
	t.strictEqual( typeof detect(), 'boolean', 'detection result is a boolean' );
	t.end();
});

tape( 'if `Uint32Array` is supported, detection result is `true`', function test( t ) {
	var mocked;
	if ( hasUint32Array ) {
		t.strictEqual( detect(), true, 'detection result is `true`' );
	} else {
		t.strictEqual( detect(), false, 'detection result is `false`' );
	}
	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': Mock,
		'@stdlib/assert/is-uint32array': isArray
	});
	t.strictEqual( mocked(), true, 'detection result is `true` (mocked)' );

	t.end();

	function isArray() {
		return true;
	}

	function Mock() {
		return [
			1,
			3,
			4294967293,
			0,
			1
		];
	}
});

tape( 'if `Uint32Array` is not supported, detection result is `false`', function test( t ) {
	var mocked;
	if ( hasUint32Array ) {
		t.strictEqual( detect(), true, 'detection result is `true`' );
	} else {
		t.strictEqual( detect(), false, 'detection result is `false`' );
	}
	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': {}
	});
	t.strictEqual( mocked(), false, 'detection result is `false`' );

	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': Mock1
	});
	t.strictEqual( mocked(), false, 'detection result is `false`' );

	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': Mock2,
		'@stdlib/assert/is-uint32array': isArray
	});
	t.strictEqual( mocked(), false, 'detection result is `false`' );

	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': Mock3,
		'@stdlib/assert/is-uint32array': isArray
	});
	t.strictEqual( mocked(), false, 'detection result is `false`' );

	mocked = proxyquire( './../lib/detect_uint32array_support.js', {
		'./uint32array.js': Mock4
	});
	t.strictEqual( mocked(), false, 'detection result is `false`' );

	t.end();

	function isArray() {
		return true;
	}

	function Mock1() {
		// Not a typed array:
		return [];
	}

	function Mock2() {
		// Does not truncate...
		return [ 1, 3.14, -3.14, 0, 1 ];
	}

	function Mock3() {
		// Does not wrap around...
		return [ 1, 3, 4294967293, 4294967296, 4294967297 ];
	}

	function Mock4() {
		throw new Error( 'beep' );
	}
});
