// @stdlib/random/base/arcsine/test/test.factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/process/env' );
var kstest = require( '@stdlib/stats/kstest' );
var round = require( '@stdlib/math/base/special/round' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var isUint32Array = require( '@stdlib/assert/is-uint32array' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var arcsine;
	var r;
	var i;

	// When binding distribution parameters...
	arcsine = factory( -10.0, 10.0 );
	for ( i = 0; i < 100; i++ ) {
		r = arcsine();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameters...
	arcsine = factory();
	for ( i = 0; i < 100; i++ ) {
		r = arcsine( -10.0, 10.0 );
		t.equal( typeof r, 'number', 'returns a number' );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var arcsine1;
	var arcsine2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	arcsine1 = factory( 0.0, 4.0, {
		'seed': seed
	});
	arcsine2 = factory( 0.0, 4.0, {
		'seed': seed
	});

	t.notEqual( arcsine1, arcsine2, 'separate generators' );

	for ( i = 0; i < 100; i++ ) {
		r1 = arcsine1();
		r2 = arcsine2();
		t.equal( r1, r2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var arcsine = factory();
	t.equal( arcsine.NAME, 'arcsine', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var arcsine = factory();
	t.equal( typeof arcsine.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var arcsine = factory({
		'seed': 12345
	});
	t.equal( isUint32Array( arcsine.seed ), true, 'has `seed` property' );
	t.equal( arcsine.seed[ 0 ], 12345, 'equal to provided seed' );
	t.end();
});

tape( 'attached to the returned function is the generator state', function test( t ) {
	var arcsine = factory();
	t.equal( isUint32Array( arcsine.state ), true, 'has `state` property' );
	t.end();
});

tape( 'attached to the returned function is the generator state length', function test( t ) {
	var arcsine = factory();
	t.equal( typeof arcsine.STATE_LENGTH, 'number', 'has `STATE_LENGTH` property' );
	t.end();
});

tape( 'attached to the returned function is the generator state size', function test( t ) {
	var arcsine = factory();
	t.equal( typeof arcsine.STATE_SIZE, 'number', 'has `STATE_SIZE` property' );
	t.end();
});

tape( 'the function throws an error if minimum support `a` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, 2.0 );
		};
	}
});

tape( 'the function throws an error if maximum support `b` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 2.0, value );
		};
	}
});

tape( 'the function throws an error if minimum support `a` is not less than maximum support `b`', function test( t ) {
	var values;
	var i;

	values = [
		[ 0.0, 0.0 ],
		[ -2.0, -4.0 ],
		[ 2.0, 1.0 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( arr ) {
		return function badValue() {
			factory( arr[0], arr[1] );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 0.0, 1.0, value );
		};
	}
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `a` is `NaN`', function test( t ) {
	var arcsine;
	var r;

	arcsine = factory();
	r = arcsine( NaN, 5.0 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `b` is `NaN`', function test( t ) {
	var arcsine;
	var r;

	arcsine = factory();
	r = arcsine( 0.0, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if both `a` and `b` are `NaN`', function test( t ) {
	var arcsine;
	var r;

	arcsine = factory();
	r = arcsine( NaN, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if minimum support `a` is not less than maximum support `b`', function test( t ) {
	var arcsine;
	var values;
	var r;
	var i;

	arcsine = factory();
	values = [
		[ 2.0, 1.0 ],
		[ -3.0, -4.0 ],
		[ 0.0, 0.0 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		r = arcsine( values[ i ][ 0 ], values[ i ][ 1 ] );
		t.strictEqual( isnan( r ), true, 'returns `NaN` when provided '+values[ i ] );
	}
	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from an arcsine distribution', opts, function test( t ) {
	var threshold;
	var count;
	var npass;
	var N;
	var a;
	var b;
	var x;

	threshold = 0.10;

	a = 2.0;
	b = 4.0;

	x = new Array( 1e3 );
	N = 500;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var rejected;
		var arcsine;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			arcsine = factory( a, b );
			t.ok( true, 'seed: '+arcsine.seed );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = arcsine();
				if ( x[ j ] < a || x[ j ] > b ) {
					t.ok( false, 'returned a number outside support: '+x[ j ] );
				}
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'arcsine', a, b ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Arcsine('+a+','+b+')) is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});

tape( 'the returned function supports setting the generator state', function test( t ) {
	var arcsine;
	var state;
	var arr;
	var i;

	arcsine = factory( 2.0, 4.0 );

	// Move to a future state...
	for ( i = 0; i < 100; i++ ) {
		arcsine();
	}
	// Capture the current state:
	state = arcsine.state;

	// Move to a future state...
	arr = [];
	for ( i = 0; i < 100; i++ ) {
		arr.push( arcsine() );
	}
	// Set the state:
	arcsine.state = state;

	// Replay previously generated values...
	for ( i = 0; i < 100; i++ ) {
		t.equal( arcsine(), arr[ i ], 'returns expected value. i: '+i+'.' );
	}
	t.end();
});
