/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import ctors = require( './index' );


// TESTS //

// The function returns a function or null..
{
	ctors( 'float64' ); // $ExpectType Float64ArrayConstructor
	ctors( 'float32' ); // $ExpectType Float32ArrayConstructor
	ctors( 'complex128' ); // $ExpectType Complex128ArrayConstructor
	ctors( 'complex64' ); // $ExpectType Complex64ArrayConstructor
	ctors( 'bool' ); // $ExpectType BooleanArrayConstructor
	ctors( 'int32' ); // $ExpectType Int32ArrayConstructor
	ctors( 'int16' ); // $ExpectType Int16ArrayConstructor
	ctors( 'int8' ); // $ExpectType Int8ArrayConstructor
	ctors( 'uint32' ); // $ExpectType Uint32ArrayConstructor
	ctors( 'uint16' ); // $ExpectType Uint16ArrayConstructor
	ctors( 'uint8' ); // $ExpectType Uint8ArrayConstructor
	ctors( 'uint8c' ); // $ExpectType Uint8ClampedArrayConstructor
	ctors( 'float' ); // $ExpectType Function | null
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	ctors(); // $ExpectError
	ctors( 'int32', 3 ); // $ExpectError
}
