// @stdlib/string/remove-first/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var removeFirst = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof removeFirst, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			removeFirst( value );
		};
	}
});

tape( 'the function returns an empty string if provided an empty string', function test( t ) {
	t.strictEqual( removeFirst( '' ), '', 'returns empty string' );
	t.end();
});

tape( 'the function removes the first character of a given string', function test( t ) {
	var out;

	out = removeFirst( 'hello world' );
	t.strictEqual( out, 'ello world', 'removes first character' );

	out = removeFirst( '!!!' );
	t.strictEqual( out, '!!', 'removes character' );

	out = removeFirst( 'Hello World' );
	t.strictEqual( out, 'ello World', 'removes character' );

	t.end();
});
