/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import str2enum = require( './index' );


// TESTS //

// The function returns a number or null...
{
	str2enum( 'upper' ); // $ExpectType number | null
}

// The compiler throws an error if not provided a string...
{
	str2enum( 10 ); // $ExpectError
	str2enum( true ); // $ExpectError
	str2enum( false ); // $ExpectError
	str2enum( null ); // $ExpectError
	str2enum( undefined ); // $ExpectError
	str2enum( [] ); // $ExpectError
	str2enum( {} ); // $ExpectError
	str2enum( ( x: number ): number => x ); // $ExpectError
}
