<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Linebreak Style

> [ESLint rule][eslint-rules] to enforce that linebreaks follow a given or detected style in JSDoc comments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-linebreak-style' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that linebreaks follow a given or detected style in JSDoc comments. The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-linebreak-style]:

-   'unix' (for `\n`, denoted as `␊`)
-   'windows' (for `\r\n`, denoted as `␍␊`)
-   'consistent' (default)

**Bad**:

<!-- eslint-disable -->

```javascript
/**
* Boop beep.␊
*␊
* ## Boop␊
*␊
* #### Beep␍␊
*␍␊
* @returns {string} a value␍␊
*␍␊
* @example␍␊
* var str = beep();␍␊
* // returns 'boop'␍␊
*/
function beep() {
    return 'boop';
}
```

**Good**:

<!-- eslint-disable -->

```javascript
/**
* Boop beep.␊
*␊
* ## Boop␊
*␊
* ### Beep␊
*␊
* @returns {string} a value␊
*␊
* @example␊
* var str = beep();␊
* // returns 'boop'␊
*/
function beep() {
    return 'boop';
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-linebreak-style' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* -   Beep',
    '* -   Boop',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\r\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-linebreak-style', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-linebreak-style': [ 'error', 'unix' ]
    }
});
console.log( result );
// => [...]
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-linebreak-style]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-linebreak-style

</section>

<!-- /.links -->
