/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import delay = require( './index' );


// TESTS //

// The function returns a timeout object...
{
	delay( () => {}, 2000 ); // $ExpectType TimeoutObject
}

// The compiler throws an error if the function is provided a first argument which is not a function...
{
	delay( 'beep', 2000 ); // $ExpectError
	delay( 5, 2000 ); // $ExpectError
	delay( true, 2000 ); // $ExpectError
	delay( false, 2000 ); // $ExpectError
	delay( null, 2000 ); // $ExpectError

	delay( [], 2000 ); // $ExpectError
	delay( {}, 2000 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	delay( () => {}, 'beep' ); // $ExpectError
	delay( () => {}, true ); // $ExpectError
	delay( () => {}, false ); // $ExpectError
	delay( () => {}, null ); // $ExpectError
	delay( () => {}, [] ); // $ExpectError
	delay( () => {}, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an incorrect number of arguments...
{
	delay(); // $ExpectError
	delay( () => {} ); // $ExpectError
	delay( () => {}, 2000, 'beep' ); // $ExpectError
}


