/*
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import groupIn = require( './index' );

/**
* Indicator function.
*
* @param v - value
* @returns result
*/

const indicator = ( v: string ): string => v[0];

/**
* A class for instantiating objects with `a` and `b` dummy properties.
*/


// TESTS //

// The function returns an object...
{
	groupIn( ['beep', 'boop', 'foo', 'bar'], indicator ); // $ExpectType ValuesResults<string>
	groupIn( [], indicator ); // $ExpectType ValuesResults<string>

	const opts = {
		'returns': 'indices' as 'indices'
	};
	groupIn( ['beep', 'boop', 'foo', 'bar'], opts, indicator ); // $ExpectType IndicesResults
}

// The compiler throws an error if the function is provided a last argument which is not a function...
{
	const arr = ['beep', 'boop', 'foo', 'bar'];
	groupIn( arr, false ); // $ExpectError
	groupIn( arr, true ); // $ExpectError
	groupIn( arr, 32 ); // $ExpectError
	groupIn( arr, 'abc' ); // $ExpectError
	groupIn( arr, [] ); // $ExpectError
	groupIn( arr, {} ); // $ExpectError

	groupIn( arr, {}, false ); // $ExpectError
	groupIn( arr, {}, true ); // $ExpectError
	groupIn( arr, {}, 32 ); // $ExpectError
	groupIn( arr, {}, 'abc' ); // $ExpectError
	groupIn( arr, {}, [] ); // $ExpectError
	groupIn( arr, {}, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an options argument which is not an object...
{
	const arr = ['beep', 'boop', 'foo', 'bar'];
	groupIn( arr, null, indicator ); // $ExpectError
}

// The compiler throws an error if the function is provided a `returns` option which is not one of 'indices', 'values', or '*'...
{
	const arr = ['beep', 'boop', 'foo', 'bar'];
	groupIn( arr, { 'returns': '5' }, indicator ); // $ExpectError
	groupIn( arr, { 'returns': 123 }, indicator ); // $ExpectError
	groupIn( arr, { 'returns': [] }, indicator ); // $ExpectError
	groupIn( arr, { 'returns': {} }, indicator ); // $ExpectError
	groupIn( arr, { 'returns': ( x: number ): number => x }, indicator ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid number of arguments...
{
	const arr = ['beep', 'boop', 'foo', 'bar'];
	groupIn(); // $ExpectError
	groupIn( arr ); // $ExpectError
	groupIn( arr, {}, indicator, 16 ); // $ExpectError
}
