// @stdlib/math/base/dist/lognormal/ctor/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var isFunction = require( '@stdlib/assert/is-function' );
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var quantile = require( '@stdlib/math/base/dist/lognormal/quantile' );
var cdf = require( '@stdlib/math/base/dist/lognormal/cdf' );
var pdf = require( '@stdlib/math/base/dist/lognormal/pdf' );
var kurtosis = require( '@stdlib/math/base/dist/lognormal/kurtosis' );
var skewness = require( '@stdlib/math/base/dist/lognormal/skewness' );
var variance = require( '@stdlib/math/base/dist/lognormal/variance' );
var entropy = require( '@stdlib/math/base/dist/lognormal/entropy' );
var median = require( '@stdlib/math/base/dist/lognormal/median' );
var mode = require( '@stdlib/math/base/dist/lognormal/mode' );
var mean = require( '@stdlib/math/base/dist/lognormal/mean' );
var LogNormal = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof LogNormal, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided a `sigma` argument which is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		NaN,
		true,
		false,
		undefined,
		null,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			// eslint-disable-next-line no-new
			new LogNormal( value, 1.0 );
		};
	}
});

tape( 'the function throws an error if provided a `sigma` argument which is not a positive number', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		-5.0,
		0.0,
		NaN,
		true,
		false,
		undefined,
		null,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			// eslint-disable-next-line no-new
			new LogNormal( 2.0, value );
		};
	}
});

tape( 'if provided arguments, the function requires both `mu` and `sigma`', function test( t ) {
	t.throws( foo, TypeError, 'throws an error' );
	t.end();

	function foo() {
		// eslint-disable-next-line no-new
		new LogNormal( 2.0 );
	}
});

tape( 'the function returns a new distribution instance (default parameters)', function test( t ) {
	var lognormal = new LogNormal();
	t.strictEqual( lognormal instanceof LogNormal, true, 'returns an instance');
	t.end();
});

tape( 'the function returns a new distribution instance (custom parameters)', function test( t ) {
	var lognormal = new LogNormal( 2.0, 4.0 );
	t.strictEqual( lognormal instanceof LogNormal, true, 'returns an instance');
	t.end();
});

tape( 'the function can be invoked without the new operator', function test( t ) {
	// eslint-disable-next-line new-cap
	var lognormal = LogNormal();
	t.strictEqual( lognormal instanceof LogNormal, true, 'returns an instance');

	// eslint-disable-next-line new-cap
	lognormal = LogNormal( 2.0, 4.0 );
	t.strictEqual( lognormal instanceof LogNormal, true, 'returns an instance');

	t.end();
});

tape( 'the created distribution has a property for getting and setting `mu`', function test( t ) {
	var lognormal;

	lognormal = new LogNormal( 2.0, 4.0 );
	t.strictEqual( hasOwnProp( lognormal, 'mu' ), true, 'has property' );
	t.strictEqual( lognormal.mu, 2.0, 'returns expected value' );

	lognormal.mu = 3.0;
	t.strictEqual( lognormal.mu, 3.0, 'returns expected value' );

	t.end();
});

tape( 'the created distribution throws an error if one attempts to set `mu` to a value which is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		NaN,
		true,
		false,
		undefined,
		null,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var lognormal = new LogNormal();
			lognormal.mu = value;
		};
	}
});

tape( 'the created distribution has a property for getting and setting `sigma`', function test( t ) {
	var lognormal;

	lognormal = new LogNormal( 2.0, 4.0 );
	t.strictEqual( hasOwnProp( lognormal, 'sigma' ), true, 'has property' );
	t.strictEqual( lognormal.sigma, 4.0, 'returns expected value' );

	lognormal.sigma = 3.0;
	t.strictEqual( lognormal.sigma, 3.0, 'returns expected value' );

	t.end();
});

tape( 'the created distribution throws an error if one attempts to set `sigma` to a value which is not a positive number', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		-5.0,
		0.0,
		NaN,
		true,
		false,
		undefined,
		null,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var lognormal = new LogNormal();
			lognormal.sigma = value;
		};
	}
});

tape( 'the distribution prototype has a property for retrieving the distribution entropy', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'entropy' ), true, 'has property' );

	lognormal = new LogNormal();
	t.strictEqual( lognormal.entropy, entropy( 0.0, 1.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution kurtosis', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'kurtosis' ), true, 'has property' );

	lognormal = new LogNormal( 2.0, 4.0 );
	t.strictEqual( lognormal.kurtosis, kurtosis( 2.0, 4.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution mean', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'mean' ), true, 'has property' );

	lognormal = new LogNormal();
	t.strictEqual( lognormal.mean, mean( 0.0, 1.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution median', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'median' ), true, 'has property' );

	lognormal = new LogNormal( 5.0, 2.0 );
	t.strictEqual( lognormal.median, median( 5.0, 2.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution mode', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'mode' ), true, 'has property' );

	lognormal = new LogNormal( 2.0, 3.0 );
	t.strictEqual( lognormal.mode, mode( 2.0, 3.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution skewness', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'skewness' ), true, 'has property' );

	lognormal = new LogNormal( 0.5, 0.5 );
	t.strictEqual( lognormal.skewness, skewness( 0.5, 0.5 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a property for retrieving the distribution variance', function test( t ) {
	var lognormal;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'variance' ), true, 'has property' );

	lognormal = new LogNormal( 3.0, 1.0 );
	t.strictEqual( lognormal.variance, variance( 3.0, 1.0 ), 'returns expected value' );

	t.end();
});

tape( 'the distribution prototype has a method for evaluating the cumulative distribution function (CDF)', function test( t ) {
	var lognormal;
	var y;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'cdf' ), true, 'has property' );
	t.strictEqual( isFunction( LogNormal.prototype.cdf ), true, 'has method' );

	lognormal = new LogNormal();
	y = lognormal.cdf( 0.5 );

	t.strictEqual( y, cdf( 0.5, 0.0, 1.0 ), 'returns expected value' );
	t.end();
});

tape( 'the distribution prototype has a method for evaluating the probability density function (PDF)', function test( t ) {
	var lognormal;
	var y;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'pdf' ), true, 'has property' );
	t.strictEqual( isFunction( LogNormal.prototype.pdf ), true, 'has method' );

	lognormal = new LogNormal();
	y = lognormal.pdf( 0.2 );

	t.strictEqual( y, pdf( 0.2, 0.0, 1.0 ), 'returns expected value' );
	t.end();
});

tape( 'the distribution prototype has a method for evaluating the quantile function', function test( t ) {
	var lognormal;
	var y;

	t.strictEqual( hasOwnProp( LogNormal.prototype, 'quantile' ), true, 'has property' );
	t.strictEqual( isFunction( LogNormal.prototype.quantile ), true, 'has method' );

	lognormal = new LogNormal();
	y = lognormal.quantile( 0.8 );

	t.strictEqual( y, quantile( 0.8, 0.0, 1.0 ), 'returns expected value' );
	t.end();
});
