<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# uppercase-required-constants

> [ESLint rule][eslint-rules] enforcing that required `stdlib` constants have uppercase variable names.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/uppercase-required-constants' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that required `stdlib` constants have uppercase variable names.

**Bad**:

<!-- eslint-disable stdlib/uppercase-required-constants -->

```javascript
var float64SqrtEpsilon = require( '@stdlib/constants/float64/sqrt-eps' );
```

**Good**:

```javascript
var FLOAT64_SQRT_EPSILON = require( '@stdlib/constants/float64/sqrt-eps' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/uppercase-required-constants' );

var linter = new Linter();
var result;

var code = 'var pi = require( \'@stdlib/constants/float64/pi\' );';

linter.defineRule( 'uppercase-required-constants', rule );

result = linter.verify( code, {
    'rules': {
        'uppercase-required-constants': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'uppercase-required-constants',
            'severity': 2,
            'message': 'stdlib constants must be uppercase',
            'line': 1,
            'column': 1,
            'nodeType': 'VariableDeclaration',
            'source': 'var pi = require( \'@stdlib/constants/float64/pi\' );',
            'endLine': 1,
            'endColumn': 57
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
