<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# vars-order

> [ESLint rule][eslint-rules] enforcing variable declarations inside of functions to be ordered by length.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/vars-order' );
```

#### rule

[ESLint rule][eslint-rules] enforcing variable declarations inside of functions to be ordered by length.

**Bad**:

<!-- eslint-disable stdlib/vars-order -->

```javascript
/**
* Returns a pseudorandom number drawn from a triangular distribution with minimum support `a`, maximum support `b` and mode `c`.
*
* @private
* @param {Function} rand - PRNG for generating uniformly distributed numbers
* @param {number} a - minimum support
* @param {number} b - maximum support
* @param {number} c - mode
* @returns {number} pseudorandom number
*/
function triangular( rand, a, b, c ) {
    var x;
    var fc;
    var u;
    fc = (c - a) / (b - a);
    u = rand();
    if ( u < fc ) {
        x = (b - a) * (c - a);
        return a + sqrt( x * u );
    }
    x = (b - a) * (b - c);
    return b - sqrt( x * (1.0 - u) );
}
```

**Good**:

```javascript
/**
* Returns a pseudorandom number drawn from a triangular distribution with minimum support `a`, maximum support `b` and mode `c`.
*
* @private
* @param {Function} rand - PRNG for generating uniformly distributed numbers
* @param {number} a - minimum support
* @param {number} b - maximum support
* @param {number} c - mode
* @returns {number} pseudorandom number
*/
function triangular( rand, a, b, c ) {
    var fc;
    var x;
    var u;
    fc = (c - a) / (b - a);
    u = rand();
    if ( u < fc ) {
        x = (b - a) * (c - a);
        return a + sqrt( x * u );
    }
    x = (b - a) * (b - c);
    return b - sqrt( x * (1.0 - u) );
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/vars-order' );

var linter = new Linter();
var result;

var code = [
    'function fizzBuzz() {',
    '  var i;',
    '  var out;',
    '',
    '  for ( i = 1; i <= 100; i++ ) {',
    '    out = ( i % 5 === 0 ) ? "Buzz" : ( i % 3 === 0 ) ? "Fizz" : i;',
    '    console.log( out );',
    '  }',
    '}'
].join( '\n' );

linter.defineRule( 'vars-order', rule );

result = linter.verify( code, {
    'rules': {
        'vars-order': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'vars-order',
            'severity': 2,
            'message': 'Variable declarations inside of functions should by ordered by length (in decreasing order)',
            'line': 1,
            'column': 1,
            'nodeType': 'ExpressionStatement',
            'source': 'function fizzBuzz() {',
            'endLine': 9,
            'endColumn': 2
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
