// @stdlib/repl/ctor/lib/create_context.js
/**
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable no-underscore-dangle */

'use strict';

// MODULES //

var Module = require( 'module' );
var vm = require( 'vm' );
var Console = require( 'console' ).Console;
var logger = require( 'debug' );
var isString = require( '@stdlib/assert/is-string' ).isPrimitive;
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var getGlobal = require( '@stdlib/utils/global' );
var objectKeys = require( '@stdlib/utils/keys' );
var defineProperty = require( '@stdlib/utils/define-property' );
var setNonEnumerable = require( '@stdlib/utils/define-nonenumerable-property' );
var setNonEnumerableReadOnly = require( '@stdlib/utils/define-nonenumerable-read-only-property' );
var setNonEnumerableReadOnlyAccessor = require( '@stdlib/utils/define-nonenumerable-read-only-accessor' ); // eslint-disable-line id-length
var setConfigurableReadOnly = require( './define_configurable_read_only_property.js' );
var setConfigurableReadOnlyAccessor = require( './define_configurable_read_only_accessor.js' ); // eslint-disable-line id-length
var createRequire = require( './create_require.js' );
var HELP_TEXT = require( './help_text.js' );
var GLOBALS = require( './globals.js' );


// VARIABLES //

var debug = logger( 'repl:context' );

// Cache references to module methods (WARNING: we use private methods exposed on the `Module` object; thus, breakage is possible!):
var resolveLookupPaths = Module._resolveLookupPaths;


// MAIN //

/**
* Creates a REPL context.
*
* @private
* @param {REPL} repl - REPL instance
* @returns {Object} REPL context
*/
function createContext( repl ) {
	var context;
	var Global;
	var keys;
	var cmds;
	var cmd;
	var i;

	debug( 'Creating REPL execution context...' );

	Global = getGlobal();
	if ( repl._sandbox ) {
		// Create a sandboxed context:
		context = vm.createContext();

		// Assign globals from the current global context to the sandboxed context (note: shallow copy!)...
		keys = objectKeys( GLOBALS );
		for ( i = 0; i < keys.length; i++ ) {
			setNonEnumerable( context, keys[ i ], GLOBALS[ keys[ i ] ] );
		}

		// Create a circular reference as in Node.js:
		context.global = context;

		// Create a new `console` interface:
		setNonEnumerable( context, 'console', new Console( repl._ostream ) );

		// Create a new `module` object:
		setNonEnumerable( context, 'module', new Module( '<repl>' ) );

		// Create a new `require` function:
		setNonEnumerable( context, 'require', createRequire( context.module ) );
	} else {
		context = Global;

		// Create a new `module` object:
		defineProperty( context, 'module', {
			'configurable': true,
			'writable': false,
			'enumerable': false,
			'value': new Module( '<repl>' )
		});

		// Create a new `require` function:
		defineProperty( context, 'require', {
			'configurable': true,
			'writable': false,
			'enumerable': false,
			'value': createRequire( context.module )
		});
	}
	context.module.paths = resolveLookupPaths( '<repl>', module, true ) || [];

	// TODO: lazy load project APIs (configurable, enumerable(?) accessors which overwrite themselves to become read-only properties; see https://github.com/nodejs/node/blob/0c1e93b9efadfc9fae74907a631908477c7d085e/lib/internal/modules/cjs/helpers.js#L102)

	// Define a list of REPL-specific commands (NOTE: keep in alphabetical order):
	cmds = [
		// [ <command>, <function>, <is_accessor> ]
		[ 'ans', onAns, true ],
		[ 'clear', onClear, false ],
		[ 'help', onHelp, false ],
		[ 'load', onLoad, false ],
		[ 'quit', onQuit, false ]
	];

	// TODO: add `example` command

	// TODO: add `namespace` command

	// TODO: add `save` command

	// TODO: add `saveStart` command

	// TODO: add `saveStop` command

	// TODO: add `citation` command

	// TODO: add `license` command

	// TODO: add `copyright` command

	// TODO: add `contributors` command

	// TODO: add `credits` command

	// TODO: add `donate` command

	// Add commands requiring privileged access to internal instance variables...
	if ( repl._sandbox ) {
		for ( i = 0; i < cmds.length; i++ ) {
			cmd = cmds[ i ];
			if ( cmd[ 2 ] ) {
				setNonEnumerableReadOnlyAccessor( context, cmd[ 0 ], cmd[ 1 ] );
			} else {
				setNonEnumerableReadOnly( context, cmd[ 0 ], cmd[ 1 ] );
			}
		}
	} else {
		for ( i = 0; i < cmds.length; i++ ) {
			cmd = cmds[ i ];

			// Avoid overwriting properties in a non-sandboxed environment in order to avoid unintended side-effects...
			if ( !hasOwnProp( context, cmd[ 0 ] ) ) {
				if ( cmd[ 2 ] ) {
					setConfigurableReadOnlyAccessor( context, cmd[ 0 ], cmd[ 1 ] ); // eslint-disable-line max-len
				} else {
					setConfigurableReadOnly( context, cmd[ 0 ], cmd[ 1 ] );
				}
			}
		}
	}
	return context;

	/**
	* Returns the result of the last successfully executed command.
	*
	* @private
	* @returns {*} result
	*/
	function onAns() {
		return repl._ans;
	}

	/**
	* Clears a REPL's local context.
	*
	* @private
	*/
	function onClear() {
		repl.clear();
	}

	/**
	* Prints help text.
	*
	* @private
	* @param {*} [alias] - variable alias or value
	* @returns {void}
	*/
	function onHelp( alias ) {
		if ( arguments.length === 0 ) {
			repl._ostream.write( HELP_TEXT );
			return;
		}
		// TODO: add support for printing help text; first look through project namespace and then REPL-specific namespace
		console.log( String( alias ) );
	}

	/**
	* Loads and evaluates a JavaScript file line-by-line.
	*
	* @private
	* @param {string} fpath - file path
	* @returns {void}
	*/
	function onLoad( fpath ) {
		var err;
		if ( !isString( fpath ) ) {
			err = new TypeError( 'invalid argument. Must provide a string. Value: `'+fpath+'`.' );
			debug( 'Error: %s', err.message );
			repl._ostream.write( 'Error: '+err.message+'\n' );
			return;
		}
		debug( 'Loading a file...' );
		repl.load( fpath );
	}

	/**
	* Quits a REPL.
	*
	* @private
	*/
	function onQuit() {
		debug( 'Quitting the REPL...' );
		repl.close();
	}
}


// EXPORTS //

module.exports = createContext;
