// @stdlib/assert/is-regexp/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var hasSymbols = require( '@stdlib/utils/detect-symbol-support' )();
var hasToStringTag = require( '@stdlib/utils/detect-tostringtag-support' )();
var isRegExp = require( './../lib' );


// VARIABLES //

var opts = {
	'skip': !hasSymbols
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof isRegExp, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if provided a regular expression', function test( t ) {
	t.strictEqual( isRegExp( /\.+/ ), true, 'returns true' );
	t.strictEqual( isRegExp( new RegExp( '\\.+' ) ), true, 'returns true' );
	t.end();
});

tape( 'if `Symbol.toStringTag` is supported, the function guards against objects masquerading as regular expressions', opts, function test( t ) {
	var isRegExp;
	var mock;

	isRegExp = proxyquire( './../lib/is_regexp.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	mock = {
		'toString': toString,
		'exec': exec
	};
	if ( hasToStringTag ) {
		mock[ Symbol.toStringTag ] = 'RegExp';
	}
	t.strictEqual( isRegExp( mock ), false, 'returns false' );

	t.end();

	function detect() {
		return true;
	}
	function toString() {
		return '/beep/';
	}
	function exec() {
		return null;
	}
});

tape( 'if `Symbol.toStringTag` is supported, the function returns `true` if provided a regular expression', opts, function test( t ) {
	var isRegExp = proxyquire( './../lib/is_regexp.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	t.strictEqual( isRegExp( /beep/ ), true, 'returns true' );

	t.end();

	function detect() {
		return true;
	}
});

tape( 'if `Symbol.toStringTag` is not supported, the function attempts to determine the native class', function test( t ) {
	var isRegExp = proxyquire( './../lib/is_regexp.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	t.strictEqual( isRegExp( /beep/ ), true, 'returns true' );
	t.strictEqual( isRegExp( {} ), false, 'returns false' );

	t.end();

	function detect() {
		return false;
	}
});

tape( 'the function returns `false` if not provided a regular expression', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		'/.+/',
		5,
		NaN,
		null,
		undefined,
		true,
		false,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.strictEqual( isRegExp( values[i] ), false, 'returns false when provided '+values[i] );
	}
	t.end();
});
