<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-tag-spacing

> [ESLint rule][eslint-rules] to enforce a padding style for JSDoc tags.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-spacing' );
```

#### rule

[ESLint rule][eslint-rules] to enforce a padding style for JSDoc tags.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-tag-spacing, valid-jsdoc -->

```javascript
/**
* Squares a number.
*
*   @param{number} x - input number
* @throws { TypeError }must provide a number
*@returns   {number} `x` squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x * x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @throws {TypeError} must provide number
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*/
function square( x ) {
    return x * x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-tag-spacing' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '* ',
    '* @param { number } x - input number',
    '* @returns{number} x squared',
    '*',
    '* @example',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-tag-spacing', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-tag-spacing': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-tag-spacing',
            'severity': 2,
            'message': 'there should be no spaces inside of JSDoc type definitions',
            'line': 4,
            'column': 1,
            'nodeType': null,
            'source': '* @param { number } x - input number',
            'endLine': 4,
            'endColumn': 37
        },
        {
            'ruleId': 'jsdoc-tag-spacing',
            'severity': 2,
            'message': '"@returns" tag should be separated from opening curly brace by one space',
            'line': 5,
            'column': 1,
            'nodeType': null,
            'source': '* @returns{number} x squared',
            'endLine': 5,
            'endColumn': 29
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
