// @stdlib/utils/reduce-right-async/lib/reduce_right.js
'use strict';

// MODULES //

var factory = require( './factory.js' );


// MAIN //

/**
* Applies a function against an accumulator and each element in a collection and return the accumulated result, iterating from right to left.
*
* ## Notes
*
* * If a provided function calls the provided callback with a truthy error argument, the function suspends execution and immediately calls the `done` callback for subsequent error handling.
* * This function does __not__ guarantee that execution is asynchronous. To do so, wrap the `done` callback in a function which either executes at the end of the current stack (e.g., `nextTick`) or during a subsequent turn of the event loop (e.g., `setIntermediate`, `setTimeout`).
*
*
* @param {Collection} collection - input collection
* @param {*} initial - initial value
* @param {Options} [options] - function options
* @param {*} [options.thisArg] - execution context
* @param {PositiveInteger} [options.limit] - maximum number of pending invocations at any one time
* @param {boolean} [options.series=true] - boolean indicating whether to wait for a previous invocation to complete before invoking a provided function for the next element in a collection
* @param {Function} fcn - function to invoke for each element in a collection
* @param {Callback} done - function to invoke upon completion
* @throws {TypeError} first argument must be a collection
* @throws {TypeError} options argument must be an object
* @throws {TypeError} must provide valid options
* @throws {TypeError} second-to-last argument must be a function
* @throws {TypeError} last argument must be a function
* @returns {void}
*
* @example
* var readFile = require( '@stdlib/fs/read-file' );
*
* function done( error, acc ) {
*     if ( error ) {
*         throw error;
*     }
*     console.log( acc.count );
* }
*
* function read( acc, file, next ) {
*     var opts = {
*         'encoding': 'utf8'
*     };
*     readFile( file, opts, onFile );
*
*     function onFile( error ) {
*         if ( error ) {
*             return next( null, acc );
*         }
*         acc.count += 1;
*         next( null, acc );
*     }
* }
*
* var files = [
*     './beep.js',
*     './boop.js'
* ];
* var acc = {
*     'count': 0
* };
* reduceRightAsync( files, acc, read, done );
*/
function reduceRightAsync( collection, initial, options, fcn, done ) {
	if ( arguments.length < 5 ) {
		return factory( options )( collection, initial, fcn );
	}
	factory( options, fcn )( collection, initial, done );
} // end FUNCTION reduceRightAsync()


// EXPORTS //

module.exports = reduceRightAsync;
