<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Table Cell Padding

> [ESLint rule][eslint-rules] to require Markdown table cells in JSDoc descriptions to be correctly padded.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-table-cell-padding' );
```

#### rule

[ESLint rule][eslint-rules] to require Markdown table cells in JSDoc descriptions to be correctly padded.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-table-cell-padding, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* | x  | y   |
* |any |boop |
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Boop beep.
*
* | x   | y    |
* | any | boop |
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-table-cell-padding]. 

-   `padded`: require table cells to be padded with spaces (default).
-   `consistent`:  detects the first used cell padding style and requires subsequent cells to use same style.
-   `compact`: require table cells to be next to the pipes without any padding.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-table-cell-padding' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Squares a number.',
    '*',
    '* | Input | Output |',
    '* | ----- | ------ |',
    '* |2      | 4      |',
    '*',
    '* @param {number} x - input value',
    '* @returns {number} output value',
    '*',
    '* @example',
    '* var out = square( 2 );',
    '* // returns 4',
    '*/',
    'function square( x ) {',
    '\treturn x*x;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-table-cell-padding', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-table-cell-padding': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-table-cell-padding',
            'severity': 2,
            'message': 'Cell should be padded',
            'line': 6,
            'column': 4,
            'nodeType': null,
            'source': '* |2      | 4      |',
            'endLine': 15,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-table-cell-padding]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-table-cell-padding

</section>

<!-- /.links -->
