// @stdlib/utils/group-by/lib/return_pairs.js
'use strict';

// MODULES //

var hasOwnProp = require( '@stdlib/assert/has-own-property' );


// MAIN //

/**
* Groups values according to an indicator function and outputs results as index-element values.
*
* ## Notes
*
* * We need to cache the collection value to prevent the edge case where, during the invocation of the indicator function, the element at index `i` is swapped for some other value. For some, that might be a feature; here, we take the stance that one should be less clever.
* * Checking for an "own" property is necessary to guard against the edge case where an indicator function returns a group identifier which matches a method or property on the `Object` prototype.
*
*
* @private
* @param {Collection} collection - collection to group
* @param {Options} opts - function options
* @param {*} [opts.thisArg] - execution context
* @param {Function} indicator - indicator function specifying which group an element in the input collection belongs to
* @returns {Object} group results
*
* @example
* function indicator( v ) {
*     return v[ 0 ];
* }
* var arr = [ 'beep', 'boop', 'foo', 'bar' ];
*
* var out = groupBy( arr, {}, indicator );
* // returns { 'b': [ [ 0, 'beep' ], [ 1, 'boop' ], [ 3, 'bar' ] ], 'f': [ [ 2, 'foo' ] ] }
*/
function groupBy( collection, opts, indicator ) {
	var thisArg;
	var out;
	var len;
	var g;
	var v;
	var i;

	thisArg = opts.thisArg;
	len = collection.length;

	out = {};
	for ( i = 0; i < len; i++ ) {
		v = collection[ i ];
		g = indicator.call( thisArg, v, i );
		if ( hasOwnProp( out, g ) ) {
			out[ g ].push( [ i, v ] );
		} else {
			out[ g ] = [ [ i, v ] ];
		}
	}
	return out;
} // end FUNCTION groupBy()


// EXPORTS //

module.exports = groupBy;
