// @stdlib/assert/tools/typed-array-function/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var isEven = require( '@stdlib/math/base/assert/is-even' );
var noop = require( '@stdlib/utils/noop' );
var typedarrayfcn = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof typedarrayfcn, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a predicate function', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		undefined,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws TypeError' );
	}
	function badValue( value ) {
		return function badValue() {
			typedarrayfcn( value );
		};
	}
	t.end();
});

tape( 'the function returns a function', function test( t ) {
	var f = typedarrayfcn( isEven );
	t.strictEqual( typeof f, 'function' );

	f = typedarrayfcn( noop );
	t.strictEqual( typeof f, 'function' );
	t.end();
});

tape( 'the function returns a function that tests a typed array', function test( t ) {
	var bool;
	var arr;
	var f;

	f = typedarrayfcn( isEven );

	arr = new Int32Array( [ 2, 2, 2, 2 ] );
	bool = f( arr );

	t.strictEqual( bool, true );

	arr = new Int32Array( [ 2, 2, 2, 3 ] );
	bool = f( arr );

	t.strictEqual( bool, false );

	t.end();
});

tape( 'the function returns a function which returns `false` if not supplied a typed array', function test( t ) {
	var values;
	var f;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		undefined,
		{},
		[],
		[ 3, 3, 3, 3 ],
		function noop() {}
	];

	f = typedarrayfcn( isEven );

	for ( i = 0; i < values.length; i++ ) {
		t.strictEqual( f( values[ i ] ), false, 'returns false' );
	}
	t.end();
});

tape( 'the function returns a function which returns `false` if provided an empty typed array', function test( t ) {
	var bool;
	var f;

	f = typedarrayfcn( isEven );
	bool = f( new Int32Array() );

	t.strictEqual( bool, false );

	t.end();
});
