<!--

@license Apache-2.0

Copyright (c) 2021 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# namespace-export-all

> [ESLint rule][eslint-rules] to enforce that a namespace `index.js` exports all packages in the respective namespace directory.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/namespace-export-all' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that a namespace `index.js` exports all packages in the respective namespace directory. The rule is applied when a file contains a comment with the following content: 

> `When adding modules to the namespace, ensure that they are added in alphabetical order according to module name.`.

Assuming a `@stdlib/constants/array` folder with `max-array-length` and `max-typed-array-length` package directories, the rule will validate as follows:

**Bad**:

```javascript
/*
* When adding modules to the namespace, ensure that they are added in alphabetical order according to module name.
*/

// MODULES //

var setReadOnly = require( '@stdlib/utils/define-read-only-property' );


// MAIN //

var ns = {};

/**
* @name MAX_ARRAY_LENGTH
* @memberof ns
* @readonly
* @constant
* @type {number}
* @see {@link module:@stdlib/constants/array/max-array-length}
*/
setReadOnly( ns, 'MAX_ARRAY_LENGTH', require( '@stdlib/constants/array/max-array-length' ) );
```

**Good**:

```javascript
/*
* When adding modules to the namespace, ensure that they are added in alphabetical order according to module name.
*/

// MODULES //

var setReadOnly = require( '@stdlib/utils/define-read-only-property' );


// MAIN //

var ns = {};

/**
* @name MAX_ARRAY_LENGTH
* @memberof ns
* @readonly
* @constant
* @type {number}
* @see {@link module:@stdlib/constants/array/max-array-length}
*/
setReadOnly( ns, 'MAX_ARRAY_LENGTH', require( '@stdlib/constants/array/max-array-length' ) );

/**
* @name MAX_TYPED_ARRAY_LENGTH
* @memberof ns
* @readonly
* @constant
* @type {number}
* @see {@link module:@stdlib/constants/array/max-typed-array-length}
*/
setReadOnly( ns, 'MAX_TYPED_ARRAY_LENGTH', require( '@stdlib/constants/array/max-typed-array-length' ) );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var join = require( 'path' ).join;
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/namespace-export-all' );

var linter = new Linter();
var result;
var config;
var code;

code = [
    '/*',
    '* When adding modules to the namespace, ensure that they are added in alphabetical order according to module name.',
    '*/',
    '',
    '// MODULES //',
    '',
    'var setReadOnly = require( \'@stdlib/utils/define-read-only-property\' );',
    '',
    '',
    '// MAIN //',
    '',
    '/**',
    '* Top-level namespace.',
    '*',
    '* @namespace ns',
    '*/',
    'var ns = {};',
    '',
    '/**',
    '* @name error2json',
    '* @memberof ns',
    '* @readonly',
    '* @type {Function}',
    '* @see {@link module:@stdlib/error/to-json}',
    '*/',
    'setReadOnly( ns, \'error2json\', require( \'@stdlib/error/to-json\' ) );'
].join( '\n' );

linter.defineRule( 'namespace-export-all', rule );

config = {
    'rules': {
        'namespace-export-all': 'error'
    }
};
result = linter.verify( code, config, {
    'filename': join( __dirname, 'fixtures', 'lib', 'index.js' )
});
console.log( result );
/* =>
    {
        'ruleId': 'namespace-export-all',
        'severity': 2,
        'message': '`reviver` is not exported from `index.js`',
        'line': 7,
        'column': 1,
        'nodeType': 'Program',
        'endLine': 26,
        'endColumn': 69
    }
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
