/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dlaset = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 5, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( true, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( false, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( null, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( void 0, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( [], 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( {}, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( ( x: number ): number => x, 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 5, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', true, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', false, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', null, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', void 0, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', [], 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', {}, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', ( x: number ): number => x, 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', '5', 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', true, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', false, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', null, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', void 0, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', [], 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', {}, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', ( x: number ): number => x, 3, 3.1, 3.7, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', 3, '5', 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, true, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, false, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, null, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, void 0, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, [], 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, {}, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, ( x: number ): number => x, 3.1, 3.7, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', 3, 3, '5', 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, true, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, false, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, null, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, void 0, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, [], 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, {}, 3.7, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, ( x: number ): number => x, 3.7, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', 3, 3, 3.1, '5', A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, true, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, false, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, null, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, void 0, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, [], A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, {}, A, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, ( x: number ): number => x, A, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a Float64Array...
{

	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, '5', 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, 5, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, true, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, false, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, null, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, void 0, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, [], 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, {}, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, ( x: number ): number => x, 3 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, '5' ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, true ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, false ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, null ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, void 0 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, [] ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, {} ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset(); // $ExpectError
	dlaset( 'row-major' ); // $ExpectError
	dlaset( 'row-major', 'lower' ); // $ExpectError
	dlaset( 'row-major', 'lower', 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7 ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A ); // $ExpectError
	dlaset( 'row-major', 'lower', 3, 3, 3.1, 3.7, A, 3, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 5, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( true, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( false, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( null, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( void 0, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( [], 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( {}, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( ( x: number ): number => x, 3, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', '5', 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', true, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', false, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', null, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', void 0, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', [], 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', {}, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', ( x: number ): number => x, 3, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, '5', 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, true, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, false, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, null, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, void 0, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, [], 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, {}, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, ( x: number ): number => x, 3.1, 3.7, A, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, '5', 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, true, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, false, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, null, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, void 0, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, [], 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, {}, 3.7, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, ( x: number ): number => x, 3.7, A, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, 3.1, '5', A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, true, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, false, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, null, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, void 0, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, [], A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, {}, A, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, ( x: number ): number => x, A, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float64Array...
{

	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, '5', 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, 5, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, true, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, false, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, null, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, void 0, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, [], 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, {}, 3, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, ( x: number ): number => x, 3, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, '5', 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, true, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, false, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, null, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, void 0, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, [], 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, {}, 1, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, '5', 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, true, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, false, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, null, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, void 0, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, [], 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, {}, 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, '5' ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, true ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, false ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, null ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, void 0 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, [] ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, {} ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );

	dlaset.ndarray(); // $ExpectError
	dlaset.ndarray( 'lower' ); // $ExpectError
	dlaset.ndarray( 'lower', 3 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1 ); // $ExpectError
	dlaset.ndarray( 'lower', 3, 3, 3.1, 3.7, A, 3, 1, 0, 10 ); // $ExpectError
}
