// @stdlib/string/replace/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var replace = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof replace, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the `str` value is not a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){},
		/ /g
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			replace( value, '', '' );
		};
	}
});

tape( 'the function throws an error if the `search` value is not a string or regular expression', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			replace( 'string', value, '' );
		};
	}
});

tape( 'the function throws an error if the `newval` value is not a function or string primitive', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			replace( 'string', 'str', value );
		};
	}
});

tape( 'the function replaces all occurrences of a string search value', function test( t ) {
	var out;

	out = replace( 'abc abc abc', 'b', '' );
	t.equal( out, 'ac ac ac', 'replaces all occurrences of letter b' );

	out = replace( 'abc abc abc', 'b', 'cd' );
	t.equal( out, 'acdc acdc acdc', 'replaces all occurrences of letter b by cd' );

	t.end();
});

tape( 'the function replaces all matches for a regular expression', function test( t ) {
	var expected;
	var out;

	out = replace( 'aBcDeFgHiJkLmNoPqRsTuVwXYZ', /[A-Z]+/g, '' );
	expected = 'acegikmoqsuw';
	t.equal( out, expected, 'replaces letters matching the regular expression' );
	t.end();
});

tape( 'the function replaces found matches by values created from a replacer function', function test( t ) {
	var expected;
	var out;
	var str;

	str = 'Oranges and lemons say the bells of St. Clement\'s';
	out = replace( str, /([^\s]+)/gi, function replacer( match, p1 ) {
		return '/' + p1 + '/';
	});
	expected = '/Oranges/ /and/ /lemons/ /say/ /the/ /bells/ /of/ /St./ /Clement\'s/';
	t.equal( out, expected, 'replaces all matches by values from replacer function' );
	t.end();
});
