// @stdlib/math/base/blas/dcopy/test/test.dcopy.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var floor = require( '@stdlib/math/base/special/floor' );
var dcopy = require( './../lib/dcopy.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof dcopy, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 5', function test( t ) {
	t.equal( dcopy.length, 5, 'arity of 5' );
	t.end();
});

tape( 'the function copies elements from `x` into `y`', function test( t ) {
	var x = [ 1, 2, 3, 4, 5 ];
	var y = [ 6, 7, 8, 9, 10 ];

	dcopy( x.length, x, 1, y, 1 );

	t.deepEqual( y, x, 'deep equal' );
	t.notEqual( y, x, 'different references' );

	t.end();
});

tape( 'the function supports an `x` stride', function test( t ) {
	var expected;
	var x;
	var y;
	var N;

	x = [ 1, 2, 3, 4, 5 ];
	y = [ 6, 7, 8, 9, 10 ];
	N = 3;

	dcopy( N, x, 2, y, 1 );

	expected = [ 1, 3, 5, 9, 10 ];

	t.deepEqual( y, expected, 'deep equal' );
	t.end();
});

tape( 'the function supports a `y` stride', function test( t ) {
	var expected;
	var x;
	var y;
	var N;

	x = [ 1, 2, 3, 4, 5 ];
	y = [ 6, 7, 8, 9, 10 ];
	N = 3;

	dcopy( N, x, 1, y, 2 );

	expected = [ 1, 7, 2, 9, 3 ];

	t.deepEqual( y, expected, 'deep equal' );
	t.end();
});

tape( 'the function returns a reference to the destination array', function test( t ) {
	var x = [ 1, 2, 3, 4, 5 ];
	var y = [ 6, 7, 8, 9, 10 ];

	var out = dcopy( x.length, x, 1, y, 1 );

	t.equal( out, y, 'same reference' );
	t.end();
});

tape( 'if provided an `N` parameter less than or equal to 0, the function returns `y` unchanged', function test( t ) {
	var expected;
	var x;
	var y;

	x = [ 1, 2, 3, 4, 5 ];
	y = [ 6, 7, 8, 9, 10 ];

	expected = [ 6, 7, 8, 9, 10 ];

	dcopy( -1, x, 1, y, 1 );
	t.deepEqual( y, expected, 'returns `y` unchanged' );

	dcopy( 0, x, 1, y, 1 );
	t.deepEqual( y, expected, 'returns `y` unchanged' );

	t.end();
});

tape( 'the function supports negative strides', function test( t ) {
	var expected;
	var x;
	var y;
	var N;

	x = [ 1, 2, 3, 4, 5 ];
	y = [ 6, 7, 8, 9, 10 ];
	N = 3;

	dcopy( N, x, -2, y, -1 );

	expected = [ 1, 3, 5, 9, 10 ];

	t.deepEqual( y, expected, 'deep equal' );
	t.end();
});

tape( 'the function supports complex access patterns', function test( t ) {
	var expected;
	var x;
	var y;
	var N;

	x = [ 1, 2, 3, 4, 5, 6 ];
	y = [ 7, 8, 9, 10, 11, 12 ];
	N = 3;

	dcopy( N, x, 2, y, -1 );

	expected = [ 5, 3, 1, 10, 11, 12 ];

	t.deepEqual( y, expected, 'deep equal' );
	t.end();
});

tape( 'the function supports view offsets', function test( t ) {
	var expected;
	var x0;
	var y0;
	var x1;
	var y1;
	var N;

	// Initial arrays...
	x0 = new Float64Array( [ 1, 2, 3, 4, 5, 6 ] );
	y0 = new Float64Array( [ 7, 8, 9, 10, 11, 12 ] );

	// Create offset views...
	x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // begin at 2nd element
	y1 = new Float64Array( y0.buffer, y0.BYTES_PER_ELEMENT*3 ); // begin at the 4th element

	N = floor( x0.length / 2 );

	dcopy( N, x1, -2, y1, 1 );
	expected = new Float64Array( [ 7, 8, 9, 6, 4, 2 ] );

	t.deepEqual( y0, expected, 'deep equal' );
	t.end();
});

tape( 'if both strides are equal to `1`, the function efficiently copies elements from `x` into `y`', function test( t ) {
	var x;
	var y;
	var i;

	x = new Array( 100 );
	y = new Array( x.length );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = i;
		y[ i ] = x.length - i;
	}

	dcopy( x.length, x, 1, y, 1 );

	t.deepEqual( y, x, 'deep equal' );
	t.notEqual( y, x, 'different references' );

	x = new Array( 120 );
	y = new Array( x.length );
	for ( i = 0; i < x.length; i++ ) {
		x[ i ] = i*2;
		y[ i ] = x.length - i;
	}

	dcopy( x.length, x, 1, y, 1 );

	t.deepEqual( y, x, 'deep equal' );
	t.notEqual( y, x, 'different references' );

	t.end();
});
