// @stdlib/ndarray/ctor/test/test.instance.set_no_codegen_3d.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var ctor = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof ctor, 'function', 'main export is a function' );
	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (3d; codegen=false)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 2 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		if ( dim === 1 ) {
			return j;
		}
		return k;

		function i() {
			arr.set( value, 0, 0, 10.0 );
		}

		function j() {
			arr.set( 0, value, 0, 10.0 );
		}

		function k() {
			arr.set( 0, 0, value, 10.0 );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 0, 0, 0, 5.0 );
	arr.set( 0, 0, 1, 6.0 );
	arr.set( 0, 1, 0, 7.0 );
	arr.set( 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 0, 0, 0, 5.0 );
	arr.set( 0, 0, 1, 6.0 );
	arr.set( 0, 1, 0, 7.0 );
	arr.set( 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'wrap',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'wrap',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; mode=clamp; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'clamp',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 8.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 6.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 2.0, 3.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; mode=clamp; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'clamp',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 8.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 6.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 2.0, 3.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; mode=throw; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'throw',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; mode=throw; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'throw',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; mode=default; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; mode=default; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; submode=[wrap]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'wrap', 'wrap' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; submode=[wrap,wrap]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'wrap', 'wrap' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; submode=[clamp,clamp,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'clamp', 'clamp', 'clamp' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 8.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 6.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 2.0, 3.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; submode=[clamp,clamp,clamp,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'clamp', 'clamp', 'clamp', 'clamp' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );
	arr.set( 0, 10, 5, 6.0 );
	arr.set( 0, -1, -2, 7.0 );
	arr.set( 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 8.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 6.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 2.0, 3.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; submode=[throw]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'submode': [ 'throw' ],
		'mode': 'clamp',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; submode=[throw,throw,throw,throw]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'submode': [ 'throw', 'throw', 'throw', 'throw' ],
		'mode': 'clamp',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		[ 2, 2, 2, 5.0 ],
		[ -4, -2, 11, 6.0 ],
		[ 0, 5, 0, 7.0 ],
		[ 0, 15, -1, 8.0 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 4.0, 'returns expected value' );

	t.deepEqual( buffer, [ 1.0, 2.0, 3.0, 4.0 ], 'has expected values' );

	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major; submode=[wrap,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'wrap', 'clamp' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );    // (0,1,0)
	arr.set( 0, 10, 5, 6.0 );   // (0,1,1)
	arr.set( 0, -1, -2, 7.0 );  // (0,0,0)
	arr.set( 0, -5, -11, 8.0 ); // (0,0,1)

	t.strictEqual( arr.get( 0, 0, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 8.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 6.0, 'returns expected value' );

	t.deepEqual( buffer, [ 7.0, 8.0, 5.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major; submode=[clamp,wrap,wrap]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'clamp', 'wrap', 'wrap' ],
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	arr.set( 2, 2, 2, 5.0 );    // (0,0,0)
	arr.set( 0, 10, 5, 6.0 );   // (0,0,1)
	arr.set( 0, -1, -2, 7.0 );  // (0,1,0)
	arr.set( 0, -5, -11, 8.0 ); // (0,1,1)

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});
