// @stdlib/ndarray/ctor/test/test.instance.get_no_codegen_3d.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var ctor = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof ctor, 'function', 'main export is a function' );
	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method which throws an error if not provided an integer value (3d; codegen=false)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 2 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		if ( dim === 1 ) {
			return j;
		}
		return k;

		function i() {
			arr.get( value, 0, 0 );
		}

		function j() {
			arr.get( 0, value, 0 );
		}

		function k() {
			arr.get( 0, 0, value );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 0, 0 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'mode': 'wrap'
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'mode': 'wrap'
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; mode=clamp; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'mode': 'clamp'
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; mode=clamp; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'mode': 'clamp'
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; mode=throw; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'throw',
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; mode=throw; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'throw',
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; mode=default; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; mode=default; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; submode=[wrap]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'submode': [ 'wrap' ]
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; submode=[wrap,wrap]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'submode': [ 'wrap', 'wrap' ]
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 2.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; submode=[clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'submode': [ 'clamp' ]
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; submode=[clamp,clamp,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false,
		'submode': [ 'clamp', 'clamp', 'clamp' ]
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; submode=[throw,throw]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'wrap',
		'submode': [ 'throw', 'throw' ],
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'row-major';
	strides = [ 3, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; submode=[throw,throw,throw]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var values;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'mode': 'clamp',
		'submode': [ 'throw', 'throw', 'throw' ],
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0 ] );
	shape = [ 1, 3, 1 ];
	order = 'column-major';
	strides = [ 1, 1, 3 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	values = [
		[ 2, 6, 100 ],
		[ 0, -2, -4 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided arguments: ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.get.apply( arr, value );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; row-major; submode=[wrap,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'wrap', 'clamp' ],
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 3.0, 'returns expected value' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `get` method for retrieving an array element using subscripts (3d; column-major; submode=[wrap,wrap,clamp]; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'submode': [ 'wrap', 'wrap', 'clamp' ],
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( arr.get( 2, 6, 100 ), 3.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, -2, -4 ), 1.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 2, 0 ), 1.0, 'returns expected value' );

	t.end();
});
