<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-require-absolute-path

> [ESLint rule][eslint-rules] disallowing `require()` statements containing absolute file paths.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-require-absolute-path' );
```

#### rule

[ESLint rule][eslint-rules] disallowing `require()` statements containing absolute file paths.

**Bad**:

<!-- run-disable -->

<!-- eslint-disable stdlib/no-require-absolute-path -->

```javascript
var stdlib = require( '/path/to/stdlib' );
```

**Good**:

```javascript
var stdlib = require( '@stdlib' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-require-absolute-path' );

var linter = new Linter();
var result;

var code = 'var stdlib = require( \'/path/to/stdlib/\' );';

linter.defineRule( 'no-require-absolute-path', rule );

result = linter.verify( code, {
    'rules': {
        'no-require-absolute-path': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-require-absolute-path',
            'severity': 2,
            'message': 'require() call contains an absolute file path',
            'line': 1,
            'column': 14,
            'nodeType': 'CallExpression',
            'source': 'var stdlib = require( \'/path/to/stdlib/\' );',
            'endLine': 1,
            'endColumn': 43
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
