// @stdlib/math/base/random/discrete-uniform/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/utils/env' );
var round = require( '@stdlib/math/base/special/round' );
var chi2gof = require( '@stdlib/math/stats/chi2gof' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var isInteger = require( '@stdlib/assert/is-integer' ).isPrimitive;
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var discreteUniform;
	var r;
	var i;

	// When binding distribution parameters...
	discreteUniform = factory( -100, 100 );
	for ( i = 0; i < 100; i++ ) {
		r = discreteUniform();
		t.equal( typeof r, 'number', 'returns a number' );
		t.equal( isInteger( r ), true, 'returns an integer' );
	}

	// Without binding distribution parameters...
	discreteUniform = factory();
	for ( i = 0; i < 100; i++ ) {
		r = discreteUniform( -100, 100 );
		t.equal( typeof r, 'number', 'returns a number' );
		t.equal( isInteger( r ), true, 'returns an integer' );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var discreteUniform1;
	var discreteUniform2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	discreteUniform1 = factory( 0, 40, {
		'seed': seed
	});
	discreteUniform2 = factory( 0, 40, {
		'seed': seed
	});

	t.notEqual( discreteUniform1, discreteUniform2, 'separate generators' );

	for ( i = 0; i < 100; i++ ) {
		r1 = discreteUniform1();
		r2 = discreteUniform2();
		t.equal( r1, r2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var discreteUniform = factory();
	t.equal( discreteUniform.NAME, 'discrete-uniform', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var discreteUniform = factory();
	t.equal( typeof discreteUniform.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var discreteUniform = factory({
		'seed': 12345
	});
	t.equal( typeof discreteUniform.SEED, 'number', 'has `SEED` property' );
	t.equal( discreteUniform.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'the function throws an error if minimum support `a` is not an integer primitive', function test( t ) {
	var values;
	var i;

	values = [
		3.14,
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, 20 );
		};
	}
});

tape( 'the function throws an error if maximum support `b` is not an integer primitive', function test( t ) {
	var values;
	var i;

	values = [
		30.14,
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 20, value );
		};
	}
});

tape( 'the function throws an error if minimum support `a` is not less than or equal to maximum support `b`', function test( t ) {
	var values;
	var i;

	values = [
		[ 0, -1 ],
		[ -20, -40 ],
		[ 20, 10 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( arr ) {
		return function badValue() {
			factory( arr[0], arr[1] );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 0, 10, value );
		};
	}
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `a` is `NaN`', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( NaN, 50 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `b` is `NaN`', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( 0, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if both `a` and `b` are `NaN`', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( NaN, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `a` is not an integer value', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( 3.14, 50 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `b` is not an integer value', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( 0, 3.14 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if both `a` and `b` are not integer values', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();
	r = discreteUniform( -3.14, 3.14 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if minimum support `a` is not less than or equal to maximum support `b`', function test( t ) {
	var discreteUniform;
	var values;
	var r;
	var i;

	discreteUniform = factory();
	values = [
		[ 20, 10 ],
		[ -30, -40 ],
		[ 0, -1 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		r = discreteUniform( values[ i ][ 0 ], values[ i ][ 1 ] );
		t.strictEqual( isnan( r ), true, 'returns `NaN` when provided '+values[ i ] );
	}
	t.end();
});

tape( 'if provided a maximum support equal to the minimum support, the function returns a function that returns the minimum support', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory( 12, 12 );

	r = discreteUniform();
	t.strictEqual( r, 12, 'returns minimum support' );

	r = discreteUniform();
	t.strictEqual( r, 12, 'returns minimum support' );

	r = discreteUniform();
	t.strictEqual( r, 12, 'returns minimum support' );

	discreteUniform = factory( -10, -10 );

	r = discreteUniform();
	t.strictEqual( r, -10, 'returns minimum support' );

	r = discreteUniform();
	t.strictEqual( r, -10, 'returns minimum support' );

	r = discreteUniform();
	t.strictEqual( r, -10, 'returns minimum support' );

	t.end();
});

tape( 'if provided a maximum support equal to the minimum support, the function returns a function that returns the minimum support', function test( t ) {
	var discreteUniform;
	var r;

	discreteUniform = factory();

	r = discreteUniform( 12, 12 );
	t.strictEqual( r, 12, 'returns minimum support' );

	r = discreteUniform( 2, 2 );
	t.strictEqual( r, 2, 'returns minimum support' );

	r = discreteUniform( -4, -4 );
	t.strictEqual( r, -4, 'returns minimum support' );

	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from a discrete uniform distribution', opts, function test( t ) {
	var threshold;
	var repeats;
	var count;
	var npass;
	var freq;
	var N;
	var a;
	var b;
	var x;

	threshold = 0.15;

	a = -10;
	b = 10;

	freq = new Array( b - a + 1 );
	repeats = 100;
	N = 1e3;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var discreteUniform;
		var rejected;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < repeats; i++ ) {
			discreteUniform = factory( a, b );
			t.ok( true, 'seed: '+discreteUniform.SEED );

			// Reset the `freq` array...
			for ( j = 0; j < freq.length; j++ ) {
				freq[ j ] = 0;
			}
			for ( j = 0; j < N; j++ ) {
				x = discreteUniform();
				freq[ x-a ] += 1;
			}
			// Test using chi-square goodness-of-fit test:
			pValue = chi2gof( freq, 'discrete-uniform', a, b, {
				'simulate': true,
				'iterations': 500
			}).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample sizes and few repeats...
		bool = ( rejected / repeats < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from DiscreteUniform('+a+','+b+') is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+repeats+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});
