// @stdlib/_tools/tests/bundle/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var mkdir = require( 'fs' ).mkdirSync;
var rmdir = require( 'fs' ).rmdirSync;
var tape = require( 'tape' );
var readFile = require( '@stdlib/fs/read-file' ).sync;
var unlink = require( '@stdlib/fs/unlink' ).sync;
var noop = require( '@stdlib/utils/noop' );
var evil = require( '@stdlib/utils/eval' );
var bundle = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof bundle, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the first argument is not a string primitive', function test( t ) {
	var values;
	var i;
	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			bundle( value, noop );
		};
	}
});

tape( 'the function throws an error if the last argument is not a callback function', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var dir = join( __dirname, 'fixtures' );
			bundle( dir, value );
		};
	}
});

tape( 'the function throws an error if the last argument is not a callback function (options)', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var dir = join( __dirname, 'fixtures' );
			bundle( dir, {}, value );
		};
	}
});

tape( 'the function generates a bundle', function test( t ) {
	var dir = join( __dirname, 'fixtures', 'main' );
	bundle( dir, onBundle );

	function onBundle( error, output ) {
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		if ( output === false ) {
			t.fail( 'failed to generate a bundle' );
		} else {
			try {
				evil( output.toString() );
				t.fail( 'should throw an error' );
			} catch ( err ) {
				t.equal( err.message, 'BEEP', 'has expected error message' );
			}
		}
		t.end();
	}
});

tape( 'the function generates a bundle (custom glob pattern)', function test( t ) {
	var opts;
	var dir;

	dir = join( __dirname, 'fixtures', 'main' );
	opts = {
		'pattern': '*.js'
	};
	bundle( dir, opts, onBundle );

	function onBundle( error, output ) {
		if ( error ) {
			t.fail( error.message );
		}
		if ( output === false ) {
			t.fail( 'failed to generate a bundle' );
		} else {
			try {
				evil( output.toString() );
				t.fail( 'should throw an error' );
			} catch ( err ) {
				t.equal( err.message, 'BEEP', 'has expected error message' );
			}
		}
		t.end();
	}
});

tape( 'the function supports providing an output file path', function test( t ) {
	var opath;
	var opts;
	var dir;

	dir = join( __dirname, 'tmp' );
	mkdir( dir );

	opath = join( dir, 'bundle.js' );
	opts = {
		'out': opath,
		'pattern': '*.js'
	};
	bundle( join( __dirname, 'fixtures', 'main' ), opts, onWrite );

	function onWrite( error ) {
		var out;
		if ( error ) {
			t.fail( error.message );
		}
		out = readFile( opath, {
			'encoding': 'utf8'
		});

		try {
			evil( out.toString() );
			t.fail( 'should throw an error' );
		} catch ( err ) {
			t.equal( err.message, 'BEEP', 'has expected error message' );
		}

		unlink( opath );
		rmdir( dir );

		t.end();
	}
});

tape( 'the function returns an error if unable to generate a bundle', function test( t ) {
	var dir = join( __dirname, 'fixtures', 'bad-path' );
	bundle( dir, onBundle );

	function onBundle( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});

tape( 'the function returns an error if unable to write to an output file (e.g., due to a non-existing output directory)', function test( t ) {
	var opath;
	var opts;
	var dir;

	dir = join( __dirname, 'fixtures', 'main' );
	opath = join( __dirname, 'fixtures', 'nonexistant_directory', 'bundle.js' );
	opts = {
		'out': opath
	};
	bundle( dir, opts, onWrite );

	function onWrite( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});
