/*
* @license Apache-2.0
*
* Copyright (c) 2020 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/// <reference types="@stdlib/types"/>

import { ndarray } from '@stdlib/types/ndarray';
import fibonacci = require( './index' );

/**
* Returns a mock ndarray object.
*
* @returns ndarray
*/
function array(): ndarray {
	const buf = [ 0, 1, 1, 2, 3, 5, 8, 13, 21, 34 ];
	const x: ndarray = {
		'byteLength': null,
		'BYTES_PER_ELEMENT': null,
		'data': buf,
		'dtype': 'generic',
		'flags': {
			'ROW_MAJOR_CONTIGUOUS': true,
			'COLUMN_MAJOR_CONTIGUOUS': false
		},
		'length': 10,
		'ndims': 1,
		'offset': 0,
		'order': 'row-major',
		'shape': [ 10 ],
		'strides': [ 1 ],
		'get': ( i: number ): number => {
			return buf[ i ];
		},
		'set': ( i: number, v: number ): ndarray => {
			buf[ i ] = v;
			return x;
		}
	};
	return x;
}

// TESTS //

// The function returns a number if provided a number...
{
	fibonacci( 5 ); // $ExpectType number
}

// The function returns an array-like object if provided an array-like object...
{
	const x = new Float64Array( 10 );

	fibonacci( x ); // $ExpectType ArrayLike<number>
}

// The function returns an ndarray if provided an ndarray...
{
	const x = array();

	fibonacci( x ); // $ExpectType ndarray
	fibonacci( x, {} ); // $ExpectType ndarray
	fibonacci( x, { 'order': 'row-major' } ); // $ExpectType ndarray
	fibonacci( x, { 'dtype': 'float64' } ); // $ExpectType ndarray
}

// The compiler throws an error if the function is provided a first argument which is neither an ndarray, array-like object, nor number...
{
	fibonacci( '5' ); // $ExpectError
	fibonacci( true ); // $ExpectError
	fibonacci( false ); // $ExpectError
	fibonacci( null ); // $ExpectError
	fibonacci( undefined ); // $ExpectError
	fibonacci( {} ); // $ExpectError
	fibonacci( [ '5' ] ); // $ExpectError
	fibonacci( ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a valid options object...
{
	const x = array();

	fibonacci( x, '5' ); // $ExpectError
	fibonacci( x, true ); // $ExpectError
	fibonacci( x, false ); // $ExpectError
	fibonacci( x, [ '5' ] ); // $ExpectError
	fibonacci( x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid `order` option...
{
	const x = array();
	fibonacci( x, { 'order': '5' } ); // $ExpectError
	fibonacci( x, { 'order': true } ); // $ExpectError
	fibonacci( x, { 'order': false } ); // $ExpectError
	fibonacci( x, { 'order': null } ); // $ExpectError
	fibonacci( x, { 'order': {} } ); // $ExpectError
	fibonacci( x, { 'order': [ '5' ] } ); // $ExpectError
	fibonacci( x, { 'order': ( x: number ): number => x } ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid `dtype` option...
{
	const x = array();
	fibonacci( x, { 'dtype': '5' } ); // $ExpectError
	fibonacci( x, { 'dtype': true } ); // $ExpectError
	fibonacci( x, { 'dtype': false } ); // $ExpectError
	fibonacci( x, { 'dtype': null } ); // $ExpectError
	fibonacci( x, { 'dtype': {} } ); // $ExpectError
	fibonacci( x, { 'dtype': [ '5' ] } ); // $ExpectError
	fibonacci( x, { 'dtype': ( x: number ): number => x } ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	fibonacci(); // $ExpectError
}

// Attached to the main function is an `assign` method which returns an array-like object if provided an array-like object...
{
	const x = new Float64Array( 10 );
	const y = new Float64Array( 10 );

	fibonacci.assign( x, y ); // $ExpectType ArrayLike<number>
}

// Attached to the main function is an `assign` method which returns an ndarray if provided an ndarray...
{
	const x = array();
	const y = array();

	fibonacci.assign( x, y ); // $ExpectType ndarray
}

// The compiler throws an error if the `assign` method is provided a first argument which is neither an ndarray nor array-like object...
{
	const y = array();

	fibonacci.assign( '5', y ); // $ExpectError
	fibonacci.assign( true, y ); // $ExpectError
	fibonacci.assign( false, y ); // $ExpectError
	fibonacci.assign( null, y ); // $ExpectError
	fibonacci.assign( undefined, y ); // $ExpectError
	fibonacci.assign( {}, y ); // $ExpectError
	fibonacci.assign( ( x: number ): number => x, y ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a second argument which is neither an ndarray nor array-like object...
{
	const x = array();

	fibonacci.assign( x, '5' ); // $ExpectError
	fibonacci.assign( x, true ); // $ExpectError
	fibonacci.assign( x, false ); // $ExpectError
	fibonacci.assign( x, null ); // $ExpectError
	fibonacci.assign( x, undefined ); // $ExpectError
	fibonacci.assign( x, {} ); // $ExpectError
	fibonacci.assign( x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided an unsupported number of arguments...
{
	const x = array();
	const y = array();

	fibonacci.assign(); // $ExpectError
	fibonacci.assign( x ); // $ExpectError
	fibonacci.assign( x, y, y ); // $ExpectError
}
