// @stdlib/lapack/base/dptts2/test/test.dptts2.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var abs = require( '@stdlib/math/base/special/abs' );
var EPS = require( '@stdlib/constants/float64/eps' );
var dptts2 = require( './../lib/dptts2.js' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof dptts2, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 7', function test( t ) {
	t.strictEqual( dptts2.length, 7, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided a first argument which is not a valid order', function test( t ) {
	var values;
	var E;
	var D;
	var B;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop',
		-5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	E = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	D = new Float64Array( [ 2.0, 3.0 ] );
	B = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dptts2( value, 3, 2, E, D, B, 3 );
		};
	}
});

tape( 'the function solves a tridiagonal system of the form `A * X = B` using `L * D * L^T` factorization of `A` (row-major)', function test( t ) {
	var expected;
	var out;
	var D;
	var E;
	var B;

	D = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	E = new Float64Array( [ 2.0, 3.0 ] );
	B = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );

	expected = new Float64Array( [ 11.0, 38.0, -5.0, -18.0, 2.0, 6.0 ] );
	out = dptts2( 'row-major', 3, 2, D, E, B, 2 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 1.0e-7 );

	t.end();
});

tape( 'the function solves a tridiagonal system of the form `A * X = B` using `L * D * L^T` factorization of `A` (column-major)', function test( t ) {
	var expected;
	var out;
	var D;
	var E;
	var B;

	D = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	E = new Float64Array( [ 2.0, 3.0 ] );
	B = new Float64Array( [ 1.0, 3.0, 5.0, 2.0, 4.0, 6.0 ] );

	expected = new Float64Array( [ 11.0, -5.0, 2.0, 38.0, -18.0, 6.0 ] );
	out = dptts2( 'column-major', 3, 2, D, E, B, 3 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 1.0e-7 );

	t.end();
});

tape( 'the function returns B unchanged when `N==1`', function test( t ) {
	var expected;
	var out;
	var D;
	var E;
	var B;

	D = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	E = new Float64Array( [ 2.0, 3.0 ] );
	B = new Float64Array( [ 1.0, 3.0, 5.0, 2.0, 4.0, 6.0 ] );

	expected = new Float64Array( [ 1.0, 3.0, 5.0, 2.0, 4.0, 6.0 ] );
	out = dptts2( 'column-major', 1, 2, D, E, B, 3 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 1.0e-7 );

	t.end();
});

tape( 'the function returns B unchanged when `N<1`', function test( t ) {
	var expected;
	var out;
	var D;
	var E;
	var B;

	D = new Float64Array( [ 1.0, 1.0, 1.0 ] );
	E = new Float64Array( [ 2.0, 3.0 ] );
	B = new Float64Array( [ 1.0, 3.0, 5.0, 2.0, 4.0, 6.0 ] );

	expected = new Float64Array( [ 1.0, 3.0, 5.0, 2.0, 4.0, 6.0 ] );
	out = dptts2( 'column-major', 0, 2, D, E, B, 3 );
	t.strictEqual( out, B, 'returns expected value' );
	isApprox( t, B, expected, 1.0e-7 );

	t.end();
});
