// @stdlib/utils/is-type-error/test/test.js
'use strict';

// MODULES //

var vm = require( 'vm' ); // TODO: handle in-browser tests
var tape = require( 'tape' );
var inherit = require( '@stdlib/utils/inherit' );
var isTypeError = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof isTypeError, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if provided a `TypeError` object', function test( t ) {
	var values;
	var i;

	values = [
		new TypeError( 'type error' )
	];

	for ( i = 0; i < values.length; i++ ) {
		t.strictEqual( isTypeError( values[ i ] ), true, 'returns true when provided '+values[i] );
	}
	t.end();
});

tape( 'the function returns `true` if provided an object inheriting from `TypeError`', function test( t ) {
	function CustomError( msg ) {
		TypeError.call( this );
		this.name = 'CustomError';
		this.message = msg;
		return this;
	}

	inherit( CustomError, TypeError );

	t.strictEqual( isTypeError( new CustomError( 'custom error' ) ), true, 'returns true when provided a value which inherits from TypeError' );
	t.end();
});

tape( 'the function returns `true` if provided a `TypeError` object from a different realm', function test( t ) {
	var error = vm.runInNewContext( 'new TypeError()' );
	t.strictEqual( isTypeError( error ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `true` if provided an object from a different realm which inherits from a `TypeError` object', function test( t ) {
	var error = vm.runInNewContext( 'function Err() { return this; }; Err.prototype = new TypeError(); new Err();' );
	t.strictEqual( isTypeError( error ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if not provided a `TypeError` object', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		undefined,
		[],
		{},
		function noop() {},
		function error() {},
		new Date(),
		new RegExp( '.*' ),
		new Error( 'error' ),
		new EvalError( 'eval error' ),
		new ReferenceError( 'reference error' ),
		new SyntaxError( 'syntax error' ),
		new RangeError( 'range error' ),
		new URIError( 'URI error' )
	];

	for ( i = 0; i < values.length; i++ ) {
		t.strictEqual( isTypeError( values[ i ] ), false, 'returns false when provided '+values[i] );
	}
	t.end();
});
