// @stdlib/lapack/base/dpotf2/test/test.dpotf2.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var EPS = require( '@stdlib/constants/float64/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var dpotf2 = require( './../lib/dpotf2.js' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof dpotf2, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 5', function test( t ) {
	t.strictEqual( dpotf2.length, 5, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided an invalid first argument', function test( t ) {
	var values;
	var A;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop',
		-5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	A = new Float64Array( [ 1.0, 0.0, 0.0, 4.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dpotf2( value, 'upper', 2, A, 2 );
		};
	}
});

tape( 'the function throws an error if provided an invalid second argument', function test( t ) {
	var values;
	var A;
	var i;

	values = [
		'foo',
		'bar',
		'beep',
		'boop',
		-5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	A = new Float64Array( [ 1.0, 2.0, 0.0, 4.0 ] );

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			dpotf2( 'row-major', value, 2, A, 2 );
		};
	}
});

tape( 'the function correctly computes cholesky factorization of a square symmetric matrix ( row-major )', function test( t ) {
	var expected;
	var out;
	var A;

	A = new Float64Array( [ 1.0, 0.0, 0.0, 4.0 ] );
	expected = new Float64Array( [ 1.0, 0.0, 0.0, 2.0 ] );

	out = dpotf2( 'row-major', 'upper', 2, A, 2 );
	t.strictEqual( out, 0, 'returns expected value' );
	isApprox( t, A, expected, 1.0 );

	A = new Float64Array( [ 1.0, 0.0, 0.0, 4.0 ] );
	out = dpotf2( 'row-major', 'lower', 2, A, 2 );
	t.strictEqual( out, 0, 'returns expected value' );
	isApprox( t, A, expected, 1.0 );
	t.end();
});

tape( 'the function correctly computes cholesky factorization of a square symmetric matrix ( column-major )', function test( t ) {
	var expected;
	var out;
	var A;

	A = new Float64Array( [ 1.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0, 3.0 ] );
	expected = new Float64Array( [ 1.0, 0.0, 0.0, 0.0, 1.4142135623730951, 0.0, 0.0, 0.0, 1.7320508075688772 ] );

	out = dpotf2( 'column-major', 'upper', 3, A, 3 );
	t.strictEqual( out, 0, 'returns expected value' );
	isApprox( t, A, expected, 1.0 );

	A = new Float64Array( [ 1.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0, 3.0 ] );
	out = dpotf2( 'column-major', 'lower', 3, A, 3 );
	t.strictEqual( out, 0, 'returns expected value' );
	isApprox( t, A, expected, 1.0 );
	t.end();
});
