// @stdlib/math/base/tools/evalrational/test/test.factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var factory = require( './../lib/factory.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a function', function test( t ) {
	var evalrational = factory( [ 1.0, 2.0, 3.0 ], [ 1.0, 2.0, 3.0 ] );
	t.equal( typeof evalrational, 'function', 'returns a function' );
	t.end();
});

tape( 'if provided two empty coefficient arrays, the generated `evalrational` function always returns `NaN`', function test( t ) {
	var evalrational;
	var v;
	var i;

	evalrational = factory( [], [] );
	for ( i = 0; i < 100; i++ ) {
		v = evalrational( i );
		t.equal( isnan( v ), true, 'returns NaN' );
	}
	t.end();
});

tape( 'if provided coefficient arrays of different lengths, the generated `evalrational` function always returns `NaN`', function test( t ) {
	var evalrational1 = factory( [ 2.0, 1.0, 2.0 ], [ 3.0, 1.0 ] );
	var evalrational2 = factory( [ 0.5, 2.0 ], [ 2.0, 3.0, 1.0 ] );
	var v;
	var i;

	for ( i = 0; i < 100; i++ ) {
		v = evalrational1( i );
		t.equal( isnan( v ), true, 'returns NaN' );
		v = evalrational2( i );
		t.equal( isnan( v ), true, 'returns NaN' );
	}
	t.end();
});

tape( 'if provided only one coefficient for both arrays, the generated `evalrational` function always returns the ratio of the two coefficients', function test( t ) {
	var evalrational;
	var v;
	var i;

	evalrational = factory( [2.0], [4.0] );
	for ( i = 0; i < 100; i++ ) {
		v = evalrational( i );
		t.equal( v, 0.5, 'returns coefficient ratio' );
	}
	t.end();
});

tape( 'if the value at which to evaluate a rational function is `0`, the generated `evalrational` function returns the ratio of the first coefficients', function test( t ) {
	var evalrational;
	var v;

	evalrational = factory( [ 3.0, 2.0, 1.0 ], [ 0.5, 2.0, 1.0 ] );

	v = evalrational( 0.0 );
	t.equal( v, 6.0, 'returns coefficient ratio' );

	t.end();
});

tape( 'the generated `evalrational` function evaluates a rational function', function test( t ) {
	var evalrational;
	var v;

	evalrational = factory( [ -6.0, -5.0 ], [ 3.0, 0.5 ] );

	v = evalrational( 6.0 );
	t.equal( v, -6.0, 'returns -6' );

	v = evalrational( 2.0 );
	t.equal( v, -4.0, 'returns -4' );

	v = evalrational( 0.5 ); // -8.5 / 3.25
	t.equal( v, -2.6153846153846154, 'returns -2.6153846153846154' );

	v = evalrational( 1.0 ); // -11 / 3.5
	t.equal( v, -3.142857142857143, 'returns -3.142857142857143' );

	t.end();
});

tape( 'the generated function handles large `x` values', function test( t ) {
	var evalrational;
	var v1;
	var v2;
	var P;
	var Q;
	var x;
	var i;

	// 6x^5 + 5x^4 + 4x^3 + 3x^2 + 2x^1 + 1x^0
	P = [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ];

	// 1x^5 + 2x^4 + 3x^3 + 4x^2 + 5x^1 + 6x^0
	Q = [ 6.0, 5.0, 4.0, 3.0, 2.0, 1.0 ];

	evalrational = factory( P, Q );

	x = 1.0e100;
	for ( i = 0; i < 1000; i++ ) {
		x *= 2.0;
		v1 = evalrational( x );
		v2 = evalrational( -x );
		t.equal( v1, v2, 'for x=+-'+x+', returns '+v1 );
	}
	t.end();
});

tape( 'the generated function returns analytically incorrect results for certain coefficient and `x` combinations', function test( t ) {
	var evalrational;
	var P;
	var Q;
	var v;

	// Case 1: large negative `x`

	// 1x^2 + 1e308x^1 + 0x^0
	P = [ 0.0, 1.0e308, 1.0 ];

	// 0x^2 + 0x^2 + 1x^0
	Q = [ 1.0, 0, 0 ];

	evalrational = factory( P, Q );

	// (-1e308)^2 + (1e308)(-1e308) = +inf - inf => indeterminate => NaN
	v = evalrational( -1.0e308 );
	t.notOk( isnan( v ), 'should be NaN' );

	// Case 2: large positive `x`

	// 1x^2 - 1e308x^1 + 0x^0
	P = [ 0.0, -1.0e308, 1.0 ];

	// 0x^2 + 0x^2 + 1x^0
	Q = [ 1.0, 0.0, 0.0 ];

	evalrational = factory( P, Q );

	// (1e308)^2 - (1e308)(1e308) = +inf - inf => indeterminate => NaN
	v = evalrational( 1.0e308 );
	t.notOk( isnan( v ), 'should be NaN' );

	t.end();
});

tape( 'the generated `evalrational` function evaluates a rational function (large number of coefficients)', function test( t ) {
	var evalrational;
	var denom;
	var num;
	var i;
	var P;
	var Q;
	var v;

	P = new Array( 1000 );
	Q = new Array( P.length );
	num = 0;
	denom = 0;
	for ( i = 0; i < P.length; i++ ) {
		P[ i ] = i;
		num += i;
		Q[ i ] = i + 1;
		denom += i + 1;
	}
	evalrational = factory( P, Q );

	v = evalrational( 1.0 );
	t.equal( v, num / denom, 'returns expected value' );

	t.end();
});
