<!--

@license Apache-2.0

Copyright (c) 2023 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# No Empty Comments

> [ESLint rule][eslint-rules] to disallow empty comments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-empty-comments' );
```

#### rule

[ESLint rule][eslint-rules] to disallow empty comments.

**Bad**:

<!-- eslint-disable stdlib/no-empty-comments -->

```javascript
function square( x ) {
    var out;

    /* */
    out = x*x; //
    return out;
}
```

**Good**:

```javascript
function square( x ) {
    var out;

    // Square the number:
    out = x*x; // x^2
    return out;
}
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   Comments are considered empty if they contain only whitespace characters.

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-empty-comments' );

var linter = new Linter();
var result;
var code;

code = [
    'function pow2( x ) {',
    '  var out;',
    '',
    '  /* */',
    '  out = x*x; //',
    '  return out;',
    '}'
].join( '\n' );

linter.defineRule( 'no-empty-comments', rule );

result = linter.verify( code, {
    'rules': {
        'no-empty-comments': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-empty-comments',
            'severity': 2,
            'message': 'Empty comments are not allowed',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'endLine': 4,
            'endColumn': 8
        },
        {
            'ruleId': 'no-empty-comments',
            'severity': 2,
            'message': 'Empty comments are not allowed',
            'line': 5,
            'column': 14,
            'nodeType': null,
            'endLine': 5,
            'endColumn': 16
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
