/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import sgbmv = require( './index' );


// TESTS //

// The function returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1  ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 10, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( true, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( false, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( null, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( undefined, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( [], 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( {}, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( ( x: number ): number => x, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a character...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 10, 10, 1, 1, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', true, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', false, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', null, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', undefined, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', [ '1' ], 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', {}, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', ( x: number ): number => x, 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', '10', 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', true, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', false, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', null, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', undefined, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', [], 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', {}, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', ( x: number ): number => x, 10, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, '10', 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, true, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, false, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, null, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, undefined, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, [], 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, {}, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, ( x: number ): number => x, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, '10', 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, true, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, false, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, null, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, undefined, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, [], 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, {}, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, ( x: number ): number => x, 1, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, '10', 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, true, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, false, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, null, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, undefined, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, [], 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, {}, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, ( x: number ): number => x, 1.0, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, '10', A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, true, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, false, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, null, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, undefined, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, [], A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, {}, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, ( x: number ): number => x, A, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, 10, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, '10', 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, true, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, false, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, null, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, undefined, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, [ '1' ], 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, {}, 10, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, ( x: number ): number => x, 10, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, '10', x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, true, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, false, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, null, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, undefined, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, [], x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, {}, x, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, ( x: number ): number => x, x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a Float32Array...
{
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, 10, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, '10', 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, true, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, false, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, null, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, undefined, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, [ '1' ], 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, {}, 1, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, ( x: number ): number => x, 1, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eleventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, '10', 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, true, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, false, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, null, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, undefined, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, [], 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, {}, 1.0, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, ( x: number ): number => x, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, '10', y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, true, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, false, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, null, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, undefined, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, [], y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, {}, y, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, ( x: number ): number => x, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a thirteen argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, 10, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, '10', 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, true, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, false, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, null, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, undefined, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, [ '1' ], 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, {}, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, ( x: number ): number => x, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourteen argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, '10' ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, true ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, false ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, null ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, undefined ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, [] ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, {} ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv(); // $ExpectError
	sgbmv( 'row-major' ); // $ExpectError
	sgbmv( 'row-major', 'N' ); // $ExpectError
	sgbmv( 'row-major', 'N', 10 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0 ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y ); // $ExpectError
	sgbmv( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 1.0, y, 1, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float32Array...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 10, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( true, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( false, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( null, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( undefined, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( [], 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( {}, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( ( x: number ): number => x, 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a character...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 10, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', true, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', false, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', null, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', undefined, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', [ '1' ], 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', {}, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', ( x: number ): number => x, 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', '10', 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', true, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', false, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', null, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', undefined, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', [], 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', {}, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', ( x: number ): number => x, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, '10', 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, true, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, false, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, null, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, undefined, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, [], 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, {}, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, ( x: number ): number => x, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, '10', 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, true, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, false, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, null, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, undefined, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, [], 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, {}, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, ( x: number ): number => x, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, '10', 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, true, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, false, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, null, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, undefined, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, [], 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, {}, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, ( x: number ): number => x, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, '10', A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, true, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, false, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, null, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, undefined, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, [], A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, {}, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, ( x: number ): number => x, A, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eighth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, 10, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, '10', 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, true, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, false, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, null, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, undefined, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, [ '1' ], 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, {}, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, ( x: number ): number => x, 10, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, '10', x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, true, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, false, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, null, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, undefined, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, [], x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, {}, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, ( x: number ): number => x, x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a Float32Array...
{
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, 10, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, '10', 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, true, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, false, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, null, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, undefined, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, [ '1' ], 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, {}, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, ( x: number ): number => x, 1, 0, 1.0, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a eleventh argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, '10', 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, true, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, false, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, null, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, undefined, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, [], 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, {}, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, ( x: number ): number => x, 0, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, '10', 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, true, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, false, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, null, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, undefined, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, [], 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, {}, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, ( x: number ): number => x, 1.0, y, 1 ); // $ExpectError
}

// The compiler throws an error if the function is provided a thirteenth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, '10', y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, true, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, false, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, null, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, undefined, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, [], y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, {}, y, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, ( x: number ): number => x, y, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourteenth argument which is not a Float32Array...
{
	const x = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, 10, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, '10', 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, true, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, false, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, null, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, undefined, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, [ '1' ], 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, {}, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifteenth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, '10', 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, true, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, false, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, null, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, undefined, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, [], 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, {}, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixteenth argument which is not a number...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, '10' ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, true ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, false ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, null ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, undefined ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, [] ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, {} ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const x = new Float32Array( 10 );
	const y = new Float32Array( 10 );
	const A = new Float32Array( 20 );

	sgbmv.ndarray(); // $ExpectError
	sgbmv.ndarray( 'row-major' ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N' ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1 ); // $ExpectError
	sgbmv.ndarray( 'row-major', 'N', 10, 10, 1, 1, 1.0, A, 10, x, 1, 0, 1.0, y, 1, 0, 10 ); // $ExpectError
}
