<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# require-globals

> [ESLint rule][eslint-rules] enforcing that specified global variables are explicitly required.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-globals' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that specified global variables are explicitly required. To specify a list of globals, provide a `globals` option.

**Bad**:

<!-- eslint-disable stdlib/require-globals -->

```javascript
/*
    Options: [ 'error', {
        'globals': [
            'Float32Array'
        ]
    }]
*/
var arr = new Float32Array();
```

**Good**:

<!-- eslint-disable stdlib/require-globals -->

```javascript
/*
    Options: [ 'error', {
        'globals': [
            'Float32Array'
        ]
    }]
*/
var Float32Array = require( '@stdlib/array/float32' );

var arr = new Float32Array();

arr = new Int32Array();
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-globals' );

var linter = new Linter();
var result;

var code = 'var arr = new Int32Array( [1,2,3] );';

linter.defineRule( 'require-globals', rule );

result = linter.verify( code, {
    'rules': {
        'require-globals': [ 'error', {
            'globals': [
                'Int32Array'
            ]
        }]
    }
});
/* returns
    [
        {
            'ruleId': 'require-globals',
            'severity': 2,
            'message': 'Int32Array is used without loading it beforehand via require()',
            'line': 1,
            'column': 15,
            'nodeType': 'Identifier',
            'source': 'var arr = new Int32Array( [1,2,3] );',
            'endLine': 1,
            'endColumn': 25
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
