<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-dynamic-exports

> [ESLint rule][eslint-rules] enforcing that only statically analyzable values are exported.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-dynamic-exports' );
```

#### rule

[ESLint rule][eslint-rules] enforcing only statically analyzable values are exported. Statically analyable values include function expressions, variables, and string and number literals. The rule validates both Node.js `module.exports` and ECMAScript 6 default exports. 

**Bad**:

<!-- eslint-disable stdlib/no-dynamic-exports -->

```javascript
var normal = require( '@stdlib/random/base/normal' );
var beta = require( '@stdlib/random/base/beta' );


// EXPORTS //

module.exports = {
    'normal': normal,
    'beta': beta
};
```

**Good**:

```javascript
var normal = require( '@stdlib/random/base/normal' );
var beta = require( '@stdlib/random/base/beta' );


// VARIABLES //

var ns = {
    'normal': normal,
    'beta': beta
};


// EXPORTS //

module.exports = ns;
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-dynamic-exports' );

var linter = new Linter();
var result;
var code;

code = [
    'function fun() {',
    '  return 3.14;',
    '}',
    'module.exports = fun();'
].join( '\n' );

linter.defineRule( 'no-dynamic-exports', rule );

result = linter.verify( code, {
    'rules': {
        'no-dynamic-exports': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-dynamic-exports',
            'severity': 2,
            'message': 'Only variables, literals, and functions should be exported',
            'line': 1,
            'column': 1,
            'nodeType': 'AssignmentExpression',
            'source': 'module.exports = fun();',
            'endLine': 1,
            'endColumn': 23
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
