// @stdlib/utils/open-url/test/test.open_url.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var noop = require( '@stdlib/utils/noop' );
var openURL = require( './../lib/open_url.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof openURL, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a valid URI', function test( t ) {
	var values;
	var i;

	values = [
		'beep',
		'',
		'///boop',
		5,
		NaN,
		true,
		null,
		undefined,
		[],
		{},
		function(){}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			openURL( value );
		};
	}
});

tape( 'if the platform is `Darwin`, the function invokes the `open` command', function test( t ) {
	var openURL;
	var url;

	openURL = proxyquire( './../lib/open_url.js', {
		'@stdlib/assert/is-darwin': true,
		'@stdlib/assert/is-windows': false,
		'child_process': {
			'spawn': spawn
		}
	});

	url = 'https://google.com';
	openURL( url );

	t.end();

	function spawn( cmd, args ) {
		t.equal( cmd, 'open', 'calls `spawn` with `open` command' );
		t.equal( args[0], url, 'equal URL' );
		return proc();
	}
	function proc() {
		return {
			'unref': noop
		};
	}
});

tape( 'if the platform is `Windows`, the function invokes the `cmd` command', function test( t ) {
	var openURL;
	var url;

	openURL = proxyquire( './../lib/open_url.js', {
		'@stdlib/assert/is-darwin': false,
		'@stdlib/assert/is-windows': true,
		'child_process': {
			'spawn': spawn
		}
	});

	url = 'https://google.com?q=beep&b=boop';
	openURL( url );

	t.end();

	function spawn( cmd, args ) {
		t.equal( cmd, 'cmd', 'calls `spawn` with `open` command' );
		t.equal( args[0], '/c', 'first arg is `/c`' );
		t.equal( args[1], 'start', 'second arg is `start`' );
		t.equal( args[2], '""', 'empty title' );
		t.equal( args[3], 'https://google.com?q=beep^&b=boop', 'escaped ampersands' );
		return proc();
	}
	function proc() {
		return {
			'unref': noop
		};
	}
});

tape( 'if the platform is neither `Darwin` or `Windows`, the function invokes the `xdg-open` command', function test( t ) {
	var openURL;
	var url;

	openURL = proxyquire( './../lib/open_url.js', {
		'@stdlib/assert/is-darwin': false,
		'@stdlib/assert/is-windows': false,
		'child_process': {
			'spawn': spawn
		}
	});

	url = 'https://google.com';
	openURL( url );

	t.end();

	function spawn( cmd, args ) {
		t.equal( cmd, 'xdg-open', 'calls `spawn` with `xdg-open` command' );
		t.equal( args[0], url, 'equal URL' );
		return proc();
	}
	function proc() {
		return {
			'unref': noop
		};
	}
});
