/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import readNDJSON = require( './index' );


// TESTS //

/**
* Call back function.
*
* @param {Error|null} error - error
* @param {string|Buffer} file - file path
* @returns {None} - none
*
*/

function onLoad( error: Error | null, file: string | Buffer ): void {
	if ( error || !file ) {
		throw error;
	}
};


// TESTS //

// The function does not have a return value...
{
	readNDJSON( 'package.ndjson', onLoad ); // $ExpectType void
}

// The compiler throws an error if the function is provided a first argument which is not a string, buffer, or file descriptor...
{
	readNDJSON( false, onLoad ); // $ExpectError
	readNDJSON( true, onLoad ); // $ExpectError
	readNDJSON( null, onLoad ); // $ExpectError
	readNDJSON( undefined, onLoad ); // $ExpectError
	readNDJSON( [], onLoad ); // $ExpectError
	readNDJSON( {}, onLoad ); // $ExpectError
	readNDJSON( ( x: number ): number => x, onLoad ); // $ExpectError
}

// The compiler throws an error if the function is provided a callback argument which is not a function with the expected signature...
{
	readNDJSON( '/path/to/package.ndjson', 'abc' ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', 1 ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', false ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', true ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', null ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', undefined ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', [] ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', {} ); // $ExpectError
	readNDJSON( '/path/to/package.ndjson', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an options argument which is not an object or string...
{
	readNDJSON( 'package.ndjson', false, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', true, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', null, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', undefined, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', 123, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', [], onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', ( x: number ): number => x, onLoad ); // $ExpectError
}

// The compiler throws an error if the function is provided an `encoding` option which is not a string or null...
{
	readNDJSON( 'package.ndjson', { 'encoding': 123 }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'encoding': true }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'encoding': false }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'encoding': [] }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'encoding': {} }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'encoding': ( x: number ): number => x }, onLoad ); // $ExpectError
}

// The compiler throws an error if the function is provided a `flag` option which is not a string...
{
	readNDJSON( 'package.ndjson', { 'flag': 123 }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': true }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': false }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': null }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': [] }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': {} }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'flag': ( x: number ): number => x }, onLoad ); // $ExpectError
}

// The compiler throws an error if the function is provided a `reviver` option which is not a function...
{
	readNDJSON( 'package.ndjson', { 'reviver': 123 }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': 'abc' }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': true }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': false }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': null }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': [] }, onLoad ); // $ExpectError
	readNDJSON( 'package.ndjson', { 'reviver': {} }, onLoad ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	readNDJSON(); // $ExpectError
	readNDJSON( 'C:\\foo\\bar\\baz\\package.ndjson' ); // $ExpectError
}

// Attached to main export is a `sync` method which returns a string or an error...
{
	readNDJSON.sync( 'package.ndjson' ); // $ExpectType string | Error
}

// The compiler throws an error if the `sync` method is provided a first argument which is not a string, buffer, or file descriptor...
{
	readNDJSON.sync( false ); // $ExpectError
	readNDJSON.sync( true ); // $ExpectError
	readNDJSON.sync( null ); // $ExpectError
	readNDJSON.sync( undefined ); // $ExpectError
	readNDJSON.sync( [] ); // $ExpectError
	readNDJSON.sync( {} ); // $ExpectError
	readNDJSON.sync( ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `sync` method is provided an options argument which is not an object or string...
{
	readNDJSON.sync( 'package.ndjson', null ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', true ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', false ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', 123 ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', [] ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `sync` method is provided an `encoding` option which is not a string or null...
{
	readNDJSON.sync( 'package.ndjson', { 'encoding': 123 } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'encoding': true } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'encoding': false } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'encoding': [] } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'encoding': {} } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'encoding': ( x: number ): number => x } ); // $ExpectError
}

// The compiler throws an error if the `sync` method is provided a `flag` option which is not a string...
{
	readNDJSON.sync( 'package.ndjson', { 'flag': 123 } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': true } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': false } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': null } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': [] } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': {} } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'flag': ( x: number ): number => x } ); // $ExpectError
}

// The compiler throws an error if the `sync` method is provided a `reviver` option which is not a function...
{
	readNDJSON.sync( 'package.ndjson', { 'reviver': 'abc' } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': 123 } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': true } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': false } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': null } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': [] } ); // $ExpectError
	readNDJSON.sync( 'package.ndjson', { 'reviver': {} } ); // $ExpectError
}

// The compiler throws an error if the `sync` method is provided an unsupported number of arguments...
{
	readNDJSON.sync(); // $ExpectError
}
