// @stdlib/strided/dispatch/lib/main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2020 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var isNonNegativeInteger = require( '@stdlib/assert/is-nonnegative-integer' ).isPrimitive;
var isPositiveInteger = require( '@stdlib/assert/is-positive-integer' );
var isInteger = require( '@stdlib/assert/is-integer' ).isPrimitive;
var isFunctionArray = require( '@stdlib/assert/is-function-array' );
var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
var isArrayLikeObject = require( '@stdlib/assert/is-array-like-object' );
var abs = require( '@stdlib/math/base/special/abs' );
var dtype = require( '@stdlib/ndarray/base/buffer-dtype' );
var indexOfTypes = require( './index_of_types.js' );


// MAIN //

/**
* Returns a strided array function interface which performs multiple dispatch.
*
* @param {FunctionArray} fcns - list of strided array functions
* @param {StringArray} types - one-dimensional list of arguments for each strided array function
* @param {(ArrayLikeObject|null)} data - strided array function data (e.g., callbacks)
* @param {PositiveInteger} nargs - total number of strided array function input arguments (including strides and offsets)
* @param {NonNegativeInteger} nin - number of input arrays
* @param {PositiveInteger} nout - number of output arrays
* @throws {TypeError} first argument must be an array of functions
* @throws {TypeError} second argument must be an array of strings
* @throws {TypeError} third argument must be an array-like object or `null`
* @throws {Error} third and first arguments must have the same number of elements
* @throws {TypeError} fourth argument must be a positive integer
* @throws {TypeError} fifth argument must be a nonnegative integer
* @throws {TypeError} sixth argument must be a positive integer
* @throws {Error} fourth argument must be compatible with the specified number of input and output arrays
* @throws {Error} number of types must match the number of functions times the total number of array arguments for each function
* @returns {Function} strided array function interface
*
* @example
* // TODO
*/
function dispatch( fcns, types, data, nargs, nin, nout ) {
	var strideArgs;
	var hasOffsets;
	var narrays;
	var nfcns;
	var iout;

	if ( !isFunctionArray( fcns ) ) {
		throw new TypeError( 'invalid argument. First argument must be an array of functions. Value: `' + fcns + '`.' );
	}
	if ( !isStringArray( types ) ) {
		throw new TypeError( 'invalid argument. Second argument must be an array of strings. Value: `' + types + '`.' );
	}
	if ( !isArrayLikeObject( data ) && data !== null ) {
		throw new TypeError( 'invalid argument. Third argument must be an array-like object or `null`. Value: `' + data + '`.' );
	}
	if ( !isPositiveInteger( nargs ) ) {
		throw new TypeError( 'invalid argument. Fourth argument must be a positive integer. Value: `' + nargs + '`.' );
	}
	if ( !isNonNegativeInteger( nin ) ) {
		throw new TypeError( 'invalid argument. Fifth argument must be a nonnegative integer. Value: `' + nin + '`.' );
	}
	if ( !isPositiveInteger( nout ) ) {
		throw new TypeError( 'invalid argument. Sixth argument must be a positive integer. Value: `' + nout + '`.' );
	}
	nfcns = fcns.length;
	if ( data && data.length !== nfcns ) {
		throw new Error( 'invalid argument. The third argument must have the same number of elements as the first argument.' );
	}
	narrays = nin + nout;
	if ( types.length !== nfcns*narrays ) {
		throw new Error( 'invalid argument. Unexpected number of types. A type must be specified for each input and output array for each provided function.' );
	}
	// Determine whether the strided array interface includes offsets:
	if ( (narrays*2)+1 === nargs ) {
		hasOffsets = false;
	} else if ( (narrays*3)+1 === nargs ) {
		hasOffsets = true;
	} else {
		throw new Error( 'invalid argument. Fourth argument is incompatible with the number of input and output arrays.' );
	}
	// Determine the "stride" for accessing related arguments:
	if ( hasOffsets ) {
		strideArgs = 3;
	} else {
		strideArgs = 2;
	}
	// Compute the index of the first output strided array argument:
	iout = ( nin*strideArgs ) + 1;

	return dispatcher;

	/**
	* Strided array function interface which performs multiple dispatch.
	*
	* @private
	* @param {NonNegativeInteger} N - number of indexed elements
	* @param {...(ArrayLikeObject|integer|NonNegativeInteger)} args - array arguments (arrays, strides, and offsets)
	* @throws {Error} insufficient arguments
	* @throws {Error} too many arguments
	* @throws {TypeError} first argument must be a nonnegative integer
	* @throws {TypeError} input array strides must be integers
	* @throws {TypeError} output array strides must be integers
	* @throws {TypeError} input array offsets must be nonnegative integers
	* @throws {TypeError} output array offsets must be nonnegative integers
	* @throws {TypeError} input array arguments must be array-like objects
	* @throws {TypeError} output array arguments must be array-like objects
	* @throws {RangeError} input array arguments must have sufficient elements based on the associated stride and the number of indexed elements
	* @throws {RangeError} output array arguments must have sufficient elements based on the associated stride and the number of indexed elements
	* @throws {TypeError} unable to resolve a strided array function supporting the provided array argument data types
	* @returns {(ArrayLikeObject|Array<ArrayLikeObject>)} destination array(s)
	*/
	function dispatcher() {
		var strides;
		var offsets;
		var arrays;
		var dtypes;
		var shape;
		var argc;
		var idx;
		var fcn;
		var N;
		var v;
		var i;
		var j;

		argc = arguments.length;
		if ( argc < nargs ) {
			throw new Error( 'invalid invocation. Insufficient arguments.' );
		}
		if ( argc > nargs ) {
			throw new Error( 'invalid invocation. Too many arguments.' );
		}
		N = arguments[ 0 ];
		if ( !isNonNegativeInteger( N ) ) {
			throw new TypeError( 'invalid argument. First argument must be a nonnegative integer.' );
		}
		shape = [ N ];

		// Strides for both input and output strided arrays are every `strideArgs` arguments beginning from the third argument...
		strides = [];
		for ( i = 2; i < nargs; i += strideArgs ) {
			v = arguments[ i ];
			if ( !isInteger( v ) ) {
				if ( i < iout ) {
					throw new TypeError( 'invalid argument. Input array stride argument must be an integer.' );
				} else {
					throw new TypeError( 'invalid argument. Output array stride argument must be an integer.' );
				}
			}
			strides.push( v );
		}
		if ( hasOffsets ) {
			// Offsets for both input and output strided arrays are every `strideArgs` arguments beginning from the fourth argument...
			offsets = [];
			for ( i = 3; i < nargs; i += strideArgs ) {
				v = arguments[ i ];
				if ( !isNonNegativeInteger( v ) ) {
					if ( i < iout ) {
						throw new TypeError( 'invalid argument. Input array offset argument must be a nonnegative integer.' );
					} else {
						throw new TypeError( 'invalid argument. Output array offset argument must be a nonnegative integer.' );
					}
				}
				offsets.push( v );
			}
		}
		// Input and output strided arrays are every `strideArgs` arguments beginning from the second argument...
		arrays = [];
		dtypes = [];
		for ( i = 1; i < nargs; i += strideArgs ) {
			v = arguments[ i ];
			if ( !isArrayLikeObject( v ) ) {
				if ( i < iout ) {
					throw new TypeError( 'invalid argument. Input array argument must be an array-like object.' );
				} else {
					throw new TypeError( 'invalid argument. Output array argument must be an array-like object.' );
				}
			}
			j = (i-1) / strideArgs;
			if ( (N-1)*abs(strides[j]) > v.length ) {
				if ( i < iout ) {
					throw new RangeError( 'invalid argument. Input array argument has insufficient elements based on the associated stride and the number of indexed elements.' );
				} else {
					throw new RangeError( 'invalid argument. Output array argument has insufficient elements based on the associated stride and the number of indexed elements.' );
				}
			}
			arrays.push( v );
			dtypes.push( dtype( v ) );
		}
		// Resolve the strided array function satisfying the input array types:
		idx = indexOfTypes( nfcns, narrays, types, narrays, 1, 0, dtypes, 1, 0 ); // eslint-disable-line max-len

		// Check whether we were able to successfully resolve a strided array function:
		if ( idx < 0 ) {
			throw new TypeError( 'invalid arguments. Unable to resolve a strided array function supporting the provided array argument data types.' );
		}
		// Retrieve the strided array function:
		fcn = fcns[ idx ];

		// Evaluate the strided array function:
		if ( data ) {
			if ( hasOffsets ) {
				fcn( arrays, shape, strides, offsets, data[ idx ] );
			} else {
				fcn( arrays, shape, strides, data[ idx ] );
			}
		} else if ( hasOffsets ) {
			fcn( arrays, shape, strides, offsets );
		} else {
			fcn( arrays, shape, strides );
		}
		if ( nout === 1 ) {
			return arrays[ narrays-1 ];
		}
		return arrays.slice( nin+1 );
	}
}


// EXPORTS //

module.exports = dispatch;
