// @stdlib/assert/is-plain-object/test/test.is_plain_object.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var isPlainObject = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof isPlainObject, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if provided a plain object', function test( t ) {
	t.equal( isPlainObject( {} ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `true` if provided an object with no prototype', function test( t ) {
	var bool;

	bool = isPlainObject( Object.create( null ) );
	t.equal( bool, true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if not provided an object', function test( t ) {
	var values;
	var i;
	values = [
		'5',
		5,
		true,
		undefined,
		null,
		NaN,
		[],
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.equal( isPlainObject( values[i] ), false, 'returns false' );
	}
	t.end();
});

tape( 'the function returns `false` if provided a value having an own `constructor` property', function test( t ) {
	var obj = {
		'constructor': Object
	};

	t.equal( isPlainObject( obj ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype lacks a `isPrototypeOf` property', function test( t ) {
	function Foo() {
		return this;
	}
	Foo.prototype = Object.create( Object.prototype );
	Foo.prototype.constructor = Foo;
	delete Foo.prototype.isPrototypeOf;

	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype has a `isPrototypeOf` property which is not a function', function test( t ) {
	function Foo() {
		return this;
	}
	Foo.prototype = Object.create( Object.prototype );
	Foo.prototype.constructor = Foo;
	Foo.prototype.isPrototypeOf = 'beep';

	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype lacks a `constructor` property', function test( t ) {
	function Foo(){
		return this;
	}
	Foo.prototype = {};

	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype has a constructor property which is not a function', function test( t ) {
	function Foo() {
		return this;
	}
	Foo.prototype = Object.create( {} );
	Foo.prototype.constructor = 'beep';

	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype has a constructor property whose native class does not match a function\'s native class', function test( t ) {
	var isPlainObject;

	function Foo() {
		return this;
	}
	Foo.prototype = Object.create( {} );
	Foo.prototype.constructor = Foo;

	isPlainObject = proxyquire( './../lib/is_plain_object.js', {
		'@stdlib/utils/native-class': nativeClass
	});
	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();

	function nativeClass() {
		return '[object String]';
	}
});

tape( 'the function returns `true` if provided a value whose prototype equals the global `Object` prototype (same realm)', function test( t ) {
	function Foo() {
		return this;
	}
	Foo.prototype = Object.prototype;
	Foo.prototype.constructor = Foo;

	t.equal( isPlainObject( new Foo() ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if provided a value whose prototype does not equal the global `Object` prototype and has non-own properties (cross-realm)', function test( t ) {
	function Foo() {
		this.foo = 'bar';
		this.a = 'b';
		this.c = 'd';
		return this;
	}
	Foo.prototype = Object.create( Function.prototype );
	Foo.prototype.constructor = Foo;
	Foo.prototype.isPrototypeOf = function beep(){};
	Foo.prototype.boop = 'boop';

	t.equal( isPlainObject( new Foo() ), false, 'returns false' );
	t.end();
});

tape( 'the function returns `true` if provided a value whose prototype does not equal the global `Object` prototype and has only own properties (cross-realm)', function test( t ) {
	function Foo() {
		this.foo = 'bar';
		this.a = 'b';
		this.c = 'd';
		return this;
	}
	Foo.prototype = Object.create( Object.prototype );

	Object.defineProperty( Foo.prototype, 'constructor', {
		'configurable': true,
		'enumerable': false,
		'value': Foo
	});
	Object.defineProperty( Foo.prototype, 'isPrototypeOf', {
		'configurable': true,
		'enumerable': false,
		'value': function foo(){}
	});

	t.equal( isPlainObject( new Foo() ), true, 'returns true' );
	t.end();
});
