// @stdlib/_tools/github/rank-users/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var copy = require( '@stdlib/utils/copy' );
var rank = require( './../lib/main.js' );


// FIXTURES //

var getOpts = require( './fixtures/opts.js' );
var data = require( './fixtures/data.json' );
var results = require( './fixtures/results.json' );
var info = require( './fixtures/info.json' );


// FUNCTIONS //

function setup() {
	return copy( data );
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof rank, 'function', 'export is a function' );
	t.end();
});

tape( 'function returns an error to a provided callback if an error is encountered when running analysis', function test( t ) {
	var opts;
	var rank;

	rank = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	opts = getOpts();
	rank( setup(), opts, done );

	function factory( opts, clbk ) {
		return function rank() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk({
					'status': 404,
					'message': 'beep'
				});
			}
		};
	}

	function done( error ) {
		t.equal( error.status, 404, 'equal status' );
		t.equal( error.message, 'beep', 'equal message' );
		t.end();
	}
});

tape( 'function returns analysis results to a provided callback', function test( t ) {
	var expected;
	var opts;
	var rank;

	rank = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = results;

	opts = getOpts();
	rank( setup(), opts, done );

	function factory( opts, clbk ) {
		return function rank() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function returns rate limit info to a provided callback', function test( t ) {
	var expected;
	var opts;
	var rank;

	rank = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = info;

	opts = getOpts();
	rank( setup(), opts, done );

	function factory( opts, clbk ) {
		return function rank() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, data, info ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( info, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function options are optional', function test( t ) {
	var expected;
	var rank;

	rank = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = results;

	rank( setup(), done );

	function factory( opts, clbk ) {
		return function rank() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, data ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( data, expected, 'deep equal' );
		t.end();
	}
});
