<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-leading-description-sentence

> [ESLint rule][eslint-rules] to enforce that JSDoc descriptions start with an uppercase letter or number and end with a period.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-leading-description-sentence' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that JSDoc descriptions start with an uppercase letter or number and end with a period.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-leading-description-sentence -->

```javascript
/**
* returns the number 3.14
*
* @returns {number} a number
*
* @example
* var y = foo();
* // returns 3.14
*/
function foo() {
    return 3.14;
}
```

**Good**:

```javascript
/**
* Returns the number 3.14.
*
* @returns {number} a number
*
* @example
* var y = foo();
* // returns 3.14
*/
function foo() {
    return 3.14;
}
```

To allow uncapitalized words at the beginning of a JSDoc description, set the `whitelist` option to a list of uncapitalized exceptions.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-leading-description-sentence' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* squares a number.',
    '* ',
    '* @param {number} x - input number',
    '* @returns {number} x squared',
    '*',
    '* @example',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '    return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-leading-description-sentence', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-leading-description-sentence': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-leading-description-sentence',
            'severity': 2,
            'message': 'Description must start with an uppercase letter or number and end with a period',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'source': '/**',
            'endLine': 10,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
