// @stdlib/math/base/tools/continued-fraction/lib/basic.js
'use strict';

// MODULES //

var abs = require( '@stdlib/math/base/special/abs' );
var TOLERANCE = require( '@stdlib/math/constants/float64-eps' );
var TINY = require( '@stdlib/math/constants/float32-smallest-normal' );


// CONSTANTS //

var MAX_ITER = 1000000;


// FUNCTIONS //

/**
* Evaluates
*           a1
*      ---------------
*      b1 +     a2
*           ----------
*            b2 +   a3
*                -----
*                b3 + ...
*
* @private
* @param {Function} generator - function giving terms of continued fraction expansion
* @param {PositiveNumber} factor - further terms are only added as long as factor*result is smaller than the next term
* @param {PositiveInteger} maxIter - maximum number of iterations
* @returns {number} evaluated expansion
*/
function continued_fraction_a( gen, factor, maxIter ) {
	var delta;
	var a0;
	var C;
	var D;
	var f;
	var v;

	v = gen();
	f = v[ 1 ];
	a0 = v[ 0 ];
	if ( f === 0 ) {
		f = TINY;
	}
	C = f;
	D = 0.0;

	do {
		v = gen();
		if ( v ) {
			D = v[ 1 ] + v[ 0 ] * D;
			if ( D === 0.0 ) {
				D = TINY;
			}
			C = v[ 1 ] + v[ 0 ] / C;
			if ( C === 0.0 ) {
				C = TINY;
			}
			D = 1.0 / D;
			delta = C * D;
			f = f * delta;
		}
	} while( v && ( abs( delta - 1.0 ) > factor ) && --maxIter );

	return a0 / f;
} // end FUNCTION continued_fraction_a()

/**
* Evaluates
*	  b0 +	   a1
*	  ---------------
*	  b1 +	 a2
*		   ----------
*		   b2 +   a3
*				-----
*				b3 + ...
*
* @private
* @param {Function} generator - function giving terms of continued fraction expansion
* @param {PositiveNumber} factor - further terms are only added as long as factor*result is smaller than the next term
* @param {PositiveInteger} maxIter - maximum number of iterations
* @returns {number} evaluated expansion
*/
function continued_fraction_b( gen, factor, maxIter ) {
	var delta;
	var C;
	var D;
	var f;
	var v;

	v = gen();
	f = v[ 1 ];
	if( f === 0.0 ) {
		f = TINY;
	}
	C = f;
	D = 0.0;
	do {
		v = gen();
		if ( v ) {
			D = v[ 1 ] + v[ 0 ] * D;
			if ( D === 0.0 ) {
				D = TINY;
			}
			C = v[ 1 ] + v[ 0 ] / C;
			if ( C === 0.0 ) {
				C = TINY;
			}
			D = 1.0 / D;
			delta = C * D;
			f = f * delta;
		}
	} while( v && ( abs( delta - 1.0 ) > factor ) && --maxIter );
	return f;
} // end FUNCTION continued_fraction_b()


// CONTINUED FRACTION //

/**
* Evaluates the continued fraction approximation for the supplied series generator using the modified Lentz algorithm.
*
* #### References
* * Lentz, W.J. 1976, Applied Optics, vol. 15, pp. 668-671.
*
* @param {Function} generator - function returning terms of continued fraction expansion
* @param {Object} [opts] - function options
* @param {PositiveInteger} [opts.maxIter=1000000] - maximum number of iterations
* @param {PositiveNumber} [opts.tolerance=2.22e-16] - further terms are only added as long as the next term is greater than current term times the tolerance
* @param {boolean} [opts.keep=false] - whether to keep the leading b term
* @returns {number} value of continued fraction
*
* @example*
* // Continued fraction for (e-1)^(-1):
* var gen = generator()
* var out = continuedFraction( gen );
* // returns ~0.582
*
* function generator() {
*    var i = 0;
*    return function() {
*        i++;
*        return [ i, i ];
*    };
* }
*/
function continued_fraction( generator, options ) {
	var maxIter;
	var opts;
	var eps;

	opts = {};
	if ( arguments.length > 1 ) {
		opts = options;
	}
	eps = opts.tolerance || TOLERANCE;
	maxIter = opts.maxIter || MAX_ITER;

	if ( opts.keep ) {
		return continued_fraction_b( generator, eps, maxIter );
	}
	return continued_fraction_a( generator, eps, maxIter );
} // end FUNCTION continued_fraction()


// EXPORTS //

module.exports = continued_fraction;
