<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Fenced Code Marker

> [ESLint rule][eslint-rules] to enforce a specific Markdown fenced code marker in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-fenced-code-marker' );
```

#### rule

[ESLint rule][eslint-rules] to enforce a specific Markdown fenced code marker in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-fenced-code-marker, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep boop.
*
* ~~~javascript
* y = x;
* ~~~
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

````javascript
/**
* Beep boop.
*
* ```javascript
* y = x;
* ```
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
````

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-fenced-code-marker]. By default, the rule enforces that `` ` `` be used as the Markdown fenced code marker in JSDoc descriptions.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-fenced-code-marker' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep *boop*.',
    '*',
    '* ~~~javascript',
    '* y = x;',
    '* ~~~',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-fenced-code-marker', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-fenced-code-marker': [ 'error', '`' ]
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-fenced-code-marker',
            'severity': 2,
            'message': 'Fenced code should use ` as a marker',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* ~~~javascript',
            'endLine': 14,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-fenced-code-marker]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-fenced-code-marker

</section>

<!-- /.links -->
