// @stdlib/_tools/github/delete-token/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var noop = require( '@stdlib/utils/noop' );
var del = require( './../lib/main.js' );


// FIXTURES //

var getOpts = require( './fixtures/opts' );
var info = require( './fixtures/info' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof del, 'function', 'export is a function' );
	t.end();
});

tape( 'function throws if provided an invalid options argument', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			del( 1, {
				'username': value,
				'password': 'boop'
			}, noop );
		};
	}
});

tape( 'function throws if provided a callback argument which is not a function', function test( t ) {
	var values;
	var opts;
	var i;

	values = [
		'5',
		5,
		NaN,
		null,
		void 0,
		true,
		[],
		{}
	];

	opts = getOpts();
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			del( 1, opts, value );
		};
	}
});

tape( 'function throws if provided a token id argument which is not a nonnegative integer', function test( t ) {
	var values;
	var opts;
	var i;

	values = [
		'5',
		-5,
		3.14,
		NaN,
		null,
		void 0,
		true,
		[],
		{},
		function noop() {}
	];

	opts = getOpts();
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			del( value, opts, noop );
		};
	}
});

tape( 'function returns an error to a provided callback if an error is encountered when deleting a token (no rate limit info)', function test( t ) {
	var opts;
	var del;

	del = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	opts = getOpts();
	del( 1, opts, done );

	function query( id, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( new Error( 'beep' ) );
		}
	}

	function done( error ) {
		t.ok( error instanceof Error, 'error instance' );
		t.equal( error.message, 'beep' );
		t.end();
	}
});

tape( 'function returns an error to a provided callback if an error is encountered when deleting a token', function test( t ) {
	var opts;
	var del;

	del = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	opts = getOpts();
	del( 1, opts, done );

	function query( id, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( new Error( 'beep' ), info );
		}
	}

	function done( error ) {
		t.ok( error instanceof Error, 'error instance' );
		t.equal( error.message, 'beep' );
		t.end();
	}
});

tape( 'function returns rate limit info to a provided callback', function test( t ) {
	var expected;
	var opts;
	var del;

	del = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	expected = info;

	opts = getOpts();
	del( 1, opts, done );

	function query( id, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( null, info );
		}
	}

	function done( error, info ) {
		if ( error ) {
			t.fail( error.message );
		}
		t.deepEqual( info, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function returns rate limit info to a provided callback (error)', function test( t ) {
	var expected;
	var opts;
	var del;

	del = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	expected = info;

	opts = getOpts();
	del( 1, opts, done );

	function query( id, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( new Error( 'beep' ), info );
		}
	}

	function done( error, info ) {
		t.ok( error instanceof Error, 'error instance' );
		t.equal( error.message, 'beep' );

		t.deepEqual( info, expected, 'deep equal' );

		t.end();
	}
});
