// @stdlib/assert/is-string/test/test.object.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var hasSymbols = require( '@stdlib/utils/detect-symbol-support' )();
var hasToStringTag = require( '@stdlib/utils/detect-tostringtag-support' )();
var isString = require( './../lib/object.js' );


// VARIABLES //

var opts = {
	'skip': !hasSymbols
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof isString, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if provided a string object', function test( t ) {
	t.equal( isString( new String( 'a' ) ), true, 'returns true' );
	t.end();
});

tape( 'the function returns `false` if provided a primitive string', function test( t ) {
	t.equal( isString( 'a' ), false, 'returns false' );
	t.end();
});

tape( 'if `Symbol.toStringTag` is supported, the function guards against objects masquerading as `String` objects', opts, function test( t ) {
	var isString;
	var mock;

	isString = proxyquire( './../lib/object.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	mock = {
		'toString': toString,
		'valueOf': valueOf
	};
	if ( hasToStringTag ) {
		mock[ Symbol.toStringTag ] = 'String';
	}
	t.equal( isString( mock ), false, 'returns false' );
	t.end();

	function detect() {
		return true;
	}
	function toString() {
		return 'a';
	}
	function valueOf() {
		return 'a';
	}
});

tape( 'if `Symbol.toStringTag` is not supported, the function attempts to determine the native class', function test( t ) {
	var isString = proxyquire( './../lib/object.js', {
		'@stdlib/utils/detect-tostringtag-support': detect
	});

	t.equal( isString( new String( 'a' ) ), true, 'returns true' );
	t.equal( isString( {} ), false, 'returns false' );

	t.end();

	function detect() {
		return false;
	}
});

tape( 'function returns `false` if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		true,
		undefined,
		[],
		{},
		new Date(),
		/./,
		new RegExp( '.' ),
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.equal( isString( values[i] ), false, 'returns false when provided '+values[i] );
	}
	t.end();
});
