// @stdlib/random/base/exponential/test/test.factory.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/process/env' );
var kstest = require( '@stdlib/stats/kstest' );
var now = require( '@stdlib/time/now' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var isUint32Array = require( '@stdlib/assert/is-uint32array' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var exponential;
	var r;
	var i;

	// When binding distribution parameter...
	exponential = factory( 2.0 );
	for ( i = 0; i < 100; i++ ) {
		r = exponential();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameter...
	exponential = factory();
	for ( i = 0; i < 100; i++ ) {
		r = exponential( 100.0 );
		t.equal( typeof r, 'number', 'returns a number' );
	}

	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var exponential1;
	var exponential2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = now();

	exponential1 = factory( 35.0, {
		'seed': seed
	});
	exponential2 = factory( 35.0, {
		'seed': seed
	});

	t.notEqual( exponential1, exponential2, 'separate generators' );

	for ( i = 0; i < 100; i++ ) {
		r1 = exponential1();
		r2 = exponential2();
		t.equal( r1, r2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var exponential = factory();
	t.equal( exponential.NAME, 'exponential', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var exponential = factory();
	t.equal( typeof exponential.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var exponential = factory({
		'seed': 12345
	});
	t.equal( isUint32Array( exponential.seed ), true, 'has `seed` property' );
	t.equal( exponential.seed[ 0 ], 12345, 'equal to provided seed' );
	t.end();
});

tape( 'attached to the returned function is the generator state', function test( t ) {
	var exponential = factory();
	t.equal( isUint32Array( exponential.state ), true, 'has `state` property' );
	t.end();
});

tape( 'attached to the returned function is the generator state length', function test( t ) {
	var exponential = factory();
	t.equal( typeof exponential.STATE_LENGTH, 'number', 'has `STATE_LENGTH` property' );
	t.end();
});

tape( 'attached to the returned function is the generator state size', function test( t ) {
	var exponential = factory();
	t.equal( typeof exponential.byteLength, 'number', 'has `byteLength` property' );
	t.end();
});

tape( 'the function throws an error if provided a rate which is not a positive number (no options)', function test( t ) {
	var values;
	var i;

	values = [
		-2.0,
		0.0,
		NaN,
		'5',
		null,
		true,
		false,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided a rate which is not a positive number (options)', function test( t ) {
	var values;
	var i;

	values = [
		-2.0,
		0.0,
		NaN,
		'5',
		null,
		true,
		false,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, {} );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		false,
		void 0,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 10.0, value );
		};
	}
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` when provided a rate equal to `NaN`', function test( t ) {
	var exponential;
	var r;

	exponential = factory();
	r = exponential( NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a PRNG that returns `NaN` when provided a rate which is not a positive number', function test( t ) {
	var exponential;
	var r;

	exponential = factory();

	r = exponential( 0 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = exponential( -1 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	r = exponential( -3.14 );
	t.strictEqual( isnan( r ), true, 'returns NaN' );

	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from an exponential distribution', opts, function test( t ) {
	var threshold;
	var lambda;
	var count;
	var npass;
	var N;
	var x;

	threshold = 0.10;

	lambda = 4.0;

	x = new Array( 1e3 );
	N = 300;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var exponential;
		var rejected;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			exponential = factory( lambda );
			t.ok( true, 'seed: '+exponential.seed );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = exponential();
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'exponential', lambda ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Exponential('+lambda+') is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});

tape( 'the returned function supports setting the generator state', function test( t ) {
	var exponential;
	var state;
	var arr;
	var i;

	exponential = factory( 4.0 );

	// Move to a future state...
	for ( i = 0; i < 100; i++ ) {
		exponential();
	}
	// Capture the current state:
	state = exponential.state;

	// Move to a future state...
	arr = [];
	for ( i = 0; i < 100; i++ ) {
		arr.push( exponential() );
	}
	// Set the state:
	exponential.state = state;

	// Replay previously generated values...
	for ( i = 0; i < 100; i++ ) {
		t.equal( exponential(), arr[ i ], 'returns expected value. i: '+i+'.' );
	}
	t.end();
});
