// @stdlib/math/base/random/uniform/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var kstest = require( '@stdlib/math/generics/statistics/kstest' );
var round = require( '@stdlib/math/base/special/round' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( process.env.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var uniform;
	var r;
	var i;

	// When binding distribution parameters...
	uniform = factory( -10.0, 10.0 );
	for ( i = 0; i < 100; i++ ) {
		r = uniform();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameters...
	uniform = factory();
	for ( i = 0; i < 100; i++ ) {
		r = uniform( -10.0, 10.0 );
		t.equal( typeof r, 'number', 'returns a number' );
	}
	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var uniform1;
	var uniform2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	uniform1 = factory( 0.0, 4.0, {
		'seed': seed
	});
	uniform2 = factory( 0.0, 4.0, {
		'seed': seed
	});

	t.notEqual( uniform1, uniform2, 'separate generators' );

	for ( i = 0; i < 100; i++ ) {
		r1 = uniform1();
		r2 = uniform2();
		t.equal( r1, r2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var uniform = factory();
	t.equal( uniform.NAME, 'uniform', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var uniform = factory();
	t.equal( typeof uniform.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var uniform = factory({
		'seed': 12345
	});
	t.equal( typeof uniform.SEED, 'number', 'has `SEED` property' );
	t.equal( uniform.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'the function throws an error if minimum support `a` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			factory( value, 2.0 );
		};
	}
});

tape( 'the function throws an error if maximum support `b` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			factory( 2.0, value );
		};
	}
});

tape( 'the function throws an error if minimum support `a` is not less than maximum support `b`', function test( t ) {
	var values;
	var i;

	values = [
		[ 0.0, 0.0 ],
		[ -2.0, -4.0 ],
		[ 2.0, 1.0 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( arr ) {
		return function() {
			factory( arr[0], arr[1] );
		};
	}
});

tape( 'the function throws an error if the options argument is not an object (when `a` and `b` are not supplied)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if the options argument is not an object (when `a` and `b` are supplied)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			factory( 0.0, 1.0, value );
		};
	}
});

tape( 'when called without parameters, the function returns a function that throws an error if minimum support `a` value is not a number primitive', function test( t ) {
	var uniform;
	var values;
	var i;

	uniform = factory();
	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			uniform( value, 1.0 );
		};
	}
});

tape( 'when called without parameters, the function returns a function that throws an error if maximum support `b` is not a number primitive', function test( t ) {
	var uniform;
	var values;
	var i;

	uniform = factory();
	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function(){}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function() {
			uniform( 1.0, value );
		};
	}
});

tape( 'when called without parameters, the function returns a function that throws an error if minimum support `a` is not less than maximum support `b`', function test( t ) {
	var uniform;
	var values;
	var i;

	uniform = factory();
	values = [
		[ 2.0, 1.0 ],
		[ -3.0, -4.0 ],
		[ 0.0, 0.0 ],
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( arr ) {
		return function() {
			uniform( arr[0], arr[1] );
		};
	}
});

tape( 'the function returns a function for creating random numbers from a uniform distribution', opts, function test( t ) {
	var rejected;
	var uniform;
	var pValue;
	var N;
	var a;
	var b;
	var i;
	var j;
	var x;

	a = 2.0;
	b = 4.0;
	uniform = factory( a, b, {
		'seed': 12345
	});
	rejected = 0;
	x = new Array( 1e3 );
	N = 300;
	for ( i = 0; i < N; i++ ) {
		for ( j = 0; j < x.length; j++ ) {
			x[ j ] = uniform();
			if ( x[ j ] < a || x[ j ] > b ) {
				t.ok( false, 'returned a number outside support: '+x[ j ] );
			}
		}
		// Test using Kolmogorov-Smirnov goodness-of-fit test:
		pValue = kstest( x, 'uniform', 2.0, 4.0 ).pValue;
		t.equal( typeof pValue, 'number', 'returns a p-value' );
		if ( pValue < 0.05 ) {
			rejected += 1;
		}
	}
	// Account for small sample sizes and few repeats...
	t.ok( rejected / N < 0.07, 'null hypothesis (i.e., that numbers are drawn from Uniform(2,4)) is rejected in less than 7% of cases ('+rejected+' of '+N+')' );
	t.end();
});
