// @stdlib/repl/server/test/test.pkg2alias.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var pkg2alias = require( './../lib/functions/pkg2alias.js' );


// FIXTURES //

function namespace() {
	var ns = [
		{
			'alias': 'beep',
			'path': '@stdlib/beep',
			'value': {}
		},
		{
			'alias': 'boop',
			'path': '@stdlib/foo/boop',
			'value': []
		},
		{
			'value': {}
		}
	];
	ns[ '@noCallThru' ] = true;
	return ns;
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof pkg2alias, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			pkg2alias( value );
		};
	}
});

tape( 'the function resolves a package name to an alias', function test( t ) {
	var pkg2alias;
	var mock;
	var ns;

	mock = {
		'log': log
	};

	ns = namespace();

	pkg2alias = proxyquire( './../lib/functions/pkg2alias.js', {
		'./../namespace.js': ns,
		'./../console.js': mock
	});

	pkg2alias( '@stdlib/beep' );

	function log( str ) {
		t.strictEqual( str, 'beep', 'logs expected value' );
		t.end();
	}
});

tape( 'the function resolves a package name to an alias', function test( t ) {
	var pkg2alias;
	var mock;
	var ns;

	mock = {
		'log': log
	};

	ns = namespace();

	pkg2alias = proxyquire( './../lib/functions/pkg2alias.js', {
		'./../namespace.js': ns,
		'./../console.js': mock
	});

	pkg2alias( '@stdlib/foo/boop' );

	function log( str ) {
		t.strictEqual( str, 'boop', 'logs expected value' );
		t.end();
	}
});

tape( 'if unable to resolve an alias, the function logs a message', function test( t ) {
	var pkg2alias;
	var mock;
	var ns;

	mock = {
		'log': log
	};

	ns = namespace();

	pkg2alias = proxyquire( './../lib/functions/pkg2alias.js', {
		'./../namespace.js': ns,
		'./../console.js': mock
	});

	pkg2alias( 'adkfjadkfjsldjjfasdlfsdj' );

	function log( str ) {
		t.notEqual( str, 'beep', 'logs a different string' );
		t.notEqual( str, 'boop', 'logs a different string' );
		t.end();
	}
});
