// @stdlib/iter/cuevery-by/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var array2iterator = require( '@stdlib/array/to-iterator' );
var noop = require( '@stdlib/utils/noop' );
var iterCuEveryBy = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof iterCuEveryBy, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided an iterator', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			iterCuEveryBy( value, noop );
		};
	}
});

tape( 'the function throws an error if not provided a predicate function', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			iterCuEveryBy( array2iterator( [ 1, 2, 3 ] ), value );
		};
	}
});

tape( 'if provided an "empty" iterator, the function returns an iterator which always returns `true`', function test( t ) {
	var arr;
	var it;
	var v;

	arr = array2iterator( [] );
	it = iterCuEveryBy( arr, predicate );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().done;
	t.strictEqual( v, true, 'returns expected value' );

	t.end();

	function predicate() {
		return true;
	}
});

tape( 'the function returns an iterator which returns `true` if every iterated value passes a test implemented by a predicate function', function test( t ) {
	var arr;
	var it;
	var v;

	arr = array2iterator( [ 1, 1, 1, 1, 1 ] );
	it = iterCuEveryBy( arr, predicate );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().done;
	t.strictEqual( v, true, 'returns expected value' );

	t.end();

	function predicate( v ) {
		return ( v > 0 );
	}
});

tape( 'the function returns an iterator which returns `false` if at least one iterated value fails a test implemented by a predicate function', function test( t ) {
	var arr;
	var it;
	var v;

	arr = array2iterator( [ 1, 1, 0 ] );
	it = iterCuEveryBy( arr, predicate );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, false, 'returns expected value' );

	v = it.next().done;
	t.strictEqual( v, true, 'returns expected value' );

	t.end();

	function predicate( v ) {
		return ( v > 0 );
	}
});

tape( 'the function supports providing an execution context', function test( t ) {
	var arr;
	var ctx;
	var it;
	var v;

	ctx = {
		'count': 0
	};

	arr = array2iterator( [ 1, 1, 1, 0, 1 ] );
	it = iterCuEveryBy( arr, predicate, ctx );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, true, 'returns expected value' );

	v = it.next().value;
	t.strictEqual( v, false, 'returns expected value' );

	v = it.next().done;
	t.strictEqual( v, true, 'returns expected value' );

	t.strictEqual( ctx.count, 4, 'returns expected value' );

	t.end();

	function predicate( v ) {
		this.count += 1; // eslint-disable-line no-invalid-this
		return ( v > 0 );
	}
});
