# Heading Punctuation

> [ESLint rule][eslint-rules] to prevent Markdown headings in JSDoc descriptions from ending with specified characters.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-heading-punctuation' );
```

#### rule

[ESLint rule][eslint-rules] to prevent Markdown headings in JSDoc descriptions from ending with specified characters.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-no-heading-punctuation, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* ## Boop:
*
* Beep.
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Boop beep.
*
* ## Boop
*
* Beep.
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-no-heading-punctuation]. By default, the rule forbids the use of `.,;:!?` at the end of a Markdown heading in JSDoc descriptions.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-heading-punctuation' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* ## Beep?',
    '*',
    '* boop',
    '*',
    '*  ## Boop!',
    '*',
    '* beep',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-no-heading-punctuation', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-no-heading-punctuation': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-no-heading-punctuation',
            'severity': 2,
            'message': 'Don’t add a trailing `?` to headings',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* ## Beep?',
            'endLine': 18,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-no-heading-punctuation',
            'severity': 2,
            'message': 'Don’t add a trailing `!` to headings',
            'line': 8,
            'column': 3,
            'nodeType': null,
            'source': '* ## Boop!',
            'endLine': 18,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-no-heading-punctuation]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-no-heading-punctuation

</section>

<!-- /.links -->
