// @stdlib/assert/is-electron-renderer/test/test.is_electron_renderer.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var ENV = require( '@stdlib/utils/env' );
var isElectronRenderer = require( './../lib/is_electron_renderer.js' );


// FIXTURES //

function Process() {
	this.type = 'renderer';
	this.versions = {
		'electron': '1.0.0',
		'chrome': '42.1.34'
	};
	this.env = ENV;
	return this;
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof isElectronRenderer, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns `true` if runtime is the Electron renderer process', function test( t ) {
	var isElectronRenderer;
	var proc;
	var bool;

	proc = new Process();

	isElectronRenderer = proxyquire( './../lib/is_electron_renderer.js', {
		'@stdlib/assert/is-electron': true,
		'./process.js': proc
	});

	bool = isElectronRenderer();
	t.strictEqual( bool, true, 'returns true' );

	t.end();
});

tape( 'the function returns `false` if runtime is not the Electron renderer process (not Electron)', function test( t ) {
	var isElectronRenderer;
	var proc;
	var bool;

	proc = new Process();

	isElectronRenderer = proxyquire( './../lib/is_electron_renderer.js', {
		'@stdlib/assert/is-electron': false,
		'./process.js': proc
	});

	bool = isElectronRenderer();
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if runtime is not the Electron renderer process (process type does not equal `renderer`)', function test( t ) {
	var isElectronRenderer;
	var proc;
	var bool;

	proc = new Process();
	proc.type = 'browser';

	isElectronRenderer = proxyquire( './../lib/is_electron_renderer.js', {
		'@stdlib/assert/is-electron': true,
		'./process.js': proc
	});

	bool = isElectronRenderer();
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});
