// @stdlib/_tools/browserify/file-list/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var mkdir = require( 'fs' ).mkdirSync;
var rmdir = require( 'fs' ).rmdirSync;
var tape = require( 'tape' );
var unlink = require( '@stdlib/fs/unlink' ).sync;
var readFile = require( '@stdlib/fs/read-file' ).sync;
var noop = require( '@stdlib/utils/noop' );
var evil = require( '@stdlib/utils/eval' );
var bundle = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof bundle, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the first argument is not an array of strings', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[],
		[ 1, 2, 3 ],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			bundle( value, noop );
		};
	}
});

tape( 'the function throws an error if the output file path argument is not a string primitive', function test( t ) {
	var values;
	var i;
	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[],
		[ 1, 2, 3 ],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var fpath = join( __dirname, 'fixtures', 'index.js' );
			bundle( [ fpath ], value, noop );
		};
	}
});

tape( 'the function throws an error if provided a callback argument which is not a function', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		false,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var fpath = join( __dirname, 'fixtures', 'index.js' );
			bundle( [ fpath ], value );
		};
	}
});

tape( 'the function generates a bundle', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'index.js' );

	bundle( [ fpath ], onBundle );

	function onBundle( error, output ) {
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		} else {
			try {
				evil( output.toString() );
				t.fail( 'should throw an error' );
			} catch ( err ) {
				t.equal( err.message, 'BEEP', 'has expected message' );
			}
		}
		t.end();
	}
});

tape( 'the function supports specifying an output file path', function test( t ) {
	var fpath;
	var opath;
	var dir;

	dir = join( __dirname, 'tmp' );
	mkdir( dir );

	fpath = join( __dirname, 'fixtures', 'index.js' );
	opath = join( dir, 'bundle.js' );

	bundle( [ fpath ], opath, onWrite );

	function onWrite( error ) {
		var out;
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		out = readFile( opath, {
			'encoding': 'utf8'
		});
		try {
			evil( out.toString() );
			t.fail( 'should throw an error' );
		} catch ( err ) {
			t.equal( err.message, 'BEEP', 'has expected message' );
		}

		// Clean-up:
		unlink( opath );
		rmdir( dir );

		t.end();
	}
});

tape( 'if unable to generate a bundle, the function invokes the provided callback with an error argument', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'bad_path.js' );

	bundle( [ fpath ], onBundle );

	function onBundle( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});

tape( 'if unable to write a bundle to an output file path, the function invokes the callback with an error argument', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'index.js' );
	var opath = join( __dirname, 'fixtures', 'nonexistant_directory', 'bundle.js' );

	bundle( [ fpath ], opath, onWrite );

	function onWrite( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});
