// @stdlib/_tools/browserify/file-list/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var join = require( 'path' ).join;
var unlink = require( '@stdlib/fs/unlink' ).sync;
var readFile = require( '@stdlib/fs/read-file' ).sync;
var noop = require( '@stdlib/utils/noop' );
var bundle = require( './../lib' );


// FIXTURES //

var EXPECTED = readFile( join( __dirname, 'fixtures', 'expected.txt' ) );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof bundle, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the first argument is not an array of strings', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		{},
		[],
		[ 1, 2, 3 ],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			bundle( value, noop );
		};
	}
});

tape( 'the function throws an error if the `dest` argument is not a string primitive', function test( t ) {
	var values;
	var i;
	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		{},
		[],
		[ 1, 2, 3 ],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var fpath = join( __dirname, 'fixtures', 'index.js' );
			var files = [ fpath ];
			bundle( files, value, noop );
		};
	}
});

tape( 'the function throws an error if provided a callback argument which is not a function', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var fpath = join( __dirname, 'fixtures', 'index.js' );
			var files = [ fpath ];
			bundle( files, value );
		};
	}
});

tape( 'the function creates a bundle using `browserify` and calls the callback with the generated output', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'index.js' );
	var files = [ fpath ];
	bundle( files, onRead );

	function onRead( error, output ) {
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		t.deepEqual( output, EXPECTED, 'returns bundled code' );
		t.end();
	}
});

tape( 'the function creates a bundle using `browserify` and writes the results to disk', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'index.js' );
	var files = [ fpath ];
	var opath = join( __dirname, 'fixtures', 'bundle.js' );
	bundle( files, opath, onWrite );

	function onWrite( error ) {
		var out;
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		out = readFile( opath );
		t.deepEqual( out, EXPECTED, 'written bundle contains expected contents' );

		// Remove created file:
		unlink( opath );
		t.end();
	}
});

tape( 'the function invokes the callback with an error if bundling fails', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'error.js' );
	var files = [ fpath ];
	bundle( files, onRead );

	function onRead( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});

tape( 'the function invokes the callback with an error if writing the bundle to disk fails', function test( t ) {
	var fpath = join( __dirname, 'fixtures', 'index.js' );
	var files = [ fpath ];
	var opath = join( __dirname, 'fixtures', 'nonexisting', 'bundle.js' );
	bundle( files, opath, onWrite );

	function onWrite( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});
