// @stdlib/blas/base/sgemv/test/test.ndarray.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float32Array = require( '@stdlib/array/float32' );
var scopy = require( '@stdlib/blas/base/scopy' );
var EPS = require( '@stdlib/constants/float32/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var sgemv = require( './../lib/ndarray.js' );


// FIXTURES //

var stridesxoyo = require( './fixtures/julia/strides_xoyo.json' );
var stridesxoyt = require( './fixtures/julia/strides_xoyt.json' );
var stridexpyp = require( './fixtures/julia/strides_xpyp.json' );
var stridexnyp = require( './fixtures/julia/strides_xnyp.json' );
var stridexpyn = require( './fixtures/julia/strides_xpyn.json' );
var stridexnyn = require( './fixtures/julia/strides_xnyn.json' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof sgemv, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 14', function test( t ) {
	t.strictEqual( sgemv.length, 14, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations `y = alpha*A*x + beta*y` or `y = alpha*A**T*x + beta*y` (sx=1, sy=1)', function test( t ) {
	var expected;
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridesxoyo.order;
	trans = stridesxoyo.trans;

	m = stridesxoyo.M;
	n = stridesxoyo.N;
	lda = stridesxoyo.lda;

	alpha = stridesxoyo.alpha;
	beta = stridesxoyo.beta;

	a = new Float32Array( stridesxoyo.A );
	x = new Float32Array( stridesxoyo.x );
	y = new Float32Array( stridesxoyo.y );

	expected = new Float32Array( stridesxoyo.y_out );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, 1, 0, beta, y, 1, 0 );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function performs one of the matrix-vector operations `y = alpha*A*x + beta*y` or `y = alpha*A**T*x + beta*y` (sx=1, sy=2)', function test( t ) {
	var expected;
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridesxoyt.order;
	trans = stridesxoyt.trans;

	m = stridesxoyt.M;
	n = stridesxoyt.N;
	lda = stridesxoyt.lda;

	alpha = stridesxoyt.alpha;
	beta = stridesxoyt.beta;

	a = new Float32Array( stridesxoyt.A );
	x = new Float32Array( stridesxoyt.x );
	y = new Float32Array( stridesxoyt.y );

	expected = new Float32Array( stridesxoyt.y_out );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, 1, 0, beta, y, 2, 0 );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function supports `y` offset', function test( t ) {
	var expected;
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridexpyn.order;
	trans = stridexpyn.trans;

	m = stridexpyn.M;
	n = stridexpyn.N;
	lda = stridexpyn.lda;

	alpha = stridexpyn.alpha;
	beta = stridexpyn.beta;

	a = new Float32Array( stridexpyn.A );
	x = new Float32Array( stridexpyn.x );
	y = new Float32Array( stridexpyn.y );

	expected = new Float32Array( stridexpyn.y_out );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, 1, 0, beta, y, -2, 2 );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function supports `x` offset', function test( t ) {
	var expected;
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridexnyp.order;
	trans = stridexnyp.trans;

	m = stridexnyp.M;
	n = stridexnyp.N;
	lda = stridexnyp.lda;

	alpha = stridexnyp.alpha;
	beta = stridexnyp.beta;

	a = new Float32Array( stridexnyp.A );
	x = new Float32Array( stridexnyp.x );
	y = new Float32Array( stridexnyp.y );

	expected = new Float32Array( stridexnyp.y_out );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, -1, 2, beta, y, 2, 0 );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'the function returns a reference to the second input array', function test( t ) {
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridexpyp.order;
	trans = stridexpyp.trans;

	m = stridexpyp.M;
	n = stridexpyp.N;
	lda = stridexpyp.lda;

	alpha = stridexpyp.alpha;
	beta = stridexpyp.beta;

	a = new Float32Array( stridexpyp.A );
	x = new Float32Array( stridexpyp.x );
	y = new Float32Array( stridexpyp.y );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, 1, 0, beta, y, 1, 0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});

tape( 'if provided an `M` or `N` parameter less than to `0`, the function leaves both input arrays unchanged', function test( t ) {
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var xe;
	var ye;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridexpyp.order;
	trans = stridexpyp.trans;

	m = stridexpyp.M;
	n = stridexpyp.N;
	lda = stridexpyp.lda;

	alpha = stridexpyp.alpha;
	beta = stridexpyp.beta;

	a = new Float32Array( stridexpyp.A );
	x = new Float32Array( stridexpyp.x );
	y = new Float32Array( stridexpyp.y );

	xe = new Float32Array( x.length );
	scopy( x.length, x, 1, xe, 1 );

	ye = new Float32Array( y.length );
	scopy( y.length, y, 1, ye, 1 );

	sgemv( order, trans, -1, n, alpha, a, lda, x, 1, 0, beta, y, 1, 0 );
	t.deepEqual( x, xe, 'returns expected value' );
	t.deepEqual( y, ye, 'returns expected value' );

	sgemv( order, trans, m, -1, alpha, a, lda, x, 1, 0, beta, y, 1, 0 );
	t.deepEqual( x, xe, 'returns expected value' );
	t.deepEqual( y, ye, 'returns expected value' );

	t.end();
});

tape( 'the function supports complex access patterns', function test( t ) {
	var expected;
	var alpha;
	var order;
	var trans;
	var beta;
	var lda;
	var out;
	var a;
	var m;
	var n;
	var x;
	var y;

	order = stridexnyn.order;
	trans = stridexnyn.trans;

	m = stridexnyn.M;
	n = stridexnyn.N;
	lda = stridexnyn.lda;

	alpha = stridexnyn.alpha;
	beta = stridexnyn.beta;

	a = new Float32Array( stridexnyn.A );
	x = new Float32Array( stridexnyn.x );
	y = new Float32Array( stridexnyn.y );

	expected = new Float32Array( stridexnyn.y_out );

	out = sgemv( order, trans, m, n, alpha, a, lda, x, -1, 1, beta, y, -2, 4 );
	isApprox( t, y, expected, 2.0 );
	t.strictEqual( out, y, 'returns expected value' );
	t.end();
});
