// @stdlib/string/len/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2020 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var len = require( './../lib' );


// FIXTURES //

var testData = require( './fixtures/test_data.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof len, 'function', 'main export is a function' );
	t.end();
});

tape( 'the main function throws an error if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			len( value );
		};
	}
});

tape( 'the main function returns 0 if provided an empty string', function test( t ) {
	t.strictEqual( len( '' ), 0, 'returns 0 if provided an empty string' );
	t.end();
});

tape( 'the main function returns the Unicode aware length of a given string', function test( t ) {
	var out;

	out = len( 'hello world' );
	t.strictEqual( out, 11, 'returns length' );

	out = len( '!!!' );
	t.strictEqual( out, 3, 'returns length' );

	out = len( 'à¤…à¤¨à¥à¤šà¥à¤›à¥‡à¤¦' );
	t.strictEqual( out, 5, 'returns Unicode aware length' );

	out = len( 'ðŸŒ·' );
	t.strictEqual( out, 1, 'returns Unicode aware length' );

	t.end();
});

tape( 'the main function returns the Unicode aware length for official Unicode 13 testcases', function test( t ) {
	var out;
	var i;

	i = 0;
	testData.forEach( function iterer( { input, expected } ) {
		out = len( input );
		t.strictEqual( out, expected.length, 'passes Unicode testcase: '+i );
		i += 1;
	});

	t.end();
});
