/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dtrsm = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 5, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( true, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( false, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( null, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( void 0, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( [], 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( {}, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( ( x: number ): number => x, 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 5, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', true, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', false, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', null, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', void 0, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', [], 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', {}, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', ( x: number ): number => x, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 5, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', true, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', false, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', null, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', void 0, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', [], 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', {}, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', ( x: number ): number => x, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 5, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', true, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', false, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', null, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', void 0, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', [], 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', {}, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', ( x: number ): number => x, 'non-unit', 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 5, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', true, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', false, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', null, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', void 0, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', [], 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', {}, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', ( x: number ): number => x, 2, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', '5', 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', true, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', false, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', null, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', void 0, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', [], 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', {}, 2, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', ( x: number ): number => x, 2, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, '5', 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, true, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, false, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, null, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, void 0, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, [], 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, {}, 6.0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, ( x: number ): number => x, 6.0, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, '5', A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, true, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, false, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, null, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, void 0, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, [], A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, {}, A, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, ( x: number ): number => x, A, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a Float64Array...
{
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, '5', 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, 5, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, true, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, false, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, null, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, void 0, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, [], 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, {}, 2, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, ( x: number ): number => x, 2, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, '5', B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, true, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, false, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, null, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, void 0, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, [], B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, {}, B, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, ( x: number ): number => x, B, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, '5', 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 5, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, true, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, false, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, null, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, void 0, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, [], 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, {}, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, ( x: number ): number => x, 2 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, '5' ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, true ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, false ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, null ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, void 0 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, [] ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, {} ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm(); // $ExpectError
	dtrsm( 'row-major' ); // $ExpectError
	dtrsm( 'row-major', 'left' ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper' ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose' ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit' ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2 ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B ); // $ExpectError
	dtrsm( 'row-major', 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, B, 2, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectType Float64Array
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 5, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( true, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( false, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( null, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( void 0, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( [], 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( {}, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( ( x: number ): number => x, 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 5, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', true, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', false, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', null, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', void 0, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', [], 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', {}, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', ( x: number ): number => x, 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 5, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', true, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', false, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', null, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', void 0, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', [], 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', {}, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', ( x: number ): number => x, 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a string...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 5, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', true, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', false, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', null, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', void 0, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', [], 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', {}, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', ( x: number ): number => x, 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', '5', 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', true, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', false, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', null, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', void 0, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', [], 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', {}, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', ( x: number ): number => x, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, '5', 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, true, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, false, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, null, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, void 0, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, [], 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, {}, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, ( x: number ): number => x, 6.0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, '5', A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, true, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, false, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, null, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, void 0, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, [], A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, {}, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, ( x: number ): number => x, A, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a Float64Array...
{
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, '5', 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, 5, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, true, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, false, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, null, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, void 0, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, [], 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, {}, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, ( x: number ): number => x, 2, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, '5', 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, true, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, false, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, null, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, void 0, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, [], 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, {}, 1, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, ( x: number ): number => x, 1, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, '5', 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, true, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, false, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, null, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, void 0, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, [], 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, {}, 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, ( x: number ): number => x, 0, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, '5', B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, true, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, false, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, null, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, void 0, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, [], B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, {}, B, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, ( x: number ): number => x, B, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a twelfth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, '5', 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, 5, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, true, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, false, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, null, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, void 0, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, [], 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, {}, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, ( x: number ): number => x, 2, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a thirteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, '5', 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, true, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, false, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, null, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, void 0, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, [], 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, {}, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, '5', 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, true, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, false, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, null, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, void 0, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, [], 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, {}, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifteenth argument which is not a number...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, '5' ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, true ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, false ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, null ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, void 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, [] ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, {} ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const A = new Float64Array( [ 1.0, 3.0, 0.0, 4.0 ] );
	const B = new Float64Array( [ 5.0, 7.0, 0.0, 8.0 ] );

	dtrsm.ndarray(); // $ExpectError
	dtrsm.ndarray( 'left' ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper' ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose' ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit' ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1 ); // $ExpectError
	dtrsm.ndarray( 'left', 'upper', 'no-transpose', 'non-unit', 2, 2, 6.0, A, 2, 1, 0, B, 2, 1, 0, 10 ); // $ExpectError
}
