// @stdlib/types/ndarray/ctor/test/test.instance.set.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var hasProp = require( '@stdlib/assert/has-property' );
var isFunction = require( '@stdlib/assert/is-function' );
var ctor = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof ctor, 'function', 'main export is a function' );
	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (1d)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;
	var i;

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 4 ];
	order = 'row-major';
	strides = [ 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			arr.set( value, 10.0 );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (2d)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;
	var i;

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		return j;

		function i() {
			arr.set( value, 0, 10.0 );
		}

		function j() {
			arr.set( 0, value, 10.0 );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (3d)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;
	var i;

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 2 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		if ( dim === 1 ) {
			return j;
		}
		return k;

		function i() {
			arr.set( value, 0, 0, 10.0 );
		}

		function j() {
			arr.set( 0, value, 0, 10.0 );
		}

		function k() {
			arr.set( 0, 0, value, 10.0 );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (4d)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;
	var i;

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 2 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 3 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		if ( dim === 1 ) {
			return j;
		}
		if ( dim === 2 ) {
			return k;
		}
		return l;

		function i() {
			arr.set( value, 0, 0, 0, 10.0 );
		}

		function j() {
			arr.set( 0, value, 0, 0, 10.0 );
		}

		function k() {
			arr.set( 0, 0, value, 0, 10.0 );
		}

		function l() {
			arr.set( 0, 0, 0, value, 10.0 );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (>4d)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;
	var i;
	var j;

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 1, 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		for ( j = 0; j < shape.length; j++ ) {
			t.throws( badValue( values[ i ], j ), TypeError, 'throws an error when provided ' + values[ i ] );
		}
	}
	t.end();

	function badValue( value, dim ) {
		return function badValue() {
			var args = new Array( shape.length+1 );
			var i;

			for ( i = 0; i < args.length; i++ ) {
				if ( i === dim ) {
					args[ i ] = value;
				} else {
					args[ i ] = 0;
				}
			}
			args[ i ] = 10.0;
			arr.set.apply( arr, args );
		};
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (1d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 4 ];
	order = 'row-major';
	strides = [ 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 5.0 );
	arr.set( 1, 6.0 );
	arr.set( 2, 7.0 );
	arr.set( 3, 8.0 );

	t.strictEqual( arr.get( 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 2 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 3 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (1d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 4 ];
	order = 'column-major';
	strides = [ 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 5.0 );
	arr.set( 1, 6.0 );
	arr.set( 2, 7.0 );
	arr.set( 3, 8.0 );

	t.strictEqual( arr.get( 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 2 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 3 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'row-major';
	strides = [ 2, -1 ];
	offset = 1;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 6.0, 5.0, 8.0, 7.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'row-major';
	strides = [ -2, 1 ];
	offset = 2;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 7.0, 8.0, 5.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'row-major';
	strides = [ -2, -1 ];
	offset = 3;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 7.0, 6.0, 5.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'column-major';
	strides = [ 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'column-major';
	strides = [ -1, 2 ];
	offset = 1;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 7.0, 5.0, 8.0, 6.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'column-major';
	strides = [ 1, -2 ];
	offset = 2;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 6.0, 8.0, 5.0, 7.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (2d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 2, 2 ];
	order = 'column-major';
	strides = [ -1, -2 ];
	offset = 3;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 5.0 );
	arr.set( 0, 1, 6.0 );
	arr.set( 1, 0, 7.0 );
	arr.set( 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 8.0, 6.0, 7.0, 5.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; row-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 0, 5.0 );
	arr.set( 0, 0, 1, 6.0 );
	arr.set( 0, 1, 0, 7.0 );
	arr.set( 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (3d; column-major)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var arr;
	var f;

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 0, 5.0 );
	arr.set( 0, 0, 1, 6.0 );
	arr.set( 0, 1, 0, 7.0 );
	arr.set( 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});
