// @stdlib/types/ndarray/ctor/test/test.instance.set_no_codegen_4d.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var hasOwnProp = require( '@stdlib/assert/has-own-property' );
var hasProp = require( '@stdlib/assert/has-property' );
var isFunction = require( '@stdlib/assert/is-function' );
var ctor = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof ctor, 'function', 'main export is a function' );
	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method which throws an error if not provided an integer value for a dimension index (4d; codegen=false)', function test( t ) {
	var strides;
	var values;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;
	var i;

	opts = {
		'codegen': false
	};

	dtype = 'float64';
	buffer = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );
	shape = [ 1, 1, 2, 2 ];
	order = 'row-major';
	strides = [ 2, 1, 1, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	values = [
		'5',
		3.14,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ], 0 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 1 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 2 ), TypeError, 'throws an error when provided ' + values[ i ] );
		t.throws( badValue( values[ i ], 3 ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value, dim ) {
		if ( dim === 0 ) {
			return i;
		}
		if ( dim === 1 ) {
			return j;
		}
		if ( dim === 2 ) {
			return k;
		}
		return l;

		function i() {
			arr.set( value, 0, 0, 0, 10.0 );
		}

		function j() {
			arr.set( 0, value, 0, 0, 10.0 );
		}

		function k() {
			arr.set( 0, 0, value, 0, 10.0 );
		}

		function l() {
			arr.set( 0, 0, 0, value, 10.0 );
		}
	}
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (4d; row-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 0, 0, 5.0 );
	arr.set( 0, 0, 0, 1, 6.0 );
	arr.set( 0, 0, 1, 0, 7.0 );
	arr.set( 0, 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (4d; column-major; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 0, 0, 0, 0, 5.0 );
	arr.set( 0, 0, 0, 1, 6.0 );
	arr.set( 0, 0, 1, 0, 7.0 );
	arr.set( 0, 0, 1, 1, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (4d; row-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'wrap',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 1, 2, 2 ];
	order = 'row-major';
	strides = [ 4, 4, 2, 1 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 2, 2, 2, 2, 5.0 );
	arr.set( 30, 20, 10, 11, 6.0 );
	arr.set( 0, 0, -1, -2, 7.0 );
	arr.set( 0, 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 6.0, 7.0, 8.0 ], 'has expected values' );

	t.end();
});

tape( 'an ndarray constructor returns an instance which has a `set` method for setting an array element using subscripts (4d; column-major; mode=wrap; codegen=false)', function test( t ) {
	var strides;
	var buffer;
	var offset;
	var dtype;
	var order;
	var shape;
	var opts;
	var arr;
	var f;

	opts = {
		'mode': 'wrap',
		'codegen': false
	};

	dtype = 'generic';
	buffer = [ 1.0, 2.0, 3.0, 4.0 ];
	shape = [ 1, 1, 2, 2 ];
	order = 'column-major';
	strides = [ 1, 1, 1, 2 ];
	offset = 0;

	f = ctor( dtype, shape.length, opts );
	arr = f( buffer, shape, strides, offset, order );

	t.strictEqual( hasOwnProp( arr, 'set' ), false, 'does not have own property' );
	t.strictEqual( hasProp( arr, 'set' ), true, 'has property' );
	t.strictEqual( isFunction( arr.set ), true, 'has method' );

	arr.set( 2, 2, 2, 2, 5.0 );
	arr.set( 30, 20, 10, 11, 6.0 );
	arr.set( 0, 0, -1, -2, 7.0 );
	arr.set( 0, 0, -5, -11, 8.0 );

	t.strictEqual( arr.get( 0, 0, 0, 0 ), 5.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 0, 1 ), 6.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 0 ), 7.0, 'returns expected value' );
	t.strictEqual( arr.get( 0, 0, 1, 1 ), 8.0, 'returns expected value' );

	t.deepEqual( buffer, [ 5.0, 7.0, 6.0, 8.0 ], 'has expected values' );

	t.end();
});
