<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# doctest-quote-props

> [ESLint rule][eslint-rules] to enforce that property names in return annotations are quoted using single quotes.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/doctest-quote-props' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that property names in return annotations are quoted using single quotes.

**Bad**:

<!-- eslint-disable stdlib/doctest-quote-props -->

```javascript
var copy = require( '@stdlib/utils/copy' );

var value = {
    'a': 1,
    'b': true,
    'c': [ 1, 2, 3 ]
};
var out = copy( value );
/* returns
    {
        a: 1,
        b: true,
        c: [ 1, 2, 3 ]
    }
*/
```

**Good**:

```javascript
var copy = require( '@stdlib/utils/copy' );

var value = {
    'a': 1,
    'b': true,
    'c': [ 1, 2, 3 ]
};
var out = copy( value );
/* returns
    {
        'a': 1,
        'b': true,
        'c': [ 1, 2, 3 ]
    }
*/
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/doctest-quote-props' );

var linter = new Linter();
var result;
var code;

code = [
    ' var target = {',
    '     \'a\': \'beep\'',
    ' };',
    ' var source = {',
    '     \'a\': \'boop\',',
    '     \'b\': \'bap\'',
    ' };',
    '',
    ' var out = merge( target, source );',
    ' // returns { \'a\': \'boop\', b: \'bap\' }'
].join( '\n' );

linter.defineRule( 'doctest-quote-props', rule );

result = linter.verify( code, {
    'rules': {
        'doctest-quote-props': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'doctest-quote-props',
            'severity': 2,
            'message': 'Use single quotes around property names',
            'line': 10,
            'column': 2,
            'nodeType': null,
            'source': ' // returns { \'a\': \'boop\', b: \'bap\' }',
            'endLine': 10,
            'endColumn': 38
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
