// @stdlib/_tools/licenses/insert-header-file-list/test/test.sync.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var IS_BROWSER = require( '@stdlib/assert/is-browser' );
var readFile = require( '@stdlib/fs/read-file' ).sync;
var writeFile = require( '@stdlib/fs/write-file' ).sync;
var unlink = require( '@stdlib/fs/unlink' ).sync;
var exists = require( '@stdlib/fs/exists' ).sync;
var licenseHeader = require( '@stdlib/_tools/licenses/header' );
var insertHeader = require( './../lib/sync.js' );


// VARIABLES //

var opts = {
	'skip': IS_BROWSER
};
var FILE = join( __dirname, 'fixtures', 'file.js.txt' );
var TMP = join( __dirname, 'fixtures', 'tmp.js.txt' );
var EXPECTED_HEADER_STRING = join( __dirname, 'fixtures', 'expected_header_string.js.txt' );
var EXPECTED_HEADER_OBJECT = join( __dirname, 'fixtures', 'expected_header_object.js.txt' );


// FUNCTIONS //

/**
* Creates a temporary file.
*
* @private
* @throws {Error} unexpected error
* @returns {string} temporary file filename
*/
function tmp() {
	var err = writeFile( TMP, readFile( FILE ) );
	if ( err ) {
		throw err;
	}
	return TMP;
} // end FUNCTION tmp()

/**
* Removes a temporary file.
*
* @private
* @throws {Error} unexpected error
*/
function cleanup() {
	var err = unlink( TMP );
	if ( err ) {
		throw err;
	}
} // end FUNCTION cleanup()


// TESTS //

tape( 'main export is a function', opts, function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof insertHeader, 'function', 'main export is a function' );
	t.end();
});

tape( 'if provided a first argument which is not a list of files, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			insertHeader( value, {} );
		};
	}
});

tape( 'if provided a header argument which is not either a string or an object mapping file types to license header strings, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{
			'foo': null
		},
		{
			'js': []
		},
		{
			'js': {}
		},
		{
			'awk': '# @license Apache-2.0',
			'js': '// @license Apache-2.0',
			'jl': 3.14
		},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			insertHeader( [ __filename ], value );
		};
	}
});

tape( 'the function returns an error if an error is encountered when reading a file', opts, function test( t ) {
	var files;
	var err;

	files = [ 'dakfjdlfjsdfjasdjf/aldfkajdlfjsadl.jsdfad' ]; // File should not exist!

	err = insertHeader( files, {} );
	t.strictEqual( err instanceof Error, true, 'returns an error' );

	t.end();
});

tape( 'the function returns an error if an error is encountered when writing a file', opts, function test( t ) {
	var insertHeader;
	var header;
	var files;
	var err;

	insertHeader = proxyquire( './../lib/sync.js', {
		'@stdlib/fs/write-file': {
			'sync': writeFile
		}
	});

	files = [ __filename ];
	header = {
		'js': '// @license Apache-2.0\n'
	};
	err = insertHeader( files, header );
	t.strictEqual( err instanceof Error, true, 'returns an error' );

	t.end();

	function writeFile() {
		return new Error( 'beep' );
	}
});

tape( 'the function returns `null` if provided an empty file array', opts, function test( t ) {
	var err = insertHeader( [], {} );
	t.strictEqual( err, null, 'returns expected value' );
	t.end();
});

tape( 'the function supports providing an empty header string', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp();

	// Create an empty header string:
	header = '';

	// Attempt to insert headers:
	err = insertHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has not changed:
	actual = readFile( f, 'utf8' );
	expected = readFile( FILE, 'utf8' );
	t.strictEqual( actual, expected, 'file is unchanged' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing an empty header object', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp();

	// Create an empty header object:
	header = {};

	// Attempt to insert headers:
	err = insertHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has not changed:
	actual = readFile( f, 'utf8' );
	expected = readFile( FILE, 'utf8' );
	t.strictEqual( actual, expected, 'file is unchanged' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header string', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp();

	// Create a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to insert headers:
	err = insertHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFile( f, 'utf8' );
	expected = readFile( EXPECTED_HEADER_STRING, 'utf8' );
	t.strictEqual( actual, expected, 'file contains header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp();

	// Create a header object:
	header = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to insert headers:
	err = insertHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFile( f, 'utf8' );
	expected = readFile( EXPECTED_HEADER_OBJECT, 'utf8' );
	t.strictEqual( actual, expected, 'file contains header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});

tape( 'the function supports providing a header object containing a `default` header string', opts, function test( t ) {
	var expected;
	var actual;
	var header;
	var err;
	var f;

	// Create a temporary file:
	f = tmp();

	// Create a header object:
	header = {
		'f': licenseHeader( 'Apache-2.0', 'f', {
			'year': 2050,
			'copyright': 'The Stdlib Authors and Contributors'
		}),
		'default': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to insert headers:
	err = insertHeader( [ f ], header );
	t.strictEqual( err, null, 'return null' );

	// Check that the temporary file has changed:
	actual = readFile( f, 'utf8' );
	expected = readFile( EXPECTED_HEADER_OBJECT, 'utf8' );
	t.strictEqual( actual, expected, 'file contains header' );

	// Cleanup:
	cleanup();
	t.strictEqual( exists( f ), false, 'tmp file removed' );

	t.end();
});
