/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import cartesianSquare = require( './index' );


// TESTS //

// The function returns an array of arrays...
{
	cartesianSquare( [ 1, 2, 3, 4 ] ); // $ExpectType number[][]
	cartesianSquare<number>( [ 1, 2, 3, 4 ] ); // $ExpectType number[][]
	cartesianSquare<string>( [ '1', '2', '3', '4' ] ); // $ExpectType string[][]
}

// The compiler throws an error if the function is provided a first argument which is not an array-like object...
{
	cartesianSquare( 1 ); // $ExpectError
	cartesianSquare( true ); // $ExpectError
	cartesianSquare( false ); // $ExpectError
	cartesianSquare( null ); // $ExpectError
	cartesianSquare( void 0 ); // $ExpectError
	cartesianSquare( {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	cartesianSquare(); // $ExpectError
	cartesianSquare( [], [] ); // $ExpectError
	cartesianSquare( [], [], [] ); // $ExpectError
}
