<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Ordered List Marker Value

> [ESLint rule][eslint-rules] to enforce a specified Markdown ordered list marker style in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-ordered-list-marker-value' );
```

#### rule

[ESLint rule][eslint-rules] to enforce a specified Markdown ordered list marker style in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-ordered-list-marker-value, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Beep boop.
*
* 1.  Foo
* 1.  Bar
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Beep boop.
*
* 1.  Foo
* 2.  Bar
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-ordered-list-marker-value]. 

-   `ordered`: increment list item markers by one (default).
-   `single`: list item markers should all have the same starting value.
-   `one`: list item markers should always be equal to `1`.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-ordered-list-marker-value' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep *boop*.',
    '*',
    '* 1.  Foo',
    '* 1.  Bar',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-ordered-list-marker-value', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-ordered-list-marker-value': [ 'error', 'ordered' ]
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-ordered-list-marker-value',
            'severity': 2,
            'message': 'Marker should be `2`, was `1`',
            'line': 5,
            'column': 3,
            'nodeType': null,
            'source': '* 1.  Bar',
            'endLine': 13,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-ordered-list-marker-value]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-ordered-list-marker-value

</section>

<!-- /.links -->
