<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# return-annotations

> [ESLint rule][eslint-rules] to enforce conventions for return annotations.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/return-annotations' );
```

#### rule

[ESLint rule][eslint-rules] to enforce conventions for return annotations. Return annotations may start with `returns`, `throws`, or `=>`. `returns` may only follow variable declarations or assignment expressions, whereas `=>` may only follow `console.log` calls.

**Bad**:

<!-- eslint-disable stdlib/return-annotations -->

```javascript
var x = 3.0;
// => 3.0

console.log( 'Hello World' );
// returns 'Hello World'

// => null
```

**Good**:

```javascript
var x = 3.0;
// returns 3.0

console.log( 'Hello World' );
// => 'Hello World'
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/return-annotations' );

var linter = new Linter();
var result;
var code;

code = [
    'var isLowercase = require( \'./../lib\' );',
    '',
    'console.log( isLowercase( \'hello\' ) );',
    '// returns true',
    '',
    'console.log( isLowercase( \'Hello\' ) );',
    '// => false'
].join( '\n' );

linter.defineRule( 'return-annotations', rule );

result = linter.verify( code, {
    'rules': {
        'return-annotations': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'return-annotations',
            'severity': 2,
            'message': 'Only include `// returns` after variable declarations or assignment expressions (use `=>` after `console.log`)',
            'line': 4,
            'column': 1,
            'nodeType': null,
            'source': '  // returns true',
            'endLine': 4,
            'endColumn': 16
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
