/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import safeCasts = require( './index' );


// TESTS //

// The function returns a casting table..
{
	safeCasts(); // $ExpectType Table
}

// The function returns a list of data types when provided a recognized data type...
{
	safeCasts( 'float32' ); // $ExpectType DataType[]
}

// The function returns null when provided an unrecognized data type...
{
	safeCasts( 'float' ); // $ExpectType null
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	safeCasts( 123 );  // $ExpectError
	safeCasts( true ); // $ExpectError
	safeCasts( false ); // $ExpectError
	safeCasts( {} ); // $ExpectError
	safeCasts( [] ); // $ExpectError
	safeCasts( null ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	safeCasts( 'float32', 3 ); // $ExpectError
}
