// @stdlib/repl/test/test.context.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var deepGet = require( '@stdlib/utils/deep-get' );
var deepSet = require( '@stdlib/utils/deep-set' );
var createContext = require( './../lib/context.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof createContext, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns the input object', function test( t ) {
	var obj;
	var out;

	obj = {};
	out = createContext( obj );

	t.strictEqual( obj, out, 'returns input object' );
	t.end();
});

tape( 'the function sets property paths on the input object', function test( t ) {
	var createContext;
	var expected;
	var actual;
	var obj;
	var ns;

	ns = [
		{
			'alias': 'a.b.c',
			'value': []
		},
		{
			'alias': 'a.d.e',
			'value': {}
		},
		{
			'alias': 'a.b.f',
			'value': []
		},
		{
			'alias': 'g',
			'value': {}
		}
	];

	createContext = proxyquire( './../lib/context.js', {
		'./namespace': ns
	});

	obj = {};
	createContext( obj );

	expected = ns[ 0 ].value;
	actual = deepGet( obj, 'a.b.c' );
	t.strictEqual( actual, expected, 'sets nested value' );

	expected = ns[ 1 ].value;
	actual = deepGet( obj, 'a.d.e' );
	t.strictEqual( actual, expected, 'sets nested value' );

	expected = ns[ 2 ].value;
	actual = deepGet( obj, 'a.b.f' );
	t.strictEqual( actual, expected, 'sets nested value' );

	expected = ns[ 3 ].value;
	actual = deepGet( obj, 'g' );
	t.strictEqual( actual, expected, 'sets nested value' );

	t.end();
});

tape( 'the function throws an error if provided a duplicate alias', function test( t ) {
	var createContext;
	var obj;
	var ns;

	ns = [
		{
			'alias': 'a',
			'value': []
		},
		{
			'alias': 'b',
			'value': {}
		},
		{
			'alias': 'a',
			'value': []
		}
	];

	obj = {};

	createContext = proxyquire( './../lib/context.js', {
		'./namespace': ns
	});

	t.throws( foo, Error, 'throws an error' );

	t.end();

	function foo() {
		createContext( obj );
	}
});

tape( 'the function throws an error if provided a duplicate alias (nested)', function test( t ) {
	var createContext;
	var obj;
	var ns;

	ns = [
		{
			'alias': 'a.b.c',
			'value': []
		},
		{
			'alias': 'a.d.e',
			'value': {}
		},
		{
			'alias': 'a.d.e',
			'value': []
		}
	];

	obj = {};

	createContext = proxyquire( './../lib/context.js', {
		'./namespace': ns
	});

	t.throws( foo, Error, 'throws an error' );

	t.end();

	function foo() {
		createContext( obj );
	}
});

tape( 'the function sets context properties as read-only', function test( t ) {
	var createContext;
	var paths;
	var obj;
	var ns;
	var i;

	ns = [
		{
			'alias': 'a.b.c',
			'value': []
		},
		{
			'alias': 'a.d.e',
			'value': {}
		},
		{
			'alias': 'a.b.f',
			'value': []
		}
	];

	createContext = proxyquire( './../lib/context.js', {
		'./namespace': ns
	});

	obj = {};
	createContext( obj );

	paths = [
		'a.b',
		'a.b.c',
		'a',
		'a.d',
		'a.d.e',
		'a.b.f'
	];

	for ( i = 0; i < paths.length; i++ ) {
		t.throws( setProps( paths[i] ), Error, 'throws an error' );
	}
	t.end();

	function setProps( path ) {
		return function set() {
			deepSet( obj, path, 'beep' );
		};
	}
});
