// @stdlib/repl/test/test.example.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var noop = require( '@stdlib/utils/noop' );
var example = require( './../lib/example.js' );


// FIXTURES //

function namespace() {
	return [
		{
			'alias': 'beep',
			'value': {}
		},
		{
			'alias': 'boop',
			'value': []
		},
		{
			'alias': 'bop',
			'value': {}
		}
	];
}

function examples() {
	return {
		'beep': 'x = "Beep!"',
		'boop': 'y = "Boop!"'
	};
}

function repl() {
	return {
		'write': noop,
		'displayPrompt': noop
	};
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof example, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a function', function test( t ) {
	t.strictEqual( typeof example( repl() ), 'function', 'returns a function' );
	t.end();
});

tape( 'the returned function executes an example (alias)', function test( t ) {
	var example;
	var count;
	var mock;
	var ns;
	var eg;
	var f;

	mock = repl();
	mock.write = write;

	ns = namespace();
	eg = examples();

	example = proxyquire( './../lib/example.js', {
		'./namespace': ns,
		'./examples.js': eg
	});
	f = example( mock );

	count = 0;
	f( 'beep' );

	function write( str ) {
		count += 1;
		if ( count === 1 ) {
			t.strictEqual( str, '\n', 'passes expected value' );
		} else {
			t.strictEqual( str, eg[ 'beep' ], 'passes expected value' );
			t.end();
		}
	}
});

tape( 'the returned function executes an example (known value reference)', function test( t ) {
	var example;
	var count;
	var mock;
	var ns;
	var eg;
	var f;

	mock = repl();
	mock.write = write;

	ns = namespace();
	eg = examples();

	example = proxyquire( './../lib/example.js', {
		'./namespace': ns,
		'./examples.js': eg
	});
	f = example( mock );

	count = 0;
	f( ns[ 1 ].value );

	function write( str ) {
		count += 1;
		if ( count === 1 ) {
			t.strictEqual( str, '\n', 'passes expected value' );
		} else {
			t.strictEqual( str, eg[ 'boop' ], 'passes expected value' );
			t.end();
		}
	}
});

tape( 'if unable to resolve an example, the returned function logs a message (alias)', function test( t ) {
	var example;
	var mock;
	var ns;
	var eg;
	var f;

	mock = {
		'log': log
	};
	ns = namespace();
	eg = examples();

	example = proxyquire( './../lib/example.js', {
		'./namespace': ns,
		'./examples.js': eg,
		'./console.js': mock
	});
	f = example( repl );

	f( 'adkfjadkfjsldjjfasdlfsdj' );

	function log( str ) {
		t.notEqual( str, eg[ 'beep' ], 'passes different string' );
		t.notEqual( str, eg[ 'boop' ], 'passes different string' );
		t.end();
	}
});

tape( 'if unable to resolve an example, the returned function logs a message (value)', function test( t ) {
	var example;
	var mock;
	var ns;
	var eg;
	var f;

	mock = {
		'log': log
	};
	ns = namespace();
	eg = examples();

	example = proxyquire( './../lib/example.js', {
		'./namespace': ns,
		'./examples.js': eg,
		'./console.js': mock
	});
	f = example( repl );

	f( ns[ 2 ].value );

	function log( str ) {
		t.notEqual( str, eg[ 'beep' ], 'passes different string' );
		t.notEqual( str, eg[ 'boop' ], 'passes different string' );
		t.end();
	}
});
