// @stdlib/blas/base/sspr2/test/test.sspr2.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable max-len */

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float32Array = require( '@stdlib/array/float32' );
var sspr2 = require( './../lib/sspr2.js' );


// FIXTURES //

var rl = require( './fixtures/row_major_l.json' );
var ru = require( './fixtures/row_major_u.json' );
var rxpyp = require( './fixtures/row_major_xpyp.json' );
var rxnyp = require( './fixtures/row_major_xnyp.json' );
var rxpyn = require( './fixtures/row_major_xpyn.json' );
var rxnyn = require( './fixtures/row_major_xnyn.json' );

var cl = require( './fixtures/column_major_l.json' );
var cu = require( './fixtures/column_major_u.json' );
var cxpyp = require( './fixtures/column_major_xpyp.json' );
var cxnyp = require( './fixtures/column_major_xnyp.json' );
var cxpyn = require( './fixtures/column_major_xpyn.json' );
var cxnyn = require( './fixtures/column_major_xnyn.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof sspr2, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 9', function test( t ) {
	t.strictEqual( sspr2.length, 9, 'returns expected value' );
	t.end();
});

tape( 'the function throws an error if provided an invalid first argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspr2( value, data.uplo, data.N, data.alpha, new Float32Array( data.x ), data.strideX, new Float32Array( data.y ), data.strideY, data.AP );
		};
	}
});

tape( 'the function throws an error if provided an invalid second argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		'foo',
		'bar',
		'beep',
		'boop'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspr2( data.order, value, data.N, data.alpha, new Float32Array( data.x ), data.strideX, new Float32Array( data.y ), data.strideY, data.AP );
		};
	}
});

tape( 'the function throws an error if provided an invalid third argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		-1,
		-2,
		-3
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspr2( data.order, data.uplo, value, data.alpha, new Float32Array( data.x ), data.strideX, new Float32Array( data.y ), data.strideY, data.AP );
		};
	}
});

tape( 'the function throws an error if provided an invalid sixth argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		0
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspr2( data.order, data.uplo, data.N, data.alpha, new Float32Array( data.x ), value, new Float32Array( data.y ), data.strideY, data.AP );
		};
	}
});

tape( 'the function throws an error if provided an invalid eighth argument', function test( t ) {
	var values;
	var data;
	var i;

	data = rl;

	values = [
		0
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), RangeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sspr2( data.order, data.uplo, data.N, data.alpha, new Float32Array( data.x ), data.strideX, new Float32Array( data.y ), value, data.AP );
		};
	}
});

tape( 'the function performs the symmetric rank 2 operation `A := α*x*y**T + α*y*x**T + A` (row-major, lower)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the symmetric rank 2 operation `A := α*x*y**T + α*y*x**T + A` (column-major, lower)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the symmetric rank 2 operation `A := α*x*y**T + α*y*x**T + A` (row-major, upper)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = ru;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function performs the symmetric rank 2 operation `A := α*x*y**T + α*y*x**T + A` (column-major, upper)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cu;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function returns a reference to the packed form of a symmetric matrix `A`', function test( t ) {
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rl;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );

	t.end();
});

tape( 'if `N` is zero or `α` is zero, respectively, the function returns `AP` unchanged (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = ru;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP );

	out = sspr2( data.order, data.uplo, 0, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	out = sspr2( data.order, data.uplo, data.N, 0.0, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'if `N` is zero or `α` is zero, respectively, the function returns `AP` unchanged (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cu;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP );

	out = sspr2( data.order, data.uplo, 0, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	out = sspr2( data.order, data.uplo, data.N, 0.0, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying `x` and `y` strides (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxpyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying `x` and `y` strides (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxpyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `x` stride (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxnyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `x` stride (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxnyp;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `y` stride (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxpyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports specifying a negative `y` stride (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxpyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports complex access patterns (row-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = rxnyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});

tape( 'the function supports complex access patterns (column-major)', function test( t ) {
	var expected;
	var data;
	var out;
	var ap;
	var x;
	var y;

	data = cxnyn;

	ap = new Float32Array( data.AP );
	x = new Float32Array( data.x );
	y = new Float32Array( data.y );

	expected = new Float32Array( data.AP_out );

	out = sspr2( data.order, data.uplo, data.N, data.alpha, x, data.strideX, y, data.strideY, ap );
	t.strictEqual( out, ap, 'returns expected value' );
	t.deepEqual( out, expected, 'returns expected value' );

	t.end();
});
