// @stdlib/_tools/licenses/update-header-file-list/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var IS_BROWSER = require( '@stdlib/assert/is-browser' );
var readFileSync = require( '@stdlib/fs/read-file' ).sync;
var writeFileSync = require( '@stdlib/fs/write-file' ).sync;
var unlinkSync = require( '@stdlib/fs/unlink' ).sync;
var exists = require( '@stdlib/fs/exists' ).sync;
var noop = require( '@stdlib/utils/noop' );
var licenseHeader = require( '@stdlib/_tools/licenses/header' );
var updateHeader = require( './../lib/main.js' );


// VARIABLES //

var opts = {
	'skip': IS_BROWSER
};
var TMP = join( __dirname, 'fixtures', 'tmp.js.txt' );
var EXPECTED = join( __dirname, 'fixtures', 'expected.js.txt' );
var EXPECTED_SHEBANG = join( __dirname, 'fixtures', 'expected_shebang.js.txt' );
var EXPECTED_NO_HEADER = join( __dirname, 'fixtures', 'expected_no_header.js.txt' );
var SINGLE_LINE_HEADER = join( __dirname, 'fixtures', 'single_line_header.js.txt' );
var MULTI_LINE_HEADER = join( __dirname, 'fixtures', 'multiline_header.js.txt' );
var SHEBANG_SINGLE_LINE_HEADER = join( __dirname, 'fixtures', 'shebang_single_line_header.js.txt' ); // eslint-disable-line id-length
var SHEBANG_MULTI_LINE_HEADER = join( __dirname, 'fixtures', 'shebang_multiline_header.js.txt' ); // eslint-disable-line id-length


// FUNCTIONS //

/**
* Creates a temporary file.
*
* @private
* @param {boolean} bool - boolean indicating whether a temporary file should include a "shebang"
* @param {string} type - header type ('single-line' or 'multi-line')
* @param {string} [f] - filename
* @throws {Error} unexpected error
* @returns {string} temporary file filename
*/
function tmp( bool, type ) {
	var err;
	var out;
	var f;
	if ( arguments.length > 2 ) {
		out = arguments[ 2 ];
	} else {
		out = TMP;
	}
	if ( bool ) {
		if ( type === 'single-line' ) {
			f = SHEBANG_SINGLE_LINE_HEADER;
		} else {
			f = SHEBANG_MULTI_LINE_HEADER;
		}
		err = writeFileSync( out, readFileSync( f ) );
	} else {
		if ( type === 'single-line' ) {
			f = SINGLE_LINE_HEADER;
		} else {
			f = MULTI_LINE_HEADER;
		}
		err = writeFileSync( out, readFileSync( f ) );
	}
	if ( err ) {
		throw err;
	}
	return out;
}

/**
* Removes a temporary file.
*
* @private
* @param {string} [f] - filename
* @throws {Error} unexpected error
*/
function cleanup() {
	var err;
	var f;
	if ( arguments.length ) {
		f = arguments[ 0 ];
	} else {
		f = TMP;
	}
	err = unlinkSync( f );
	if ( err ) {
		throw err;
	}
}


// TESTS //

tape( 'main export is a function', opts, function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof updateHeader, 'function', 'main export is a function' );
	t.end();
});

tape( 'if provided a first argument which is not a list of files, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			updateHeader( value, {}, {}, noop );
		};
	}
});

tape( 'if provided a second argument which is not either a string, regular expression, or an object mapping file types to license header strings or regular expressions, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{
			'foo': null
		},
		{
			'js': []
		},
		{
			'js': {}
		},
		{
			'awk': '# @license Apache-2.0',
			'js': '// @license Apache-2.0',
			'jl': 3.14
		},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			updateHeader( [ __filename ], value, {}, noop );
		};
	}
});

tape( 'if provided a third argument which is not either a string or an object mapping file types to license header strings, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{
			'foo': null
		},
		{
			'js': []
		},
		{
			'js': {}
		},
		{
			'awk': '# @license Apache-2.0',
			'js': '// @license Apache-2.0',
			'jl': 3.14
		},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			updateHeader( [ __filename ], {}, value, noop );
		};
	}
});

tape( 'if provided a callback argument which is not a function, the function throws an error', opts, function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			updateHeader( [ __filename ], {}, {}, value );
		};
	}
});

tape( 'the function returns an error if an error is encountered when reading a file', opts, function test( t ) {
	var files;
	var old;

	files = [ 'dakfjdlfjsdfjasdjf/aldfkajdlfjsadl.jsdfad' ]; // File should not exist!
	old = {
		'jsdfad': 'beep boop'
	};
	updateHeader( files, old, {}, done );

	function done( err ) {
		t.strictEqual( err instanceof Error, true, 'returns an error' );
		t.end();
	}
});

tape( 'the function returns an error if an error is encountered when writing a file', opts, function test( t ) {
	var updateHeader;
	var header;
	var files;

	updateHeader = proxyquire( './../lib/main.js', {
		'@stdlib/fs/write-file': writeFile
	});

	files = [ __filename ];
	header = {
		'js': '// @license Apache-2.0\n'
	};
	updateHeader( files, {}, header, done );

	function done( err ) {
		t.strictEqual( err instanceof Error, true, 'returns an error' );
		t.end();
	}

	function writeFile() {
		var clbk = arguments[ arguments.length-1 ];
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( new Error( 'beep' ) );
		}
	}
});

tape( 'the function supports providing an empty file array', opts, function test( t ) {
	updateHeader( [], {}, {}, done );

	function done( err ) {
		if ( err ) {
			t.fail( err.message );
		}
		t.end();
	}
});

tape( 'the function supports providing empty header strings', opts, function test( t ) {
	var header;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create an empty header string:
	header = '';

	// Attempt to update headers:
	updateHeader( [ f, f, f ], header, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has not changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( SINGLE_LINE_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file is unchanged' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing empty header objects', opts, function test( t ) {
	var header;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create an empty header object:
	header = {};

	// Attempt to update headers:
	updateHeader( [ f, f, f ], header, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has not changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( SINGLE_LINE_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file is unchanged' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header strings', opts, function test( t ) {
	var old;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header string:
	old = '// @license Apache-2.0\n';

	// Attempt to update headers:
	updateHeader( [ f ], old, '', done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_NO_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing a header regular expression', opts, function test( t ) {
	var old;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header regular expression:
	old = /\/\/ @license Apache-2\.0\n\n/;

	// Attempt to update headers:
	updateHeader( [ f ], old, '', done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_NO_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header strings (shebang)', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Specify a header string:
	old = '// @license Apache-2.0\n';

	// Create a header string:
	header = licenseHeader( 'Apache-2.0', 'js', {
		'year': 2018,
		'copyright': 'The Stdlib Authors'
	});

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing a header regular expression (shebang)', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Specify a header regexp:
	old = /\n\/\/ @license Apache-2\.0\n\n/;

	// Create a header string:
	header = licenseHeader( 'Apache-2.0', 'js', {
		'year': 2018,
		'copyright': 'The Stdlib Authors'
	});

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header strings (file lacking filename extension)', opts, function test( t ) {
	var old;
	var f;

	// Create a temporary file without a filename extension:
	f = tmp( false, 'single-line', join( __dirname, 'fixtures', 'tmp' ) );

	// Create a header string:
	old = '// @license Apache-2.0\n';

	// Attempt to update headers:
	updateHeader( [ f ], old, '', done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_NO_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup( f );
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header objects', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file:
	f = tmp( false, 'multi-line' );

	// Create a header object:
	old = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Specify a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing a header object (regular expression)', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file:
	f = tmp( false, 'single-line' );

	// Create a header object:
	old = {
		'txt': /\/\/ @license Apache-2\.0\n\n/
	};

	// Create a header string:
	header = '// @license Apache-2.0\n';

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header object (shebang)', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Create a header object:
	old = {
		'txt': '// @license Apache-2.0\n'
	};

	// Create a header object:
	header = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing a header object (shebang; regular expression)', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file which includes a shebang:
	f = tmp( true, 'single-line' );

	// Create a header object:
	old = {
		'txt': /\n\/\/ @license Apache-2\.0\n\n/
	};

	// Create a header object:
	header = {
		'txt': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_SHEBANG, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing a header object (file lacking filename extension)', opts, function test( t ) {
	var old;
	var f;

	// Create a temporary file without a filename extension:
	f = tmp( false, 'multi-line', join( __dirname, 'fixtures', 'tmp' ) );

	// Create a header object:
	old = {
		'tmp': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Attempt to update headers:
	updateHeader( [ f ], old, '', done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED_NO_HEADER, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup( f );
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});

tape( 'the function supports providing header objects containing a `default` header', opts, function test( t ) {
	var header;
	var old;
	var f;

	// Create a temporary file:
	f = tmp( false, 'multi-line' );

	// Create a header object:
	old = {
		'f': licenseHeader( 'Apache-2.0', 'f', {
			'year': 2050,
			'copyright': 'The Stdlib Authors and Contributors'
		}),
		'default': licenseHeader( 'Apache-2.0', 'js', {
			'year': 2018,
			'copyright': 'The Stdlib Authors'
		})
	};

	// Create a header object:
	header = {
		'f': '!> @license Apache-2.0\n',
		'default': '// @license Apache-2.0\n'
	};

	// Attempt to update headers:
	updateHeader( [ f ], old, header, done );

	function done( err ) {
		var expected;
		var actual;
		if ( err ) {
			t.fail( err.message );
		} else {
			// Check that the temporary file has changed:
			actual = readFileSync( f, 'utf8' );
			expected = readFileSync( EXPECTED, 'utf8' );
			t.strictEqual( actual, expected, 'file does not contain header' );
		}
		// Cleanup:
		cleanup();
		t.strictEqual( exists( f ), false, 'tmp file updated' );

		t.end();
	}
});
