/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import dlange = require( './index' );


// TESTS //

// The function returns a Float64Array...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', 1, 2, 2, A, 2, work ); // $ExpectType number
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 5, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( true, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( false, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( null, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( void 0, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( [], 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( {}, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( ( x: number ): number => x, 'M', 1, 2, 2, A, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a string...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 5, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', true, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', false, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', null, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', void 0, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', [], 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', {}, 1, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', ( x: number ): number => x, 1, 2, 2, A, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', '5', 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', true, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', false, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', null, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', void 0, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', [], 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', {}, 2, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', ( x: number ): number => x, 2, 2, A, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', 1, '5', 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, true, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, false, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, null, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, void 0, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, [], 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, {}, 2, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, ( x: number ): number => x, 2, A, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', 1, 2, '5', A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, true, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, false, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, null, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, void 0, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, [], A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, {}, A, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, ( x: number ): number => x, A, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a Float64Array...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );

	dlange( 'row-major', 'M', 1, 2, 2, '5', 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, 5, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, true, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, false, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, null, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, void 0, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, [], 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, {}, 2, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, ( x: number ): number => x, 2, work ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', 1, 2, 2, A, '5', work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, true, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, false, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, null, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, void 0, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, [], work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, {}, work ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, ( x: number ): number => x, work ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange( 'row-major', 'M', 1, 2, 2, A, 2, '5' ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, 5 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, true ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, false ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, null ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, void 0 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, [] ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, {} ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const work = new Float64Array( [ 0.0, 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange(); // $ExpectError
	dlange( 'row-major' ); // $ExpectError
	dlange( 'row-major', 'M' ); // $ExpectError
	dlange( 'row-major', 'M', 1 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2 ); // $ExpectError
	dlange( 'row-major', 'M', 1, 2, 2, A, 2, work, 10 ); // $ExpectError
}

// Attached to main export is an `ndarray` method which returns a Float64Array...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectType number
}

// The compiler throws an error if the function is provided a first argument which is not a string...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 5, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( true, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( false, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( null, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( void 0, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( [], 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( {}, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( ( x: number ): number => x, 1, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', '5', 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', true, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', false, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', null, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', void 0, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', [], 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', {}, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', ( x: number ): number => x, 2, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, '5', 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, true, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, false, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, null, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, void 0, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, [], 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, {}, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, ( x: number ): number => x, 2, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fourth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, '5', A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, true, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, false, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, null, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, void 0, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, [], A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, {}, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, ( x: number ): number => x, A, 2, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a fifth argument which is not a Float64Array...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, '5', 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, 5, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, true, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, false, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, null, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, void 0, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, [], 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, {}, 2, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, ( x: number ): number => x, 2, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a sixth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, '5', 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, true, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, false, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, null, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, void 0, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, [], 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, {}, 1, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, ( x: number ): number => x, 1, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a seventh argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, '5', 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, true, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, false, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, null, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, void 0, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, [], 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, {}, 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, ( x: number ): number => x, 0, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eighth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, '5', work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, true, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, false, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, null, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, void 0, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, [], work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, {}, work, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, ( x: number ): number => x, work, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a ninth argument which is not a Float64Array...
{
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, '5', 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, 5, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, true, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, false, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, null, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, void 0, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, [], 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, {}, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, ( x: number ): number => x, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a tenth argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, '5', 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, true, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, false, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, null, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, void 0, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, [], 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, {}, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an eleventh argument which is not a number...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, '5' ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, true ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, false ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, null ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, void 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, [] ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, {} ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `ndarray` method is provided an unsupported number of arguments...
{
	const work = new Float64Array( [ 0.0, 0.0 ] );
	const A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0 ] );

	dlange.ndarray(); // $ExpectError
	dlange.ndarray( 'M' ); // $ExpectError
	dlange.ndarray( 'M', 1 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1 ); // $ExpectError
	dlange.ndarray( 'M', 1, 2, 2, A, 2, 1, 0, work, 1, 0, 10 ); // $ExpectError
}
