# no-dynamic-require

> [ESLint rule][eslint-rules] enforcing that `require()` calls have only string literals as arguments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-dynamic-require' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that `require()` calls have only string literals as arguments.

**Bad**:

<!-- eslint-disable stdlib/no-dynamic-require -->

```javascript
var pkg = '@stdlib/math/base/special/betainc';
var betainc = require( pkg );
```

**Good**:

``` javascript 
var betainc = require( '@stdlib/math/base/special/betainc' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-dynamic-require' );

var linter = new Linter();
var result;
var code;

code = [
    'var pkg = \'@stdlib/math/base/special/betainc\';',
    'var betainc = require( pkg );'
].join( '\n' );

linter.defineRule( 'no-dynamic-require', rule );

result = linter.verify( code, {
    'rules': {
        'no-dynamic-require': 'error'
    }
});
/* returns
    [
        {
            ruleId: 'no-dynamic-require',
            severity: 2,
            message: 'require() calls should only use string literals',
            line: 2,
            column: 15,
            nodeType: 'CallExpression',
            source: 'var betainc = require( pkg );',
            endLine: 2,
            endColumn: 29
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
