<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-main-export

> [ESLint rule][eslint-rules] to enforce that the JSDoc comment documenting a package's main export adheres to `stdlib` conventions and contains the correct package path.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-main-export' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that the JSDoc comment documenting a package's main export adheres to `stdlib` conventions and contains the correct package path.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-main-export -->

```javascript
/**
* Tests if a value is an array.
*
* @example
* var bool = isArray( [] );
* // returns true
*
* bool = isArray( {} );
* // returns false
*/
```

**Good**:

```javascript
/**
* Test if a value is an array.
*
* @module @stdlib/assert/is-array
*
* @example
* var isArray = require( '@stdlib/assert/is-array' );
*
* var bool = isArray( [] );
* // returns true
*
* bool = isArray( {} );
* // returns false
*/
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-main-export' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Tests if a value is an array.',
    '*',
    '* @module @stdlib/assert/isarray',
    '*',
    '* @example',
    '* var isArray = require( \'@stdlib/assert/is-aray\' );',
    '*',
    '* var bool = isArray( [] );',
    '* // returns true',
    '*',
    '* bool = isArray( {} );',
    '* // returns false',
    '*/'
].join( '\n' );

linter.defineRule( 'jsdoc-main-export', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-main-export': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-main-export',
            'severity': 2,
            'message': 'Begin description with `Test` instead of `Tests`',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-main-export',
            'severity': 2,
            'message': 'Expected `@stdlib/assert/is-array` for `@module` tag, but encountered `@stdlib/assert/isarray`',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-main-export',
            'severity': 2,
            'message': 'Example code of main export should first require `@stdlib/assert/is-array`, but encountered `@stdlib/assert/is-aray`',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'endLine': 14,
            'endColumn': 3 } ]
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
