/*
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import variance = require( './index' );


// TESTS //

// The function returns a number...
{
	variance( 10, 0.3 ); // $ExpectType number
}

// The compiler throws an error if the function is provided a first argument which is not a number...
{
	variance( true, 0.5 ); // $ExpectError
	variance( false, 0.5 ); // $ExpectError
	variance( '5', 0.5 ); // $ExpectError
	variance( [], 0.5 ); // $ExpectError
	variance( {}, 0.5 ); // $ExpectError
	variance( ( x: number ): number => x, 0.5 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	variance( 8, true ); // $ExpectError
	variance( 8, false ); // $ExpectError
	variance( 8, '5' ); // $ExpectError
	variance( 8, [] ); // $ExpectError
	variance( 8, {} ); // $ExpectError
	variance( 8, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	variance(); // $ExpectError
	variance( 8 ); // $ExpectError
}
