<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# doctest

> [ESLint rule][eslint-rules] to ensure return annotations match the actual output.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/doctest' );
```

#### rule

[ESLint rule][eslint-rules] to ensure return annotations match the actual output. Return annotations start with `returns`, `throws`, or `=>`. `returns` may follow variable declarations or assignment expressions, whereas `=>` follow `console.log` calls.

**Bad**:

<!-- eslint-disable stdlib/doctest -->

```javascript
var x = 3.0;
// returns 2.0

console.log( 'Hello World' );
// => 'hello world'
```

**Good**:

```javascript
var x = 3.0;
// returns 3.0

console.log( 'Hello World' );
// => 'Hello World'
```

Node.js script files, detected by looking for the appropriate shebang at the beginning of the respective file (`#!/usr/bin/env node`), are not linted to prevent unintended side-effects from executing scripts.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/doctest' );

var linter = new Linter();
var result;
var code;

code = [
    'var isLowercase = require( \'@stdlib/assert/is-lowercase\' );',
    '',
    'console.log( isLowercase( \'hello\' ) );',
    '// returns false',
    '',
    'console.log( isLowercase( \'Hello\' ) );',
    '// => true'
].join( '\n' );

linter.defineRule( 'doctest', rule );

result = linter.verify( code, {
    'rules': {
        'doctest': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'doctest',
            'severity': 2,
            'message': 'Displayed return value is `false`, but function returns `true` instead',
            'line': 3,
            'column': 1,
            'nodeType': null,
            'endLine': 3,
            'endColumn': 58
        },
        {
            'ruleId': 'doctest',
            'severity': 2,
            'message': 'Displayed return value is `true`, but function returns `false` instead',
            'line': 6,
            'column': 1,
            'nodeType': null,
            'endLine': 6,
            'endColumn': 50
        }
    ]
*/
```

</section>

<!-- /.examples -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
