// @stdlib/_tools/pkgs/standalones/test/test.async.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var noop = require( '@stdlib/utils/noop' );
var isArray = require( '@stdlib/assert/is-array' );
var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
var ls = require( './../lib/async.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof ls, 'function', 'main export is a function' );
	t.end();
});

tape( 'if provided a `options` argument which is not an object, the function throws an error', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ls( value, noop );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	t.throws( foo, TypeError, 'throws error' );
	t.end();
	function foo() {
		var opts = {
			'dir': null
		};
		ls( opts, noop );
	}
});

tape( 'if provided a callback argument which is not a function, the function throws an error (no options)', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ls( value );
		};
	}
});

tape( 'if provided a callback argument which is not a function, the function throws an error (options)', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ls( {}, value );
		};
	}
});

tape( 'the function returns an error to a provided callback if an error is encountered while searching a directory', function test( t ) {
	var ls = proxyquire( './../lib/async.js', {
		'@stdlib/_tools/pkgs/names': pkgNames
	});

	ls( clbk );

	function pkgNames() {
		var cb = arguments[ arguments.length-1 ];
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			cb( new Error( 'beep' ) );
		}
	}

	function clbk( error ) {
		t.ok( error, 'returns an error' );
		t.end();
	}
});

tape( 'the function returns a string array', function test( t ) {
	ls( clbk );
	function clbk( error, names ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.equal( isStringArray( names ), true, 'returns a string array' );
		t.end();
	}
});

tape( 'the function returns a string array (dir option)', function test( t ) {
	var opts = {
		'dir': './@stdlib/math/base'
	};
	ls( opts, clbk );
	function clbk( error, names ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.equal( isStringArray( names ), true, 'returns a string array' );
		t.end();
	}
});

tape( 'the function returns a string array (pattern option)', function test( t ) {
	var opts = {
		'pattern': '**/math/**/package.json'
	};
	ls( opts, clbk );
	function clbk( error, names ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.equal( isStringArray( names ), true, 'returns a string array' );
		t.end();
	}
});

tape( 'the function returns an empty array if unable to resolve any packages (ignore option)', function test( t ) {
	var opts = {
		'dir': './@stdlib/math',
		'ignore': [
			'**/@stdlib/math/**' // same as the directory we are searching!
		]
	};
	ls( opts, clbk );
	function clbk( error, names ) {
		if ( error ) {
			t.ok( false, error.message );
		}
		t.equal( isArray( names ), true, 'returns an array' );
		t.equal( names.length, 0, 'returns an empty array' );
		t.end();
	}
});
