// @stdlib/fs/resolve-parent-paths/test/test.sync.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var resolve = require( 'path' ).resolve;
var basename = require( 'path' ).basename;
var tape = require( 'tape' );
var IS_BROWSER = require( '@stdlib/assert/is-browser' );
var cwd = require( '@stdlib/process/cwd' );
var resolveParentPaths = require( './../lib/sync.js' );


// VARIABLES //

// Don't run tests in the browser...for now...
var opts = {
	'skip': IS_BROWSER // FIXME
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof resolveParentPaths, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided a `paths` argument which is not an array of strings', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		{},
		function noop() {},
		'beep',
		[ 1, 2 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			resolveParentPaths( value );
		};
	}
});

tape( 'the function throws an error if provided a `paths` argument which is not an array of strings (options)', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		{},
		function noop() {},
		'beep',
		[ 1, 2 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			resolveParentPaths( value, {} );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		null,
		void 0,
		true,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			resolveParentPaths( [ 'beep' ], value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			resolveParentPaths( [ 'beep' ], {
				'dir': value
			});
		};
	}
});

tape( 'the function resolves a path from a set of paths by walking parent directories', opts, function test( t ) {
	var expected;
	var actual;
	var base;
	var dir;

	dir = cwd();
	base = basename( dir );

	expected = dir;
	actual = resolveParentPaths( [ base ] );
	t.strictEqual( actual, expected, 'returns path' );

	expected = dir;
	actual = resolveParentPaths( [ 'beep-boop!!!hello world!?!', base ] );
	t.strictEqual( actual, expected, 'returns path' );

	expected = dir;
	actual = resolveParentPaths( [ base, 'beep-boop!!!hello world!?!' ] );
	t.strictEqual( actual, expected, 'returns path' );

	t.end();
});

tape( 'the function resolves a path from a set of paths by walking parent directories (dir option)', opts, function test( t ) {
	var expected;
	var actual;
	var opts;
	var dir;

	opts = {
		'dir': __dirname
	};
	dir = resolve( __dirname, '../package.json' );

	expected = dir;
	actual = resolveParentPaths( [ 'package.json' ], opts );
	t.strictEqual( actual, expected, 'returns path' );

	expected = dir;
	actual = resolveParentPaths( [ 'beep-boop!!!hello world!?!', 'package.json' ], opts );
	t.strictEqual( actual, expected, 'returns path' );

	expected = dir;
	actual = resolveParentPaths( [ 'package.json', 'beep-boop!!!hello world!?!' ], opts );
	t.strictEqual( actual, expected, 'returns path' );

	t.end();
});

tape( 'the function returns `null` if unable to resolve a parent path', opts, function test( t ) {
	var actual = resolveParentPaths( [ 'beep-boop!!!hello world!?!' ] );

	t.strictEqual( actual, null, 'returns null' );
	t.end();
});

tape( 'the function returns `null` if unable to resolve a parent path (options)', opts, function test( t ) {
	var actual;
	var opts;

	opts = {
		'dir': __dirname
	};
	actual = resolveParentPaths( [ 'beep-boop!!!hello world!?!' ], opts );

	t.strictEqual( actual, null, 'returns null' );
	t.end();
});
