// @stdlib/lapack/base/dptcon/test/test.dptcon.js
/**
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var Float64Array = require( '@stdlib/array/float64' );
var EPS = require( '@stdlib/constants/float64/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var dptcon = require( './../lib/dptcon.js' );


// FUNCTIONS //

/**
* Tests for element-wise approximate equality.
*
* @private
* @param {Object} t - test object
* @param {Collection} actual - actual values
* @param {Collection} expected - expected values
* @param {number} rtol - relative tolerance
*/
function isApprox( t, actual, expected, rtol ) {
	var delta;
	var tol;
	var i;

	t.strictEqual( actual.length, expected.length, 'returns expected value' );
	for ( i = 0; i < expected.length; i++ ) {
		if ( actual[ i ] === expected[ i ] ) {
			t.strictEqual( actual[ i ], expected[ i ], 'returns expected value' );
		} else {
			delta = abs( actual[ i ] - expected[ i ] );
			tol = rtol * EPS * abs( expected[ i ] );
			t.ok( delta <= tol, 'within tolerance. actual: '+actual[ i ]+'. expected: '+expected[ i ]+'. delta: '+delta+'. tol: '+tol+'.' );
		}
	}
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof dptcon, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 6', function test( t ) {
	t.strictEqual( dptcon.length, 6, 'returns expected value' );
	t.end();
});

tape( 'the function correctly computest the reciprocal of condition number', function test( t ) {
	var expectedW;
	var rcond;
	var work;
	var out;
	var D;
	var E;

	D = new Float64Array( [ 3.0, 2.0, 1.0 ] );
	E = new Float64Array( [ 1.0, 1.0 ] );
	rcond = new Float64Array( 1 );
	work = new Float64Array( 3 );

	expectedW = new Float64Array( [ 4.333333333333333, 4.0, 3.0 ] );

	out = dptcon( 3, D, E, 3.0, rcond, work );
	t.strictEqual( out, 0, 'returns expected value' );
	t.strictEqual( rcond[ 0 ], 0.07692307692307693, 'returns expected value' );
	isApprox( t, work, expectedW, 1.0 );
	t.end();
});

tape( 'the function returns -1 if provided a negative value for `N`', function test( t ) {
	var out;
	var D;

	out = dptcon( -1, D, D, 3.0, D, D );
	t.strictEqual( out, -1, 'returns expected value' );
	t.end();
});

tape( 'the function returns -4 if provided a negative value for `anorm`', function test( t ) {
	var out;
	var D;

	out = dptcon( 3, D, D, -3.0, D, D );
	t.strictEqual( out, -4, 'returns expected value' );
	t.end();
});

tape( 'the function sets rcond to 1 if `N` is 0', function test( t ) {
	var rcond;
	var out;

	rcond = new Float64Array( 1 );
	out = dptcon( 0, new Float64Array( 0 ), new Float64Array( 0 ), 3.0, rcond, new Float64Array( 0 ) );
	t.strictEqual( out, 0, 'returns expected value' );
	t.strictEqual( rcond[ 0 ], 1.0, 'returns expected value' );
	t.end();
});

tape( 'the function sets rcond to 0 if `anorm` is 0', function test( t ) {
	var rcond;
	var out;

	rcond = new Float64Array( 1 );
	out = dptcon( 3, new Float64Array( 3 ), new Float64Array( 2 ), 0.0, rcond, new Float64Array( 3 ) );
	t.strictEqual( out, 0, 'returns expected value' );
	t.strictEqual( rcond[ 0 ], 0.0, 'returns expected value' );
	t.end();
});
