/*
* @license Apache-2.0
*
* Copyright (c) 2022 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import filledBy = require( './index' );

/**
* Callback function.
*
* @param i - current array index
* @returns return value
*/
function clbk( i: number ): number {
	return i;
}


// TESTS //

// The function returns an array...
{
	filledBy( 3, clbk ); // $ExpectType any[]
	filledBy( 3, clbk, {} ); // $ExpectType any[]
}

// The compiler throws an error if the function is provided a first argument which is not a number...
{
	filledBy( 'abc', clbk ); // $ExpectError
	filledBy( true, clbk ); // $ExpectError
	filledBy( false, clbk ); // $ExpectError
	filledBy( null, clbk ); // $ExpectError
	filledBy( [], clbk ); // $ExpectError
	filledBy( {}, clbk ); // $ExpectError
	filledBy( ( x: number ): number => x, clbk ); // $ExpectError

	filledBy( 'abc', clbk, {} ); // $ExpectError
	filledBy( true, clbk, {} ); // $ExpectError
	filledBy( false, clbk, {} ); // $ExpectError
	filledBy( null, clbk, {} ); // $ExpectError
	filledBy( [], clbk, {} ); // $ExpectError
	filledBy( {}, clbk, {} ); // $ExpectError
	filledBy( ( x: number ): number => x, clbk, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a callback...
{
	filledBy( 3, 'abc' ); // $ExpectError
	filledBy( 3, true ); // $ExpectError
	filledBy( 3, false ); // $ExpectError
	filledBy( 3, null ); // $ExpectError
	filledBy( 3, [] ); // $ExpectError
	filledBy( 3, {} ); // $ExpectError

	filledBy( 3, 'abc', {} ); // $ExpectError
	filledBy( 3, true, {} ); // $ExpectError
	filledBy( 3, false, {} ); // $ExpectError
	filledBy( 3, null, {} ); // $ExpectError
	filledBy( 3, [], {} ); // $ExpectError
	filledBy( 3, {}, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	filledBy(); // $ExpectError
	filledBy( 3 ); // $ExpectError
	filledBy( 3, clbk, 2, 2 ); // $ExpectError
}
