// @stdlib/math/special/abs/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* eslint-disable stdlib/capitalized-comments */

'use strict';

// MODULES //

var tape = require( 'tape' );
var uniform = require( '@stdlib/random/base/uniform' ).factory;
var Float64Array = require( '@stdlib/array/float64' );
var Float32Array = require( '@stdlib/array/float32' );
var ndarray = require( '@stdlib/ndarray/ctor' );
var shape2strides = require( '@stdlib/ndarray/base/shape2strides' );
var base = require( '@stdlib/math/base/special/abs' );
var abs = require( './../lib' );


// VARIABLES //

var rand = uniform( -10.0, 10.0 );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof abs, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function has an arity of 1', function test( t ) {
	t.strictEqual( abs.length, 1, 'has expected arity' );
	t.end();
});

tape( 'the function throws an error if provided a value other than a number, array-like object, or ndarray', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		void 0,
		true,
		false,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			abs( value );
		};
	}
});

tape( 'the function computes the absolute value (number)', function test( t ) {
	var v;
	var i;

	for ( i = 0; i < 100; i++ ) {
		v = rand();
		t.strictEqual( abs( v ), base( v ), 'returns expected value when provided ' + v );
	}
	t.end();
});

tape( 'the function computes the absolute value for each element in an array-like object', function test( t ) {
	var expected;
	var x;
	var v;
	var i;

	// dtype: float64
	x = new Float64Array( 10 );
	expected = new Float64Array( x.length );
	for ( i = 0; i < x.length; i++ ) {
		v = rand();
		x[ i ] = v;
		expected[ i ] = base( v );
	}
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	// dtype: float32
	x = new Float32Array( 10 );
	expected = new Float32Array( x.length );
	for ( i = 0; i < x.length; i++ ) {
		v = rand();
		x[ i ] = v;
		expected[ i ] = base( v );
	}
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	// dtype: generic
	x = [];
	expected = [];
	for ( i = 0; i < 10; i++ ) {
		v = rand();
		x.push( v );
		expected.push( base( v ) );
	}
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	// dtype: generic (array-like object)
	x = {
		'length': 10
	};
	expected = [];
	for ( i = 0; i < x.length; i++ ) {
		v = rand();
		x[ i ] = v;
		expected.push( base( v ) );
	}
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	t.end();
});

tape( 'the function computes the absolute value for each element in an ndarray', function test( t ) {
	var expected;
	var ctor;
	var xbuf;
	var ebuf;
	var sh;
	var st;
	var x;
	var i;

	// dtype: float64
	ctor = ndarray( 'float64', 3 );

	xbuf = new Float64Array( 100 );
	for ( i = 0; i < xbuf.length; i++ ) {
		xbuf[ i ] = rand();
	}
	sh = [ 25, 1, 4 ];
	st = shape2strides( sh, 'row-major' );
	x = ctor( xbuf, sh, st, 0, 'row-major' );

	ebuf = new Float64Array( x.length );
	for ( i = 0; i < ebuf.length; i++ ) {
		ebuf[ i ] = base( xbuf[ i ] );
	}
	expected = ctor( ebuf, sh, st, 0, 'row-major' );
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	// dtype: float32
	ctor = ndarray( 'float32', 3 );

	xbuf = new Float32Array( 100 );
	for ( i = 0; i < xbuf.length; i++ ) {
		xbuf[ i ] = rand();
	}
	sh = [ 25, 1, 4 ];
	st = shape2strides( sh, 'row-major' );
	x = ctor( xbuf, sh, st, 0, 'row-major' );

	ebuf = new Float32Array( x.length );
	for ( i = 0; i < ebuf.length; i++ ) {
		ebuf[ i ] = base( xbuf[ i ] );
	}
	expected = ctor( ebuf, sh, st, 0, 'row-major' );
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	// dtype: generic
	ctor = ndarray( 'generic', 3 );

	xbuf = [];
	for ( i = 0; i < 100; i++ ) {
		xbuf.push( rand() );
	}
	sh = [ 25, 1, 4 ];
	st = shape2strides( sh, 'row-major' );
	x = ctor( xbuf, sh, st, 0, 'row-major' );

	ebuf = [];
	for ( i = 0; i < xbuf.length; i++ ) {
		ebuf.push( base( xbuf[ i ] ) );
	}
	expected = ctor( ebuf, sh, st, 0, 'row-major' );
	t.deepEqual( abs( x ), expected, 'returns expected value' );

	t.end();
});

// TODO: add tests for complex numbers
