// @stdlib/math/statistics/ttest2/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var incrspace = require( '@stdlib/math/utils/incrspace' );
var abs = require( '@stdlib/math/base/special/abs' );
var NINF = require( '@stdlib/math/constants/float64-ninf' );
var PINF = require( '@stdlib/math/constants/float64-pinf' );
var EPS = require( '@stdlib/math/constants/float64-eps' );
var ttest2 = require( './../lib' );


// FUNCTIONS //

function almostEqual( a, b, tol ) {
	var delta = abs( a - b );
	tol *= abs( b );
	return ( delta <= tol );
}


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof ttest2, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the `x` argument is not a numeric array', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		[ 1, 2, 'not a numeric array, hehe' ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ttest2( value, [1, 2, 3] );
		};
	}
});

tape( 'the function throws an error if the `y` argument is not a numeric array', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		[ 1, 2, 'not a numeric array, hehe' ],
		true,
		undefined,
		null,
		NaN,
		function noop() {},
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ttest2( [1, 2, 3], value );
		};
	}
});

tape( 'the function throws an error if provided an invalid `alternative` option', function test( t ) {
	var values;
	var i;

	values = [
		5,
		true,
		undefined,
		null,
		NaN,
		[],
		{}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}

	t.throws(
		badValue( 'not one of less, greater or two-sided' ),
		Error,
		'throws an error when provided not less, greater or two-sided'
	);
	t.end();

	function badValue( value ) {
		return function badValue() {
			ttest2( [ 0.2, 0.5, 0.75 ], [ 0.5, 0.4, 0.3 ], {
				'alternative': value
			});
		};
	}
});

tape( 'the function throws an error if provided a `variance` option besides `equal` or `unequal`', function test( t ) {
	var values;
	var i;

	values = [
		'same',
		'different'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), Error, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ttest2( [ 0.2, 0.5, 0.75 ], [ 0.5, 0.4, 0.3 ], {
				'variance': value
			});
		};
	}
});

tape( 'the function throws an error if the `alpha` option is a numeric value outside `[0,1]`', function test( t ) {
	var values;
	var i;

	values = [
		1.1,
		-0.1
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			ttest2( [ 0.2, 0.5, 0.75 ], [ 0.5, 0.5, 0.9 ], {
				'alpha': value
			});
		};
	}
});

tape( 'the function by default performs a Welch two-sample t-test assuming unequal population variances', function test( t ) {
	var out;
	var a;
	var b;

	a = incrspace( 1, 11, 1 );
	b = incrspace( 7, 21, 1 );
	out = ttest2( a, b );

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.000018552818325118146, 2.0 * EPS ), 'returns correct p-value' );
	t.ok( almostEqual( out.statistic, -5.43492976389405857, 2.0 * EPS ), 'returns correct test statistic' );

	t.ok( almostEqual( out.ci[0], -11.05280172515816162, 2.0 * EPS ), 'returns correct lower bound of CI' );
	t.ok( almostEqual( out.ci[1], -4.94719827484183927, 2.0 * EPS ), 'returns correct upper bound of CI' );

	t.end();
});

tape( 'the function performs a two-sample t-test assuming equal population variances if `variance` option is set to `equal`', function test( t ) {
	var out;
	var a;
	var b;

	a = [ 2, 1, 0, 4, 3 ];
	b = [ 1, 2, 3, 5, 0, 0, 2, 1, 2 ];
	out = ttest2( a, b, {
		'variance': 'equal'
	});

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.803898298238838, 2.0 * EPS ), 'returns correct p-value' );
	t.ok( almostEqual( out.statistic, 0.253863804753896805, 2.0 * EPS ), 'returns correct test statistic' );

	t.ok( almostEqual( out.ci[0], -1.68502339431739845, 2.0 * EPS ), 'returns correct lower bound of CI' );
	t.ok( almostEqual( out.ci[1], 2.12946783876184309, 2.0 * EPS ), 'returns correct upper bound of CI' );

	t.end();
});

tape( 'the function performs a two-sample t-test assuming the true difference in means being equal to the value of `difference` option', function test( t ) {
	var out;
	var a;
	var b;

	a = [ 2, 1, 0, 4, 3 ];
	b = [ 4, 2, 8, 0, 6 ];
	out = ttest2( a, b, {
		'difference': -2.0
	});

	// Tested against R:
	t.equal( out.pValue, 1.0, 'returns correct p-value' );
	t.equal( out.statistic, 0.0, 'returns correct test statistic' );

	t.ok( almostEqual( out.ci[0], -5.887741643736976, 2.0 * EPS ), 'returns correct lower bound of CI' );
	t.ok( almostEqual( out.ci[1], 1.8877416437369758, 2.0 * EPS ), 'returns correct upper bound of CI' );

	t.end();
});

tape( 'the function performs a one-sided Welch two-sample t-test assuming unequal population variances', function test( t ) {
	var out;
	var a;
	var b;

	a = incrspace( 1, 11, 1 );
	b = incrspace( 7, 21, 1 );
	out = ttest2( a, b, {
		'alternative': 'less'
	});

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.00000927640916255905, 3e-15 ), 'returns correct p-value' );
	t.ok( almostEqual( out.statistic, -5.43492976389405857, 2.0 * EPS ), 'returns correct test statistic' );

	t.equal( out.ci[0], NINF, 'returns -Infinity as lower bound of CI' );
	t.ok( almostEqual( out.ci[1], -5.4723420893421153, 2.0 * EPS ), 'returns correct upper bound of CI' );

	out = ttest2( a, b, {
		'alternative': 'greater'
	});

	// Tested against R:
	t.ok( almostEqual( out.pValue, 0.999990723590837, 6.0 * EPS ), 'returns correct p-value' );
	t.ok( almostEqual( out.statistic, -5.43492976389405857, 2.0 * EPS ), 'returns correct test statistic' );

	t.ok( almostEqual( out.ci[0], -10.5276579106578865, 2.0 * EPS ), 'returns correct lower bound of CI' );
	t.equal( out.ci[1], PINF, 'returns -Infinity as upper bound of CI' );

	t.end();
});

tape( 'the function returns an object with a .print() method for generating a formatted output of results of two-sample t-test', function test( t ) {
	var table;
	var out;
	var x;
	var y;

	x = [ 0, 1, 2, 3, 4 ];
	y = [ 3, 2, 1, 5, 6, 2, 3 ];
	out = ttest2( x, y );
	table = out.print();
	/*
		Welch two-sample t-test

		Alternative hypothesis: True difference in means is 0

			pValue: 0.2697
			statistic: -1.1732
			95% confidence interval: [-3.3333,1.0476]

		Test Decision: Fail to reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );

	out = ttest2( x, y, {
		'alternative': 'less'
	});
	table = out.print();
	/*
		Welch two-sample t-test

		Alternative hypothesis: True difference in means is less than 0

			pValue: 0.1348
			statistic: -1.1732
			95% confidence interval: [-Infinity,0.6349]

		Test Decision: Fail to reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );

	out = ttest2( x, y, {
		'alternative': 'greater'
	});
	table = out.print();
	/*
		Welch two-sample t-test

		Alternative hypothesis: True difference in means is greater than 0

			pValue: 0.8652
			statistic: -1.1732
			95% confidence interval: [-2.9206,Infinity]

		Test Decision: Fail to reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that accepts a `numdigits` option to control the number of decimal degits displayed', function test( t ) {
	var table;
	var out;
	var x;
	var y;

	x = [ 0, 1, 2, 3, 4, 6, 9, 6, 4, 3 ];
	y = [ 2, 3, 2, 1, 2, 2, 1, 1 ];
	out = ttest2( x, y );
	table = out.print({
		'numdigits': 6
	});
	/*
		Welch two-sample t-test

		Alternative hypothesis: True difference in means is not equal to 0

			pValue: 0.040258
			statistic: 2.337457
			95% confidence interval: [0.109742,3.990258]

		Test Decision: Reject null in favor of alternative at 5% significance level
	*/
	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});

tape( 'the function returns an object with a .print() method that accepts an `options` object', function test( t ) {
	var table;
	var out;
	var x;
	var y;

	x = [ 0, 1, 2, 3, 4 ];
	y = [ 2, 3, 2, 1, 2 ];
	out = ttest2( x, y );
	table = out.print( {} );

	t.equal( typeof table, 'string', 'returns a pretty-printed table' );
	t.end();
});


tape( 'the function returns an object with a .print() method that throws an error if `options` is not a simple object', function test( t ) {
	var values;
	var out;
	var i;
	var x;
	var y;

	x = [ 0, 1, 2, 3, 4 ];
	y = [ 2, 3, 2, 1, 2 ];
	out = ttest2( x, y );

	values = [
		'abc',
		4,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print( value );
		};
	}
});

tape( 'the function returns an object with a .print() method that throws an error if the `numdigits` option is not a positive integer', function test( t ) {
	var values;
	var out;
	var i;
	var x;
	var y;

	x = [ 0, 1, 2, 3, 4 ];
	y = [ 2, 3, 2, 1, 2 ];
	out = ttest2( x, y );

	values = [
		'abc',
		2.4,
		0.0,
		-1.0,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			out.print({
				'numdigits': value
			});
		};
	}
});
