<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-require-throws-tags

> [ESLint rule][eslint-rules] enforcing that JSDoc comments of functions are not missing `@throws` tags.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-require-throws-tags' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that JSDoc comments of functions are not missing `@throws` tags.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-require-throws-tags -->

```javascript
var ceil = require( '@stdlib/math/base/special/ceil' );


// MAIN //

/**
* Generates a linearly spaced numeric array using an increment of one.
*
* @param {number} x1 - first array value
* @param {number} x2 - array element bound
* @returns {Array} linearly spaced numeric array
*/
function incrspace( x1, x2 ) {
    var arr;
    var len;
    var inc;
    var i;
    if ( !isNumber( x1 ) || isnan( x1 ) ) {
        throw new TypeError( 'invalid argument. Start must be numeric. Value: `' + x1 + '`.' );
    }
    if ( !isNumber( x2 ) || isnan( x2 ) ) {
        throw new TypeError( 'invalid argument. Stop must be numeric. Value: `' + x2 + '`.' );
    }
    inc = 1;
    len = ceil( ( x2-x1 ) / inc );
    if ( len > MAX_LENGTH ) {
        throw new RangeError( 'invalid arguments. Generated array exceeds maximum array length.' );
    }
    if ( len <= 1 ) {
        return [ x1 ];
    }
    arr = [];
    arr.push( x1 );
    for ( i = 1; i < len; i++ ) {
        arr.push( x1 + (inc*i) );
    }
    return arr;
}
```

**Good**:

```javascript
var ceil = require( '@stdlib/math/base/special/ceil' );


// MAIN //

/**
* Generates a linearly spaced numeric array using an increment of one.
*
* @param {number} x1 - first array value
* @param {number} x2 - array element bound
* @throws {TypeError} first argument must be numeric
* @throws {TypeError} second argument must be numeric
* @throws {RangeError} length of created array must be less than `4294967295` (`2**32 - 1`)
* @returns {Array} linearly spaced numeric array
*/
function incrspace( x1, x2 ) {
    var arr;
    var len;
    var inc;
    var i;
    if ( !isNumber( x1 ) || isnan( x1 ) ) {
        throw new TypeError( 'invalid argument. Start must be numeric. Value: `' + x1 + '`.' );
    }
    if ( !isNumber( x2 ) || isnan( x2 ) ) {
        throw new TypeError( 'invalid argument. Stop must be numeric. Value: `' + x2 + '`.' );
    }
    inc = 1;
    len = ceil( ( x2-x1 ) / inc );
    if ( len > MAX_LENGTH ) {
        throw new RangeError( 'invalid arguments. Generated array exceeds maximum array length.' );
    }
    if ( len <= 1 ) {
        return [ x1 ];
    }
    arr = [];
    arr.push( x1 );
    for ( i = 1; i < len; i++ ) {
        arr.push( x1 + (inc*i) );
    }
    return arr;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-require-throws-tags' );

var linter = new Linter();
var result;

var code = [
    ' /**',
    ' * Attempts to parse a string as JSON.',
    ' *',
    ' * @param {string} str - string to parse',
    ' * @param {Function} reviver - transformation function',
    ' * @returns {(*|Error)} parsed value or parse error',
    ' *',
    ' * @example',
    ' * var obj = parseJSON( \'{"beep":"boop"}\' );',
    ' * // returns {\'beep\':\'boop\'}',
    ' */',
    ' function parseJSON( str, reviver ) {',
    '   if ( !isString( str ) ) {',
    '     throw new TypeError( \'invalid argument. First argument must be a string. Value: `\' + str + \'`.\' );',
    '   }',
    '   if ( arguments.length > 1 ) {',
    '     if ( !isFunction( reviver ) ) {',
    '       throw new TypeError( \'invalid argument. Reviver argument must be a function. Value: `\' + reviver + \'`.\' );',
    '     }',
    '   }',
    '   try {',
    '     return JSON.parse( str, reviver );',
    '   } catch ( error ) {',
    '     return error;',
    '   }',
    ' }'
].join( '\n' );

linter.defineRule( 'jsdoc-require-throws-tags', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-require-throws-tags': [ 'error' ]
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-require-throws-tags',
            'severity': 2,
            'message': '`parseJSON` is missing `@throws {TypeError}` annotation(s)',
            'line': 12,
            'column': 2,
            'nodeType': 'null',
            'source': ' function parseJSON( str, reviver ) {',
            'endLine': 26,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
