// @stdlib/_tools/tests/html/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2018 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var join = require( 'path' ).join;
var tape = require( 'tape' );
var unlink = require( '@stdlib/fs/unlink' ).sync;
var readFile = require( '@stdlib/fs/read-file' ).sync;
var noop = require( '@stdlib/utils/noop' );
var bundle = require( './../lib' );


// FIXTURES //

var EXPECTED = readFile( join( __dirname, 'fixtures', 'expected.txt' ) );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof bundle, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the first argument is not a bundle URL string', function test( t ) {
	var values;
	var i;
	values = [
		5,
		NaN,
		null,
		void 0,
		true,
		{},
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			bundle( value, noop );
		};
	}
});

tape( 'the function throws an error if provided a callback argument which is not a function', function test( t ) {
	var values;
	var i;
	values = [
		5,
		'abc',
		NaN,
		null,
		void 0,
		true,
		{},
		[]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			bundle( '/foo/bar/beep.js', value );
		};
	}
});

tape( 'the function creates an HTML file for running tests and calls the callback with the generated output', function test( t ) {
	bundle( '/foo/bar/beep.js', onRead );

	function onRead( error, output ) {
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		t.deepEqual( output.toString(), EXPECTED.toString(), 'returns HTML' );
		t.end();
	}
});

tape( 'the function creates an HTML file for running tests and writes the results to disk', function test( t ) {
	var opath = join( __dirname, 'fixtures', 'tests.html' );
	var opts = {
		'out': opath
	};
	bundle( '/foo/bar/beep.js', opts, onWrite );

	function onWrite( error ) {
		var out;
		if ( error ) {
			t.ok( false, 'did not expect an error' );
		}
		out = readFile( opath );
		t.deepEqual( out.toString(), EXPECTED.toString(), 'written file contains expected contents' );

		// Remove created file:
		unlink( opath );
		t.end();
	}
});

tape( 'the function invokes the callback with an error if writing to disk fails', function test( t ) {
	var opath = join( __dirname, 'fixtures', 'nonexisting', 'tests.html' );
	var opts = {
		'out': opath
	};
	bundle( '/foo/bar/beep.js', opts, onWrite );

	function onWrite( error ) {
		if ( error ) {
			t.ok( true, error.message );
		} else {
			t.ok( false, 'expected an error' );
		}
		t.end();
	}
});
