// @stdlib/string/reverse/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var reverseString = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof reverseString, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			reverseString( value );
		};
	}
});

tape( 'the function returns an empty string if provided an empty string', function test( t ) {
	t.strictEqual( reverseString( '' ), '', 'returns empty string' );
	t.end();
});

tape( 'the function reverses a string', function test( t ) {
	var out;

	out = reverseString( 'hello world' );
	t.strictEqual( out, 'dlrow olleh', 'reverses a string' );

	out = reverseString( '!!!' );
	t.strictEqual( out, '!!!', 'reverses a string' );

	out = reverseString( 'Hello World' );
	t.strictEqual( out, 'dlroW olleH', 'reverses a string' );

	t.end();
});
