<!--

@license Apache-2.0

Copyright (c) 2022 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# require-leading-slash

> [ESLint rule][eslint-rules] disallowing `require()` calls of relative paths which do not begin with a leading `./`.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-leading-slash' );
```

#### rule

[ESLint rule][eslint-rules] disallowing `require()` calls of relative paths which do not begin with a leading `./`.

**Bad**:

<!-- run-disable -->

<!-- eslint-disable stdlib/require-leading-slash -->

```javascript
var stdlib = require( '../lib/index.js' );
```

**Good**:

```javascript
var stdlib = require( './../lib/index.js' );
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-leading-slash' );

var linter = new Linter();
var result;

var code = 'var stdlib = require( \'../lib/index.js\' );';

linter.defineRule( 'require-leading-slash', rule );

result = linter.verify( code, {
    'rules': {
        'require-leading-slash': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'require-leading-slash',
            'severity': 2,
            'message': 'require() call contains a relative path which does not begin with a leading `./`',
            'line': 1,
            'column': 14,
            'nodeType': 'CallExpression',
            'source': 'var stdlib = require( \'../lib/index.js\' );',
            'endLine': 1,
            'endColumn': 42
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
