<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-no-duplicate-tags

> [ESLint rule][eslint-rules] to enforce that a JSDoc tag is only used once in a JSDoc comment.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-duplicate-tags' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that a JSDoc tag is only used once in a JSDoc comment. By default, the following tags may appear multiple times in a JSDoc comment:

-   @arg
-   @argument
-   @augments
-   @author
-   @borrows
-   @example
-   @extends
-   @param
-   @property
-   @see
-   @throws
-   @todo
-   @tutorial

To specify the list of tags which are allowed to appear multiple times, set the `exceptions` option.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-no-duplicate-tags, valid-jsdoc -->

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @returns {number} x squared
* @returns {number} x squared
*
* @example
* var y = square( 2.0 );
* // returns 4.0
*
* @example
* var y = square( 4.0 );
* // returns 16.0
*/
function square( x ) {
    return x*x;
}
```

**Good**:

```javascript
/**
* Squares a number.
*
* @param {number} x - input number
* @returns {number} x squared
    *
* @example
* var y = square( 2.0 );
* // returns 4.0
*
* @example
* var y = square( 4.0 );
* // returns 16.0
*/
function square( x ) {
    return x*x;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-duplicate-tags' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Squares a number.',
    '* ',
    '* @params {number} x - input number',
    '* @returns {number} x squared',
    '* @returns {number} x squared',
    '*',
    '* @example',
    '* var y = square( 2.0 );',
    '* // returns 4.0',
    '*/',
    'function square( x ) {',
    '  return x*x;',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-no-duplicate-tags', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-no-duplicate-tags': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-no-duplicate-tags',
            'severity': 2,
            'message': '"@returns" should only be used once',
            'line': 1,
            'column': 1,
            'nodeType': null,
            'source': '/**',
            'endLine': 11,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
