<!--

@license Apache-2.0

Copyright (c) 2023 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# No Multiple Blank Lines

> [ESLint rule][eslint-rules] to prevent multiple blank lines in JSDoc comments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-multiple-blank-lines' );
```

#### rule

[ESLint rule][eslint-rules] to prevent multiple blank lines in JSDoc comments.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-no-multiple-blank-lines -->

```javascript
/**
* Boop beep.
*
*
* @returns {string} a value
*
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Boop beep.
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-no-multiple-blank-lines' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*',
    '*',
    '* @example',
    '* var out = beep( "baz" );',
    '* // returns "beepbaz"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-no-multiple-blank-lines', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-no-multiple-blank-lines': [ 'error', 2 ]
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-no-multiple-blank-lines',
            'severity': 2,
            'message': 'JSDoc comments should not have multiple subsequent blank lines',
            'line': 1,
            'column': 1,
            'nodeType': 'Block',
            'endLine': 16,
            'endColumn': 3,
            'fix': {
                'range': [...],
                'text': '...'
            }
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
