// @stdlib/math/ml/online-binary-classification/test/test.learner.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var exp = require( '@stdlib/math/base/special/exp' );
var onlineBinaryClassification = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof onlineBinaryClassification, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if the `options` object is not a simple object', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		void 0,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function call() {
			onlineBinaryClassification( value );
		};
	}
});

tape( 'the `loss` option of the function specifies the used loss function', function test( t ) {
	var values;
	var len;
	var i;

	values = [
		'hinge',
		'log',
		'modifiedHuber',
		'perceptron',
		'squaredHinge'
	];
	len = values.length;

	for ( i = 0; i < len; i++ ) {
		onlineBinaryClassification({
			'loss': values[ i ]
		});
	}

	t.end();
});

tape( 'the function throws an error if supplied an unknown `loss` option', function test( t ) {
	t.throws( badValue, Error, 'throws an error when provided unknown loss' );
	t.end();

	function badValue() {
		onlineBinaryClassification({
			'loss': 'hilbert'
		});
	}
});

tape( 'the `learningRate` option can be used to set the strength schedule of the gradient descent algorithm', function test( t ) {
	var values;
	var len;
	var i;

	values = [
		'basic',
		'constant',
		'pegasos'
	];
	for ( i = 0; i < len; i++ ) {
		onlineBinaryClassification({
			'learningRate': values[ i ]
		});
	}
	t.end();
});

tape( 'the constructor throws an error if supplied an unknown `learningRate` option', function test( t ) {
	t.throws( badValue, Error, 'throws an error when provided unknown learning rate' );
	t.end();

	function badValue() {
		onlineBinaryClassification({
			'learningRate': 'scaling'
		});
	}
});

tape( 'the function has an `intercept` option determining whether the classification model should include an implicit intercept / bias term', function test( t ) {
	var model;

	model = onlineBinaryClassification({
		'intercept': true
	});
	model.update( [ 1.0, 1.0, 1.0 ], -1.0 );
	t.equal( model.coefs.length, 4, 'the model contains a coefficient for the intercept' );

	model = onlineBinaryClassification({
		'intercept': false
	});
	model.update( [ 1.0, 1.0, 1.0 ], 1.0 );
	t.equal( model.coefs.length, 3, 'the model does not contain an intercept coefficient' );

	t.end();
});

tape( 'the `update` method throws an error if the `x` value is not an array', function test( t ) {
	var values;
	var model;
	var i;
	var y;

	model = onlineBinaryClassification();
	values = [
		'abc',
		5,
		null,
		true,
		void 0,
		NaN,
		{},
		function noop() {}
	];
	y = 1;

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function update() {
			model.update( value, y );
		};
	}
});

tape( 'the `update` method throws an error if `x` is an array with an invalid number of elements', function test( t ) {
	var model;
	var i;
	var y;
	var x;

	model = onlineBinaryClassification();
	y = 1;
	x = [ 1.0, 1.0 ];
	model.update( x, y );

	for ( i = 0; i < 10; i++ ) {
		x.push( 1.0 );
		t.throws( badValue, TypeError, 'throws an error when provided an array of any length other than two' );
	}
	t.end();

	function badValue() {
		model.update( x, y );
	}
});

tape( 'the `update` method throws an error if the `y` value is not either +1 or -1', function test( t ) {
	var values;
	var model;
	var i;

	model = onlineBinaryClassification();
	values = [
		'abc',
		5,
		null,
		true,
		void 0,
		NaN,
		{},
		function noop() {},
		0,
		3.2,
		-1.2
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function update() {
			model.update( [ 2.0, 3.0 ], value );
		};
	}
});

tape( 'by default, the `predict` method calculates the inner product of the weights and feature vector `x`', function test( t ) {
	var expected;
	var weights;
	var actual;
	var model;
	var i;
	var x;

	model = onlineBinaryClassification();
	model.update( [ 0.0, 0.0 ], 1 );
	weights = model.coefs;

	x = [ 2.0, 2.0 ];
	actual = model.predict( x );

	// Add intercept...
	x.push( 1.0 );
	expected = 0;
	for ( i = 0; i < weights.length; i++ ) {
		expected += x[ i ] * weights[ i ];
	}
	t.deepEqual( actual, expected, 'deep equal' );

	t.end();
});

tape( 'the `predict` method calculates prediction probabilities when `type` is set to `probability`', function test( t ) {
	var expected;
	var weights;
	var actual;
	var model;
	var wx;
	var i;
	var x;

	model = onlineBinaryClassification();
	model.update( [ 0.0, 0.0 ], 1 );
	weights = model.coefs;

	x = [ 2.0, 2.0 ];
	actual = model.predict( x, 'probability' );
	// Add intercept...
	x.push( 1.0 );
	wx = 0;
	for ( i = 0; i < weights.length; i++ ) {
		wx += x[ i ] * weights[ i ];
	}
	expected = 1.0 / ( 1.0 + exp( -wx ) );

	t.deepEqual( actual, expected, 'deep equal' );

	t.end();
});

tape( 'the `predict` method throws an error if the `x` value is not an array', function test( t ) {
	var values;
	var model;
	var i;

	model = onlineBinaryClassification();
	values = [
		'abc',
		5,
		null,
		true,
		void 0,
		NaN,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function predict() {
			model.predict( value );
		};
	}
});

tape( 'the `predict` method throws an error if the `x` is an array with an invalid number of elements', function test( t ) {
	var model;
	var i;
	var x;

	model = onlineBinaryClassification();
	x = [ 1.0, 1.0 ];
	model.update( x, -1 );

	for ( i = 0; i < 10; i++ ) {
		x.push( 1.0 );
		t.throws( badValue, TypeError, 'throws an error when provided an array of any length other than two' );
	}
	t.end();

	function badValue() {
		model.predict( x );
	}
});

tape( 'the `predict` method throws an error if the `type` value is not either `link` or `probability`', function test( t ) {
	var values;
	var model;
	var i;

	model = onlineBinaryClassification();
	model.update( [ 2.0, 3.0 ], -1 );

	values = [
		'abc',
		5,
		null,
		true,
		void 0,
		NaN,
		{},
		function noop() {},
		'strings'
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function predict() {
			model.predict( [ 1.0, 1.0 ], value );
		};
	}
});

tape( 'the `predict` method throws an error if the `type` is `probability` for a model with `log` or `modifiedHuber` loss', function test( t ) {
	var model;

	model = onlineBinaryClassification({
		'loss': 'hinge'
	});
	model.update( [ 2.0, 3.0 ], -1 );

	t.throws( badValue(), Error, 'throws an error' );
	t.end();

	function badValue() {
		return function predict() {
			model.predict( [ 1.0, 2.0 ], 'probability' );
		};
	}
});
