<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# empty-line-before-comment

> [ESLint rule][eslint-rules] to enforce empty lines before comments.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/empty-line-before-comment' );
```

#### rule

[ESLint rule][eslint-rules] to enforce empty lines before comments.

**Bad**:

<!-- eslint-disable stdlib/empty-line-before-comment -->

```javascript
function square( x ) {
    var out;
    // Square the number:
    out = x*x;
    return out;
}
```

**Good**:

```javascript
function square( x ) {
    var out;

    // Square the number:
    out = x*x;
    return out;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/empty-line-before-comment' );

var linter = new Linter();
var result;
var code;

code = [
    'function square( x ) {',
    '  var out;',
    '  // Square the number:',
    '  out = x*x;',
    '  return out;',
    '}'
].join( '\n' );

linter.defineRule( 'empty-line-before-comment', rule );

result = linter.verify( code, {
    'rules': {
        'empty-line-before-comment': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'empty-line-before-comment',
            'severity': 2,
            'message': 'Missing empty line before comment',
            'line': 3,
            'column': 3,
            'nodeType': null,
            'source': '  // Square the number:',
            'endLine': 3,
            'endColumn': 24
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
