// @stdlib/_tools/utils/uncapitalize-pkg-description/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var uncapitalizeDescription = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof uncapitalizeDescription, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if not provided a string', function test( t ) {
	var values;
	var i;

	values = [
		5,
		null,
		true,
		void 0,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			uncapitalizeDescription( value );
		};
	}
});

tape( 'the function returns an empty string if provided an empty string', function test( t ) {
	t.equal( uncapitalizeDescription( '' ), '', 'returns empty string' );
	t.end();
});

tape( 'the function turns the first character of a `stdlib` package description to lowercase', function test( t ) {
	var descr;
	var out;

	descr = 'Replace search occurrences with a replacement string.';
	out = uncapitalizeDescription( descr );
	t.equal( out, 'replace search occurrences with a replacement string.', 'turns first character lowercase' );

	descr = 'Poisson distribution constructor.';
	out = uncapitalizeDescription( descr );
	t.equal( out, descr, 'leaves proper nouns intact' );

	descr = 'Test for deep equality between two values.';
	out = uncapitalizeDescription( descr );
	t.equal( out, 'test for deep equality between two values.', 'turns first character lowercase' );

	t.end();
});
