<!--

@license Apache-2.0

Copyright (c) 2023 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# first-unit-test

> [ESLint rule][eslint-rules] to enforce that a [tape][tape] file starts with the expected test.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/first-unit-test' );
```

#### rule

[ESLint rule][eslint-rules] to ensure that each [tape][tape] file starts with the expected test. Specifically, the file should start with a test for the main export, with the first line of this test calling `t.ok( true, __filename )` per project conventions. The rule helps to enforce a consistent structure for tests across the project and makes it easier to identify the file associated with a failing test.

**Bad**:

<!-- eslint-disable -->

```javascript
var tape = require( 'tape' );
var identity = require( '@stdlib/utils/identity-function' );

tape( 'the function works correctly', function test( t ) {
    t.strictEqual( identity( true ), true, 'returns true' );
    t.end();
});
```

**Good**:

<!-- eslint-disable -->

```javascript
var tape = require( 'tape' );
var identity = require( '@stdlib/utils/identity-function' );

tape( 'main export is a function', function test( t ) {
    t.ok( true, __filename );
    t.strictEqual( typeof identity, 'function', 'main export is a function' );
    t.end();
});
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint-disable  node/no-unpublished-require -->

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/first-unit-test' );

var linter = new Linter();
var result;
var config;
var code;

code = [
    '// MODULES //',
    '',
    'var tape = require( \'tape\' );',
    'var noop = require( \'./../lib\' );',
    '',
    '',
    '// TESTS //',
    '',
    'tape( \'main export is a function\', function test( t ) {',
    '    t.strictEqual( typeof noop, \'function\', \'main export is a function\' );',
    '    t.end();',
    '});'
].join( '\n' );

linter.defineRule( 'first-unit-test', rule );

config = {
    'rules': {
        'first-unit-test': 'error'
    }
};
result = linter.verify( code, config, {
    'filename': 'test.factory.js'
});
console.log( result );
/* =>
    {
        'ruleId': 'first-unit-test',
        'severity': 2,
        'message': 'The first test in a test file should start with a call to `t.ok( true, __filename );`',
        'line': 3,
        'column': 1,
        'nodeType': 'Program',
        'endLine': 12,
        'endColumn': 4,
        'fix': { 'range': [...], 'text': 't.ok( true, __filename );\n' }
    }
*/

```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[tape]: https://github.com/ljharb/tape

</section>

<!-- /.links -->
