<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-doctest-quote-props

> [ESLint rule][eslint-rules] to enforce that property names in return annotations inside of example code are quoted using single quotes.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest-quote-props' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that property names in return annotations inside of example code are quoted using single quotes.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-doctest, stdlib/jsdoc-doctest-quote-props -->

```javascript
var hasOwnProp = require( '@stdlib/assert/has-own-property' );

/**
* Converts the first letter of each object key to uppercase.
*
* @param {Object} obj - source object
* @returns {Object} new object
*
* @example
* var obj1 = {
*     'aa': 1,
*     'bb': 2
* };
*
* var obj2 = capitalizeKeys( obj1 );
* // returns { "Aa": 1, Bb: 2 }
*/
function capitalizeKeys( obj ) {
    var out;
    var key;
    var k;
    out = {};
    for ( key in obj ) {
        if ( hasOwnProp( obj, key ) ) {
            if ( key === '' ) {
                out[ key ] = obj[ key ];
            } else {
                k = key.charAt( 0 ).toUpperCase() + key.slice( 1 );
                out[ k ] = obj[ key ];
            }
        }
    }
    return out;
}
```

**Good**:

```javascript
var hasOwnProp = require( '@stdlib/assert/has-own-property' );

/**
* Converts the first letter of each object key to uppercase.
*
* @param {Object} obj - source object
* @returns {Object} new object
*
* @example
* var obj1 = {
*     'aa': 1,
*     'bb': 2
* };
*
* var obj2 = capitalizeKeys( obj1 );
* // returns { 'Aa': 1, 'Bb': 2 }
*/
function capitalizeKeys( obj ) {
    var out;
    var key;
    var k;
    out = {};
    for ( key in obj ) {
        if ( hasOwnProp( obj, key ) ) {
            if ( key === '' ) {
                out[ key ] = obj[ key ];
            } else {
                k = key.charAt( 0 ).toUpperCase() + key.slice( 1 );
                out[ k ] = obj[ key ];
            }
        }
    }
    return out;
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-doctest-quote-props' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Copies or deep clones a value to an arbitrary depth.',
    '*',
    '* @param {*} value - value to copy',
    '* @param {NonNegativeInteger} [level=+infinity] - copy depth',
    '* @returns {*} value copy',
    '*',
    '* @example',
    '* var value = [',
    '*   {',
    '*     \'a\': 1,',
    '*     \'b\': true,',
    '*     \'c\': [ 1, 2, 3 ]',
    '*   }',
    '* ];',
    '* var out = copy( value );',
    '* // returns [ { a: 1, b: true, c: [ 1, 2, 3 ] } ]',
    '*/',
    'function copy( value, level ) {',
    '  var out;',
    '  if ( arguments.length > 1 ) {',
    '    if ( !isNonNegativeInteger( level ) ) {',
    '      throw new TypeError( \'invalid argument. `level` must be a nonnegative integer. Value: `\' + level + \'`.\' );',
    '    }',
    '    if ( level === 0 ) {',
    '      return value;',
    '    }',
    '  } else {',
    '    level = PINF;',
    '  }',
    '  out = ( isArray( value ) ) ? new Array( value.length ) : {};',
    '  return deepCopy( value, out, [value], [out], level );',
    '}'
].join( '\n' );

linter.defineRule( 'jsdoc-doctest-quote-props', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-doctest-quote-props': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-doctest-quote-props',
            'severity': 2,
            'message': 'Use single quotes around property names',
            'line': 19,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'source': 'function copy( value, level ) {',
            'endLine': 33,
            'endColumn': 2
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
