// @stdlib/assert/deep-has-own-property/test/test.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var deepHasOwn = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof deepHasOwn, 'function', 'export is a function' );
	t.end();
});

tape( 'the function exports a factory function', function test( t ) {
	t.strictEqual( typeof deepHasOwn.factory, 'function', 'exports a factory function' );
	t.end();
});

tape( 'the function returns `false` if provided `null` or `non-object` for the first argument', function test( t ) {
	var values;
	var i;

	values = [
		'3',
		3,
		NaN,
		true,
		false,
		null,
		void 0,
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.strictEqual( deepHasOwn( values[ i ], 'a.b.c' ), false, 'returns false when provided '+values[ i ] );
	}
	t.end();
});

tape( 'the function throws an error if provided a key path argument which is neither a string primitive or a key array', function test( t ) {
	var values;
	var i;

	values = [
		3,
		NaN,
		true,
		false,
		null,
		void 0,
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws error when provided '+values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var obj = {
				'a': {
					'b': 'c'
				}
			};
			deepHasOwn( obj, value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object', function test( t ) {
	var values;
	var i;

	values = [
		'3',
		3,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws error when provided '+values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var obj = {
				'a': {
					'b': 'c'
				}
			};
			deepHasOwn( obj, 'a.b', value );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		3,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws error when provided '+values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var opts;
			var obj;

			obj = {
				'a': {
					'b': 'c'
				}
			};
			opts = {
				'sep': value
			};
			deepHasOwn( obj, 'a.b', opts );
		};
	}
});

tape( 'the function returns `true` if an object has a specified path (delimited string)', function test( t ) {
	var bool;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	bool = deepHasOwn( obj, 'a.b.c' );
	t.strictEqual( bool, true, 'returns true' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	bool = deepHasOwn( obj, 'a.0.b.0.c' );
	t.strictEqual( bool, true, 'returns true' );

	t.end();
});

tape( 'the function returns `true` if an object has a specified path (key array)', function test( t ) {
	var bool;
	var path;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	path = [ 'a', 'b', 'c' ];
	bool = deepHasOwn( obj, path );
	t.strictEqual( bool, true, 'returns true' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	path = [ 'a', '0', 'b', '0', 'c' ];
	bool = deepHasOwn( obj, path );
	t.strictEqual( bool, true, 'returns true' );

	t.end();
});

tape( 'the function returns `true` if an object has a specified path (non-default delimiter)', function test( t ) {
	var bool;
	var opts;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	opts = {
		'sep': '/'
	};
	bool = deepHasOwn( obj, 'a/b/c', opts );
	t.strictEqual( bool, true, 'returns true' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	opts = {
		'sep': '/'
	};
	bool = deepHasOwn( obj, 'a/0/b/0/c', opts );
	t.strictEqual( bool, true, 'returns true' );

	t.end();
});

tape( 'the function returns `false` if an object does not have a specified path (delimited string)', function test( t ) {
	var bool;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	bool = deepHasOwn( obj, 'a.b.c.d.e' );
	t.strictEqual( bool, false, 'returns false' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	bool = deepHasOwn( obj, '0.a.0.b.0.e' );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if an object does not have a specified path (key array)', function test( t ) {
	var bool;
	var path;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	path = [ 'a', 'b', 'c', 'd', 'e' ];
	bool = deepHasOwn( obj, path );
	t.strictEqual( bool, false, 'returns false' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	path = [ '0', 'a', '0', 'b', '0', 'e' ];
	bool = deepHasOwn( obj, path );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if an object does not have a specified path (non-default delimiter)', function test( t ) {
	var bool;
	var opts;
	var obj;

	obj = {
		'a': {
			'b': {
				'c': 'd'
			}
		}
	};
	opts = {
		'sep': '/'
	};
	bool = deepHasOwn( obj, 'a/b/c/d/e', opts );
	t.strictEqual( bool, false, 'returns false' );

	obj = {
		'a': [
			{
				'b': [
					{
						'c': 'd'
					},
					{
						'e': 'f'
					}
				]
			}
		]
	};
	opts = {
		'sep': '/'
	};
	bool = deepHasOwn( obj, '0/a/0/b/0/e', opts );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if provided an empty object (no options)', function test( t ) {
	var bool;

	bool = deepHasOwn( {}, 'a.b.c' );
	t.strictEqual( bool, false, 'returns false' );

	bool = deepHasOwn( {}, '' );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if provided an empty object (options)', function test( t ) {
	var bool;
	var opts;

	opts = {
		'sep': '/'
	};

	bool = deepHasOwn( {}, 'a/b/c', opts );
	t.strictEqual( bool, false, 'returns false' );

	bool = deepHasOwn( {}, '', opts );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if provided an empty path (no options)', function test( t ) {
	var bool;
	var obj;

	obj = {
		'a': 'b'
	};
	bool = deepHasOwn( obj, '' );
	t.strictEqual( bool, false, 'returns false' );

	bool = deepHasOwn( {}, '' );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});

tape( 'the function returns `false` if provided an empty path (options)', function test( t ) {
	var bool;
	var opts;
	var obj;

	opts = {
		'sep': '/'
	};
	obj = {
		'a': 'b'
	};
	bool = deepHasOwn( obj, '', opts );
	t.strictEqual( bool, false, 'returns false' );

	bool = deepHasOwn( {}, '', opts );
	t.strictEqual( bool, false, 'returns false' );

	t.end();
});
