// @stdlib/_tools/pkgs/wasm/test/test.sync.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var resolve = require( 'path' ).resolve;
var proxyquire = require( 'proxyquire' );
var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
var isArray = require( '@stdlib/assert/is-array' );
var findWASM = require( './../lib/sync.js' );


// VARIABLES //

var dir = resolve( __dirname, '..', '..' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof findWASM, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	t.throws( foo, TypeError, 'throws error' );
	t.end();
	function foo() {
		var opts = {
			'dir': null
		};
		findWASM( opts );
	}
});

tape( 'the function returns a string array (if at least one package containing WebAssembly is detected)', function test( t ) {
	var findWASM;
	var count;
	var pkgs;
	var opts;

	findWASM = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		},
		'@stdlib/fs/read-json': {
			'sync': readJSON
		}
	});

	count = 0;
	opts = {
		'dir': dir
	};
	pkgs = findWASM( opts );

	t.strictEqual( isStringArray( pkgs ), true, 'returns a string array' );
	t.strictEqual( pkgs.length, 1, 'expected length' );

	t.end();

	function glob() {
		return [
			'/beep/boop/package.json',
			'/a/b/c/d/e/f/g/h/beep/boop/bop/package.json'
		];
	}

	function readJSON() {
		count += 1;
		if ( count < 2 ) {
			return {
				'__stdlib__': {
					'wasm': true
				}
			};
		}
		return {};
	}
});

tape( 'the function returns an empty array if unable to resolve any packages containing WebAssembly', function test( t ) {
	var findWASM;
	var pkgs;

	findWASM = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		},
		'@stdlib/fs/read-json': {
			'sync': readJSON
		}
	});

	pkgs = findWASM();

	t.strictEqual( isArray( pkgs ), true, 'returns an array' );
	t.strictEqual( pkgs.length, 0, 'returns an empty array' );

	t.end();

	function glob() {
		return [ '/beep/boop/package.json' ];
	}

	function readJSON() {
		return {};
	}
});

tape( 'the function returns an empty array if unable to resolve any packages', function test( t ) {
	var findWASM;
	var pkgs;

	findWASM = proxyquire( './../lib/sync.js', {
		'glob': {
			'sync': glob
		}
	});

	pkgs = findWASM();

	t.strictEqual( isArray( pkgs ), true, 'returns an array' );
	t.strictEqual( pkgs.length, 0, 'returns an empty array' );

	t.end();

	function glob() {
		return [];
	}
});
