// @stdlib/utils/try-then-async/lib/try_then_async.js
'use strict';

// MODULES //

var isFunction = require( '@stdlib/assert/is-function' );


// MAIN //

/**
* If a function does not return an error, invokes a callback with the function result; otherwise, invokes a second function `y`.
*
* @param {Function} x - function to invoke
* @param {Function} y - function to invoke if `x` returns an error
* @param {Function} done - callback to invoke upon completion
* @throws {TypeError} first argument must be a function
* @throws {TypeError} second argument must be a function
* @throws {TypeError} last argument must be a function
*
* @example
* var randu = require( '@stdlib/math/base/random/randu' );
*
* function x( clbk ) {
*     setTimeout( onTimeout, 0 );
*     function onTimeout() {
*         if ( randu() > 0.5 ) {
*             return clbk( null, 1.0 );
*         }
*         clbk( new Error( 'beep' ) );
*     }
* }
*
* function y( clbk ) {
*     setTimeout( onTimeout, 0 );
*     function onTimeout() {
*         clbk( null, -1.0 );
*     }
* }
*
* function done( error, result ) {
*     if ( error ) {
*         throw error;
*     }
*     console.log( result );
* }
*
* trythenAsync( x, y, done );
*/
function trythenAsync( x, y, done ) {
	if ( !isFunction( x ) ) {
		throw new TypeError( 'invalid input argument. First argument must be a function. Value: `'+x+'`.' );
	}
	if ( !isFunction( y ) ) {
		throw new TypeError( 'invalid input argument. Second argument must be a function. Value: `'+y+'`.' );
	}
	if ( !isFunction( done ) ) {
		throw new TypeError( 'invalid input argument. Last argument must be a function. Value: `'+done+'`.' );
	}
	x( clbk1 );

	/**
	* Callback invoked by `x`.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {...*} args - function results
	* @returns {void}
	*/
	function clbk1( error ) {
		var nargs;
		var args;
		var i;
		if ( error ) {
			if ( y.length === 1 ) {
				return y( clbk2 );
			}
			return y( error, clbk2 );
		}
		nargs = arguments.length;
		args = new Array( nargs );
		args[ 0 ] = null;
		for ( i = 1; i < nargs; i++ ) {
			args[ i ] = arguments[ i ];
		}
		return done.apply( null, args );
	} // end FUNCTION clbk1()

	/**
	* Callback invoked by `y`.
	*
	* @private
	* @param {(Error|null)} error - error object
	* @param {...*} args - function results
	* @returns {void}
	*/
	function clbk2( error ) {
		var nargs;
		var args;
		var i;
		if ( error ) {
			return done( error );
		}
		nargs = arguments.length;
		args = new Array( nargs );
		args[ 0 ] = null;
		for ( i = 1; i < nargs; i++ ) {
			args[ i ] = arguments[ i ];
		}
		return done.apply( null, args );
	} // end FUNCTION clbk2()
} // end FUNCTION trythenAsync()


// EXPORTS //

module.exports = trythenAsync;
