// @stdlib/math/generics/random/sample/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var isNumberArray = require( '@stdlib/utils/is-number' ).isNumberArray;
var isStringArray = require( '@stdlib/utils/is-string' ).isStringArray;
var factory = require( './../lib/factory.js' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an argument which is neither array-like or an options object', function test( t ) {
	var values;
	var i;

	values = [
		5,
		NaN,
		null,
		undefined,
		true,
		function(){}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the returned function throws an error if not provided an array-like object', function test( t ) {
	var sample;
	var values;
	var i;

	sample = factory();

	values = [
		5,
		NaN,
		null,
		undefined,
		true,
		{},
		function(){}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sample( value );
		};
	}
});

tape( 'the returned function throws an error if provided an options object which is not an object', function test( t ) {
	var sample;
	var values;
	var i;

	sample = factory();

	values = [
		'5',
		5,
		NaN,
		null,
		undefined,
		true,
		[],
		function(){}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sample( [], value );
		};
	}
});

tape( 'the returned function throws an error if provided an invalid option', function test( t ) {
	var sample;
	var values;
	var i;

	sample = factory();

	values = [
		'5',
		5,
		NaN,
		null,
		undefined,
		[],
		{},
		function(){}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sample( [], {
				'replace': value
			});
		};
	}
});

tape( 'the returned function throws an error if provided a `size` option exceeding the input array length when sampling without recplacement', function test( t ) {
	var sample;
	var values;
	var i;

	sample = factory();

	values = [
		[ [ 1, 2, 3 ], 4 ],
		[ [ 1, 2, 3, 4, 5, 6 ], 10 ]
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws a type error' );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			sample( value[ 0 ], {
				'size': value[ 1 ],
				'replace': false
			});
		};
	}
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var sample = factory({
		'seed': 311
	});
	t.equal( typeof sample.SEED, 'number', 'has `SEED` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var sample = factory();
	t.equal( typeof sample.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'the returned function samples characters with replacement from a string', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 311
	});

	expected = [ 'f', 'e', 'e', 'e', 'b', 'a' ];
	actual = sample( 'abcdef' );

	t.equal( isStringArray( actual ), true, 'returns an array of chracters' );
	t.equal( actual.length, 6, 'returned array has six elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function samples with replacement from an array', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 311
	});

	// Simple array:
	expected = [ 4, 4, 3, 3 ];
	actual = sample( [ 1, 2, 3, 4 ] );

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 4, 'returned array has four elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	// Typed array:
	expected = [ 6, 5, 8, 8, 8 ];
	actual = sample( new Float64Array( [ 5, 6, 7, 8, 9 ] ) );

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 5, 'returned array has five elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function samples without replacement from an array', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 789
	});

	// Simple array:
	expected = [ 4, 3, 1, 2 ];
	actual = sample( [ 1, 2, 3, 4 ], {
		'replace': false
	});

	// Typed array:
	expected = [ 7, 9, 6, 5, 8 ];
	actual = sample( new Int32Array( [ 5, 6, 7, 8, 9 ] ), {
		'replace': false
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 5, 'returned array has five elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function draws a sample with the specified size from an array (with replacement)', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 789
	});

	// Simple array:
	expected = [ 0, 0, 0, 1, 0, 1, 1, 1, 1, 1 ];
	actual = sample( [ 0, 1 ], {
		'size': 10
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 10, 'returned array has ten elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	// Typed array:
	expected = [ 1, 1, 1, 0, 0, 1, 0, 1, 1, 0 ];
	actual = sample( new Int8Array( [ 0, 1 ] ), {
		'size': 10
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 10, 'returned array has ten elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function draws a sample with the specified size from an array (without replacement)', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 656
	});

	// Simple array:
	expected = [ 3, 8, 7, 5 ];
	actual = sample( [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ], {
		'size': 4,
		'replace': false
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 4, 'returned array has four elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	// Typed array:
	expected = [ 3, 5, 4, 8, 6 ];
	actual = sample( new Int16Array( [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ] ), {
		'size': 5,
		'replace': false
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 5, 'returned array has five elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function draws a sample from an array using specified probabilities (with replacement)', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 321
	});

	// Simple array:
	expected = [ 4, 5, 2, 2 ];
	actual = sample( [ 1, 2, 3, 4, 5, 6 ], {
		'size': 4,
		'probs': [ 1/12, 2/6, 2/6, 1/12, 1/12, 1/12 ]
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 4, 'returned array has four elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	// Typed array:
	expected = [ 5, 4, 5, 2, 2 ];
	actual = sample( new Int8Array( [ 1, 2, 3, 4, 5, 6 ] ), {
		'size': 5,
		'probs': [ 0.0, 1/12, 1/12, 3/6, 2/6, 0.0 ]
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 5, 'returned array has five elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	expected = [ 1, 3, 1, 1, 3 ];
	actual = sample( new Int8Array( [ 1, 2, 3, 4 ] ), {
		'size': 5,
		'probs': [ 0.5, 0.2, 0.2, 0.1 ]
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 5, 'returned array has five elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});

tape( 'the returned function draws a sample from an array using specified probabilities (without replacement)', function test( t ) {
	var expected;
	var actual;
	var sample;

	sample = factory({
		'seed': 589
	});

	// Simple array:
	expected = [ 3, 6, 5 ];
	actual = sample( [ 1, 2, 3, 4, 5, 6 ], {
		'size': 3,
		'probs': [ 1/12, 2/6, 2/6, 1/12, 1/12, 1/12 ],
		'replace': false
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 3, 'returned array has three elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	// Typed array:
	expected = [ 6, 3, 2, 1 ];
	actual = sample( new Int8Array( [ 1, 2, 3, 4, 5, 6 ] ), {
		'size': 4,
		'probs': [ 1/12, 2/6, 2/6, 1/12, 1/12, 1/12 ],
		'replace': false
	});

	t.equal( isNumberArray( actual ), true, 'returns an array of numbers' );
	t.equal( actual.length, 4, 'returned array has four elements' );
	t.deepEqual( actual, expected, 'returned array is a random draw' );

	t.end();
});
