// @stdlib/math/base/random/triangular/test/test.factory.js
'use strict';

// MODULES //

var tape = require( 'tape' );
var ENV = require( '@stdlib/utils/env' );
var kstest = require( '@stdlib/math/statistics/kstest' );
var round = require( '@stdlib/math/base/special/round' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var factory = require( './../lib/factory.js' );


// VARIABLES //

var opts = {
	'skip': ( ENV.TEST_MODE === 'coverage' )
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof factory, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns a pseudorandom number generator (no seed)', function test( t ) {
	var triangular;
	var r;
	var i;

	// When binding distribution parameters...
	triangular = factory( -10.0, 10.0, 5.0 );
	for ( i = 0; i < 100; i++ ) {
		r = triangular();
		t.equal( typeof r, 'number', 'returns a number' );
	}

	// Without binding distribution parameters...
	triangular = factory();
	for ( i = 0; i < 100; i++ ) {
		r = triangular( -10.0, 10.0, 5.0 );
		t.equal( typeof r, 'number', 'returns a number' );
	}

	t.end();
});

tape( 'the function returns a seeded pseudorandom number generator', function test( t ) {
	var triangular1;
	var triangular2;
	var seed;
	var r1;
	var r2;
	var i;

	seed = round( Date.now() / 1000 );

	triangular1 = factory( 0.0, 4.0, 2.0, {
		'seed': seed
	});
	triangular2 = factory( 0.0, 4.0, 2.0, {
		'seed': seed
	});

	t.notEqual( triangular1, triangular2, 'separate generators' );

	for ( i = 0; i < 100; i++ ) {
		r1 = triangular1();
		r2 = triangular2();
		t.equal( r1, r2, 'both return same number' );
	}
	t.end();
});

tape( 'attached to the returned function is the generator name', function test( t ) {
	var triangular = factory();
	t.equal( triangular.NAME, 'triangular', 'has `NAME` property' );
	t.end();
});

tape( 'attached to the returned function is the underlying PRNG', function test( t ) {
	var triangular = factory();
	t.equal( typeof triangular.PRNG, 'function', 'has `PRNG` property' );
	t.end();
});

tape( 'attached to the returned function is the generator seed', function test( t ) {
	var triangular = factory({
		'seed': 12345
	});
	t.equal( typeof triangular.SEED, 'number', 'has `SEED` property' );
	t.equal( triangular.SEED, 12345, 'equal to provided seed' );
	t.end();
});

tape( 'the function throws an error if minimum support `a` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value, 2.0, 1.0 );
		};
	}
});

tape( 'the function throws an error if maximum support `b` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 2.0, value, 3.0 );
		};
	}
});

tape( 'the function throws an error if mode `c` is not a number primitive', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		null,
		true,
		undefined,
		NaN,
		[],
		{},
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 2.0, 3.0, value );
		};
	}
});

tape( 'the function throws an error if the condition `a <= c <= b` is not satisfied', function test( t ) {
	var values;
	var i;

	values = [
		[ 0.0, 1.0, 2.0 ],
		[ -2.0, -4.0, -3.0 ],
		[ 2.0, 1.0, 1.5 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), RangeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( arr ) {
		return function badValue() {
			factory( arr[0], arr[1], arr[2] );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (no other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( value );
		};
	}
});

tape( 'the function throws an error if provided an options argument which is not an object (other arguments)', function test( t ) {
	var values;
	var i;

	values = [
		'abc',
		5,
		null,
		true,
		undefined,
		NaN,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			factory( 0.0, 1.0, 0.5, value );
		};
	}
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `a` is `NaN`', function test( t ) {
	var triangular;
	var r;

	triangular = factory();
	r = triangular( NaN, 5.0, 2.5 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `b` is `NaN`', function test( t ) {
	var triangular;
	var r;

	triangular = factory();
	r = triangular( 0.0, NaN, 1.0 );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if `c` is `NaN`', function test( t ) {
	var triangular;
	var r;

	triangular = factory();
	r = triangular( 0.0, 10.0, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if provided all `NaNs`', function test( t ) {
	var triangular;
	var r;

	triangular = factory();
	r = triangular( NaN, NaN, NaN );

	t.strictEqual( isnan( r ), true, 'returns NaN' );
	t.end();
});

tape( 'when called without arguments, the function returns a function that returns `NaN` if the condition `a <= c <= b` is not satisfied', function test( t ) {
	var triangular;
	var values;
	var r;
	var i;

	triangular = factory();
	values = [
		[ 2.0, 1.0, 1.5 ],
		[ -5.0, -4.0, 3.14 ],
		[ 0.0, 1.0, 5.5 ]
	];

	for ( i = 0; i < values.length; i++ ) {
		r = triangular( values[ i ][ 0 ], values[ i ][ 1 ], values[ i ][ 2 ] );
		t.strictEqual( isnan( r ), true, 'returns `NaN` when provided '+values[ i ] );
	}
	t.end();
});

tape( 'the function returns a PRNG for generating random numbers from a triangular distribution', opts, function test( t ) {
	var threshold;
	var count;
	var npass;
	var N;
	var a;
	var b;
	var c;
	var x;

	threshold = 0.10;

	a = 2.0;
	b = 4.0;
	c = 3.33;

	x = new Array( 1e3 );
	N = 500;

	count = -1;
	npass = 0;

	gof();

	function gof() {
		var triangular;
		var rejected;
		var pValue;
		var bool;
		var i;
		var j;

		count += 1;
		rejected = 0;
		for ( i = 0; i < N; i++ ) {
			triangular = factory( a, b, c );
			t.ok( true, 'seed: '+triangular.SEED );
			for ( j = 0; j < x.length; j++ ) {
				x[ j ] = triangular();
				if ( x[ j ] < a || x[ j ] > b ) {
					t.ok( false, 'returned a number outside support: '+x[ j ] );
				}
			}
			// Test using Kolmogorov-Smirnov goodness-of-fit test:
			pValue = kstest( x, 'triangular', a, b, c ).pValue;
			t.equal( typeof pValue, 'number', 'returns a p-value: '+pValue );
			if ( pValue < 0.05 ) {
				rejected += 1;
			}
		}
		// Account for small sample size and few repeats...
		bool = ( rejected / N < threshold );

		// If we succeed the first time, we are done...
		if ( count === 0 && bool ) {
			return done( bool, rejected );
		}
		// Retry mode...
		if ( bool ) {
			npass += 1;
		}
		// Retry twice...
		if ( count < 2 ) {
			return gof();
		}
		// Both retries must succeed for test to pass:
		bool = ( npass >= 2 );
		return done( bool, rejected );
	}

	function done( bool, rejected ) {
		t.ok( bool, 'null hypothesis (i.e., that numbers are drawn from Triangular('+a+','+b+','+c+')) is rejected in less than '+(threshold*100)+'% of cases ('+rejected+' of '+N+'). Repeats: '+npass+' of '+count+'.' );
		t.end();
	}
});
