<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Table Pipes

> [ESLint rule][eslint-rules] to require pipes as fences for Markdown table rows in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-table-pipes' );
```

#### rule

[ESLint rule][eslint-rules] to require pipes as fences for Markdown table rows in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-table-pipes, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* x   | y
* any | boop
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Boop beep.
*
* | x   | y    |
* | any | boop |
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-table-pipes' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Squares a number.',
    '*',
    '* Input | Output',
    '* ----- | ------',
    '* 2     | 4',
    '*',
    '* @param {number} x - input value',
    '* @returns {number} output value',
    '*',
    '* @example',
    '* var out = square( 2 );',
    '* // returns 4',
    '*/',
    'function square( x ) {',
    '\treturn x*x;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-table-pipes', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-table-pipes': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-table-pipes',
            'severity': 2,
            'message': 'Missing initial pipe in table fence',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* Input | Output',
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-table-pipes',
            'severity': 2,
            'message': 'Missing final pipe in table fence',
            'line': 4,
            'column': 17,
            'nodeType': null,
            'source': '* Input | Output',
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-table-pipes',
            'severity': 2,
            'message': 'Missing initial pipe in table fence',
            'line': 6,
            'column': 3,
            'nodeType': null,
            'source': '* 2     | 4',
            'endLine': 14,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-table-pipes',
            'severity': 2,
            'message': 'Missing final pipe in table fence',
            'line': 6,
            'column': 12,
            'nodeType': null,
            'source': '* 2     | 4',
            'endLine': 14,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
