/*
* @license Apache-2.0
*
* Copyright (c) 2024 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import replaceBeforeLast = require( './index' );


// TESTS //

// The function returns a string...
{
	replaceBeforeLast( 'beep boop', ' ', 'foo' ); // $ExpectType string
	replaceBeforeLast( 'beep boop', 'xyz', 'foo' ); // $ExpectType string
	replaceBeforeLast( 'beep boop', '', 'foo' ); // $ExpectType string
}

// The compiler throws an error if the function is provided arguments having invalid types...
{
	replaceBeforeLast( true, 'd', 'foo' ); // $ExpectError
	replaceBeforeLast( false, 'd' , 'foo' ); // $ExpectError
	replaceBeforeLast( 3, 'd' , 'foo' ); // $ExpectError
	replaceBeforeLast( [], 'd' , 'foo' ); // $ExpectError
	replaceBeforeLast( {}, 'd' , 'foo' ); // $ExpectError
	replaceBeforeLast( ( x: number ): number => x, 'd', 'foo' ); // $ExpectError

	replaceBeforeLast( 'abc', true, 'foo' ); // $ExpectError
	replaceBeforeLast( 'abc', false, 'foo' ); // $ExpectError
	replaceBeforeLast( 'abc', 5 , 'foo' ); // $ExpectError
	replaceBeforeLast( 'abc', [], 'foo' ); // $ExpectError
	replaceBeforeLast( 'abc', {} , 'foo' ); // $ExpectError
	replaceBeforeLast( 'abc', ( x: number ): number => x , 'foo' ); // $ExpectError

	replaceBeforeLast( 'abc', 'd', true ); // $ExpectError
	replaceBeforeLast( 'abc', 'd', false ); // $ExpectError
	replaceBeforeLast( 'abc', 'd', 5 ); // $ExpectError
	replaceBeforeLast( 'abc', 'd', [] ); // $ExpectError
	replaceBeforeLast( 'abc', 'd', {} ); // $ExpectError
	replaceBeforeLast( 'abc', 'd', ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided insufficient arguments...
{
	replaceBeforeLast(); // $ExpectError
	replaceBeforeLast( 'abc' ); // $ExpectError
	replaceBeforeLast( 'abc', 'd' ); // $ExpectError
}
