// @stdlib/_tools/github/user-details/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var userinfo = require( './../lib/main.js' );


// FIXTURES //

var getOpts = require( './fixtures/opts.js' );
var data = require( './fixtures/data.json' );
var info = require( './fixtures/info.json' );
var results = {
	'meta': {
		'total': 2,
		'success': 2,
		'failure': 0
	},
	'data': {
		'kgryte': data[ 0 ],
		'planeshifter': data[ 1 ]
	},
	'failures': {}
};


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.equal( typeof userinfo, 'function', 'export is a function' );
	t.end();
});

tape( 'function returns an error to a provided callback if an error is encountered when fetching user details', function test( t ) {
	var userinfo;
	var opts;

	userinfo = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	opts = getOpts();
	userinfo( opts, done );

	function factory( opts, clbk ) {
		return function userinfo() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk({
					'status': 404,
					'message': 'beep'
				});
			}
		};
	}

	function done( error ) {
		t.equal( error.status, 404, 'equal status' );
		t.equal( error.message, 'beep', 'equal message' );
		t.end();
	}
});

tape( 'functions returns a username hash containing user details to a provided callback', function test( t ) {
	var expected;
	var userinfo;
	var opts;

	userinfo = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = results;

	opts = getOpts();
	userinfo( opts, done );

	function factory( opts, clbk ) {
		return function userinfo() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, results ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( results, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function returns rate limit info to a provided callback', function test( t ) {
	var expected;
	var userinfo;
	var opts;

	userinfo = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = info;

	opts = getOpts();
	userinfo( opts, done );

	function factory( opts, clbk ) {
		return function userinfo() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, results, info ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( info, expected, 'deep equal' );
		t.end();
	}
});

tape( 'function supports requesting user details without authentication (no token)', function test( t ) {
	var expected;
	var userinfo;
	var opts;

	userinfo = proxyquire( './../lib/main.js', {
		'./factory.js': factory
	});

	expected = results;

	opts = getOpts();
	delete opts.token;

	userinfo( opts, done );

	function factory( opts, clbk ) {
		return function userinfo() {
			setTimeout( onTimeout, 0 );
			function onTimeout() {
				clbk( null, results, info );
			}
		};
	}

	function done( error, results ) {
		t.equal( error, null, 'does not return an error' );
		t.deepEqual( results, expected, 'deep equal' );
		t.end();
	}
});
