<!--

@license Apache-2.0

Copyright (c) 2018 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# Heading Style

> [ESLint rule][eslint-rules] to enforce that Markdown headings follow a specified style in JSDoc descriptions.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-heading-style' );
```

#### rule

[ESLint rule][eslint-rules] to enforce that Markdown headings follow a specified style in JSDoc descriptions.

**Bad**:

<!-- eslint-disable stdlib/jsdoc-heading-style, stdlib/jsdoc-markdown-remark -->

```javascript
/**
* Boop beep.
*
* ## Boop ##
*
* Beep
* ====
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

**Good**:

```javascript
/**
* Boop beep.
*
* ## Boop
*
* ## Beep
*
* @returns {string} a value
*
* @example
* var str = beep();
* // returns 'boop'
*/
function beep() {
    return 'boop';
}
```

The [rule][eslint-rules] may be configured using the same options as supported by [remark][remark-lint-heading-style]. By default, the rule requires that headings be formatted as ATX.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-heading-style' );

var linter = new Linter();
var result;
var code;

// Generate our source code:
code = [
    '/**',
    '* Beep boop.',
    '*',
    '* ## Beep ##',
    '*',
    '* boop',
    '*',
    '* Boop',
    '* ----',
    '*',
    '* beep',
    '*',
    '*',
    '* @param {string} str - input value',
    '* @returns {string} output value',
    '*',
    '* @example',
    '* var out = beep( "boop" );',
    '* // returns "beepboop"',
    '*/',
    'function beep( str ) {',
    '\treturn "beep" + str;',
    '}'
].join( '\n' );

// Register the ESLint rule:
linter.defineRule( 'jsdoc-heading-style', rule );

// Lint the code:
result = linter.verify( code, {
    'rules': {
        'jsdoc-heading-style': 'error'
    }
});
console.log( result );
/* =>
    [
        {
            'ruleId': 'jsdoc-heading-style',
            'severity': 2,
            'message': 'Headings should use atx',
            'line': 4,
            'column': 3,
            'nodeType': null,
            'source': '* ## Boop ##',
            'endLine': 19,
            'endColumn': 3
        },
        {
            'ruleId': 'jsdoc-heading-style',
            'severity': 2,
            'message': 'Headings should use atx',
            'line': 8,
            'column': 3,
            'nodeType': null,
            'source': '* Boop',
            'endLine': 19,
            'endColumn': 3
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

[remark-lint-heading-style]: https://github.com/remarkjs/remark-lint/tree/19150d94f89f7a0d94d083417890236d11839641/packages/remark-lint-heading-style

</section>

<!-- /.links -->
