/*
* @license Apache-2.0
*
* Copyright (c) 2020 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import copyWithin = require( './index' );


// TESTS //

// The function returns the updated input collection ...
{
	copyWithin( [ 2, 3 ], 1 ); // $ExpectType Collection
	copyWithin( [ 'a', 'b', 'c' ], 0, 1 ); // $ExpectType Collection
	copyWithin( [ true, false ], 1 ); // $ExpectType Collection
	copyWithin( [ true, true, 1, 2, 3 ], 1, 2, 4 ); // $ExpectType Collection
}

// The compiler throws an error if the function is provided a first argument which is not a collection...
{
	copyWithin( true ); // $ExpectError
	copyWithin( false ); // $ExpectError
	copyWithin( 5 ); // $ExpectError
}

// The compiler throws an error if the function is not provided at least one argument...
{
	copyWithin(); // $ExpectError
}
