// @stdlib/_tools/github/create-issue/test/test.main.js
/**
* @license Apache-2.0
*
* Copyright (c) 2020 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var proxyquire = require( 'proxyquire' );
var noop = require( '@stdlib/utils/noop' );
var createIssue = require( './../lib/main.js' );


// FIXTURES //

var getOpts = require( './fixtures/opts.js' );
var data = require( './fixtures/results.json' );
var info = require( './fixtures/info.json' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.equal( typeof createIssue, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided an invalid `options` argument', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			createIssue( value, noop );
		};
	}
});

tape( 'the function throws an error if provided an invalid option', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			var opts = {
				'port': value
			};
			createIssue( opts, noop );
		};
	}
});

tape( 'the function throws if provided a callback argument which is not a function (options)', function test( t ) {
	var values;
	var opts;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		{}
	];

	opts = getOpts();
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws a type error when provided ' + values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			createIssue( opts, value );
		};
	}
});

tape( 'the function returns an error to a provided callback if an error is encountered when creating an issue', function test( t ) {
	var createIssue;
	var opts;

	createIssue = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	opts = getOpts();
	createIssue( 'foo/bar', 'Bug', opts, done );

	function query( slug, title, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( new Error( 'beep' ), null, info );
		}
	}

	function done( error ) {
		t.equal( error instanceof Error, true, 'error instance' );
		t.equal( error.message, 'beep', 'returns expected value' );
		t.end();
	}
});

tape( 'the function returns response data to a provided callback', function test( t ) {
	var createIssue;
	var expected;
	var opts;

	createIssue = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	expected = data;

	opts = getOpts();
	createIssue( 'foo/bar', 'Bug', opts, done );

	function query( slug, title, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( null, data, info );
		}
	}

	function done( error, data ) {
		if ( error ) {
			t.fail( error.message );
		}
		t.deepEqual( data, expected, 'returns expected value' );
		t.end();
	}
});

tape( 'the function returns rate limit info to a provided callback', function test( t ) {
	var createIssue;
	var expected;
	var opts;

	createIssue = proxyquire( './../lib/main.js', {
		'./query.js': query
	});

	expected = info;

	opts = getOpts();
	createIssue( 'foo/bar', 'Bug', opts, done );

	function query( slug, title, opts, clbk ) {
		setTimeout( onTimeout, 0 );
		function onTimeout() {
			clbk( null, data, info );
		}
	}

	function done( error, data, info ) {
		if ( error ) {
			t.fail( error.message );
		}
		t.deepEqual( info, expected, 'returns expected value' );
		t.end();
	}
});
